// @php-wasm/cli-util/index.js
import p from "fs";
import x from "path";
import { toPosixPath as _ } from "@php-wasm/util";
import { XMLParser as T, XMLBuilder as F } from "fast-xml-parser";
import * as w from "jsonc-parser";
class W {
  constructor(e) {
    this.verbosity = e.verbosity, this.writeStream = e.writeStream || process.stdout;
  }
  get isTTY() {
    return !!this.writeStream.isTTY;
  }
  get isQuiet() {
    return this.verbosity === "quiet";
  }
  /**
   * ANSI formatting helpers.
   *
   * These only apply color codes when outputting to a terminal (TTY).
   * When piped to files or non-TTY streams, they return plain text to
   * avoid polluting logs with escape sequences.
   */
  bold(e) {
    return this.isTTY ? `\x1B[1m${e}\x1B[0m` : e;
  }
  dim(e) {
    return this.isTTY ? `\x1B[2m${e}\x1B[0m` : e;
  }
  italic(e) {
    return this.isTTY ? `\x1B[3m${e}\x1B[0m` : e;
  }
  red(e) {
    return this.isTTY ? `\x1B[31m${e}\x1B[0m` : e;
  }
  green(e) {
    return this.isTTY ? `\x1B[32m${e}\x1B[0m` : e;
  }
  yellow(e) {
    return this.isTTY ? `\x1B[33m${e}\x1B[0m` : e;
  }
  cyan(e) {
    return this.isTTY ? `\x1B[36m${e}\x1B[0m` : e;
  }
  highlight(e) {
    return this.yellow(e);
  }
  print(e) {
    this.isQuiet || this.writeStream.write(`${e}
`);
  }
  printError(e) {
    this.writeStream.write(`${this.red("Error:")} ${e}
`);
  }
  printWarning(e) {
    this.isQuiet || this.writeStream.write(`${this.yellow("Warning:")} ${e}
`);
  }
}
const A = "PHPWASMCLI", J = [
  "/dev/",
  "/home/",
  "/internal/",
  "/request/",
  "/proc/"
];
async function U(u, e, f) {
  const m = f === "win32" ? (
    // On Windows, creating a 'dir' symlink can require elevated permissions.
    // In this case, let's make junction points because they function like
    // symlinks and do not require elevated permissions.
    "junction"
  ) : "dir";
  p.symlinkSync(u, e, m);
}
async function q(u) {
  try {
    p.lstatSync(u).isSymbolicLink() && p.unlinkSync(u);
  } catch {
  }
}
function M(u, e) {
  return e.filter((f) => {
    const m = x.resolve(f.hostPath), c = x.join(u, x.sep);
    return (
      // If auto-mounting from the current directory,
      // the entire project directory can be mapped.
      m === u || m.startsWith(c)
    );
  });
}
const E = {
  ignoreAttributes: !1,
  attributeNamePrefix: "",
  preserveOrder: !0,
  cdataPropName: "__cdata",
  commentPropName: "__xmlComment",
  allowBooleanAttributes: !0,
  trimValues: !0
}, b = {
  ignoreAttributes: E.ignoreAttributes,
  attributeNamePrefix: E.attributeNamePrefix,
  preserveOrder: E.preserveOrder,
  cdataPropName: E.cdataPropName,
  commentPropName: E.commentPropName,
  suppressBooleanAttributes: !E.allowBooleanAttributes,
  format: !0,
  indentBy: "	"
}, $ = {
  allowEmptyContent: !0,
  allowTrailingComma: !0
};
function N(u, e) {
  var S, C, I, B, k, L;
  const { name: f, host: m, port: c, pathMappings: h, ideKey: o } = e, P = new T(E), v = (() => {
    try {
      return P.parse(u, !0);
    } catch {
      throw new Error("PhpStorm configuration file is not valid XML.");
    }
  })(), y = {
    server: [{}],
    ":@": {
      name: f,
      // NOTE: PhpStorm quirk: Xdebug only works when the full URL (including port)
      // is provided in `host`. The separate `port` field is ignored or misinterpreted,
      // so we rely solely on host: "host:port".
      host: `${m}:${c}`,
      use_path_mappings: "true"
    }
  };
  h && h.length && (y.server[0].path_mappings = h.map(
    (n) => ({
      mapping: [],
      ":@": {
        "local-root": `$PROJECT_DIR$/${_(
          x.relative(e.projectDir, n.hostPath)
        )}`,
        "remote-root": n.vfsPath
      }
    })
  ));
  let l = v == null ? void 0 : v.find(
    (n) => !!(n != null && n.project)
  );
  if (l) {
    const n = (S = l[":@"]) == null ? void 0 : S.version;
    if (n === void 0)
      throw new Error(
        'PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but the <project> configuration has no version number.'
      );
    if (n !== "4")
      throw new Error(
        `PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but we found a <project> configuration with version "${n}".`
      );
  }
  l === void 0 && (l = {
    project: [],
    ":@": { version: "4" }
  }, v.push(l));
  let i = (C = l.project) == null ? void 0 : C.find(
    (n) => {
      var j;
      return !!(n != null && n.component) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === "PhpServers";
    }
  );
  i === void 0 && (i = {
    component: [],
    ":@": { name: "PhpServers" }
  }, l.project === void 0 && (l.project = []), l.project.push(i));
  let r = (I = i.component) == null ? void 0 : I.find(
    (n) => !!(n != null && n.servers)
  );
  r === void 0 && (r = { servers: [] }, i.component === void 0 && (i.component = []), i.component.push(r));
  const s = (B = r.servers) == null ? void 0 : B.findIndex(
    (n) => {
      var j;
      return !!(n != null && n.server) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === f;
    }
  );
  (s === void 0 || s < 0) && (r.servers === void 0 && (r.servers = []), r.servers.push(y));
  let t = (k = l.project) == null ? void 0 : k.find(
    (n) => {
      var j;
      return !!(n != null && n.component) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === "RunManager";
    }
  );
  if (t === void 0 && (t = {
    component: [],
    ":@": { name: "RunManager" }
  }, l.project === void 0 && (l.project = []), l.project.push(t)), (((L = t.component) == null ? void 0 : L.findIndex(
    (n) => {
      var j;
      return !!(n != null && n.configuration) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === f;
    }
  )) ?? -1) < 0) {
    const n = {
      configuration: [
        {
          method: [],
          ":@": { v: "2" }
        }
      ],
      ":@": {
        name: f,
        type: "PhpRemoteDebugRunConfigurationType",
        factoryName: "PHP Remote Debug",
        filter_connections: "FILTER",
        server_name: f,
        session_id: o
      }
    };
    t.component === void 0 && (t.component = []), t.component.push(n);
  }
  const d = new F(b).build(v);
  try {
    P.parse(d, !0);
  } catch {
    throw new Error(
      "The resulting PhpStorm configuration file is not valid XML."
    );
  }
  return d;
}
function O(u, e) {
  var l, i, r, s;
  const { pathSkippings: f } = e, m = new T(E), c = (() => {
    try {
      return m.parse(u, !0);
    } catch {
      throw new Error(
        "PhpStorm PHP configuration file is not valid XML."
      );
    }
  })();
  let h = c == null ? void 0 : c.find(
    (t) => !!(t != null && t.project)
  );
  if (h) {
    const t = (l = h[":@"]) == null ? void 0 : l.version;
    if (t === void 0)
      throw new Error(
        'PhpStorm IDE integration only supports <project version="4"> in php.xml, but the <project> configuration has no version number.'
      );
    if (t !== "4")
      throw new Error(
        `PhpStorm IDE integration only supports <project version="4"> in php.xml, but we found a <project> configuration with version "${t}".`
      );
  }
  h === void 0 && (h = {
    project: [],
    ":@": { version: "4" }
  }, c.push(h));
  let o = (i = h.project) == null ? void 0 : i.find(
    (t) => {
      var a;
      return !!(t != null && t.component) && ((a = t == null ? void 0 : t[":@"]) == null ? void 0 : a.name) === "PhpStepFilterConfiguration";
    }
  );
  o === void 0 && (o = {
    component: [],
    ":@": { name: "PhpStepFilterConfiguration" }
  }, h.project === void 0 && (h.project = []), h.project.push(o));
  let P = (r = o.component) == null ? void 0 : r.find(
    (t) => !!(t != null && t.skipped_files)
  );
  if (P === void 0 && (P = { skipped_files: [] }, o.component === void 0 && (o.component = []), o.component.push(P)), f && f.length)
    for (const t of f) {
      const g = `$PROJECT_DIR$${t.endsWith("/") ? t.slice(0, -1) : t}`;
      ((s = P.skipped_files) == null ? void 0 : s.some(
        (S) => {
          var C;
          return !!(S != null && S.skipped_file) && ((C = S == null ? void 0 : S[":@"]) == null ? void 0 : C.file) === g;
        }
      )) || (P.skipped_files === void 0 && (P.skipped_files = []), P.skipped_files.push({
        skipped_file: [],
        ":@": { file: g }
      }));
    }
  const y = new F(b).build(c);
  try {
    m.parse(y, !0);
  } catch {
    throw new Error(
      "The resulting PhpStorm PHP configuration file is not valid XML."
    );
  }
  return y;
}
function X(u, e) {
  var l, i;
  const { name: f, pathMappings: m, pathSkippings: c } = e, h = [];
  let o = u, P = w.parseTree(o, h, $);
  if (P === void 0 || h.length)
    throw new Error("VS Code configuration file is not valid JSON.");
  let v = w.findNodeAtLocation(P, ["configurations"]);
  if (v === void 0 || v.children === void 0) {
    const r = w.modify(o, ["configurations"], [], {});
    o = w.applyEdits(o, r), P = w.parseTree(o, [], $), v = w.findNodeAtLocation(P, [
      "configurations"
    ]);
  }
  const y = (l = v == null ? void 0 : v.children) == null ? void 0 : l.findIndex(
    (r) => {
      var s;
      return ((s = w.findNodeAtLocation(r, ["name"])) == null ? void 0 : s.value) === f;
    }
  );
  if (y === void 0 || y < 0) {
    const r = {
      name: f,
      type: "php",
      request: "launch",
      port: 9003
    };
    m && m.length && (r.pathMappings = m.reduce((a, g) => (a[g.vfsPath] = `\${workspaceFolder}/${_(
      x.relative(e.workspaceDir, g.hostPath)
    )}`, a), {})), c && c.length && (r.skipFiles = c.map(
      (a) => a.endsWith("/") ? `${a}**` : a
    ));
    const s = ((i = v == null ? void 0 : v.children) == null ? void 0 : i.length) || 0, t = w.modify(
      o,
      ["configurations", s],
      r,
      {
        formattingOptions: {
          insertSpaces: !0,
          tabSize: 4,
          eol: `
`
        }
      }
    );
    o = D(o, t);
  }
  return o;
}
async function z({
  name: u,
  ides: e,
  host: f,
  port: m,
  cwd: c,
  mounts: h,
  pathSkippings: o,
  ideKey: P = A
}) {
  const v = h ? M(c, h) : [], y = {};
  if (e.includes("phpstorm")) {
    const l = ".idea/workspace.xml", i = x.join(
      c,
      l
    );
    if (!p.existsSync(i)) {
      if (p.existsSync(x.dirname(i)))
        p.writeFileSync(
          i,
          `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>`
        );
      else if (e.length == 1)
        throw new Error(
          "PhpStorm IDE integration requested, but no '.idea' directory was found in the current working directory."
        );
    }
    if (p.existsSync(i)) {
      const r = p.readFileSync(i, "utf8"), s = N(r, {
        name: u,
        host: f,
        port: m,
        projectDir: c,
        pathMappings: v,
        ideKey: P
      });
      p.writeFileSync(i, s), y.phpstorm = l;
    }
    if (o && o.length) {
      const r = ".idea/php.xml", s = x.join(
        c,
        r
      );
      if (p.existsSync(s) || p.existsSync(x.dirname(s)) && p.writeFileSync(
        s,
        `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>`
      ), p.existsSync(s)) {
        const t = p.readFileSync(
          s,
          "utf8"
        ), a = O(t, {
          pathSkippings: o
        });
        p.writeFileSync(s, a), y["phpstorm-php"] = r;
      }
    }
  }
  if (e.includes("vscode")) {
    const l = ".vscode/launch.json", i = x.join(
      c,
      l
    );
    if (!p.existsSync(i)) {
      if (p.existsSync(x.dirname(i)))
        p.writeFileSync(
          i,
          `{
    "configurations": []
}`
        );
      else if (e.length == 1)
        throw new Error(
          "VS Code IDE integration requested, but no '.vscode' directory was found in the current working directory."
        );
    }
    if (p.existsSync(i)) {
      const r = p.readFileSync(i, "utf-8"), s = X(r, {
        name: u,
        workspaceDir: c,
        pathMappings: v,
        pathSkippings: o
      });
      s !== r && (p.writeFileSync(i, s), y.vscode = l);
    }
  }
  return y;
}
async function K(u, e) {
  var h, o, P, v, y;
  const f = x.join(e, ".idea/workspace.xml");
  if (p.existsSync(f)) {
    const l = p.readFileSync(f, "utf8"), i = new T(E), r = (() => {
      try {
        return i.parse(l, !0);
      } catch {
        throw new Error(
          "PhpStorm configuration file is not valid XML."
        );
      }
    })(), s = r.find(
      (d) => !!(d != null && d.project)
    ), t = (h = s == null ? void 0 : s.project) == null ? void 0 : h.find(
      (d) => {
        var S;
        return !!(d != null && d.component) && ((S = d == null ? void 0 : d[":@"]) == null ? void 0 : S.name) === "PhpServers";
      }
    ), a = (o = t == null ? void 0 : t.component) == null ? void 0 : o.find(
      (d) => !!(d != null && d.servers)
    ), g = (P = a == null ? void 0 : a.servers) == null ? void 0 : P.findIndex(
      (d) => {
        var S;
        return !!(d != null && d.server) && ((S = d == null ? void 0 : d[":@"]) == null ? void 0 : S.name) === u;
      }
    );
    if (g !== void 0 && g >= 0) {
      a.servers.splice(g, 1);
      const S = new F(b).build(r);
      try {
        i.parse(S, !0);
      } catch {
        throw new Error(
          "The resulting PhpStorm configuration file is not valid XML."
        );
      }
      S === `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
	<component name="PhpServers">
		<servers></servers>
	</component>
</project>` ? p.unlinkSync(f) : p.writeFileSync(f, S);
    }
  }
  const m = x.join(e, ".idea/php.xml");
  if (p.existsSync(m)) {
    const l = p.readFileSync(m, "utf8"), i = new T(E), r = (() => {
      try {
        return i.parse(l, !0);
      } catch {
        throw new Error(
          "PhpStorm PHP configuration file is not valid XML."
        );
      }
    })(), s = r.find(
      (a) => !!(a != null && a.project)
    ), t = (v = s == null ? void 0 : s.project) == null ? void 0 : v.findIndex(
      (a) => {
        var g;
        return !!(a != null && a.component) && ((g = a == null ? void 0 : a[":@"]) == null ? void 0 : g.name) === "PhpStepFilterConfiguration";
      }
    );
    if (t !== void 0 && t >= 0) {
      s.project.splice(t, 1);
      const g = new F(b).build(r);
      try {
        i.parse(g, !0);
      } catch {
        throw new Error(
          "The resulting PhpStorm PHP configuration file is not valid XML."
        );
      }
      g === `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>` ? p.unlinkSync(m) : p.writeFileSync(m, g);
    }
  }
  const c = x.join(e, ".vscode/launch.json");
  if (p.existsSync(c)) {
    const l = [], i = p.readFileSync(c, "utf-8"), r = w.parseTree(i, l, $);
    if (r === void 0 || l.length)
      throw new Error("VS Code configuration file is not valid JSON.");
    const s = w.findNodeAtLocation(r, [
      "configurations"
    ]), t = (y = s == null ? void 0 : s.children) == null ? void 0 : y.findIndex(
      (a) => {
        var g;
        return ((g = w.findNodeAtLocation(a, ["name"])) == null ? void 0 : g.value) === u;
      }
    );
    if (t !== void 0 && t >= 0) {
      const a = w.modify(
        i,
        ["configurations", t],
        void 0,
        {
          formattingOptions: {
            insertSpaces: !0,
            tabSize: 4,
            eol: `
`
          }
        }
      ), g = D(i, a);
      g === `{
    "configurations": []
}` ? p.unlinkSync(c) : p.writeFileSync(c, g);
    }
  }
}
function Q({
  cwd: u,
  mounts: e,
  pathSkippings: f
}) {
  return { pathMappings: u && e ? M(u, e) : void 0, pathSkippings: f };
}
function D(u, e) {
  const f = [], m = w.applyEdits(u, e);
  if (f.length = 0, w.parseTree(m, f, $), f.length) {
    const c = f.map((o) => ({
      message: w.printParseErrorCode(o.error),
      offset: o.offset,
      length: o.length,
      fragment: m.slice(
        Math.max(0, o.offset - 20),
        Math.min(m.length, o.offset + o.length + 10)
      )
    })).map(
      (o) => `${o.message} at ${o.offset}:${o.length} (${o.fragment})`
    ), h = e.map(
      (o) => `At ${o.offset}:${o.length} - (${o.content})`
    );
    throw new Error(
      `VS Code configuration file (.vscode/launch.json) is not valid a JSONC after CLI modifications. This is likely a CLI bug. Please report it at https://github.com/WordPress/wordpress-playground/issues and include the contents of your ".vscode/launch.json" file. 

 Applied edits: ${h.join(
        `
`
      )}

 The errors are: ${c.join(`
`)}`
    );
  }
  return m;
}
export {
  W as CLIOutput,
  A as DEFAULT_IDE_KEY,
  J as DEFAULT_PATH_SKIPPINGS,
  z as addXdebugIDEConfig,
  K as clearXdebugIDEConfig,
  U as createTempDirSymlink,
  Q as makeXdebugConfig,
  q as removeTempDirSymlink,
  O as updatePhpStormPHPConfig,
  N as updatePhpStormWorkspaceConfig,
  X as updateVSCodeConfig
};
