// @php-wasm/node-5-2/index.js
// packages/php-wasm/node-builds/5-2/src/index.ts
import { jspi } from "wasm-feature-detect";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { existsSync } from "node:fs";
var currentDirPath = typeof __dirname !== "undefined" ? __dirname : dirname(fileURLToPath(import.meta.url));
var packageRoot = existsSync(join(currentDirPath, "jspi")) ? currentDirPath : dirname(currentDirPath);
async function getPHPLoaderModule() {
  if (await jspi()) {
    return await import("./jspi/php_5_2.js");
  } else {
    return await import("./asyncify/php_5_2.js");
  }
}
async function getIntlExtensionPath() {
  throw new Error("The intl extension is not available for PHP 5.2.");
}
async function getXdebugExtensionPath() {
  throw new Error("The Xdebug extension is not available for PHP 5.2.");
}
async function getRedisExtensionPath() {
  throw new Error("The Redis extension is not available for PHP 5.2.");
}
async function getMemcachedExtensionPath() {
  throw new Error("The Memcached extension is not available for PHP 5.2.");
}
export {
  getIntlExtensionPath,
  getMemcachedExtensionPath,
  getPHPLoaderModule,
  getRedisExtensionPath,
  getXdebugExtensionPath,
  jspi
};
