// @php-wasm/web-service-worker/index.js
import { isURLScoped as R, setURLScope as S, getURLScope as m } from "@php-wasm/scopes";
import { portToStream as C } from "@php-wasm/universal";
const L = 25e3;
let U = 0;
function M(e, t, ...r) {
  const n = E();
  return e.postMessage(
    {
      ...t,
      requestId: n
    },
    ...r
  ), n;
}
function E() {
  return ++U;
}
function H(e, t, r = L) {
  return new Promise((n, s) => {
    const o = (d) => {
      d.data.type === "response" && d.data.requestId === t && (e.removeEventListener("message", o), clearTimeout(a), n(d.data.response));
    }, a = setTimeout(() => {
      s(new Error("Request timed out")), e.removeEventListener("message", o);
    }, r);
    e.addEventListener("message", o);
  });
}
function N(e, t) {
  return {
    type: "response",
    requestId: e,
    response: t
  };
}
async function $(e) {
  let t = new URL(e.request.url);
  if (!R(t))
    try {
      const c = new URL(e.request.referrer);
      t = S(t, m(c));
    } catch {
    }
  const r = e.request.headers.get("content-type"), n = e.request.method === "POST" ? new Uint8Array(await e.request.clone().arrayBuffer()) : void 0, s = {};
  for (const c of e.request.headers.entries())
    s[c[0]] = c[1];
  let o;
  try {
    const c = {
      method: "request",
      args: [
        {
          body: n,
          url: t.toString(),
          method: e.request.method,
          headers: {
            ...s,
            Host: t.host,
            // Safari and Firefox don't make the User-Agent header
            // available in the fetch event. Let's add it manually:
            "User-agent": self.navigator.userAgent,
            "Content-type": r
          }
        }
      ]
    }, i = m(t);
    if (i === null)
      throw new Error(
        `The URL ${t.toString()} is not scoped. This should not happen.`
      );
    const p = await P(c, i);
    if (o = await H(self, p), delete o.headers["x-frame-options"], o.headers["content-security-policy"]) {
      const f = o.headers["content-security-policy"].map(
        (l) => x(
          "frame-ancestors",
          l
        )
      ).filter((l) => l.trim().length > 0);
      f.length > 0 ? o.headers["content-security-policy"] = f : delete o.headers["content-security-policy"];
    }
  } catch (c) {
    throw console.error(c, { url: t.toString() }), c;
  }
  if (o.httpStatusCode >= 300 && o.httpStatusCode <= 399 && o.headers.location) {
    const c = m(t);
    let i = new URL(
      o.headers.location[0],
      t.toString()
    );
    return c && !R(i) && (i = S(i, c)), Response.redirect(
      i.toString(),
      o.httpStatusCode
    );
  }
  const a = [101, 103, 204, 205, 304].includes(
    o.httpStatusCode
  );
  let d = null;
  return a || (o.bodyPort ? d = C(o.bodyPort) : d = o.bytes), new Response(d, {
    headers: o.headers,
    status: o.httpStatusCode
  });
}
async function P(e, t) {
  const r = E();
  for (const n of await self.clients.matchAll({
    // Sometimes the client that triggered the current fetch()
    // event is considered uncontrolled in Google Chrome. This
    // only happens on the first few fetches() after the initial
    // registration of the service worker.
    includeUncontrolled: !0
  }))
    n.postMessage({
      ...e,
      /**
       * Attach the scope with a URL starting with `/scope:` to this message.
       *
       * We need this mechanics because this worker broadcasts
       * events to all the listeners across all browser tabs. Scopes
       * helps WASM workers ignore requests meant for other WASM workers.
       */
      scope: t,
      requestId: r
    });
  return r;
}
async function w(e, t) {
  let r;
  return ["GET", "HEAD"].includes(e.method) ? r = void 0 : "body" in t ? r = t.body : !e.bodyUsed && e.body ? r = e.body : r = await e.arrayBuffer(), new Request(t.url || e.url, {
    body: r,
    method: e.method,
    headers: e.headers,
    referrer: e.referrer,
    referrerPolicy: e.referrerPolicy,
    mode: e.mode === "navigate" ? "same-origin" : e.mode,
    credentials: e.credentials,
    cache: e.cache,
    redirect: e.redirect,
    integrity: e.integrity,
    /**
     * Infer the duplex value in a way that's consistent across browsers. Web browsers
     * only support 'half' as of January 2026, but other values may be supported in the future.
     * Unfortunately, also as of January 2026, we cannot read the duplex value directly from the
     * request object:
     *
     * > Although duplex can be passed as an option when constructing a Request,
     * > it is not currently exposed as a readable property on the resulting Request
     * > object in all browsers.
     *
     * See MDN: https://developer.mozilla.org/en-US/docs/Web/API/Request/duplex
     */
    ...r instanceof ReadableStream && { duplex: "half" },
    ...t
  });
}
let h;
const j = {
  resetStreamBodySupported() {
    h = void 0;
  }
};
async function A() {
  if (h !== void 0)
    return h;
  try {
    const e = new ReadableStream({
      start(t) {
        t.close();
      }
    });
    await fetch("data:,", {
      method: "POST",
      body: e,
      duplex: "half"
    }), h = !0;
  } catch {
    h = !1;
  }
  return h;
}
function k(e) {
  const t = {};
  return e.headers.forEach((r, n) => {
    t[n] = r;
  }), t;
}
function x(e, t) {
  const r = /^[\u{9}\u{A}\u{C}\u{D}\u{20}]+/u, n = /[\u{9}\u{A}\u{C}\u{D}\u{20}]+$/u, s = /[\u{9}\u{A}\u{C}\u{D}\u{20}]/u;
  return t.split(";").filter((o) => {
    const a = o.replace(r, "").replace(n, ""), [d] = a.split(
      s,
      // The directive name is the first token.
      1
    );
    return d.toLowerCase() !== e.toLowerCase();
  }).join(";");
}
class O extends Error {
  constructor(t, r, n) {
    super(
      `Could not fetch ${t} – your network appears to be blocking this request (HTTP ${r}). This often happens on school, university, or corporate networks. Try switching to a different network or using a VPN.`
    ), this.name = "FirewallInterferenceError", this.url = t, this.status = r, this.statusText = n;
  }
}
const q = "X-Playground-Cors-Proxy", b = /* @__PURE__ */ new Map([
  [
    "api.anthropic.com",
    {
      "anthropic-dangerous-direct-browser-access": "true"
    }
  ],
  ["api.openai.com", {}],
  ["generativelanguage.googleapis.com", {}]
]);
async function W(e, t, r, n) {
  var c;
  let s = typeof e == "string" ? new Request(e, t) : e;
  const o = n ? new URL(n) : null;
  let a = o ? new URL(s.url, o) : new URL(s.url);
  if (D(a))
    return await fetch(s);
  if (B(a))
    return s = await I(
      s,
      a
    ), await fetch(s);
  if (a.protocol === "http:") {
    a.protocol = "https:";
    const i = a.toString();
    s = await w(s, { url: i }), a = new URL(i);
  }
  if (!r)
    return await fetch(s);
  if (o && a.protocol === o.protocol && a.hostname === o.hostname && a.port === o.port && a.pathname.startsWith(o.pathname))
    return await fetch(s);
  const d = s.clone();
  try {
    return await fetch(s);
  } catch {
    const i = new Headers(s.headers), p = ((c = i.get("x-cors-proxy-allowed-request-headers")) == null ? void 0 : c.split(",")) || [], f = p.includes("authorization") || p.includes("cookie"), l = i.get("content-type");
    l && l.toLowerCase().includes("multipart/form-data") && (i.set("x-cors-proxy-content-type", l), i.set("content-type", "application/octet-stream"));
    let u = null;
    const g = s.method.toUpperCase();
    g !== "GET" && g !== "HEAD" && (await A() ? u = d.body : u = await d.arrayBuffer()), u instanceof ReadableStream && new URL(r, import.meta.url).protocol === "http:" && (u = await new Response(u).arrayBuffer());
    const T = await w(s, {
      url: `${r}${s.url}`,
      headers: i,
      body: u,
      ...f && { credentials: "include" }
    }), y = await fetch(T);
    if (!y.headers.has(q))
      throw new O(
        s.url,
        y.status,
        y.statusText
      );
    return y;
  }
}
function D(e) {
  return e.hostname === "localhost" || e.hostname === "127.0.0.1" || e.hostname === "[::1]" || e.hostname === "::1";
}
function B(e) {
  return e.protocol === "https:" && b.has(e.hostname);
}
async function I(e, t) {
  const r = b.get(t.hostname);
  if (!r)
    return e;
  const n = new Headers(e.headers);
  for (const [s, o] of Object.entries(r))
    n.has(s) || n.set(s, o);
  return await w(e, {
    headers: n
  });
}
export {
  O as FirewallInterferenceError,
  j as __testing,
  H as awaitReply,
  P as broadcastMessageExpectReply,
  w as cloneRequest,
  $ as convertFetchEventToPHPRequest,
  W as fetchWithCorsProxy,
  E as getNextRequestId,
  k as getRequestHeaders,
  M as postMessageExpectReply,
  x as removeContentSecurityPolicyDirective,
  N as responseTo,
  A as supportsReadableStreamBody
};
