// @wp-playground/wordpress/index.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const f=require("@php-wasm/universal"),o=require("@php-wasm/util"),m=require("@wp-playground/common"),$=require("@php-wasm/logger");async function x(e){await e.writeFile("/internal/shared/mu-plugins/0-playground.php",`<?php

		// Save WordPress environment information to a file.
		// Named function (not a closure) so this file parses on PHP 5.2.
		function playground_save_wp_env_info() {
			if (defined('DB_ENGINE') && DB_ENGINE === 'sqlite') {
				$db_info = array(
					'type' => 'sqlite',
					'path' => FQDB,
					'driver_path' => defined('WP_MYSQL_ON_SQLITE_LOADER_PATH')
						? WP_MYSQL_ON_SQLITE_LOADER_PATH
						: dirname(SQLITE_MAIN_FILE) . '/wp-pdo-mysql-on-sqlite.php',
				);
			} else {
				$db_info = array(
					'type' => 'mysql',
					// TODO: Save MySQL connection config.
				);
			}
			$wp_env = array('db' => $db_info);
			$wp_env_php = sprintf('<?php return %s;', var_export($wp_env, true));
			$wp_env_file = '/internal/shared/wp-env.php';
			if (!file_exists($wp_env_file) || file_get_contents($wp_env_file) !== $wp_env_php ) {
				file_put_contents($wp_env_file, $wp_env_php);
			}
		}
		add_action('wp_loaded', 'playground_save_wp_env_info');

        // Needed because gethostbyname( 'wordpress.org' ) returns
        // a private network IP address for some reason.
        function playground_allowed_redirect_hosts( $deprecated = '' ) {
            return array(
                'wordpress.org',
                'api.wordpress.org',
                'downloads.wordpress.org',
            );
        }
        add_filter( 'allowed_redirect_hosts', 'playground_allowed_redirect_hosts' );

		/**
		 * Prevents wp_http_validate_url() from universally failing.
		 *
		 * wp_http_validate_url() calls gethostbyname() to verify whether the host
		 * is external. If it is internal, the URL validation fails and WordPress
		 * refuses to make a request.
		 *
		 * However, in EMscripten, gethostbyname() returns a private network IP address.
		 * This causes wp_http_validate_url() to return false for all URLs.
		 *
		 * This filter ensures that all URLs are considered external. In production
		 * environments, this would be considered a security risk. However, Playground
		 * already provides multiple code execution vectors as features (e.g. Blueprints).
		 *
		 * If someone wants to poke around local IP addresses, they already have multiple
		 * tools at their disposal. Therefore, this is not a real security risk in context
		 * of WordPress Playground or Playground CLI.
		 */
		add_filter('http_request_host_is_external', '__return_true');

		// Support pretty permalinks
        add_filter( 'got_url_rewrite', '__return_true' );

		/**
		 * Flush rewrite rules on the first real WordPress request.
		 *
		 * During boot, we set permalink_structure in the database
		 * but can't flush rewrite rules at that point because WordPress
		 * isn't fully bootstrapped — post types and taxonomies haven't
		 * been registered yet, so the generated rules are incomplete.
		 *
		 * This hook fires on 'init' at a very late priority, after all
		 * post types and taxonomies are registered. It checks if the
		 * rewrite_rules option is empty (meaning rules were never
		 * flushed) and if permalink_structure is set, then flushes once.
		 * A flag file prevents repeated flushes on subsequent requests.
		 */
		function playground_maybe_flush_rewrite_rules() {
			$flag = '/internal/shared/.rewrite-rules-flushed';
			if (file_exists($flag)) {
				return;
			}
			if (!function_exists('get_option')) {
				return;
			}
			$structure = get_option('permalink_structure');
			if (empty($structure)) {
				return;
			}
			$rules = get_option('rewrite_rules');
			if (!empty($rules)) {
				@file_put_contents($flag, '1');
				return;
			}
			global $wp_rewrite;
			if (!isset($wp_rewrite) && class_exists('WP_Rewrite')) {
				$wp_rewrite = new WP_Rewrite();
			}
			if (isset($wp_rewrite) && method_exists($wp_rewrite, 'flush_rules')) {
				$wp_rewrite->flush_rules();
			}
			@file_put_contents($flag, '1');
		}
		add_action('init', 'playground_maybe_flush_rewrite_rules', 99999);

        // Create the fonts directory if missing
        if(!file_exists(WP_CONTENT_DIR . '/fonts')) {
            mkdir(WP_CONTENT_DIR . '/fonts');
        }

        $log_file = WP_CONTENT_DIR . '/debug.log';
        if ( defined( 'WP_DEBUG_LOG' ) && WP_DEBUG_LOG ) {
            if ( is_string( WP_DEBUG_LOG ) ) {
                $log_file = WP_DEBUG_LOG;
            }
            ini_set('error_log', $log_file);
        } else {
            ini_set('log_errors', '0');
        }
        define('ERROR_LOG_FILE', $log_file);
        ?>`),await e.writeFile("/internal/shared/mu-plugins/sitemap-redirect.php",`<?php
		/**
		 * Redirect sitemap.xml to wp-sitemap.xml for non-root installations.
		 *
		 * WordPress seems to only generate the sitemap.xml → wp-sitemap.xml rewrite
		 * rule when installed at the domain root. This mu-plugin handles the
		 * redirect for non-root installations.
		 */
		if (isset($_SERVER['REQUEST_URI'])) {
			$site_url = site_url();
			$parsed = parse_url($site_url);
			$base_path = isset($parsed['path']) ? rtrim($parsed['path'], '/') : '';

			$request_uri = $_SERVER['REQUEST_URI'];
			if (
				$request_uri === $base_path . '/sitemap.xml' ||
				strpos($request_uri, $base_path . '/sitemap.xml?') === 0 ||
				strpos($request_uri, $base_path . '/sitemap.xml/') === 0
			) {
				$query_string = strpos($request_uri, '?') !== false ? substr($request_uri, strpos($request_uri, '?')) : '';
				header('Location: ' . $base_path . '/wp-sitemap.xml' . $query_string, true, 301);
				exit;
			}
		}
		`),await e.writeFile("/internal/shared/mu-plugins/inline-tinymce-content-css.php",`<?php
		function playground_inline_tinymce_content_css($settings) {
			if (empty($settings['content_css'])) return $settings;
			$css_urls = explode(',', $settings['content_css']);
			$inline_css = '';
			$doc_root = isset($_SERVER['DOCUMENT_ROOT'])
				? $_SERVER['DOCUMENT_ROOT'] : '/wordpress';
			foreach ($css_urls as $url) {
				$url = trim($url);
				if (!$url) continue;
				$parsed = parse_url($url);
				if (!isset($parsed['path'])) continue;
				$path = preg_replace('#^/scope:[^/]+#', '', $parsed['path']);
				$file = $doc_root . $path;
				if (file_exists($file)) {
					$inline_css .= @file_get_contents($file) . "\\n";
				}
			}
			if ($inline_css !== '') {
				if (!empty($settings['content_style'])) {
					$inline_css = $settings['content_style'] . "\\n" . $inline_css;
				}
				$settings['content_style'] =
					playground_escape_tinymce_content_style_for_wp_editor_serializer($inline_css);
				$settings['content_css'] = '';
			}
			return $settings;
		}
		add_filter('tiny_mce_before_init', 'playground_inline_tinymce_content_css');

		// _WP_Editors::_parse_init() wraps strings in double quotes without escaping them.
		// Encode the CSS as a JavaScript string body before WordPress adds those quotes.
		function playground_escape_tinymce_content_style_for_wp_editor_serializer($css) {
			$encoded = json_encode($css);
			if (
				is_string($encoded) &&
				substr($encoded, 0, 1) === chr(34) &&
				substr($encoded, -1) === chr(34)
			) {
				$escaped = substr($encoded, 1, -1);
			} else {
				// Fallback: json_encode() can fail on non-UTF-8 CSS bytes.
				// Escape the bytes that would break _WP_Editors::_parse_init().
				$escaped = str_replace(
					array('\\\\', chr(34), "\\r", "\\n", "\\xe2\\x80\\xa8", "\\xe2\\x80\\xa9"),
					array('\\\\\\\\', '\\\\' . chr(34), '\\\\r', '\\\\n', '\\\\u2028', '\\\\u2029'),
					$css
				);
			}
			return str_replace('</', '<\\/', $escaped);
		}
		`)}function v(e){return`
/**
 * Loads the SQLite integration plugin before WordPress is loaded
 * and without creating a drop-in "db.php" file.
 *
 * Technically, it creates a global $wpdb object whose only two
 * purposes are to:
 *
 * * Exist – because the require_wp_db() WordPress function won't
 *           connect to MySQL if $wpdb is already set.
 * * Load the SQLite integration plugin the first time it's used
 *   and replace the global $wpdb reference with the SQLite one.
 *
 * This lets Playground keep the WordPress installation clean and
 * solves dillemas like:
 *
 * * Should we include db.php in Playground exports?
 * * Should we remove db.php from Playground imports?
 * * How should we treat stale db.php from long-lived OPFS sites?
 *
 * @see https://github.com/WordPress/wordpress-playground/discussions/1379 for
 *      more context.
 */
class Playground_SQLite_Integration_Loader {
	public function __call($name, $arguments) {
		$this->load_sqlite_integration();
		if($GLOBALS['wpdb'] === $this) {
			throw new Exception('Infinite loop detected in $wpdb – SQLite integration plugin could not be loaded');
		}
		return call_user_func_array(
			array($GLOBALS['wpdb'], $name),
			$arguments
		);
	}
	public function __get($name) {
		$this->load_sqlite_integration();
		if($GLOBALS['wpdb'] === $this) {
			throw new Exception('Infinite loop detected in $wpdb – SQLite integration plugin could not be loaded');
		}
		return $GLOBALS['wpdb']->$name;
	}
	public function __set($name, $value) {
		$this->load_sqlite_integration();
		if($GLOBALS['wpdb'] === $this) {
			throw new Exception('Infinite loop detected in $wpdb – SQLite integration plugin could not be loaded');
		}
		$GLOBALS['wpdb']->$name = $value;
	}
    protected function load_sqlite_integration() {
        ${e}
    }
}
/**
 * The Query Monitor plugin short-circuits in the CLI SAPI. However, in Playground,
 * the SAPI is always "cli" at the moment. Let's set a constant to disable the CLI
 * detection.
 *
 * @see https://github.com/WordPress/sqlite-database-integration/pull/212
 * @see https://github.com/WordPress/sqlite-database-integration/pull/215
 */
define('QM_TESTS', true);
$wpdb = $GLOBALS['wpdb'] = new Playground_SQLite_Integration_Loader();

/**
 * WordPress is capable of using a preloaded global $wpdb. However, if
 * it cannot find the drop-in db.php plugin it still checks whether
 * the mysqli_connect() function exists even though it's not used.
 *
 * What WordPress demands, Playground shall provide.
 */
`}const ne=`
			if (function_exists('is_user_logged_in') && is_user_logged_in()) {
				return;
			}
			if (headers_sent()) {
				return;
			}

			// Legacy auto-login never redirects; it populates $_COOKIE
			// in-process for the current request and relies on
			// setcookie() / wp_set_auth_cookie() to persist across
			// requests via HttpCookieStore.

			// WP 2.5+
			if (function_exists('wp_set_current_user') && function_exists('wp_set_auth_cookie')) {
				$user = function_exists('get_user_by')
					? get_user_by('login', $user_name)
					: (function_exists('get_userdatabylogin')
						? get_userdatabylogin($user_name) : null);
				if (!$user) return;

				wp_set_current_user($user->ID, $user->user_login);
				// Populate $_COOKIE in-process so auth_redirect() and
				// wp_verify_nonce() see the session for the remainder
				// of this request; wp_set_auth_cookie() also emits
				// Set-Cookie for subsequent requests.
				wp_set_auth_cookie($user->ID);
				if (function_exists('wp_generate_auth_cookie')) {
					$_pg_exp = time() + 172800;
					if (defined('AUTH_COOKIE'))
						$_COOKIE[AUTH_COOKIE] = wp_generate_auth_cookie($user->ID, $_pg_exp, 'auth');
					if (defined('SECURE_AUTH_COOKIE'))
						$_COOKIE[SECURE_AUTH_COOKIE] = wp_generate_auth_cookie($user->ID, $_pg_exp, 'secure_auth');
					if (defined('LOGGED_IN_COOKIE'))
						$_COOKIE[LOGGED_IN_COOKIE] = wp_generate_auth_cookie($user->ID, $_pg_exp, 'logged_in');
				}
				return;
			}

			// WP 1.5–2.4
			if (defined('USER_COOKIE') && defined('PASS_COOKIE')) {
				$_pg_pass_cookie = md5(md5('password'));
				$_COOKIE[USER_COOKIE] = $user_name;
				$_COOKIE[PASS_COOKIE] = $_pg_pass_cookie;
				if (!headers_sent()) {
					$_pg_exp = time() + 172800;
					setcookie(USER_COOKIE, $user_name, $_pg_exp, '/');
					setcookie(PASS_COOKIE, $_pg_pass_cookie, $_pg_exp, '/');
				}
				$GLOBALS['current_user'] = null;
				if (function_exists('get_currentuserinfo')) {
					get_currentuserinfo();
				}
				return;
			}

			// WP 1.0–1.2: cookies are usually already set by
			// playground_legacy_set_auth_cookies_early() in env.php,
			// but WP 1.0–1.2 reads its user state from globals (no
			// WP_User), so populate those explicitly here.
			$cookiehash = defined('COOKIEHASH')
				? COOKIEHASH
				: (isset($GLOBALS['cookiehash']) && $GLOBALS['cookiehash']
					? $GLOBALS['cookiehash']
					: (function_exists('get_settings')
						? md5(get_settings('siteurl'))
						: ''));
			if ($cookiehash) {
				$_pg_user_cookie_name = 'wordpressuser_' . $cookiehash;
				$_pg_pass_cookie_name = 'wordpresspass_' . $cookiehash;
				$_pg_pass_cookie_value = md5(md5('password'));
				$_COOKIE[$_pg_user_cookie_name] = $user_name;
				$_COOKIE[$_pg_pass_cookie_name] = $_pg_pass_cookie_value;
				if (!headers_sent()) {
					$_pg_exp = time() + 172800;
					setcookie($_pg_user_cookie_name, $user_name, $_pg_exp, '/');
					setcookie($_pg_pass_cookie_name, $_pg_pass_cookie_value, $_pg_exp, '/');
				}
				if (function_exists('get_userdatabylogin')) {
					$userdata = get_userdatabylogin($user_name);
					if ($userdata) {
						$GLOBALS['user_login']    = $user_name;
						$GLOBALS['userdata']      = $userdata;
						$GLOBALS['user_level']    = isset($userdata->user_level) ? (int) $userdata->user_level : 10;
						$GLOBALS['user_ID']       = $userdata->ID;
						$GLOBALS['user_email']    = isset($userdata->user_email) ? $userdata->user_email : '';
						$GLOBALS['user_url']      = isset($userdata->user_url) ? $userdata->user_url : '';
						$GLOBALS['user_nickname'] = isset($userdata->user_nickname) ? $userdata->user_nickname : $user_name;
						$GLOBALS['user_pass_md5'] = md5(isset($userdata->user_pass) ? $userdata->user_pass : '');
					}
				}
				return;
			}
`;async function ie(e){await e.mkdir("/internal/shared/mu-plugins"),await e.writeFile("/internal/shared/auto_prepend_file.php",`<?php
// Polyfill the PHP 4 superglobals WP 1.0–2.5 still reads (removed
// in PHP 5.4). Bind by reference so later writes to $_COOKIE
// reach $HTTP_COOKIE_VARS, which WP 1.0's get_currentuserinfo()
// consults.
$GLOBALS['HTTP_GET_VARS']     = &$_GET;
$GLOBALS['HTTP_POST_VARS']    = &$_POST;
$GLOBALS['HTTP_COOKIE_VARS']  = &$_COOKIE;
$GLOBALS['HTTP_SERVER_VARS']  = &$_SERVER;
if (isset($_FILES))   $GLOBALS['HTTP_POST_FILES']   = &$_FILES;
if (isset($_ENV))     $GLOBALS['HTTP_ENV_VARS']     = &$_ENV;
if (isset($_SESSION)) $GLOBALS['HTTP_SESSION_VARS'] = &$_SESSION;
// Top-level names register_globals=On used to expose. WP 1.0
// reads $PHP_SELF / $REMOTE_ADDR directly instead of $_SERVER.
if (isset($_SERVER['PHP_SELF'])) $GLOBALS['PHP_SELF'] = $_SERVER['PHP_SELF'];
if (isset($_SERVER['REMOTE_ADDR'])) $GLOBALS['REMOTE_ADDR'] = $_SERVER['REMOTE_ADDR'];
if (isset($_SERVER['REQUEST_URI'])) $GLOBALS['REQUEST_URI'] = $_SERVER['REQUEST_URI'];
// Default SERVER_PROTOCOL for scripts invoked outside an HTTP
// request (e.g. php.run() during boot/fixups) — legacy WP reads
// it unconditionally in places like wp_redirect().
if (!isset($_SERVER['SERVER_PROTOCOL'])) {
	$_SERVER['SERVER_PROTOCOL'] = 'HTTP/1.1';
}
if(file_exists('/internal/shared/consts.json')) {
	$consts = json_decode(file_get_contents('/internal/shared/consts.json'), true);
	if ($consts) {
		foreach ($consts as $const => $value) {
			if (!defined($const) && is_scalar($value)) {
				define($const, $value);
			}
		}
	}
}
foreach (glob('/internal/shared/preload/*.php') as $file) {
	require_once $file;
}
// Buffer early output so a stray PHP notice doesn't commit the
// response headers before the auto-login mu-plugin gets a chance
// to call wp_set_auth_cookie() / setcookie() on the init hook —
// otherwise nonce validation breaks on POST requests. PHP flushes
// the buffer at script end so output still reaches the browser.
ob_start();
`),await e.writeFile("/internal/shared/preload/env.php",`<?php
// Reads $wp_version from the WordPress install on disk. Falls back
// to '1.0' so callers can use version_compare() unconditionally.
function _playground_detect_wp_version() {
	static $version = null;
	if ($version !== null) return $version;
	$doc_root = isset($_SERVER['DOCUMENT_ROOT'])
		? $_SERVER['DOCUMENT_ROOT'] : '/wordpress';
	$version_path = $doc_root . '/wp-includes/version.php';
	$wp_version = '1.0';
	if (file_exists($version_path)) {
		include $version_path;
	}
	$version = $wp_version;
	return $version;
}

// Returns 'wp10', 'wp12', or 'wp15' based on the WP version on
// disk — the three $wp_filter shapes apply_filters() understands.
function _playground_detect_wp_hook_format() {
	static $format = null;
	if ($format !== null) return $format;
	$wp_version = _playground_detect_wp_version();
	if (version_compare($wp_version, '1.5', '>=')) {
		$format = 'wp15';
	} elseif (version_compare($wp_version, '1.2', '>=')) {
		$format = 'wp12';
	} else {
		$format = 'wp10';
	}
	return $format;
}

// Adds filters/actions before WordPress is loaded by writing the
// $wp_filter shape the target version expects. $function_to_add
// MUST be a string (no closures).
function playground_add_filter( $tag, $function_to_add, $priority = 10, $accepted_args = 1 ) {
	global $wp_filter;
	$fmt = _playground_detect_wp_hook_format();
	if ($fmt === 'wp10') {
		$wp_filter[$tag][] = $function_to_add;
	} elseif ($fmt === 'wp12') {
		$wp_filter[$tag][$priority][] = $function_to_add;
	} else {
		$wp_filter[$tag][$priority][$function_to_add] = array(
			'function' => $function_to_add,
			'accepted_args' => $accepted_args
		);
	}
}
function playground_add_action( $tag, $function_to_add, $priority = 10, $accepted_args = 1 ) {
	playground_add_filter( $tag, $function_to_add, $priority, $accepted_args );
}

// Set WP 1.0–2.4 auth cookies before WordPress loads — by the time
// the init hook fires (and on WP 1.0–1.2 it may not fire at all on
// the front page) WordPress has already read $_COOKIE. setcookie()
// also persists them across requests via HttpCookieStore.
// WP 2.5+ uses the HMAC auth cookie scheme and doesn't read these
// wordpressuser_/wordpresspass_ cookies at all — bail there so we
// don't write inert cookies the runtime would have to clean up.
function playground_legacy_set_auth_cookies_early() {
	if (!defined('PLAYGROUND_AUTO_LOGIN_AS_USER')) return;
	if (isset($_COOKIE['playground_auto_login_already_logged_out'])) return;
	if (version_compare(_playground_detect_wp_version(), '2.5', '>=')) return;

	foreach ($_COOKIE as $name => $_) {
		if (strncmp($name, 'wordpressuser_', 14) === 0) return;
	}

	$user_name = PLAYGROUND_AUTO_LOGIN_AS_USER;
	$pass_md5 = md5(md5('password'));

	// Read siteurl from SQLite so the cookie hash matches what
	// WP 1.0–2.4 derives from get_settings('siteurl').
	$siteurl = null;
	$db_path = defined('DB_DIR') ? DB_DIR . '.ht.sqlite' : '';
	if ($db_path && class_exists('PDO') && file_exists($db_path)) {
		try {
			$pdo = new PDO('sqlite:' . $db_path);
			$stmt = $pdo->query("SELECT option_value FROM wp_options WHERE option_name = 'siteurl' LIMIT 1");
			if ($stmt) $siteurl = $stmt->fetchColumn();
			$pdo = null;
		} catch (Exception $e) {}
	}
	if (!$siteurl && defined('WP_SITEURL')) $siteurl = WP_SITEURL;
	if (!$siteurl) return;

	$cookiehash = md5($siteurl);
	$user_cookie_name = 'wordpressuser_' . $cookiehash;
	$pass_cookie_name = 'wordpresspass_' . $cookiehash;
	$_COOKIE[$user_cookie_name] = $user_name;
	$_COOKIE[$pass_cookie_name] = $pass_md5;

	if (!headers_sent()) {
		$exp = time() + 172800;
		setcookie($user_cookie_name, $user_name, $exp, '/');
		setcookie($pass_cookie_name, $pass_md5, $exp, '/');
	}
}
playground_legacy_set_auth_cookies_early();

// WP < 4.0 emits YEAR(post_date)='2026' AND MONTH(post_date)='4'
// against MySQL's loose type coercion. The SQLite driver's UDFs
// return integers and SQLite is strictly typed (4 != '4'), so
// strip quotes around numeric RHS values to keep both sides ints.
function playground_fix_sqlite_date_comparisons($query) {
	if (
		stripos($query, 'YEAR') === false &&
		stripos($query, 'MONTH') === false &&
		stripos($query, 'DAY') === false
	) {
		return $query;
	}
	return preg_replace(
		'/\\b(YEAR|MONTH|DAYOFMONTH|DAY)\\s*\\(([^)]+)\\)\\s*=\\s*\\'(\\d+)\\'/i',
		'$1($2) = $3',
		$query
	);
}
playground_add_filter( 'query', 'playground_fix_sqlite_date_comparisons' );

// WP 2.2+ checks WP_SITEURL/WP_HOME inside get_option(); WP <2.2
// doesn't, so backfill the same behaviour via the option filters
// to keep admin links on the Playground-scoped URL.
function playground_override_siteurl($value) {
	if (defined('WP_SITEURL')) {
		return WP_SITEURL;
	}
	return $value;
}
function playground_override_home($value) {
	if (defined('WP_HOME')) {
		return WP_HOME;
	}
	return $value;
}
playground_add_filter( 'option_siteurl', 'playground_override_siteurl' );
playground_add_filter( 'option_home', 'playground_override_home' );

// Load mu-plugins last so customer mu-plugins win — and so they
// can't depend on muplugins_loaded. WP < 2.8 doesn't fire that
// action at all, so init -1000 acts as a fallback (the $loaded
// flag keeps it idempotent).
playground_add_action( 'muplugins_loaded', 'playground_load_mu_plugins', 0 );
playground_add_action( 'init', 'playground_load_mu_plugins', -1000 );
function playground_load_mu_plugins() {
	static $loaded = false;
	if ($loaded) return;
	$loaded = true;
	$mu_plugins_dir = '/internal/shared/mu-plugins';
	if(!is_dir($mu_plugins_dir)){
		return;
	}
	$mu_plugins = glob( $mu_plugins_dir . '/*.php' );
	sort( $mu_plugins );
	global $wp_version;
	$is_legacy_wp = isset($wp_version) && version_compare($wp_version, '2.8', '<');
	foreach ( $mu_plugins as $mu_plugin ) {
		// Loaded separately by the preload lazy loader or db.php.
		if (strpos($mu_plugin, 'sqlite-database-integration') !== false) {
			continue;
		}
		// WP < 2.8 crashes on closures in hooks and lacks
		// site_url() (added 2.6). 1-auto-login.php is written
		// without either, so it's the only mu-plugin we load
		// on legacy WP.
		if ($is_legacy_wp) {
			if (strpos($mu_plugin, '1-auto-login.php') === false) {
				continue;
			}
		}
		require_once $mu_plugin;
	}

	// PHP 5.x's foreach over $wp_filter['init'] iterates a copy,
	// so add_action() calls made by the mu-plugin we just loaded
	// won't fire on this same init run. Call them directly.
	if ($is_legacy_wp) {
		if (function_exists('playground_auto_login_redirect_target')) {
			playground_auto_login_redirect_target();
		}
		if (function_exists('playground_auto_login')) {
			playground_auto_login();
		}
	}
}
`),await e.writeFile("/internal/shared/mu-plugins/1-auto-login.php",`<?php
		/**
		 * Returns the username to auto-login as, if any.
		 * @return string|false
		 */
		function playground_get_username_for_auto_login() {
			if ( defined('PLAYGROUND_AUTO_LOGIN_AS_USER') && !isset($_COOKIE['playground_auto_login_already_happened']) ) {
				return PLAYGROUND_AUTO_LOGIN_AS_USER;
			}
			if ( defined('PLAYGROUND_FORCE_AUTO_LOGIN_ENABLED') && isset($_GET['playground_force_auto_login_as_user']) ) {
				return $_GET['playground_force_auto_login_as_user'];
			}
			return false;
		}

		function playground_auto_login() {
			if (empty($_SERVER['REQUEST_URI'])) {
				return;
			}
			$user_name = playground_get_username_for_auto_login();
			if ( false === $user_name ) {
				return;
			}
			if ((function_exists('wp_doing_ajax') && wp_doing_ajax()) || defined('REST_REQUEST')) {
				return;
			}
			${ne}
		}
		add_action('init', 'playground_auto_login', 1);

		function playground_auto_login_redirect_target() {
			if(strpos($_SERVER['REQUEST_URI'], '?playground-redirection-handler') !== false) {
				$next = $_GET['next'];
				header('Location: ' . $next, true, 302);
				exit;
			}
		}
		add_action('init', 'playground_auto_login_redirect_target', 1);

		/**
		 * Disable the Site Admin Email Verification Screen for any session started
		 * via autologin.
		 */
		if (function_exists('add_filter')) {
			add_filter('admin_email_check_interval', 'playground_disable_admin_email_check');
		}
		function playground_disable_admin_email_check($interval) {
			if(false === playground_get_username_for_auto_login()) {
				return 0;
			}
			return $interval;
		}
		`),await x(e),await e.writeFile("/internal/shared/preload/error-handler.php",`<?php
$GLOBALS['_playground_consts'] = array();
if (file_exists('/internal/shared/consts.json')) {
	$GLOBALS['_playground_consts'] = @json_decode(file_get_contents('/internal/shared/consts.json'), true);
	if (!is_array($GLOBALS['_playground_consts'])) { $GLOBALS['_playground_consts'] = array(); }
	$GLOBALS['_playground_consts'] = array_keys($GLOBALS['_playground_consts']);
}
function _playground_error_handler($severity, $message, $file, $line) {
	$playground_consts = $GLOBALS['_playground_consts'];
${se}
	return false;
}
set_error_handler('_playground_error_handler');`)}const se=`
		// http_api_transports is deprecated since 6.4.0 but Playground's
		// networking layer still registers it for wp_http_supports().
		// @see https://core.trac.wordpress.org/ticket/37708
		if (
			strpos($message, "http_api_transports") !== false &&
			strpos($message, "since version 6.4.0 with no alternative available") !== false
		) {
			return;
		}
		// Playground predefines constants (SITE_URL, WP_DEBUG, …) that
		// wp-config.php is allowed to redefine; ours take precedence.
		if (strpos($message, "already defined") !== false) {
			foreach($playground_consts as $const) {
				if(strpos($message, "Constant $const already defined") !== false) {
					return;
				}
			}
		}
		// Legacy WP (2.0–3.5) assigns props on uninitialised vars,
		// valid in PHP 4 but E_WARNING since 5.x. Unfixable in core —
		// Playground ships unmodified WordPress releases.
		if (strpos($message, "Creating default object from empty value") !== false) {
			return;
		}
		// WP 2.8's dashboard widget calls get_error_string() on a
		// null SimplePie when feed HTTP requests fail in WASM.
		if (strpos($message, "get_error_string() on null") !== false ||
			strpos($message, "get_error_string() on a non-object") !== false) {
			return;
		}
		// Don't complain about WordPress.org connection errors when
		// the runtime isn't using fetch().
		if (
			(
				! defined('USE_FETCH_FOR_REQUESTS') ||
				! USE_FETCH_FOR_REQUESTS
			) &&
			strpos($message, "WordPress could not establish a secure connection to WordPress.org") !== false)
		{
			return;
		}
`;function I(e){return[/(^|[^\w$])((?:require|include)(?:_once)?)\s*\(\s*(['"])(\.\.\/[^'"]+)\3\s*\)/g,/(^|[^\w$])((?:require|include)(?:_once)?)\s*\(\s*(['"])(\.\/[^'"]+)\3\s*\)/g,/(^|[^\w$])((?:require|include)(?:_once)?)\s*\(\s*(['"])([a-z][\w-]*\.php)\3\s*\)/g,/(^|[^\w$])((?:require|include)(?:_once)?)\s+(['"])(\.\.\/[^'"]+)\3/g,/(^|[^\w$])((?:require|include)(?:_once)?)\s+(['"])(\.\/[^'"]+)\3/g,/(^|[^\w$])((?:require|include)(?:_once)?)\s+(['"])([a-z][\w-]*\.php)\3/g].reduce((i,n)=>i.replace(n,(s,r,a,l,_)=>`${r}${a}(${re(_)})`),e)}function re(e){let t=e,i=0;for(;t.startsWith("../");)i++,t=t.slice(3);for(;t.startsWith("./");)t=t.slice(2);let n="dirname(__FILE__)";for(let s=0;s<i;s++)n=`dirname(${n})`;return`${n} . '/${t}'`}async function ae(e,t){return k(e,t)?(await le(e,t),await _e(e,t),await ce(e,t),await ue(e,t),await de(e,t),await fe(e,t),await he(e,t),!0):!1}async function oe(e){return k(e,e.documentRoot)?(await e.run({code:`<?php
			$db_dir = getenv('DOCUMENT_ROOT') . '/wp-content/database/';
			if (!is_dir($db_dir)) { @mkdir($db_dir, 0777, true); }
			$db_path = $db_dir . '.ht.sqlite';
			$pdo = new PDO('sqlite:' . $db_path);
			$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
			$now = date('Y-m-d H:i:s');

			$tables_sql = array(
				'b2posts' => "CREATE TABLE IF NOT EXISTS b2posts (
					ID INTEGER PRIMARY KEY AUTOINCREMENT,
					post_author INTEGER NOT NULL DEFAULT 0,
					post_date TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
					post_content TEXT NOT NULL DEFAULT '',
					post_title TEXT NOT NULL DEFAULT '',
					post_category INTEGER NOT NULL DEFAULT 0,
					post_excerpt TEXT NOT NULL DEFAULT '',
					post_status TEXT NOT NULL DEFAULT 'publish',
					comment_status TEXT NOT NULL DEFAULT 'open',
					ping_status TEXT NOT NULL DEFAULT 'open',
					post_password TEXT NOT NULL DEFAULT ''
				)",
				'b2categories' => "CREATE TABLE IF NOT EXISTS b2categories (
					cat_ID INTEGER PRIMARY KEY AUTOINCREMENT,
					cat_name TEXT NOT NULL DEFAULT ''
				)",
				'b2comments' => "CREATE TABLE IF NOT EXISTS b2comments (
					comment_ID INTEGER PRIMARY KEY AUTOINCREMENT,
					comment_post_ID INTEGER NOT NULL DEFAULT 0,
					comment_author TEXT NOT NULL DEFAULT '',
					comment_author_email TEXT NOT NULL DEFAULT '',
					comment_author_url TEXT NOT NULL DEFAULT '',
					comment_author_IP TEXT NOT NULL DEFAULT '',
					comment_date TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
					comment_content TEXT NOT NULL DEFAULT '',
					comment_karma INTEGER NOT NULL DEFAULT 0
				)",
				'b2settings' => "CREATE TABLE IF NOT EXISTS b2settings (
					ID INTEGER NOT NULL DEFAULT 1,
					posts_per_page INTEGER NOT NULL DEFAULT 20,
					what_to_show TEXT NOT NULL DEFAULT 'posts',
					archive_mode TEXT NOT NULL DEFAULT 'postbypost',
					time_difference INTEGER NOT NULL DEFAULT 0,
					AutoBR INTEGER NOT NULL DEFAULT 1,
					time_format TEXT NOT NULL DEFAULT 'g:i a',
					date_format TEXT NOT NULL DEFAULT 'n/j/Y',
					PRIMARY KEY (ID)
				)",
				'b2users' => "CREATE TABLE IF NOT EXISTS b2users (
					ID INTEGER PRIMARY KEY AUTOINCREMENT,
					user_login TEXT NOT NULL DEFAULT '',
					user_pass TEXT NOT NULL DEFAULT '',
					user_firstname TEXT NOT NULL DEFAULT '',
					user_lastname TEXT NOT NULL DEFAULT '',
					user_nickname TEXT NOT NULL DEFAULT '',
					user_icq INTEGER NOT NULL DEFAULT 0,
					user_email TEXT NOT NULL DEFAULT '',
					user_url TEXT NOT NULL DEFAULT '',
					user_ip TEXT NOT NULL DEFAULT '',
					user_domain TEXT NOT NULL DEFAULT '',
					user_browser TEXT NOT NULL DEFAULT '',
					dateYMDhour TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
					user_level INTEGER NOT NULL DEFAULT 0,
					user_aim TEXT NOT NULL DEFAULT '',
					user_msn TEXT NOT NULL DEFAULT '',
					user_yim TEXT NOT NULL DEFAULT '',
					user_idmode TEXT NOT NULL DEFAULT ''
				)",
				'b2links' => "CREATE TABLE IF NOT EXISTS b2links (
					link_id INTEGER PRIMARY KEY AUTOINCREMENT,
					link_url TEXT NOT NULL DEFAULT '',
					link_name TEXT NOT NULL DEFAULT '',
					link_image TEXT NOT NULL DEFAULT '',
					link_target TEXT NOT NULL DEFAULT '',
					link_category INTEGER NOT NULL DEFAULT 0,
					link_description TEXT NOT NULL DEFAULT '',
					link_visible TEXT NOT NULL DEFAULT 'Y',
					link_owner INTEGER NOT NULL DEFAULT 1,
					link_rating INTEGER NOT NULL DEFAULT 0,
					link_updated TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
					link_rel TEXT NOT NULL DEFAULT '',
					link_notes TEXT NOT NULL DEFAULT '',
					link_rss TEXT NOT NULL DEFAULT ''
				)",
				'b2linkcategories' => "CREATE TABLE IF NOT EXISTS b2linkcategories (
					cat_id INTEGER PRIMARY KEY AUTOINCREMENT,
					cat_name TEXT NOT NULL DEFAULT '',
					auto_toggle TEXT NOT NULL DEFAULT 'N',
					show_images TEXT NOT NULL DEFAULT 'Y',
					show_description TEXT NOT NULL DEFAULT 'N',
					show_rating TEXT NOT NULL DEFAULT 'Y',
					show_updated TEXT NOT NULL DEFAULT 'Y',
					sort_order TEXT NOT NULL DEFAULT 'name',
					sort_desc TEXT NOT NULL DEFAULT 'ASC',
					text_before_link TEXT NOT NULL DEFAULT '<li>',
					text_after_link TEXT NOT NULL DEFAULT '<br />',
					text_after_all TEXT NOT NULL DEFAULT '</li>',
					list_limit INTEGER NOT NULL DEFAULT -1
				)"
			);
			foreach ($tables_sql as $sql) {
				$pdo->exec($sql);
			}

			if (!$pdo->query("SELECT COUNT(*) FROM b2categories")->fetchColumn()) {
				$pdo->exec("INSERT INTO b2categories (cat_ID, cat_name) VALUES (1, 'General')");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2settings")->fetchColumn()) {
				$pdo->exec("INSERT INTO b2settings (ID, posts_per_page, what_to_show, archive_mode, time_difference, AutoBR, time_format, date_format) VALUES (1, 20, 'posts', 'postbypost', 0, 1, 'g:i a', 'n/j/Y')");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2users")->fetchColumn()) {
				$pass = md5('password');
				$pdo->exec("INSERT INTO b2users (ID, user_login, user_pass, user_nickname, user_email, user_level, dateYMDhour, user_idmode) VALUES (1, 'admin', '{$pass}', 'admin', 'admin@localhost.com', 10, '{$now}', 'nickname')");
			} else {
				$pass = md5('password');
				$pdo->exec("UPDATE b2users SET user_pass = '{$pass}', user_level = 10 WHERE user_login = 'admin'");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2posts")->fetchColumn()) {
				$content = 'Welcome to WordPress. This is the first post. Edit or delete it, then start blogging!';
				$pdo->exec("INSERT INTO b2posts (ID, post_author, post_date, post_content, post_title, post_category, post_excerpt, post_status, comment_status, ping_status, post_password) VALUES (1, 1, '{$now}', '{$content}', 'Hello world!', 1, '', 'publish', 'open', 'open', '')");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2comments")->fetchColumn()) {
				$pdo->exec("INSERT INTO b2comments (comment_post_ID, comment_author, comment_author_email, comment_author_url, comment_author_IP, comment_date, comment_content, comment_karma) VALUES (1, 'Mr WordPress', 'mr@wordpress.org', 'http://wordpress.org', '127.0.0.1', '{$now}', 'Hi, this is a comment. To delete a comment, just log in and view the comments for this post.', 0)");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2linkcategories")->fetchColumn()) {
				$pdo->exec("INSERT INTO b2linkcategories (cat_id, cat_name) VALUES (1, 'General')");
			}
			if (!$pdo->query("SELECT COUNT(*) FROM b2links")->fetchColumn()) {
				$pdo->exec("INSERT INTO b2links (link_url, link_name, link_category, link_visible, link_owner) VALUES ('http://wordpress.org', 'WordPress', 1, 'Y', 1)");
			}
		`,env:{DOCUMENT_ROOT:e.documentRoot}}),!0):!1}function k(e,t){const i=o.joinPaths(t,"b2config.php"),n=o.joinPaths(t,"b2-include/b2vars.php");return!e.fileExists(i)||!e.fileExists(n)?!1:e.readFileAsText(n).includes("$b2_version = '0.71'")}async function le(e,t){for(const i of[o.joinPaths(t,"wp-includes"),o.joinPaths(t,"wp-content"),o.joinPaths(t,"wp-content/database")])e.isDir(i)||await e.mkdir(i)}async function _e(e,t){const i=o.joinPaths(t,"wp-includes/version.php");await e.writeFile(i,`<?php
$wp_version = '0.71';
$wp_db_version = 71;
`);const n=o.joinPaths(t,"wp-config.php");e.fileExists(n)||await e.writeFile(n,"<?php require_once dirname(__FILE__) . '/b2config.php';");const s=o.joinPaths(t,"wp-load.php");e.fileExists(s)||await e.writeFile(s,`<?php
if (!defined('ABSPATH')) {
	define('ABSPATH', dirname(__FILE__) . '/');
}
require_once ABSPATH . 'b2config.php';
`);const r=o.joinPaths(t,"wp-admin/post.php");e.fileExists(r)||await e.writeFile(r,"<?php require_once dirname(__FILE__) . '/b2edit.php';")}async function ce(e,t){const i=o.joinPaths(t,"b2config.php");let n=e.readFileAsText(i);const s=n;n.includes("pg_wp07_bootstrap")||(n=n.replace("<?php",`<?php
if (!defined('ABSPATH')) define('ABSPATH', dirname(__FILE__) . '/'); /* pg_wp07_bootstrap */
if (!defined('WPINC')) define('WPINC', 'b2-include');
if (!defined('WP_CONTENT_DIR')) define('WP_CONTENT_DIR', ABSPATH . 'wp-content');
if (!defined('DB_ENGINE')) define('DB_ENGINE', 'sqlite');
if (!isset($table_prefix)) $table_prefix = 'b2';
error_reporting(E_ALL & ~E_NOTICE & ~8192 & ~2048);
`)),n=n.replace(/\$siteurl\s*=\s*'http:\/\/example\.com';[^\n]*/,"$siteurl = defined('WP_SITEURL') ? WP_SITEURL : 'http://localhost'; // pg_wp07_siteurl").replace('$blogname = "my weblog";','$blogname = "My WordPress Website";').replace('$blogdescription = "babblings !";','$blogdescription = "Just another WordPress weblog";').replace("$admin_email = 'you@example.com';","$admin_email = 'admin@localhost.com';").replace("$abspath =  getenv('DOCUMENT_ROOT') . $relpath . '/';","$abspath = dirname(__FILE__) . '/';"),n!==s&&await e.writeFile(i,n)}async function ue(e,t){const i=o.joinPaths(t,"blog.header.php");let n=e.readFileAsText(i);const s=n;n=n.replace(`$where .= ' AND (post_status = "publish"';`,`$where .= " AND (post_status = 'publish'";`),n!==s&&await e.writeFile(i,n)}async function de(e,t){const i=o.joinPaths(t,"b2-include/wp-db.php");let n=e.readFileAsText(i);const s=n;n=pe(n),n=n.replace("$wpdb = new wpdb(DB_USER, DB_PASSWORD, DB_NAME, DB_HOST);",`if (!isset($GLOBALS['wpdb'])) {
	$GLOBALS['wpdb'] = new wpdb(DB_USER, DB_PASSWORD, DB_NAME, DB_HOST);
}
$wpdb = $GLOBALS['wpdb']; /* pg_wp07_preserve_sqlite_loader */`),n!==s&&await e.writeFile(i,n)}function pe(e){const t=[];if(e.includes("function set_prefix")||t.push(`
	function set_prefix($prefix) {
		$this->prefix = $prefix;
		$tables = array('posts', 'users', 'categories', 'comments', 'links', 'linkcategories', 'settings', 'options', 'postmeta', 'usermeta', 'terms', 'term_taxonomy', 'term_relationships');
		foreach ($tables as $table) {
			$this->$table = $prefix . $table;
		}
		return $prefix;
	}`),e.includes("function timer_start")||t.push(`
	function timer_start() {
		$this->time_start = microtime(true);
		return true;
	}`),e.includes("function timer_stop")||t.push(`
	function timer_stop() {
		return microtime(true) - $this->time_start;
	}`),e.includes("function init_charset")||t.push(`
	function init_charset() {
		if (defined('DB_CHARSET')) $this->charset = DB_CHARSET;
		if (defined('DB_COLLATE')) $this->collate = DB_COLLATE;
	}`),e.includes("function bail")||t.push(`
	function bail($message, $error_code = '500') {
		die($message);
	}`),e.includes("function check_connection")||t.push(`
	function check_connection($allow_bail = true) {
		return true;
	}`),t.length===0)return e;const i=e.match(/^(\s*})\s*\n+(\$wpdb|if\s*\(\s*!\s*isset\(\s*\$GLOBALS\['wpdb'\]\s*\))/m);if(!i||i.index===void 0)throw new Error("WP 0.7 wpdb polyfill anchor not found; b2-include/wp-db.php layout changed");const n=`
	// Polyfills added by WordPress Playground for WP 0.7.
`+t.join(`
`)+`

`;return e.substring(0,i.index)+n+e.substring(i.index)}async function fe(e,t){const i=o.joinPaths(t,"b2-include/b2template.functions.php");let n=e.readFileAsText(i);const s=n;n.includes("pg_wp07_apply_filters")||(n=n.replace(`function apply_filters($tag, $string) {
	global $b2_filter;
	if (isset($b2_filter['all'])) {
		$b2_filter['all'] = (is_string($b2_filter['all'])) ? array($b2_filter['all']) : $b2_filter['all'];
		$b2_filter[$tag] = array_merge($b2_filter['all'], $b2_filter[$tag]);
		$b2_filter[$tag] = array_unique($b2_filter[$tag]);
	}
	if (isset($b2_filter[$tag])) {
		$b2_filter[$tags] = (is_string($b2_filter[$tag])) ? array($b2_filter[$tag]) : $b2_filter[$tag];
		$functions = $b2_filter[$tag];
		foreach($functions as $function) {
			$string = $function($string);
		}
	}
	return $string;
}`,`function apply_filters($tag, $string) { /* pg_wp07_apply_filters */
	global $b2_filter;
	$functions = array();
	// WP_SQLite_DB uses apply_filters('query'); b2's all filter texturizes SQL.
	if ($tag != 'query' && isset($b2_filter['all'])) {
		$all = is_array($b2_filter['all']) ? $b2_filter['all'] : array($b2_filter['all']);
		$functions = array_merge($functions, $all);
	}
	if (isset($b2_filter[$tag])) {
		$tag_functions = is_array($b2_filter[$tag]) ? $b2_filter[$tag] : array($b2_filter[$tag]);
		$functions = array_merge($functions, $tag_functions);
	}
	$functions = array_unique($functions);
	foreach($functions as $function) {
		if (function_exists($function)) {
			$string = $function($string);
		}
	}
	return $string;
}`),n!==s&&await e.writeFile(i,n))}async function he(e,t){await $e(e,t),await ge(e,t),await me(e,t),await Te(e,t)}async function $e(e,t){const i=o.joinPaths(t,"b2login.php");let n=e.readFileAsText(i);const s=n;n.includes("pg_wp07_login_auto_login")||(n=n.replace("switch($action) {",`${D("pg_wp07_login_auto_login")}
switch($action) {`)),n.includes("pg_wp07_logout_guard")||(n=n.replace(`	setcookie('wordpressuser');
	setcookie('wordpresspass');`,`	setcookie('wordpressuser', '', time() - 31536000);
	setcookie('wordpresspass', '', time() - 31536000);
	setcookie('wordpressblogid', '', time() - 31536000);
	setcookie('wordpressuser', '', time() - 31536000, '/');
	setcookie('wordpresspass', '', time() - 31536000, '/');
	setcookie('wordpressblogid', '', time() - 31536000, '/');
	setcookie('playground_auto_login_already_logged_out', '1', time() + 172800, '/'); /* pg_wp07_logout_guard */`),n=n.replace("header('Refresh: 0;url=b2login.php');","header('Refresh: 0;url=b2login.php?loggedout=1');").replace("header('Location: b2login.php');","header('Location: b2login.php?loggedout=1');")),n!==s&&await e.writeFile(i,n)}async function ge(e,t){const i=o.joinPaths(t,"wp-admin");if(e.isDir(i))for(const n of e.listFiles(i)){if(!n.endsWith(".php"))continue;const s=o.joinPaths(i,n),r=e.readFileAsText(s),a=I(r);a!==r&&await e.writeFile(s,a)}}async function me(e,t){const i=o.joinPaths(t,"wp-admin/b2menutop.php");let n=e.readFileAsText(i);const s=n;n=n.replace('file("./b2menutop.txt")',"file(dirname(__FILE__) . '/b2menutop.txt')").replace('<a href="http://wordpress.org" rel="external"><span>WordPress</span></a>','<a href="#" rel="external"><span>WordPress</span></a>'),n!==s&&await e.writeFile(i,n)}async function Te(e,t){const i=o.joinPaths(t,"wp-admin/b2verifauth.php");let n=e.readFileAsText(i);const s=n;n.includes("pg_wp07_auto_login")||(n=n.replace("require_once('../b2config.php');",`require_once('../b2config.php');
${D("pg_wp07_auto_login")}`),n!==s&&await e.writeFile(i,n))}function D(e){return`
if (isset($_GET['loggedout']) || isset($HTTP_GET_VARS['loggedout'])) {
	$_COOKIE['playground_auto_login_already_logged_out'] = '1';
	unset($_COOKIE['wordpressuser']);
	unset($_COOKIE['wordpresspass']);
	unset($_COOKIE['wordpressblogid']);
	unset($HTTP_COOKIE_VARS['wordpressuser']);
	unset($HTTP_COOKIE_VARS['wordpresspass']);
	unset($HTTP_COOKIE_VARS['wordpressblogid']);
	if (!headers_sent()) {
		setcookie('playground_auto_login_already_logged_out', '1', time() + 172800, '/');
	}
}
if (
	defined('PLAYGROUND_AUTO_LOGIN_AS_USER') &&
	empty($_COOKIE['playground_auto_login_already_logged_out']) &&
	(!isset($action) || $action != 'logout')
) {
	$_pg_wp07_user = PLAYGROUND_AUTO_LOGIN_AS_USER;
	$_pg_wp07_pass = md5(md5('password'));
	$_COOKIE['wordpressuser'] = $_pg_wp07_user;
	$_COOKIE['wordpresspass'] = $_pg_wp07_pass;
	$_COOKIE['wordpressblogid'] = 1;
	$HTTP_COOKIE_VARS = $_COOKIE;
	$GLOBALS['HTTP_COOKIE_VARS'] = $HTTP_COOKIE_VARS;
	if (!headers_sent()) {
		$_pg_wp07_exp = time() + 172800;
		setcookie('wordpressuser', $_pg_wp07_user, $_pg_wp07_exp, '/');
		setcookie('wordpresspass', $_pg_wp07_pass, $_pg_wp07_exp, '/');
		setcookie('wordpressblogid', 1, $_pg_wp07_exp, '/');
	}
} /* ${e} */
`}async function Ee(e,t){const i=w(e,t);if(i===null)return;const n=parseFloat(i);if(!Number.isFinite(n)||n<5||n>=6.2)return;const s=o.joinPaths(t,"wp-includes/load.php");if(!e.fileExists(s))return;const r=e.readFileAsText(s),a=r.replace("extension_loaded( 'mysqli' )","function_exists( 'mysqli_connect' )");a!==r&&await e.writeFile(s,a)}function w(e,t){const i=o.joinPaths(t,"wp-includes/version.php");if(!e.fileExists(i))return null;const s=e.readFileAsText(i).match(/\$wp_version\s*=\s*['"]([^'"]+)['"]/);return s?s[1]:null}const we=22527,L="E_ALL & ~8192 & ~2048";async function Le(e,t){if(await ae(e,t))return;await xe(e,t),await Ie(e,t),await ke(e,t),await qe(e,t),await We(e,t),await Qe(e,t),await Me(e,t),await Ve(e,t),await Pe(e,t),await Oe(e,t);const i=w(e,t);if(i===null)return;const n=parseFloat(i);Number.isFinite(n)&&(n<1.2&&(await ve(e,t),await Re(e,t)),n<1.5&&await Ne(e,t),1.5<=n&&n<2&&await Xe(e,t),n<2&&(await Ae(e,t),await Ce(e,t)),2.1<=n&&n<2.3&&await Se(e,t),n<2.5&&await Be(e,t),n<2.8&&(await Ye(e,t),await Ke(e,t)),2.9<=n&&n<3.6&&await be(e,t),3.3<=n&&n<3.4&&await Ue(e,t),n>=4.7&&await ye(e,t))}async function ye(e,t){const i=o.joinPaths(t,"wp-content/themes");if(e.isDir(i))for(const n of e.listFiles(i)){const s=o.joinPaths(i,n,"searchform.php");e.fileExists(s)&&e.unlink(s)}}async function be(e,t){const i=o.joinPaths(t,"wp-admin/includes/dashboard.php");if(e.fileExists(i)){let r=e.readFileAsText(i);if(r.includes("function wp_dashboard_primary()")&&!r.includes("/* pg_no_rss */")){for(const a of["wp_dashboard_primary","wp_dashboard_secondary","wp_dashboard_plugins"])r=r.replace(new RegExp(`function ${a}\\(\\)\\s*\\{`),`function ${a}() { /* pg_no_rss */ return;`);await e.writeFile(i,r)}}const n=o.joinPaths(t,"wp-admin/admin.php");if(e.fileExists(n)){let r=e.readFileAsText(n);r.includes("do_action('admin_init');")&&!r.includes("/* pg_admin_init_cleanup */")&&(r=r.replace("do_action('admin_init');",`/* pg_admin_init_cleanup */
if (function_exists('remove_action')) {
	@remove_action('admin_init', '_maybe_update_plugins');
	@remove_action('admin_init', '_maybe_update_themes');
	@remove_action('admin_init', '_maybe_update_core');
	@remove_action('admin_init', 'wp_version_check');
	@remove_action('admin_init', 'wp_update_plugins');
	@remove_action('admin_init', 'wp_update_themes');
}
do_action('admin_init');`),await e.writeFile(n,r))}const s=o.joinPaths(t,"wp-admin/includes/update.php");if(e.fileExists(s)){let r=e.readFileAsText(s);if(!r.includes("/* pg_admin_no_updates */")){for(const a of["wp_plugin_update_rows","wp_plugin_update_row","wp_theme_update_rows","wp_theme_update_row","wp_update_plugins","wp_update_themes"]){const l=new RegExp(`function ${a}\\s*\\([^)]*\\)\\s*\\{`);l.test(r)&&(r=r.replace(l,_=>_+" /* pg_admin_no_updates */ return;"))}await e.writeFile(s,r)}}for(const r of[o.joinPaths(t,"wp-includes/SimplePie/File.php"),o.joinPaths(t,"wp-includes/class-simplepie.php")]){if(!e.fileExists(r))continue;let a=e.readFileAsText(r);a.includes("function SimplePie_File(")&&!a.includes("/* pg_no_fetch */")&&(a=a.replace(/function SimplePie_File\([^)]*\)\s*\{/,l=>l+`
		/* pg_no_fetch */
		$this->error = 'Network requests disabled in Playground';
		$this->success = false;
		return;`),await e.writeFile(r,a))}}async function Oe(e,t){const i=[["function wp_new_blog_notification","pg_no_blog_notification"],["function wp_install_maybe_enable_pretty_permalinks","pg_no_permalink_check"]],n=[o.joinPaths(t,"wp-admin/includes/upgrade.php"),o.joinPaths(t,"wp-admin/upgrade-functions.php")];for(const s of n){if(!e.fileExists(s))continue;let r=e.readFileAsText(s),a=!1;for(const[l,_]of i){if(r.includes(`/* ${_} */`))continue;const c=r.indexOf(l);if(c===-1)continue;const d=r.indexOf("{",c);d!==-1&&(r=r.substring(0,d+1)+` /* ${_} */ return;`+r.substring(d+1),a=!0)}a&&await e.writeFile(s,r)}}async function Pe(e,t){const i=o.joinPaths(t,"wp-load.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);if(!n.includes("error_reporting(")||n.includes("~8192")&&n.includes("~2048"))return;const s=n.replace(/error_reporting\(([^)]+)\)/g,(r,a)=>`error_reporting((${a}) & ~8192 & ~2048)`);await e.writeFile(i,s)}async function Ne(e,t){const i=o.joinPaths(t,"wp-admin/menu.php");if(e.fileExists(i)){const r=e.readFileAsText(i);if(!r.includes("/* pg_wp10_logo_link */")){const a='<h1 id="wphead"><a href="http://wordpress.org" rel="external">WordPress</a></h1>';if(r.includes(a)){const l=r.replace(a,'<h1 id="wphead"><a href="#" rel="external">WordPress</a></h1> <!-- pg_wp10_logo_link -->');l!==r&&await e.writeFile(i,l)}}}const n=o.joinPaths(t,"wp-admin/admin-header.php");if(e.fileExists(n)){const r=e.readFileAsText(n);if(!r.includes("/* pg_wp12_logo_link */")){const a='<a href="http://wordpress.org" rel="external"',l="</a>",_=r.indexOf(a);if(_!==-1){const c=r.indexOf(l,_);if(c!==-1){const d=r.substring(0,_)+'<a href="#">WordPress</a><!-- pg_wp12_logo_link -->'+r.substring(c+l.length);d!==r&&await e.writeFile(n,d)}}}}const s=o.joinPaths(t,"wp-admin/admin-footer.php");if(e.fileExists(s)){const r=e.readFileAsText(s);if(!r.includes("/* pg_wp10_footer_link */")){const a=r.replace('<a href="http://wordpress.org">WordPress</a>',"WordPress<!-- pg_wp10_footer_link -->").replace('<a href="http://wordpress.org/">WordPress</a>',"WordPress<!-- pg_wp10_footer_link -->");a!==r&&await e.writeFile(s,a)}}}async function Ae(e,t){const i=o.joinPaths(t,"wp-admin/edit.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);if(n.includes("/* pg_wp10_post_title_edit */"))return;let s=n;const r='<strong><a href="<?php permalink_link(); ?>" rel="permalink"><?php the_title() ?></a></strong>';s.includes(r)&&(s=s.replace(r,'<strong><a href="post.php?action=edit&amp;post=<?php echo $id /* pg_wp10_post_title_edit */ ?>"><?php the_title() ?></a></strong>'));const a='<td><a href="<?php the_permalink(); ?>" rel="permalink">';s.includes(a)&&(s=s.replace(a,'<td><a href="post.php?action=edit&amp;post=<?php echo $id /* pg_wp10_post_title_edit */ ?>">'));const l=`<td><?php the_title() ?>
		<?php if ('private' == $post->post_status) _e(' - <strong>Private</strong>'); ?></td>`;s.includes(l)&&(s=s.replace(l,`<td><a href="post.php?action=edit&amp;post=<?php echo $id /* pg_wp10_post_title_edit */ ?>"><?php the_title() ?></a>
		<?php if ('private' == $post->post_status) _e(' - <strong>Private</strong>'); ?></td>`)),s!==n&&await e.writeFile(i,s)}async function Ue(e,t){const i=o.joinPaths(t,"wp-admin/includes/screen.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);if(!n.includes("self::$this->_help_sidebar"))return;const s=n.replace(/self::\$this->_help_sidebar/g,"$this->_help_sidebar");s!==n&&await e.writeFile(i,s)}async function Se(e,t){const i=o.joinPaths(t,"wp-admin/plugins.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);if(n.includes("/* pg_wp21_active_plugins_array */"))return;const s="$current = get_option('active_plugins');";if(!n.includes(s))return;const r=n.replace(s,s+`
	if (!is_array($current)) $current = array(); /* pg_wp21_active_plugins_array */`);r!==n&&await e.writeFile(i,r)}async function Re(e,t){const i=o.joinPaths(t,"wp-login.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i),s="AND user_pass = '$password'";if(!n.includes(s)||n.includes("pg_wp10_plain_or_md5"))return;let r=n.replace(s,"AND (user_pass = '$password' OR user_pass = MD5('$password')) /* pg_wp10_plain_or_md5 */");r=r.replace("$login->user_pass == $password","($login->user_pass == $password || $login->user_pass == md5($password))"),r!==n&&await e.writeFile(i,r)}async function xe(e,t){const i=o.joinPaths(t,"wp-includes");if(!e.isDir(i))return;const n=o.joinPaths(i,"version.php");e.fileExists(n)||await e.writeFile(n,"<?php $wp_version = '1.0';")}async function ve(e,t){const i=o.joinPaths(t,"wp-blog-header.php");if(e.fileExists(i)){const s=e.readFileAsText(i),r=`$where .= ' AND (post_status = "publish"';`;s.includes(r)&&await e.writeFile(i,s.replace(r,`$where .= " AND (post_status = 'publish'";`))}const n=o.joinPaths(t,"wp-includes/vars.php");if(e.fileExists(n)){const s=e.readFileAsText(n),r="add_filter('all', 'wptexturize');";s.includes(r)&&await e.writeFile(n,s.replace(r,`// ${r} // Disabled by Playground: mangles SQL literals.`))}}async function Ie(e,t){const i=o.joinPaths(t,"wp-load.php");e.fileExists(i)||await e.writeFile(i,`<?php
if ( !defined('ABSPATH') ) {
	define('ABSPATH', dirname(__FILE__) . '/');
}
require_once(ABSPATH . 'wp-config.php');
`)}async function ke(e,t){const i=o.joinPaths(t,"wp-settings.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);let s=n;s=s.replace(/if\s*\(\s*!extension_loaded\('mysql'\)\s*\)\s*\n\s*die/,`if ( false ) // Patched for SQLite
	die`),s=s.replace(/error_reporting\(([^)]+)\)/g,(r,a)=>a.includes("~8192")&&a.includes("~2048")?r:`error_reporting((${a}) & ~8192 & ~2048)`),s=s.replace(/set_magic_quotes_runtime\(\s*0\s*\)\s*;/g,"// set_magic_quotes_runtime(0); // Removed"),s.includes("function_exists('get_magic_quotes_gpc')")||(s=s.replace(/get_magic_quotes_gpc\(\)/g,"(function_exists('get_magic_quotes_gpc') && get_magic_quotes_gpc())")),s=s.replace(/=\s*&\s*new\b/g,"= new"),s=s.replace(/\$HTTP_SERVER_VARS/g,"$_SERVER"),!s.includes("WP_CONTENT_DIR")&&s.includes("define('WPINC'")&&(s=s.replace(/define\('WPINC',\s*'wp-includes'\);/,`define('WPINC', 'wp-includes');
if (!defined('WP_CONTENT_DIR')) define('WP_CONTENT_DIR', ABSPATH . 'wp-content');`)),s=s.replace(/unset\(\s*\$wp_filter\s*,/,"unset("),s=De(s),s=Fe(s),s!==n&&await e.writeFile(i,s)}function De(e){const t=e.indexOf("installed WP");if(t===-1)return e;const i=e.lastIndexOf("die(",t);if(i===-1)return e;let n=0;for(let s=i+3;s<e.length;s++)if(e[s]==="(")n++;else if(e[s]===")"&&(n--,n===0)){let r=s+1;return e[r]===";"&&r++,e.substring(0,i)+"true; /* die removed by Playground */"+e.substring(r)}return e}function Fe(e){return e.replace("do_action('init');",`// Remove hooks that make outbound HTTP requests (crash WASM).
if (function_exists('remove_action')) {
	@remove_action('init', 'wp_cron');
	@remove_action('init', 'wp_version_check');
	@remove_action('init', 'wp_update_plugins');
	@remove_action('init', 'wp_update_themes');
	@remove_action('admin_init', '_maybe_update_plugins');
	@remove_action('admin_init', '_maybe_update_themes');
	@remove_action('admin_init', 'wp_version_check');
	@remove_action('admin_init', 'wp_update_plugins');
	@remove_action('admin_init', 'wp_update_themes');
	@remove_action('load-plugins.php', 'wp_update_plugins');
	@remove_action('load-update.php', 'wp_update_plugins');
	@remove_action('load-update.php', 'wp_update_themes');
	@remove_action('load-themes.php', 'wp_update_themes');
	@remove_action('wp_update_plugins', 'wp_update_plugins');
	@remove_action('wp_version_check', 'wp_version_check');
}
if (function_exists('add_filter')) {
	function _pg_disable_curl() { return false; }
	function _pg_disable_streams() { return false; }
	@add_filter('use_curl_transport', '_pg_disable_curl');
	@add_filter('use_streams_transport', '_pg_disable_streams');
	@add_filter('use_ftp_transport', '_pg_disable_curl');
	@add_filter('use_fsockopen_transport', '_pg_disable_streams');
}
do_action('init');`)}async function Ce(e,t){const i=o.joinPaths(t,"wp-includes/functions.php");if(!e.fileExists(i))return;let n=e.readFileAsText(i),s=!1;n.includes("$all_options->{$option->option_name}")&&!n.includes("$all_options = new stdClass")&&(n=n.replace("foreach ($options as $option) {",`$all_options = new stdClass;
	foreach ($options as $option) {`),s=!0),s&&await e.writeFile(i,n)}async function qe(e,t){const i=o.joinPaths(t,"wp-admin/install.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);let s=n;const r=o.joinPaths(t,"wp-admin");s=s.replace(/'\.\.\/(wp-config\.php)'/g,`'${t}/$1'`).replace(/'\.\.\/(wp-load\.php)'/g,`'${t}/$1'`).replace(/'\.\/(upgrade-functions\.php)'/g,`'${r}/$1'`).replace(/'(upgrade-functions\.php)'/g,`'${r}/$1'`).replace(/'\.\/(includes\/upgrade\.php)'/g,`'${r}/$1'`).replace(/'\.\.\/(wp-includes\/[^']+)'/g,`'${t}/$1'`),s=s.replace(/\$HTTP_GET_VARS/g,"$_GET").replace(/\$HTTP_POST_VARS/g,"$_POST"),s!==n&&await e.writeFile(i,s)}async function We(e,t){const i=o.joinPaths(t,"wp-includes/wp-db.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);let s=n;s.includes("isset($wpdb)")||(s=s.replace("$wpdb = new wpdb(DB_USER, DB_PASSWORD, DB_NAME, DB_HOST);","if ( !isset($wpdb) ) { $wpdb = new wpdb(DB_USER, DB_PASSWORD, DB_NAME, DB_HOST); }")),s.includes("db_connect")||(s=s.replace(/\$this->dbh\s*=\s*@mysql_connect\(\$dbhost\s*,\s*\$dbuser\s*,\s*\$dbpassword(?:\s*,\s*true)?\);/,'if (method_exists($this, "db_connect")) { $this->dbname = $dbname; $this->db_connect(); } else { $this->dbh = @mysql_connect($dbhost, $dbuser, $dbpassword); }')),s=Ge(s),s!==n&&await e.writeFile(i,s)}function Ge(e){const t=[];if(e.includes("function set_prefix")||t.push(`
	function set_prefix($prefix) {
		$this->prefix = $prefix;
		$tables = array('posts', 'users', 'categories', 'post2cat', 'comments', 'link2cat', 'links', 'options', 'postmeta', 'usermeta', 'terms', 'term_taxonomy', 'term_relationships');
		foreach ($tables as $t) {
			$this->$t = $prefix . $t;
		}
		return $prefix;
	}`),e.includes("function timer_start")||t.push(`
	function timer_start() {
		$this->time_start = microtime(true);
		return true;
	}`),e.includes("function timer_stop")||t.push(`
	function timer_stop() {
		return microtime(true) - $this->time_start;
	}`),e.includes("function init_charset")||t.push(`
	function init_charset() {
		if (defined('DB_CHARSET')) $this->charset = DB_CHARSET;
		if (defined('DB_COLLATE')) $this->collate = DB_COLLATE;
	}`),e.includes("function bail")||t.push(`
	function bail($message, $error_code = '500') {
		die($message);
	}`),e.includes("function check_connection")||t.push(`
	function check_connection($allow_bail = true) {
		return true;
	}`),t.length===0)return e;const i=e.match(/^(\s*})\s*\n+(\$wpdb|\?>\s*$|if\s*\(\s*!\s*isset\(\s*\$wpdb\s*\))/m);if(!i||i.index===void 0)return e;const n=`
	// Polyfills added by WordPress Playground.
`+t.join(`
`)+`

`;return e.substring(0,i.index)+n+e.substring(i.index)}async function Me(e,t){const i=o.joinPaths(t,"wp-admin");if(e.isDir(i))for(const r of e.listFiles(i)){if(!r.endsWith(".php"))continue;const a=o.joinPaths(i,r),l=e.readFileAsText(a),_=I(l).replace(/ABSPATH\s*\.\s*'\/wp-/g,"ABSPATH . 'wp-");_!==l&&await e.writeFile(a,_)}const n=o.joinPaths(t,"wp-admin/index.php");if(e.fileExists(n)){let r=e.readFileAsText(n);r.includes("get_settings('siteurl')")&&(r=r.replace(/get_settings\('siteurl'\)\s*\.\s*'\/wp-admin\//g,"'"),await e.writeFile(n,r))}const s=o.joinPaths(t,"wp-admin/menu.php");if(e.fileExists(s)){const r=e.readFileAsText(s),a="file('./menu.txt')";r.includes(a)&&await e.writeFile(s,r.replace(a,"file(dirname(__FILE__) . '/menu.txt')"))}}async function Be(e,t){const i=o.joinPaths(t,"wp-admin/admin-functions.php");if(!e.fileExists(i))return;const n=e.readFileAsText(i);if(!n.includes("function check_admin_referer()")||!n.includes("$_SERVER['HTTP_REFERER']"))return;const s=He(n,"check_admin_referer",`
	do_action('check_admin_referer', '');
`);s!==n&&await e.writeFile(i,s)}function He(e,t,i){const n=`function ${t}()`,s=e.indexOf(n);if(s===-1)return e;const r=e.indexOf("{",s+n.length);if(r===-1)return e;let a=1;for(let l=r+1;l<e.length;l++){const _=e[l];if(_==="{")a++;else if(_==="}"&&(a--,a===0))return e.substring(0,r+1)+i+e.substring(l)}return e}async function Xe(e,t){const i=o.joinPaths(t,"wp-admin/index.php");if(e.fileExists(i)){const n=e.readFileAsText(i),s=n.replace(/AND post_date_gmt < '\$today'/,"");s!==n&&await e.writeFile(i,s)}await je(e,t)}async function je(e,t){const i=o.joinPaths(t,"wp-includes/rss-functions.php");if(!e.fileExists(i))return;let n=e.readFileAsText(i);!/^\s*error\s*\(/m.test(n)||/^function\s+error\s*\(/m.test(n)||(n=n.replace(/^(<\?php\s*)/,`$1
if (!function_exists('error')) {
	function error($msg = '', $lvl = E_USER_WARNING) {
		if (defined('MAGPIE_DEBUG') && MAGPIE_DEBUG) {
			trigger_error($msg, $lvl);
		}
	}
}
`),await e.writeFile(i,n))}async function Ve(e,t){const i=o.joinPaths(t,"wp-login.php");if(!e.fileExists(i))return;let n=e.readFileAsText(i),s=!1;for(const r of["log","pwd"]){const a=new RegExp(`(\\bname=(['"])${r}\\2)(?![^>]*data-1p-ignore)`);a.test(n)&&(n=n.replace(a,"$1 data-1p-ignore"),s=!0)}s&&await e.writeFile(i,n)}async function Ye(e,t){const i=o.joinPaths(t,"wp-admin/admin.php");if(e.fileExists(i)){const s=e.readFileAsText(i);if(s.includes("auth_redirect()")){const r=`
// Playground: populate auth cookies and force admin user before auth_redirect.
if (defined('PLAYGROUND_AUTO_LOGIN_AS_USER')) {
	if (function_exists('is_user_logged_in') && is_user_logged_in()) {
		// On WP < 4.0, wp_set_auth_cookie() does not update $_COOKIE
		// in-process — auth_redirect() reads $_COOKIE, so re-emit.
		if (function_exists('wp_generate_auth_cookie') && defined('LOGGED_IN_COOKIE') && empty($_COOKIE[LOGGED_IN_COOKIE])) {
			$_pg_uid = wp_get_current_user()->ID;
			$_pg_exp = time() + 172800;
			$_COOKIE[AUTH_COOKIE] = wp_generate_auth_cookie($_pg_uid, $_pg_exp, 'auth');
			if (defined('SECURE_AUTH_COOKIE'))
				$_COOKIE[SECURE_AUTH_COOKIE] = wp_generate_auth_cookie($_pg_uid, $_pg_exp, 'secure_auth');
			$_COOKIE[LOGGED_IN_COOKIE] = wp_generate_auth_cookie($_pg_uid, $_pg_exp, 'logged_in');
		}
	} else {
		${F("PLAYGROUND_AUTO_LOGIN_AS_USER")}
		// WP 2.0-2.4: kses_init() runs during do_action('init') inside
		// wp-settings.php and caches $current_user as WP_User(0) when
		// no cookies were set yet. Reset and re-evaluate so capability
		// checks see the user we just authenticated.
		if (!function_exists('wp_generate_auth_cookie')) {
			$GLOBALS['current_user'] = null;
			if (function_exists('get_currentuserinfo')) {
				get_currentuserinfo();
			}
		}
	}
	// Force admin caps in-memory: if populate_roles() never ran
	// (e.g. WP 2.0, or WP 2.5 installs that crashed before writing
	// roles), the user has no caps and every current_user_can() fails.
	$_pg_cu = isset($GLOBALS['current_user']) ? $GLOBALS['current_user'] : null;
	if ($_pg_cu && isset($_pg_cu->ID) && $_pg_cu->ID > 0 && empty($_pg_cu->allcaps['read'])) {
		// Respect a DB-stored user_level so a blueprint that auto-logs
		// in as a lower-privilege user doesn't silently get level 10.
		$_pg_db_level = isset($_pg_cu->user_level)
			? (int) $_pg_cu->user_level
			: null;
		if ($_pg_db_level === null && isset($_pg_user) && $_pg_user) {
			$_pg_db_level = isset($_pg_user->user_level)
				? (int) $_pg_user->user_level
				: null;
		}
		$_pg_cu->user_level = $_pg_db_level !== null ? $_pg_db_level : 10;
		$_pg_effective_level = $_pg_cu->user_level;
		$_pg_caps = array('read');
		for ($_pg_i = 0; $_pg_i <= $_pg_effective_level; $_pg_i++) {
			$_pg_caps[] = 'level_' . $_pg_i;
		}
		if ($_pg_effective_level >= 10) {
			$_pg_caps = array_merge($_pg_caps, array(
				'switch_themes','edit_themes','activate_plugins',
				'edit_plugins','edit_users','edit_files','manage_options',
				'moderate_comments','manage_categories','manage_links',
				'upload_files','import','unfiltered_html','edit_posts',
				'edit_others_posts','edit_published_posts','publish_posts',
				'edit_pages'));
		}
		foreach ($_pg_caps as $_pg_c) {
			$_pg_cu->allcaps[$_pg_c] = true;
		}
		if ($_pg_effective_level >= 10) {
			$_pg_cu->caps = array('administrator' => true);
		}
	}
}
`,a=s.replace("auth_redirect();",r+"auth_redirect();");a!==s&&await e.writeFile(i,a)}}const n=o.joinPaths(t,"wp-admin/auth.php");if(e.fileExists(n)){const s=e.readFileAsText(n);if(s.includes("$cookiehash")&&!s.includes("Playground: bypass auth")){const r=`<?php
require_once(ABSPATH . 'wp-config.php');
// Playground: bypass auth and manually populate user globals.
global $user_login, $userdata, $user_level, $user_ID,
	$user_nickname, $user_email, $user_url, $user_pass_md5, $cookiehash;
$__pg_user_login = defined('PLAYGROUND_AUTO_LOGIN_AS_USER')
	? PLAYGROUND_AUTO_LOGIN_AS_USER
	: 'admin';
$__pg_cookiehash = defined('COOKIEHASH')
	? COOKIEHASH
	: (isset($cookiehash) && $cookiehash
		? $cookiehash
		: md5(function_exists('get_settings') ? get_settings('siteurl') : ''));
if ($__pg_cookiehash) {
	$_COOKIE['wordpressuser_' . $__pg_cookiehash] = $__pg_user_login;
}
if (function_exists('get_userdatabylogin')) {
	$__pg_userdata = get_userdatabylogin($__pg_user_login);
	if ($__pg_userdata) {
		$user_login = $__pg_user_login;
		$userdata = $__pg_userdata;
		$user_level = isset($__pg_userdata->user_level)
			? (int) $__pg_userdata->user_level
			: 10;
		$user_ID = $__pg_userdata->ID;
		$user_nickname = isset($__pg_userdata->user_nickname)
			? $__pg_userdata->user_nickname
			: $__pg_user_login;
		$user_email = isset($__pg_userdata->user_email)
			? $__pg_userdata->user_email
			: '';
		$user_url = isset($__pg_userdata->user_url)
			? $__pg_userdata->user_url
			: '';
		$user_pass_md5 = md5(
			isset($__pg_userdata->user_pass) ? $__pg_userdata->user_pass : ''
		);
	}
}
?>`;r!==s&&await e.writeFile(n,r)}}}async function Ke(e,t){const i=o.joinPaths(t,"wp-admin/admin-ajax.php");if(!e.fileExists(i))return;let n=e.readFileAsText(i);if(!n.includes("is_user_logged_in"))return;const s=`
// Playground: authenticate admin user for AJAX requests on WP < 2.8.
if (defined('PLAYGROUND_AUTO_LOGIN_AS_USER')) {
	${F("PLAYGROUND_AUTO_LOGIN_AS_USER")}
}
`;n=n.replace(/if\s*\(\s*!\s*is_user_logged_in\(\)\s*\)/,s+"if ( !is_user_logged_in() )"),await e.writeFile(i,n)}function F(e){return`
$_pg_user = null;
if (function_exists('wp_generate_auth_cookie')) {
	$_pg_user = function_exists('get_user_by')
		? get_user_by('login', ${e})
		: (function_exists('get_userdatabylogin')
			? get_userdatabylogin(${e}) : null);
	if ($_pg_user) {
		wp_set_current_user($_pg_user->ID, $_pg_user->user_login);
		$_pg_exp = time() + 172800;
		if (defined('AUTH_COOKIE'))
			$_COOKIE[AUTH_COOKIE] = wp_generate_auth_cookie($_pg_user->ID, $_pg_exp, 'auth');
		if (defined('SECURE_AUTH_COOKIE'))
			$_COOKIE[SECURE_AUTH_COOKIE] = wp_generate_auth_cookie($_pg_user->ID, $_pg_exp, 'secure_auth');
		if (defined('LOGGED_IN_COOKIE'))
			$_COOKIE[LOGGED_IN_COOKIE] = wp_generate_auth_cookie($_pg_user->ID, $_pg_exp, 'logged_in');
	}
}
`}async function Qe(e,t){const i=w(e,t);if(i===null)return;const n=parseFloat(i);if(!Number.isFinite(n)||n>=3.3)return;const s=o.joinPaths(t,"wp-admin/includes/schema.php");if(!e.fileExists(s))return;const r=e.readFileAsText(s);/\$wp_queries\s*=\s*"CREATE TABLE/.test(r)&&!r.includes("function wp_get_db_schema")&&await ze(e,t,s,r)}async function ze(e,t,i,n){const s=n.match(/\$wp_queries\s*=\s*"CREATE TABLE/);if(!s||s.index===void 0)return;const r=s.index,a='";',l=n.indexOf(a,r);if(l===-1)return;const _=l+a.length,d=`function wp_get_db_schema( $scope = 'all', $blog_id = null ) {
	global $wpdb, $wp_queries, $charset_collate;
	$charset_collate = '';
	if ( ! empty($wpdb->charset) )
		$charset_collate = "DEFAULT CHARACTER SET $wpdb->charset";
	if ( ! empty($wpdb->collate) )
		$charset_collate .= " COLLATE $wpdb->collate";
	${n.substring(r,_)}
	return $wp_queries;
}`,p=n.substring(0,r)+d+n.substring(_);await e.writeFile(i,p);const u=o.joinPaths(t,"wp-admin/includes/upgrade.php");if(e.fileExists(u)){const h=e.readFileAsText(u),P=h.replace(/(\$alterations\s*=\s*dbDelta\(\s*\$wp_queries\s*\))/g,"if ( function_exists('wp_get_db_schema') ) { $wp_queries = wp_get_db_schema(); } $1");P!==h&&await e.writeFile(u,P)}}function Ze(){return`<?php
// @playground-managed — Playground-generated db.php.
// WP < 3.0 loads only db.php and skips wp-db.php, so we pull
// in the wpdb class definition explicitly.
if (defined('ABSPATH') && defined('WPINC') && !class_exists('wpdb', false)) {
	require_once(ABSPATH . WPINC . '/wp-db.php');
}
// Old wpdb (WP < 3.0) has no db_connect() and calls mysql_connect()
// inline, so the SQLite driver never gets a chance to attach. Load
// the integration here and reinitialise to swap the dbh in place.
if (
	class_exists('wpdb', false) &&
	isset($GLOBALS['wpdb']) &&
	!($GLOBALS['wpdb'] instanceof wpdb) &&
	!method_exists('wpdb', 'db_connect') &&
	file_exists('/internal/shared/mu-plugins/sqlite-database-integration.php')
) {
	require_once '/internal/shared/mu-plugins/sqlite-database-integration.php';
	if (
		isset($GLOBALS['wpdb']) &&
		$GLOBALS['wpdb'] instanceof wpdb &&
		method_exists($GLOBALS['wpdb'], 'reinitialize_sqlite')
	) {
		$GLOBALS['wpdb']->reinitialize_sqlite();
	}
}
// Remaining mysqli_* stubs not covered by the 0-sqlite.php preload.
// WP 4.x's extension_loaded('mysqli') check expects these to exist.
if (!function_exists('mysqli_real_connect')) {
	function mysqli_real_connect() { return true; }
}
if (!function_exists('mysqli_error')) {
	function mysqli_error() { return ''; }
}
if (!function_exists('mysqli_errno')) {
	function mysqli_errno() { return 0; }
}
if (!function_exists('mysqli_query')) {
	function mysqli_query() { return false; }
}
if (!function_exists('mysqli_set_charset')) {
	function mysqli_set_charset() { return true; }
}
if (!function_exists('mysqli_select_db')) {
	function mysqli_select_db() { return true; }
}
if (!function_exists('mysqli_close')) {
	function mysqli_close() { return true; }
}
`}async function Je(e,t){if(await oe(e))return;let i=null;const n=o.joinPaths(e.documentRoot,"wp-includes/version.php");if(e.fileExists(n)){const r=e.readFileAsText(n).match(/\$wp_version\s*=\s*['"]([^'"]+)['"]/);r&&(i=r[1])}const s=i!==null&&parseFloat(i)<3.5;try{await e.run({code:`<?php
				// WP_INSTALLING bypasses WP 1.x's "not installed" die() in wp-settings.php.
				define('WP_INSTALLING', true);
				error_reporting(${L});
				ini_set('display_errors', '0');
				ob_start();
				$_pg_db_path = getenv('DOCUMENT_ROOT') . '/wp-content/database/.ht.sqlite';
				if (!file_exists($_pg_db_path)) { exit; }
				$_pg_pdo = new PDO('sqlite:' . $_pg_db_path);
				$_pg_check = $_pg_pdo->query("SELECT count(*) FROM sqlite_master WHERE type='table' AND name='wp_users'")->fetchColumn();
				$_pg_pdo = null;
				if (!$_pg_check) { exit; }
				$wp_load = getenv('DOCUMENT_ROOT') . '/wp-load.php';
				if (!file_exists($wp_load)) { exit; }
				require $wp_load;
				ob_clean();
				global $wpdb;
				if (!isset($wpdb) || !method_exists($wpdb, 'query')) { exit; }

				// Persist the scoped siteurl/home to the DB so parse_request()
				// strips the scope prefix from REQUEST_URI. Filters alone
				// (env.php) aren't enough on WP < 2.2.
				$_pg_opts = !empty($wpdb->options) ? $wpdb->options : $GLOBALS['table_prefix'] . 'options';
				try {
					$_pg_url = getenv('PLAYGROUND_SITE_URL');
					if ($_pg_url) {
						$_pg_current = $wpdb->get_var("SELECT option_value FROM {$_pg_opts} WHERE option_name = 'siteurl'");
						if ($_pg_current !== $_pg_url) {
							$wpdb->query("UPDATE {$_pg_opts} SET option_value = '{$_pg_url}' WHERE option_name = 'siteurl'");
							$wpdb->query("UPDATE {$_pg_opts} SET option_value = '{$_pg_url}' WHERE option_name = 'home'");
						}
					}
				} catch (Exception $e) {}

				// $wpdb->users exists on WP 1.5+; older WP needs the prefix.
				$users_table = !empty($wpdb->users) ? $wpdb->users : $GLOBALS['table_prefix'] . 'users';

				// WP 1.0/1.2 installers often leave the users table or admin row missing.
				$wpdb->query("CREATE TABLE IF NOT EXISTS {$users_table} (
					ID int(10) unsigned NOT NULL auto_increment,
					user_login varchar(20) NOT NULL default '',
					user_pass varchar(64) NOT NULL default '',
					user_firstname varchar(50) NOT NULL default '',
					user_lastname varchar(50) NOT NULL default '',
					user_nickname varchar(50) NOT NULL default '',
					user_icq int(10) unsigned NOT NULL default '0',
					user_email varchar(100) NOT NULL default '',
					user_url varchar(100) NOT NULL default '',
					user_ip varchar(15) NOT NULL default '',
					user_domain varchar(200) NOT NULL default '',
					user_browser varchar(200) NOT NULL default '',
					dateYMDhour datetime NOT NULL default '0000-00-00 00:00:00',
					user_level int(2) unsigned NOT NULL default '0',
					user_aim varchar(50) NOT NULL default '',
					user_msn varchar(100) NOT NULL default '',
					user_yim varchar(50) NOT NULL default '',
					user_idmode varchar(20) NOT NULL default '',
					PRIMARY KEY (ID),
					UNIQUE KEY user_login (user_login)
				)");
				if (!$wpdb->get_var("SELECT COUNT(*) FROM {$users_table}")) {
					$now = date('Y-m-d H:i:s');
					$wpdb->query(
						"INSERT INTO {$users_table} (ID, user_login, user_pass, user_email, user_level, dateYMDhour, user_nickname) " .
						"VALUES (1, 'admin', MD5('password'), 'admin@localhost.com', 10, '{$now}', 'admin')"
					);
				}
				$wpdb->query(
					"UPDATE {$users_table} SET user_pass = MD5('password') WHERE user_login = 'admin'"
				);

				// populate_roles() can fail on SQLite; seed the admin role and caps directly.
				$p = $GLOBALS['table_prefix'];
				$roles_key = $p . 'user_roles';
				try {
					$has_roles = $wpdb->get_var(
						"SELECT COUNT(*) FROM {$p}options WHERE option_name = '{$roles_key}'"
					);
				} catch (Exception $e) {
					$has_roles = 0;
				}
				if (!$has_roles) {
					$roles = array('administrator' => array(
						'name' => 'Administrator',
						'capabilities' => array(
							'switch_themes'=>true, 'edit_themes'=>true,
							'activate_plugins'=>true, 'edit_plugins'=>true,
							'edit_users'=>true, 'edit_files'=>true,
							'manage_options'=>true, 'moderate_comments'=>true,
							'manage_categories'=>true, 'manage_links'=>true,
							'upload_files'=>true, 'import'=>true,
							'unfiltered_html'=>true, 'edit_posts'=>true,
							'edit_others_posts'=>true, 'edit_published_posts'=>true,
							'publish_posts'=>true, 'edit_pages'=>true,
							'read'=>true, 'level_10'=>true, 'level_9'=>true,
							'level_8'=>true, 'level_7'=>true, 'level_6'=>true,
							'level_5'=>true, 'level_4'=>true, 'level_3'=>true,
							'level_2'=>true, 'level_1'=>true, 'level_0'=>true,
							'edit_others_pages'=>true, 'edit_published_pages'=>true,
							'publish_pages'=>true, 'delete_pages'=>true,
							'delete_others_pages'=>true, 'delete_published_pages'=>true,
							'delete_posts'=>true, 'delete_others_posts'=>true,
							'delete_published_posts'=>true, 'delete_private_posts'=>true,
							'edit_private_posts'=>true, 'read_private_posts'=>true,
							'delete_private_pages'=>true, 'edit_private_pages'=>true,
							'read_private_pages'=>true,
						)
					));
					$wpdb->query("INSERT INTO {$p}options (option_name, option_value, autoload) VALUES ('{$roles_key}', '" . addslashes(serialize($roles)) . "', 'yes')");
				}
				$um = isset($wpdb->usermeta) ? $wpdb->usermeta : $p . 'usermeta';
				try {
					$has_cap = $wpdb->get_var("SELECT COUNT(*) FROM {$um} WHERE user_id=1 AND meta_key='{$p}capabilities'");
					if (!$has_cap) {
						$cap_val = addslashes(serialize(array('administrator' => true)));
						$wpdb->query("INSERT INTO {$um} (user_id, meta_key, meta_value) VALUES (1, '{$p}capabilities', '{$cap_val}')");
					}
					$has_level = $wpdb->get_var("SELECT COUNT(*) FROM {$um} WHERE user_id=1 AND meta_key='{$p}user_level'");
					if (!$has_level) {
						$wpdb->query("INSERT INTO {$um} (user_id, meta_key, meta_value) VALUES (1, '{$p}user_level', '10')");
					}
				} catch (Exception $e) {}

				// Seed default content when the install left the posts table empty.
				$posts_table = !empty($wpdb->posts) ? $wpdb->posts : $GLOBALS['table_prefix'] . 'posts';
				$has_posts = false;
				try { $has_posts = (bool)$wpdb->get_var("SELECT COUNT(*) FROM {$posts_table}"); } catch (Exception $e) {}
				if (!$has_posts) {
					$now = date('Y-m-d H:i:s');
					$now_gmt = gmdate('Y-m-d H:i:s');
					if (isset($wpdb->categories)) {
						$wpdb->query("INSERT INTO {$wpdb->categories} (cat_ID, cat_name, category_nicename, category_description, category_parent) VALUES (1, 'Uncategorized', 'uncategorized', '', 0)");
					}
					// Columns common to WP 1.0+.
					$wpdb->query("INSERT INTO {$posts_table} (ID, post_author, post_date, post_date_gmt, post_content, post_title, post_excerpt, post_status, comment_status, ping_status, post_password, post_name, to_ping, pinged, post_modified, post_modified_gmt, post_content_filtered) VALUES (1, 1, '{$now}', '{$now_gmt}', 'Welcome to WordPress. This is your first post. Edit or delete it, then start blogging!', 'Hello world!', '', 'publish', 'open', 'open', '', 'hello-world', '', '', '{$now}', '{$now_gmt}', '')");
					if (isset($wpdb->comments)) {
						$wpdb->query("INSERT INTO {$wpdb->comments} (comment_post_ID, comment_author, comment_author_email, comment_author_url, comment_author_IP, comment_date, comment_date_gmt, comment_content, comment_karma, comment_approved, comment_agent, comment_type, comment_parent, user_id) VALUES (1, 'Mr WordPress', '', 'http://wordpress.org', '127.0.0.1', '{$now}', '{$now_gmt}', 'Hi, this is a comment. To delete a comment, just log in and view the post comments. There you will have the option to edit or delete them.', 0, '1', '', '', 0, 0)");
					}
					if (isset($wpdb->post2cat)) {
						$wpdb->query("INSERT INTO {$wpdb->post2cat} (rel_id, post_id, category_id) VALUES (1, 1, 1)");
					}
				}
			`,env:{DOCUMENT_ROOT:e.documentRoot,PLAYGROUND_SITE_URL:t||""}})}catch(r){$.logger.warn("Legacy WP post-install fixups failed (non-fatal):",r)}if(s)try{await e.run({code:`<?php
				$db_dir = getenv('DOCUMENT_ROOT') . '/wp-content/database/';
				if (!is_dir($db_dir)) { @mkdir($db_dir, 0777, true); }
				$db_path = $db_dir . '.ht.sqlite';
				$pdo = new PDO('sqlite:' . $db_path);
				$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

				$prefix = 'wp_';
				$table = $prefix . 'users';
				try {
					$count = $pdo->query("SELECT COUNT(*) FROM {$table}")->fetchColumn();
				} catch (Exception $e) {
					$pdo->exec("CREATE TABLE IF NOT EXISTS {$table} (
						ID INTEGER PRIMARY KEY AUTOINCREMENT,
						user_login TEXT NOT NULL DEFAULT '',
						user_pass TEXT NOT NULL DEFAULT '',
						user_nickname TEXT NOT NULL DEFAULT '',
						user_email TEXT NOT NULL DEFAULT '',
						user_url TEXT NOT NULL DEFAULT '',
						user_ip TEXT NOT NULL DEFAULT '',
						user_domain TEXT NOT NULL DEFAULT '',
						user_browser TEXT NOT NULL DEFAULT '',
						dateYMDhour TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						user_level INTEGER NOT NULL DEFAULT 0,
						user_idmode TEXT NOT NULL DEFAULT '',
						user_firstname TEXT NOT NULL DEFAULT '',
						user_lastname TEXT NOT NULL DEFAULT '',
						user_icq INTEGER NOT NULL DEFAULT 0,
						user_aim TEXT NOT NULL DEFAULT '',
						user_msn TEXT NOT NULL DEFAULT '',
						user_yim TEXT NOT NULL DEFAULT ''
					)");
					$count = 0;
				}
				if ($count == 0) {
					$now = date('Y-m-d H:i:s');
					// SECURITY: md5('password') matches WP 1.0-1.2's single-md5
					// scheme so auto-login works without a blueprint password.
					// Safe only inside the Playground WASM sandbox.
					$pass = md5('password');
					try {
						$col_info = $pdo->query("PRAGMA table_info({$table})")->fetchAll(PDO::FETCH_ASSOC);
						$known = array(
							'ID' => '1', 'user_login' => "'admin'",
							'user_pass' => "'{$pass}'", 'user_email' => "'admin@localhost.com'",
							'user_level' => '10', 'dateYMDhour' => "'{$now}'",
							'user_nickname' => "'admin'", 'user_nicename' => "'admin'",
							'user_registered' => "'{$now}'", 'user_status' => '0',
						);
						$ins_cols = array(); $ins_vals = array();
						foreach ($col_info as $ci) {
							$cn = $ci['name'];
							$ins_cols[] = $cn;
							if (isset($known[$cn])) {
								$ins_vals[] = $known[$cn];
							} elseif ($ci['dflt_value'] !== null) {
								$ins_vals[] = $ci['dflt_value'];
							} elseif (stripos($ci['type'], 'int') !== false) {
								$ins_vals[] = '0';
							} else {
								$ins_vals[] = "''";
							}
						}
						$pdo->exec("INSERT INTO {$table} (" . implode(',', $ins_cols) . ") VALUES (" . implode(',', $ins_vals) . ")");
					} catch (Exception $e) {}
				} else {
					// See SECURITY note above.
					$pass = md5('password');
					try { $pdo->exec("UPDATE {$table} SET user_pass = '{$pass}' WHERE user_login = 'admin'"); } catch (Exception $e) {}
				}

				// WP 1.0-1.2 install often leaves these tables missing because
				// the SQLite driver can't translate the old-style CREATE TABLEs.
				$now = date('Y-m-d H:i:s');
				$now_gmt = gmdate('Y-m-d H:i:s');
				$tables_sql = array(
					'posts' => "CREATE TABLE IF NOT EXISTS {$prefix}posts (
						ID INTEGER PRIMARY KEY AUTOINCREMENT,
						post_author INTEGER NOT NULL DEFAULT 0,
						post_date TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						post_date_gmt TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						post_content TEXT NOT NULL DEFAULT '',
						post_title TEXT NOT NULL DEFAULT '',
						post_category INTEGER NOT NULL DEFAULT 0,
						post_excerpt TEXT NOT NULL DEFAULT '',
						post_status TEXT NOT NULL DEFAULT 'publish',
						comment_status TEXT NOT NULL DEFAULT 'open',
						ping_status TEXT NOT NULL DEFAULT 'open',
						post_password TEXT NOT NULL DEFAULT '',
						post_name TEXT NOT NULL DEFAULT '',
						to_ping TEXT NOT NULL DEFAULT '',
						pinged TEXT NOT NULL DEFAULT '',
						post_modified TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						post_modified_gmt TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						post_content_filtered TEXT NOT NULL DEFAULT '',
						post_parent INTEGER NOT NULL DEFAULT 0,
						menu_order INTEGER NOT NULL DEFAULT 0,
						post_mime_type TEXT NOT NULL DEFAULT ''
					)",
					'categories' => "CREATE TABLE IF NOT EXISTS {$prefix}categories (
						cat_ID INTEGER PRIMARY KEY AUTOINCREMENT,
						cat_name TEXT NOT NULL DEFAULT '',
						category_nicename TEXT NOT NULL DEFAULT '',
						category_description TEXT NOT NULL DEFAULT '',
						category_parent INTEGER NOT NULL DEFAULT 0
					)",
					'post2cat' => "CREATE TABLE IF NOT EXISTS {$prefix}post2cat (
						rel_id INTEGER PRIMARY KEY AUTOINCREMENT,
						post_id INTEGER NOT NULL DEFAULT 0,
						category_id INTEGER NOT NULL DEFAULT 0
					)",
					'comments' => "CREATE TABLE IF NOT EXISTS {$prefix}comments (
						comment_ID INTEGER PRIMARY KEY AUTOINCREMENT,
						comment_post_ID INTEGER NOT NULL DEFAULT 0,
						comment_author TEXT NOT NULL DEFAULT '',
						comment_author_email TEXT NOT NULL DEFAULT '',
						comment_author_url TEXT NOT NULL DEFAULT '',
						comment_author_IP TEXT NOT NULL DEFAULT '',
						comment_date TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						comment_date_gmt TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						comment_content TEXT NOT NULL DEFAULT '',
						comment_karma INTEGER NOT NULL DEFAULT 0,
						comment_approved TEXT NOT NULL DEFAULT '1',
						comment_agent TEXT NOT NULL DEFAULT '',
						comment_type TEXT NOT NULL DEFAULT '',
						comment_parent INTEGER NOT NULL DEFAULT 0,
						user_id INTEGER NOT NULL DEFAULT 0
					)",
					'options' => "CREATE TABLE IF NOT EXISTS {$prefix}options (
						option_id INTEGER PRIMARY KEY AUTOINCREMENT,
						blog_id INTEGER NOT NULL DEFAULT 0,
						option_name TEXT NOT NULL DEFAULT '',
						option_can_override TEXT NOT NULL DEFAULT 'Y',
						option_type INTEGER NOT NULL DEFAULT 1,
						option_value TEXT NOT NULL DEFAULT '',
						option_width INTEGER NOT NULL DEFAULT 20,
						option_height INTEGER NOT NULL DEFAULT 8,
						option_description TEXT NOT NULL DEFAULT '',
						option_admin_level INTEGER NOT NULL DEFAULT 1,
						autoload TEXT NOT NULL DEFAULT 'yes'
					)",
					'postmeta' => "CREATE TABLE IF NOT EXISTS {$prefix}postmeta (
						meta_id INTEGER PRIMARY KEY AUTOINCREMENT,
						post_id INTEGER NOT NULL DEFAULT 0,
						meta_key TEXT NOT NULL DEFAULT '',
						meta_value TEXT NOT NULL DEFAULT ''
					)",
					'links' => "CREATE TABLE IF NOT EXISTS {$prefix}links (
						link_id INTEGER PRIMARY KEY AUTOINCREMENT,
						link_url TEXT NOT NULL DEFAULT '',
						link_name TEXT NOT NULL DEFAULT '',
						link_image TEXT NOT NULL DEFAULT '',
						link_target TEXT NOT NULL DEFAULT '',
						link_category INTEGER NOT NULL DEFAULT 0,
						link_description TEXT NOT NULL DEFAULT '',
						link_visible TEXT NOT NULL DEFAULT 'Y',
						link_owner INTEGER NOT NULL DEFAULT 1,
						link_rating INTEGER NOT NULL DEFAULT 0,
						link_updated TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
						link_rel TEXT NOT NULL DEFAULT '',
						link_notes TEXT NOT NULL DEFAULT '',
						link_rss TEXT NOT NULL DEFAULT ''
					)",
					'linkcategories' => "CREATE TABLE IF NOT EXISTS {$prefix}linkcategories (
						cat_id INTEGER PRIMARY KEY AUTOINCREMENT,
						cat_name TEXT NOT NULL DEFAULT '',
						auto_toggle TEXT NOT NULL DEFAULT 'N',
						show_images TEXT NOT NULL DEFAULT 'Y',
						show_description TEXT NOT NULL DEFAULT 'N',
						show_rating TEXT NOT NULL DEFAULT 'Y',
						show_updated TEXT NOT NULL DEFAULT 'Y',
						sort_order TEXT NOT NULL DEFAULT 'name',
						sort_desc TEXT NOT NULL DEFAULT 'ASC',
						text_before_link TEXT NOT NULL DEFAULT '<li>',
						text_after_link TEXT NOT NULL DEFAULT '<br />',
						text_after_all TEXT NOT NULL DEFAULT '</li>',
						list_limit INTEGER NOT NULL DEFAULT -1
					)",
					'optiongroups' => "CREATE TABLE IF NOT EXISTS {$prefix}optiongroups (
						group_id INTEGER PRIMARY KEY AUTOINCREMENT,
						group_name TEXT NOT NULL DEFAULT '',
						group_desc TEXT DEFAULT '',
						group_longdesc TEXT DEFAULT ''
					)",
					'optiongroup_options' => "CREATE TABLE IF NOT EXISTS {$prefix}optiongroup_options (
						group_id INTEGER NOT NULL DEFAULT 0,
						option_id INTEGER NOT NULL DEFAULT 0,
						seq INTEGER NOT NULL DEFAULT 0,
						PRIMARY KEY (group_id, option_id)
					)"
				);
				foreach ($tables_sql as $t => $sql) {
					try { $pdo->exec($sql); } catch (Exception $e) {}
				}
				// Backfill columns that WP 1.0-1.2 installs leave off but later code paths read.
				$alter_cols = array(
					'categories' => array(
						'category_nicename' => "TEXT NOT NULL DEFAULT ''",
						'category_description' => "TEXT NOT NULL DEFAULT ''",
						'category_parent' => "INTEGER NOT NULL DEFAULT 0",
						'category_count' => "INTEGER NOT NULL DEFAULT 0",
					),
					// WP 1.5+ get_comments_number() reads comment_count off wp_posts.
					'posts' => array(
						'comment_count' => "INTEGER NOT NULL DEFAULT 0",
					),
				);
				foreach ($alter_cols as $t => $cols_to_add) {
					try {
						$existing = $pdo->query("PRAGMA table_info({$prefix}{$t})")->fetchAll(PDO::FETCH_COLUMN, 1);
						foreach ($cols_to_add as $col => $type) {
							if (!in_array($col, $existing)) {
								$pdo->exec("ALTER TABLE {$prefix}{$t} ADD COLUMN {$col} {$type}");
							}
						}
					} catch (Exception $e) {}
				}
				// Dynamic column detection because the schema differs across WP 1.x.
				try {
					if (!$pdo->query("SELECT COUNT(*) FROM {$prefix}posts")->fetchColumn()) {
						$post_cols = $pdo->query("PRAGMA table_info({$prefix}posts)")->fetchAll(PDO::FETCH_COLUMN, 1);
						$post_vals = array(
							'ID' => '1', 'post_author' => '1',
							'post_date' => "'{$now}'", 'post_date_gmt' => "'{$now_gmt}'",
							'post_content' => "'Welcome to WordPress. This is your first post. Edit or delete it, then start blogging!'",
							'post_title' => "'Hello world!'", 'post_excerpt' => "''",
							'post_status' => "'publish'", 'comment_status' => "'open'",
							'ping_status' => "'open'", 'post_password' => "''",
							'post_name' => "'hello-world'", 'to_ping' => "''", 'pinged' => "''",
							'post_modified' => "'{$now}'", 'post_modified_gmt' => "'{$now_gmt}'",
							'post_content_filtered' => "''",
						);
						$ins_c = array(); $ins_v = array();
						foreach ($post_vals as $c => $v) {
							if (in_array($c, $post_cols)) { $ins_c[] = $c; $ins_v[] = $v; }
						}
						if ($ins_c) $pdo->exec("INSERT INTO {$prefix}posts (" . implode(',', $ins_c) . ") VALUES (" . implode(',', $ins_v) . ")");
					}
				} catch (Exception $e) {}
				try {
					if (!$pdo->query("SELECT COUNT(*) FROM {$prefix}categories")->fetchColumn()) {
						$pdo->exec("INSERT INTO {$prefix}categories (cat_ID, cat_name, category_nicename, category_description, category_parent) VALUES (1, 'Uncategorized', 'uncategorized', '', 0)");
					}
				} catch (Exception $e) {}
				try {
					$env_site = getenv('PLAYGROUND_SITE_URL');
					$site = $env_site ? $env_site : 'http://localhost';
					if (!$pdo->query("SELECT COUNT(*) FROM {$prefix}options WHERE option_name='siteurl'")->fetchColumn()) {
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value) VALUES ('siteurl', '{$site}')");
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value) VALUES ('blogname', 'My WordPress Website')");
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value) VALUES ('blogdescription', 'Just another WordPress weblog')");
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value) VALUES ('home', '{$site}')");
					}
					// Overwrite the placeholder 'http://localhost' with the scoped URL.
					if ($env_site) {
						$pdo->exec("UPDATE {$prefix}options SET option_value = '{$env_site}' WHERE option_name = 'siteurl'");
						$pdo->exec("UPDATE {$prefix}options SET option_value = '{$env_site}' WHERE option_name = 'home'");
					}
					// populate_options() sets template/stylesheet; backfill if it crashed.
					if (!$pdo->query("SELECT COUNT(*) FROM {$prefix}options WHERE option_name='template'")->fetchColumn()) {
						$themes_dir = getenv('DOCUMENT_ROOT') . '/wp-content/themes/';
						$tpl = 'default';
						if (is_dir($themes_dir)) {
							$entries = glob($themes_dir . '*', GLOB_ONLYDIR);
							if ($entries) {
								foreach ($entries as $e) {
									$name = basename($e);
									if ($name === '.' || $name === '..') continue;
									if (file_exists($e . '/style.css')) {
										$tpl = $name;
										break;
									}
								}
							}
						}
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value, autoload) VALUES ('template', '{$tpl}', 'yes')");
						$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value, autoload) VALUES ('stylesheet', '{$tpl}', 'yes')");
					}
					// Without a correct db_version, WP 2.0-2.5 admin redirects to upgrade.php.
					$version_path = getenv('DOCUMENT_ROOT') . '/wp-includes/version.php';
					if (file_exists($version_path)) {
						$wp_db_version = 0;
						include $version_path;
						if ($wp_db_version > 0) {
							$has_dbv = $pdo->query("SELECT COUNT(*) FROM {$prefix}options WHERE option_name='db_version'")->fetchColumn();
							if (!$has_dbv) {
								$pdo->exec("INSERT INTO {$prefix}options (option_name, option_value, autoload) VALUES ('db_version', '{$wp_db_version}', 'yes')");
							} else {
								$pdo->exec("UPDATE {$prefix}options SET option_value = '{$wp_db_version}' WHERE option_name = 'db_version'");
							}
						}
					}
				} catch (Exception $e) {}
			`,env:{DOCUMENT_ROOT:e.documentRoot,PLAYGROUND_SITE_URL:t||""}})}catch(r){$.logger.warn("Legacy WP PDO fallback failed (non-fatal):",r)}}const et=`
// Connection stubs — wpdb::__construct bails on a falsy return.
if (!function_exists('mysqli_connect')) {
	function mysqli_connect() { return true; }
}
if (!function_exists('mysqli_init')) {
	function mysqli_init() { return true; }
}
if (!function_exists('mysql_connect')) {
	function mysql_connect() { return true; }
}
if (!function_exists('mysql_select_db')) {
	function mysql_select_db() { return true; }
}
// WordPress < 3.0 wpdb::__construct calls mysql_set_charset directly.
if (!function_exists('mysql_set_charset')) {
	function mysql_set_charset() { return true; }
}
if (!defined('MYSQL_ASSOC')) define('MYSQL_ASSOC', 1);
if (!defined('MYSQL_NUM')) define('MYSQL_NUM', 2);
if (!defined('MYSQL_BOTH')) define('MYSQL_BOTH', 3);
// Functional mysql_* stubs that delegate to $wpdb (SQLite driver).
$GLOBALS['_mysql_results'] = array();
$GLOBALS['_mysql_result_id'] = 0;
if (!function_exists('mysql_query')) {
	function mysql_query($query, $link = null) {
		global $wpdb;
		if (isset($wpdb) && method_exists($wpdb, 'query')) {
			$wpdb->query($query);
			if (preg_match('/^\\s*(SELECT|SHOW|DESCRIBE|EXPLAIN)/i', $query)) {
				$rows = isset($wpdb->last_result) ? $wpdb->last_result : array();
				$id = ++$GLOBALS['_mysql_result_id'];
				$GLOBALS['_mysql_results'][$id] = array(
					'rows' => $rows,
					'index' => 0,
				);
				return $id;
			}
			return true;
		}
		return false;
	}
}
if (!function_exists('mysql_error')) {
	function mysql_error($link = null) {
		global $wpdb;
		if (isset($wpdb) && isset($wpdb->last_error)) {
			return $wpdb->last_error;
		}
		return '';
	}
}
if (!function_exists('mysql_list_tables')) {
	function mysql_list_tables($db = '', $link = null) {
		global $wpdb;
		if (isset($wpdb) && method_exists($wpdb, 'get_results')) {
			$tables = $wpdb->get_results(
				"SELECT name FROM sqlite_master WHERE type='table' ORDER BY name"
			);
			$rows = array();
			if ($tables) {
				foreach ($tables as $t) {
					$obj = new stdClass();
					$obj->name = is_object($t) ? $t->name : $t['name'];
					$rows[] = $obj;
				}
			}
			$id = ++$GLOBALS['_mysql_result_id'];
			$GLOBALS['_mysql_results'][$id] = array(
				'rows' => $rows,
				'index' => 0,
			);
			return $id;
		}
		return false;
	}
}
if (!function_exists('mysql_fetch_row')) {
	function mysql_fetch_row($result) {
		if (!isset($GLOBALS['_mysql_results'][$result])) return null;
		$r = &$GLOBALS['_mysql_results'][$result];
		if ($r['index'] >= count($r['rows'])) return null;
		$row = $r['rows'][$r['index']++];
		return array_values((array)$row);
	}
}
if (!function_exists('mysql_fetch_assoc')) {
	function mysql_fetch_assoc($result) {
		if (!isset($GLOBALS['_mysql_results'][$result])) return null;
		$r = &$GLOBALS['_mysql_results'][$result];
		if ($r['index'] >= count($r['rows'])) return null;
		return (array)$r['rows'][$r['index']++];
	}
}
if (!function_exists('mysql_fetch_array')) {
	function mysql_fetch_array($result, $result_type = null) {
		$row = mysql_fetch_assoc($result);
		if ($row === null) return null;
		if ($result_type === MYSQL_ASSOC) return $row;
		if ($result_type === MYSQL_NUM) return array_values($row);
		return array_merge(array_values($row), $row);
	}
}
if (!function_exists('mysql_fetch_object')) {
	function mysql_fetch_object($result) {
		if (!isset($GLOBALS['_mysql_results'][$result])) return null;
		$r = &$GLOBALS['_mysql_results'][$result];
		if ($r['index'] >= count($r['rows'])) return null;
		return (object)(array)$r['rows'][$r['index']++];
	}
}
if (!function_exists('mysql_num_rows')) {
	function mysql_num_rows($result) {
		if (isset($GLOBALS['_mysql_results'][$result])) {
			return count($GLOBALS['_mysql_results'][$result]['rows']);
		}
		return 0;
	}
}
if (!function_exists('mysql_get_server_info')) {
	function mysql_get_server_info() { return '8.0.0'; }
}
if (!function_exists('mysql_affected_rows')) {
	function mysql_affected_rows() {
		global $wpdb;
		if (isset($wpdb) && isset($wpdb->rows_affected)) {
			return $wpdb->rows_affected;
		}
		return 0;
	}
}
if (!function_exists('mysql_insert_id')) {
	function mysql_insert_id() {
		global $wpdb;
		if (isset($wpdb) && isset($wpdb->insert_id)) {
			return $wpdb->insert_id;
		}
		return 0;
	}
}
if (!function_exists('mysql_free_result')) {
	function mysql_free_result($result) {
		unset($GLOBALS['_mysql_results'][$result]);
		return true;
	}
}
if (!function_exists('mysql_num_fields')) {
	function mysql_num_fields($result) {
		if (isset($GLOBALS['_mysql_results'][$result])
			&& !empty($GLOBALS['_mysql_results'][$result]['rows'])) {
			return count((array)$GLOBALS['_mysql_results'][$result]['rows'][0]);
		}
		return 0;
	}
}
if (!function_exists('mysql_real_escape_string')) {
	function mysql_real_escape_string($s) { return addslashes($s); }
}
if (!function_exists('mysql_escape_string')) {
	function mysql_escape_string($s) { return addslashes($s); }
}`;async function tt(e,t,i={}){await e.isDir("/tmp/sqlite-database-integration")&&await e.rmdir("/tmp/sqlite-database-integration",{recursive:!0}),await e.mkdir("/tmp/sqlite-database-integration"),await m.unzipFile(e,t,"/tmp/sqlite-database-integration");const n="/internal/shared/sqlite-database-integration",s=`/tmp/sqlite-database-integration/${(await e.listFiles("/tmp/sqlite-database-integration"))[0]}`;await e.mv(s,n),await nt(e,n),await e.defineConstant("SQLITE_MAIN_FILE","1");let a=(await e.readFileAsText(o.joinPaths(n,"db.copy"))).replace("'{SQLITE_IMPLEMENTATION_FOLDER_PATH}'",o.phpVar(n)).replace("'{SQLITE_PLUGIN}'",o.phpVar(o.joinPaths(n,"load.php")));a=a.replace(/^add_action\(/gm,'function_exists("add_action") && add_action(');const l=o.joinPaths(await e.documentRoot,"wp-content/db.php"),_="/internal/shared/mu-plugins/sqlite-database-integration.php",c=`
if(file_exists(${o.phpVar(l)})) {
	$_pg_db_php = @file_get_contents(${o.phpVar(l)});
	if (strpos($_pg_db_php, '@playground-managed') === false) {
		return;
	}
	unset($_pg_db_php);
}
`;await e.writeFile(_,`<?php
${c}?>`+a),await e.writeFile("/internal/shared/preload/0-sqlite.php",it(c,_)),await e.writeFile("/internal/shared/mu-plugins/sqlite-test.php",`<?php
		global $wpdb;
		if(!($wpdb instanceof WP_SQLite_DB)) {
			var_dump(isset($wpdb));
			die("SQLite integration not loaded " . get_class($wpdb));
		}
		`)}async function nt(e,t){const i=o.joinPaths(t,"wp-includes/database/sqlite/class-wp-pdo-mysql-on-sqlite.php");if(!await e.fileExists(i))return;const n=await e.readFileAsText(i),s=n.replace(/\$active_sql_modes\s*=\s*array\s*\([^)]*\)\s*;/,"$active_sql_modes = array();");s!==n&&await e.writeFile(i,s)}function it(e,t){return`<?php
${e}?>
<?php
// Shim __() etc. only for WP < 1.2 (no l10n layer; the SQLite
// plugin calls __() from print_error()). WP 1.2–1.4 ship
// wp-l10n.php and WP 1.5+ ships l10n.php — defining the shims
// then would fatal on redeclare.
$_pg_doc_root = isset($_SERVER['DOCUMENT_ROOT'])
	? $_SERVER['DOCUMENT_ROOT'] : '/wordpress';
if (
	!file_exists($_pg_doc_root . '/wp-includes/l10n.php')
	&& !file_exists($_pg_doc_root . '/wp-includes/wp-l10n.php')
) {
	if (!function_exists('__')) {
		function __($text, $domain = null) { return $text; }
	}
	if (!function_exists('_e')) {
		function _e($text, $domain = null) { echo $text; }
	}
	if (!function_exists('esc_html__')) {
		function esc_html__($text, $domain = null) {
			return htmlspecialchars($text, ENT_QUOTES);
		}
	}
	if (!function_exists('esc_html_e')) {
		function esc_html_e($text, $domain = null) {
			echo htmlspecialchars($text, ENT_QUOTES);
		}
	}
}
?>
<?php
${v(`require_once ${o.phpVar(t)};
        if (
            isset($GLOBALS['wpdb']) &&
            method_exists($GLOBALS['wpdb'], 'reinitialize_sqlite')
        ) {
            $GLOBALS['wpdb']->reinitialize_sqlite();
        }`)}
${et}
if (PHP_MAJOR_VERSION < 7) {
	// E_DEPRECATED (8192) / E_STRICT (2048) are PHP 5.3+ symbols;
	// LEGACY_WP_ERROR_REPORTING_PHP_EXPR uses numeric literals.
	$level = ${L};
	error_reporting($level);
	ini_set('error_reporting', $level);
}

		`}const g=512,N="/__tar-root__",st=[40,181,47,253],rt=[80,75,3,4],y=new TextDecoder;function C(e){return st.every((t,i)=>e[i]===t)}function at(e){return rt.every((t,i)=>e[i]===t)}function q(e){const t=String(e);if(t.includes("\\"))throw new Error(`Unsafe tar entry (backslash path): ${e}`);if(t.startsWith("/"))throw new Error(`Unsafe tar entry (absolute path): ${e}`);const i=o.normalizePath(t);if(!i)return"";const n=o.resolvePathUnder(i,N);if(!n)throw new Error(`Unsafe tar entry (path traversal): ${e}`);return n.slice(N.length+1)}function W(e,t,i){const n=e.subarray(t,t+i);let s="";for(const r of n){if(r===0||r===32){if(s)break;continue}s+=String.fromCharCode(r)}return s?Number.parseInt(s,8):0}function A(e,t,i){let n=t;const s=t+i;for(;n<s&&e[n]!==0;)n+=1;return y.decode(e.subarray(t,n))}function U(e){for(let t=0;t<e.length;t+=1)if(e[t]!==0)return!1;return!0}function ot(e,t){const i=W(e,148,8);let n=0;for(let s=0;s<g;s+=1)n+=s>=148&&s<156?32:e[s];if(i!==n)throw new Error(`Malformed tar stream: invalid header checksum for ${t||"unknown"}`)}class G{constructor({onEntry:t}={}){this.leftover=new Uint8Array(0),this.state="header",this.entry=null,this.dataChunks=[],this.dataFilled=0,this.padRemaining=0,this.pendingLongName=null,this.pendingPax=null,this.zeroBlocks=0,this.maxBuffered=0,this.fileCount=0,this.dirCount=0,this.phpCount=0,this.bytesWritten=0,this.onEntry=t??(()=>{})}push(t){if(t!=null&&t.length)if(this.leftover.length===0)this.leftover=t;else{const i=new Uint8Array(this.leftover.length+t.length);i.set(this.leftover,0),i.set(t,this.leftover.length),this.leftover=i}this.track(),this.drain()}end(){var t,i,n,s;if(this.state==="data"&&this.dataFilled<(((t=this.entry)==null?void 0:t.size)??0))throw new Error(`Truncated tar stream: entry ${(i=this.entry)==null?void 0:i.name} expected ${(n=this.entry)==null?void 0:n.size} bytes, got ${this.dataFilled}`);if(this.state==="pad"&&this.padRemaining>0)throw new Error(`Truncated tar stream: entry ${((s=this.entry)==null?void 0:s.name)??"unknown"} missing ${this.padRemaining} padding bytes`);if(this.state==="header"&&this.leftover.length>0)throw new Error(`Truncated tar stream: incomplete header (${this.leftover.length} bytes)`);if(this.pendingLongName!==null)throw new Error("Malformed tar stream: GNU longlink entry was not followed by a file entry");if(this.pendingPax!==null)throw new Error("Malformed tar stream: PAX header was not followed by a file entry");if(this.zeroBlocks<2)throw new Error("Truncated tar stream: missing end-of-archive marker");return this.stats()}stats(){return{fileCount:this.fileCount,dirCount:this.dirCount,phpCount:this.phpCount,bytesWritten:this.bytesWritten,maxBuffered:this.maxBuffered}}track(t=0){const i=this.leftover.length+this.dataFilled+t;i>this.maxBuffered&&(this.maxBuffered=i)}drain(){let t=!0;for(;t;){if(t=!1,this.state==="header"){if(this.leftover.length<g)break;const i=this.leftover.subarray(0,g);if(this.leftover=this.leftover.subarray(g),U(i)){this.zeroBlocks+=1,this.zeroBlocks>=2&&(this.state="done"),t=!0;continue}this.zeroBlocks=0;const n=W(i,124,12),s=String.fromCharCode(i[156])||"0",r=A(i,0,100),a=A(i,345,155);if(ot(i,r),!Number.isFinite(n)||n<0)throw new Error(`Malformed tar stream: invalid size for ${r}`);this.entry={name:r,prefix:a,size:n,typeflag:s,isLongLink:s==="L",isPaxHeader:s==="x"},this.dataChunks=[],this.dataFilled=0,n>0?this.state="data":this.finishEntry(),t=!0;continue}if(this.state==="data"){const i=this.entry.size-this.dataFilled;if(i<=0){this.finishEntry();continue}if(this.leftover.length===0)break;const n=Math.min(i,this.leftover.length);this.dataChunks.push(this.leftover.subarray(0,n)),this.dataFilled+=n,this.leftover=this.leftover.subarray(n),this.track(),this.dataFilled===this.entry.size&&this.finishEntry(),t=!0;continue}if(this.state==="pad"){if(this.padRemaining===0){this.state="header",t=!0;continue}if(this.leftover.length===0)break;const i=Math.min(this.padRemaining,this.leftover.length);this.leftover=this.leftover.subarray(i),this.padRemaining-=i,t=!0;continue}if(this.state==="done"){if(this.leftover.length===0)break;if(!U(this.leftover))throw new Error("Malformed tar stream: non-zero data after end-of-archive marker");this.leftover=new Uint8Array(0),t=!0}}}concatData(){const t=new Uint8Array(this.dataFilled);let i=0;for(const n of this.dataChunks)t.set(n,i),i+=n.length;return t}finishEntry(){var l;const t=this.entry,i=this.concatData();this.dataChunks=[];const n=t.size%g;if(this.padRemaining=n===0?0:g-n,this.state=this.padRemaining>0?"pad":"header",this.dataFilled=0,this.entry=null,t.isLongLink){this.pendingLongName=y.decode(i).replace(/\0.*$/,"");return}if(t.isPaxHeader){this.pendingPax=ct(i);return}const s=((l=this.pendingPax)==null?void 0:l.path)??this.pendingLongName??(t.prefix?`${t.prefix}/${t.name}`:t.name);this.pendingLongName=null,this.pendingPax=null;const r=t.typeflag==="5"||s.endsWith("/"),a=q(s);if(a){if(r){this.dirCount+=1,this.onEntry({type:"dir",path:a});return}if(t.typeflag!=="0"&&t.typeflag!=="\0")throw new Error(`Unsupported tar entry type "${t.typeflag}" for ${s}`);this.fileCount+=1,this.bytesWritten+=i.length,a.endsWith(".php")&&(this.phpCount+=1),this.onEntry({type:"file",path:a,data:i})}}}async function M(e,t){const i=lt(e);if(typeof DecompressionStream<"u")try{const n=new DecompressionStream(t);return i.pipeThrough(n)}catch{}if(t==="zstd"){const{ZSTDDecoder:n}=await Promise.resolve().then(()=>require("./zstddec-stream.modern-D_CFiovn.cjs")),s=new n;await s.init();const r=await _t(e),a=s.decodeStreaming(r);let l=!1;return new ReadableStream({pull(_){if(l){_.close();return}try{for(;;){const{value:c,done:d}=a.next();if(d){l=!0,_.close();return}if(c.length!==0){_.enqueue(c);return}}}catch(c){l=!0,_.error(c)}},cancel(){var _;l=!0,(_=a.return)==null||_.call(a,void 0)}})}throw new Error(`No streaming decoder available for codec "${t}".`)}async function B(e,t,i,n={}){const{onProgress:s=()=>{},overwriteFiles:r=!0}=n,a=String(i).replace(/\/+$/,"")||"/",l=new Set,_=p=>{if(!p||l.has(p))return;t.mkdirTree(p);let u=p;for(;u&&!l.has(u);){l.add(u);const h=o.dirname(u);u=h&&h!==u?h:null}},c=new G({onEntry:p=>{var h;const u=o.joinPaths(a,p.path);if(p.type==="dir"){_(u);return}_(o.dirname(u)),!(!r&&((h=t.fileExists)!=null&&h.call(t,u)))&&(t.writeFile(u,p.data),c.fileCount%1e3===0&&s({fileCount:c.fileCount,bytes:c.bytesWritten}))}});_(a);const d=e.getReader();for(;;){const{done:p,value:u}=await d.read();if(p)break;u&&c.push(u)}return c.end()}function lt(e){return H(e)?e:new Blob([e]).stream()}async function _t(e){if(!H(e))return[e];const t=e.getReader(),i=[];try{for(;;){const{done:n,value:s}=await t.read();if(n)break;i.push(s)}return i}catch(n){throw await t.cancel(n).catch(()=>{}),n}finally{t.releaseLock()}}function H(e){return typeof ReadableStream<"u"&&e instanceof ReadableStream}function ct(e){const t=y.decode(e),i={};let n=0;for(;n<t.length;){const s=t.indexOf(" ",n);if(s===-1)throw new Error("Malformed tar stream: invalid PAX record length");const r=Number.parseInt(t.slice(n,s),10);if(!Number.isFinite(r)||r<=0)throw new Error("Malformed tar stream: invalid PAX record length");const a=n+r;if(a>t.length||t[a-1]!==`
`)throw new Error("Malformed tar stream: truncated PAX record");const l=t.slice(s+1,a-1),_=l.indexOf("=");if(_===-1)throw new Error("Malformed tar stream: invalid PAX record");i[l.slice(0,_)]=l.slice(_+1),n=a}return i}const X=`<?php

/**
 * Transforms the "wp-config.php" file.
 *
 * This parses the "wp-config.php" file contents into a token array and provides
 * methods to modify it and serialize it back to a string with the modifications.
 */
class WP_Config_Transformer {
	/**
	 * The tokens of the wp-config.php file.
	 *
	 * @var array<array|string>
	 */
	private $tokens;

	/**
	 * Constructor.
	 *
	 * @param string $content The contents of the wp-config.php file.
	 */
	public function __construct( string $content ) {
		$this->tokens = token_get_all( $content );

		// Check if the file is a valid PHP file.
		$is_valid_php_file = false;
		foreach ( $this->tokens as $token ) {
			if ( is_array( $token ) && T_OPEN_TAG === $token[0] ) {
				$is_valid_php_file = true;
				break;
			}
		}
		if ( ! $is_valid_php_file ) {
			throw new Exception( "The 'wp-config.php' file is not a valid PHP file." );
		}
	}

	/**
	 * Create a new config transformer instance from a file.
	 *
	 * @param string $path The path to the wp-config.php file.
	 * @return self        The new config transformer instance.
	 */
	public static function from_file( string $path ): self {
		if ( ! is_file( $path ) ) {
			throw new Exception( sprintf( "The '%s' file does not exist.", $path ) );
		}
		return new self( file_get_contents( $path ) );
	}

	/**
	 * Get the transformed wp-config.php file contents.
	 *
	 * @return string The transformed wp-config.php file contents.
	 */
	public function to_string(): string {
		$output = '';
		foreach ( $this->tokens as $token ) {
			$output .= is_array( $token ) ? $token[1] : $token;
		}
		return $output;
	}

	/**
	 * Save the transformed wp-config.php file contents to a file.
	 *
	 * @param string $path The path to the wp-config.php file.
	 */
	public function to_file( string $path ): void {
		$result = file_put_contents( $path, $this->to_string() );
		if ( false === $result ) {
			throw new Exception( sprintf( "Failed to write to the '%s' file.", $path ) );
		}
	}

	/**
	 * Check if a constant is defined in the wp-config.php file.
	 *
	 * @param  string $name The name of the constant.
	 * @return bool         True if the constant is defined, false otherwise.
	 */
	public function constant_exists( string $name ): bool {
		foreach ( $this->tokens as $i => $token ) {
			$is_string_token = is_array( $token ) && T_STRING === $token[0];
			if ( $is_string_token && 'define' === strtolower( $token[1] ) ) {
				$args       = $this->collect_function_call_argument_locations( $i );
				$const_name = $this->evaluate_constant_name(
					array_slice( $this->tokens, $args[0][0], $args[0][1] )
				);
				if ( $name === $const_name ) {
					return true;
				}
			}
		}
		return false;
	}

	/**
	 * Define a constant in the wp-config.php file.
	 *
	 * @param string $name  The name of the constant.
	 * @param mixed  $value The value of the constant.
	 */
	public function define_constant( string $name, $value ): void {
		// Tokenize the new constant value for insertion in the tokens array.
		$definition_tokens = token_get_all(
			sprintf(
				"<?php define( %s, %s );\\n",
				var_export( $name, true ),
				var_export( $value, true )
			)
		);

		// Full constant definition statement, e.g.: define( 'WP_DEBUG', true );\\n
		$define_tokens = array_slice( $definition_tokens, 1 );

		// The value of the constant, e.g.: "my-database-name"
		$value_tokens = array_slice( $definition_tokens, 7, -4 );

		// Collect all locations where the constant value needs to be updated.
		$updates = array();
		foreach ( $this->tokens as $i => $token ) {
			$is_string_token = is_array( $token ) && T_STRING === $token[0];
			if ( $is_string_token && 'define' === strtolower( $token[1] ) ) {
				$args       = $this->collect_function_call_argument_locations( $i );
				$const_name = $this->evaluate_constant_name(
					array_slice( $this->tokens, $args[0][0], $args[0][1] )
				);

				if ( $name === $const_name ) {
					$updates[] = $args[1];
				}
			}
		}

		// Modify the token array to define the constant. Apply updates in reverse
		// order, so splices at earlier positions don't shift indices after them.
		for ( $i = count( $updates ) - 1; $i >= 0; $i -= 1 ) {
			list ( $value_start, $value_length ) = $updates[ $i ];
			array_splice( $this->tokens, $value_start, $value_length, $value_tokens );
		}

		// If it's a new constant, inject it at the anchor location.
		if ( 0 === count( $updates ) ) {
			$anchor = $this->get_new_constant_location();
			array_splice( $this->tokens, $anchor, 0, $define_tokens );

			/*
			 * Ensure at least one newline (one "\\n") before the new constant.
			 * This must be done after inserting the constant definition in order
			 * to avoid shifting the anchor location when a new token is inserted.
			 */
			$this->ensure_newlines( $anchor - 1, 1 );
		}
	}

	/**
	 * Define multiple constants in the wp-config.php file.
	 *
	 * @param array<string, mixed> $constants An array of name-value pairs of constants to define.
	 */
	public function define_constants( array $constants ): void {
		foreach ( $constants as $name => $value ) {
			$this->define_constant( $name, $value );
		}
	}

	/**
	 * Inject code block into the wp-config.php file.
	 *
	 * @param string $code The code to inject.
	 */
	public function inject_code_block( string $code ): void {
		// Tokenize the injected code for insertion in the token array.
		$tokens      = token_get_all( sprintf( '<?php %s', trim( $code ) ) );
		$code_tokens = array_slice( $tokens, 1 );

		// Inject the code at the anchor location.
		$anchor = $this->get_injected_code_location();
		array_splice( $this->tokens, $anchor, 0, $code_tokens );

		/*
		 * Ensure empty line before and after the code block (at least two "\\n").
		 * This must be done after inserting the injected code, and the location
		 * AFTER must be updated prior to the location BEFORE, in order to avoid
		 * shifting the anchor location when a new token is inserted.
		 */
		$this->ensure_newlines( $anchor + count( $code_tokens ), 2 );
		$this->ensure_newlines( $anchor - 1, 2 );
	}

	/**
	 * Remove code block defined by two comment fragments from the wp-config.php file.
	 *
	 * @param string $from_comment_fragment A comment fragment from which to remove the code.
	 * @param string $to_comment_fragment   A comment fragment to which to remove the code.
	 */
	public function remove_code_block( string $from_comment_fragment, string $to_comment_fragment ): void {
		$start = $this->find_first_token_location( T_COMMENT, $from_comment_fragment );
		$end   = $this->find_first_token_location( T_COMMENT, $to_comment_fragment );
		if ( null === $start || null === $end ) {
			return;
		}

		// Remove the code, including the comment fragments.
		array_splice( $this->tokens, $start, $end - $start + 1 );

		// If previous and next tokens are whitespace, merge them.
		$prev = $this->tokens[ $start - 1 ];
		$next = $this->tokens[ $start ] ?? null;
		if (
			is_array( $prev ) && T_WHITESPACE === $prev[0]
			&& is_array( $next ) && T_WHITESPACE === $next[0]
		) {
			$this->tokens[ $start - 1 ][1] = $prev[1] . $next[1];
			array_splice( $this->tokens, $start, 1 );
		}

		// Remove up to two empty lines (before & after), keeping at least one.
		$token = $this->tokens[ $start - 1 ];
		if ( is_array( $token ) && T_WHITESPACE === $token[0] ) {
			$newlines = substr_count( $token[1], "\\n" );
			if ( $newlines > 2 ) {
				$limit = min( $newlines - 2, 4 );
				$value = $token[1];
				for ( $i = 0; $limit > 0; $i += 1 ) {
					if ( "\\n" === $value[ $i ] ) {
						$value  = substr_replace( $value, '', $i, 1 );
						$limit -= 1;
					}
				}
				$this->tokens[ $start - 1 ][1] = $value;
			}
		}
	}

	/**
	 * Parse arguments of a function call and collect their locations.
	 *
	 * @param  int $start             The location of the first token of the function call.
	 * @return array<array<int, int>> The arguments of the function call.
	 */
	private function collect_function_call_argument_locations( int $start ): array {
		// Find location of the opening parenthesis after the function name.
		$i = $start;
		while ( '(' !== $this->tokens[ $i ] ) {
			$i += 1;
		}
		$i += 1;

		// Collect all function call argument locations.
		$args         = array();
		$arg_start    = $this->skip_whitespace_and_comments( $i );
		$parens_level = 0;
		for ( $i = $arg_start; $i < count( $this->tokens ); $i += 1 ) {
			// Skip whitespace and comments, but preserve the index of the last
			// non-whitespace token to calculate the exact argument boundaries.
			$prev_i = $i;
			$i      = $this->skip_whitespace_and_comments( $i );
			$token  = $this->tokens[ $i ];

			if ( 0 === $parens_level && ( ',' === $token || ')' === $token ) ) {
				$args[] = array( $arg_start, $prev_i - $arg_start );
				if ( ',' === $token ) {
					// Start of the next argument.
					$arg_start = $this->skip_whitespace_and_comments( $i + 1 );
					$i         = $arg_start;
				} else {
					// End of the argument list.
					break;
				}
			} elseif ( '(' === $token || '[' === $token || '{' === $token ) {
				$parens_level += 1;
			} elseif ( ')' === $token || ']' === $token || '}' === $token ) {
				$parens_level -= 1;
			}
		}
		return $args;
	}

	/**
	 * Evaluate the constant name value from its tokens.
	 *
	 * @param  array $name_tokens The tokens containing the constant name.
	 * @return string|null        The evaluated constant name.
	 */
	private function evaluate_constant_name( array $name_tokens ): ?string {
		// Decide whether the array represents a constant name or an expression.
		$name_token = null;
		foreach ( $name_tokens as $token ) {
			if ( $this->is_whitespace( $token ) ) {
				continue;
			}
			if ( is_array( $token ) ) {
				if ( T_STRING === $token[0] || T_CONSTANT_ENCAPSED_STRING === $token[0] ) {
					$name_token = $token;
				} else {
					return null;
				}
			} elseif ( '(' !== $token && ')' !== $token ) {
				return null;
			}
		}

		if ( null === $name_token ) {
			return null;
		}

		// Get the constant name value.
		return eval( 'return ' . $name_token[1] . ';' );
	}

	/**
	 * Skip whitespace and comment tokens and return the location of the first
	 * non-whitespace and non-comment token after the specified start location.
	 *
	 * @param  int $start The start location in the token array.
	 * @return int        The location of the first non-whitespace and non-comment token.
	 */
	private function skip_whitespace_and_comments( int $start ): int {
		for ( $i = $start; $i < count( $this->tokens ); $i += 1 ) {
			if ( $this->is_whitespace( $this->tokens[ $i ] ) ) {
				continue;
			}
			break;
		}
		return $i;
	}

	/**
	 * Ensure minimum number of newlines are present at the given index.
	 *
	 * @param int $index The index of the token to ensure newlines.
	 * @param int $count The number of newlines that should be present.
	 */
	private function ensure_newlines( int $index, int $count ): void {
		$token = $this->tokens[ $index ] ?? null;
		if ( is_array( $token ) && ( T_WHITESPACE === $token[0] || T_OPEN_TAG === $token[0] ) ) {
			$newlines = substr_count( $token[1], "\\n" );
			if ( $newlines < $count ) {
				$this->tokens[ $index ][1] .= str_repeat( "\\n", $count - $newlines );
			}
		} else {
			$new_token = array( T_WHITESPACE, str_repeat( "\\n", $count ) );
			array_splice( $this->tokens, $index, 0, array( $new_token ) );
		}
	}

	/**
	 * Get the location to inject new constant definitions in the token array.
	 *
	 * @return int The location for new constant definitions in the token array.
	 */
	private function get_new_constant_location(): int {
		// First try to find the "That's all, stop editing!" comment.
		$anchor = $this->find_first_token_location( T_COMMENT, "That's all, stop editing!" );
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, try the "Absolute path to the WordPress directory." doc comment.
		$anchor = $this->find_first_token_location( T_DOC_COMMENT, 'Absolute path to the WordPress directory.' );
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, try the "Sets up WordPress vars and included files." doc comment.
		$anchor = $this->find_first_token_location( T_DOC_COMMENT, 'Sets up WordPress vars and included files.' );
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, try "require_once ABSPATH . 'wp-settings.php';".
		$anchor = $this->find_first_token_location( T_REQUIRE_ONCE );
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, fall back to the PHP opening tag.
		$open_tag_anchor = $this->find_first_token_location( T_OPEN_TAG );
		if ( null !== $open_tag_anchor ) {
			return $open_tag_anchor + 1;
		}

		// If we still don't have an anchor, the file is not a valid PHP file.
		throw new Exception( "The 'wp-config.php' file is not a valid PHP file." );
	}

	/**
	 * Get the location to inject new code in the token array.
	 *
	 * @return int The location for injected code in the token array.
	 */
	private function get_injected_code_location(): int {
		// First try to find the "/** Sets up WordPress vars and included files. */" comment.
		$anchor = $this->find_first_token_location( T_DOC_COMMENT, 'Sets up WordPress vars and included files.' );
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, try "require_once ABSPATH . 'wp-settings.php';".
		$anchor = $this->find_require_wp_settings_location();
		if ( null !== $anchor ) {
			return $anchor;
		}

		// If not found, fall back to the PHP opening tag.
		$open_tag_anchor = $this->find_first_token_location( T_OPEN_TAG );
		if ( null !== $open_tag_anchor ) {
			return $open_tag_anchor + 1;
		}

		// If we still don't have an anchor, the file is not a valid PHP file.
		throw new Exception( "The 'wp-config.php' file is not a valid PHP file." );
	}

	/**
	 * Find location of the "wp-settings.php" require statement in the token array.
	 *
	 * This method searches for the following statement:
	 *
	 *   require_once ABSPATH . 'wp-settings.php';
	 *
	 * @return int|null The location of the require statement.
	 */
	private function find_require_wp_settings_location(): ?int {
		$require_anchor = $this->find_first_token_location( T_REQUIRE_ONCE );
		if ( null === $require_anchor ) {
			return null;
		}

		$abspath = $this->tokens[ $require_anchor + 2 ] ?? null;
		$path    = $this->tokens[ $require_anchor + 6 ] ?? null;
		if (
			( is_array( $abspath ) && 'ABSPATH' === $abspath[1] )
			&& ( is_array( $path ) && "'wp-settings.php'" === $path[1] )
		) {
			return $require_anchor;
		}
		return null;
	}

	/**
	 * Find location of the first token of a given type in the token array.
	 *
	 * @param  int    $type   The type of the token.
	 * @param  string $search Optional. A search string to match against the token content.
	 * @return int|null       The location of the first token.
	 */
	private function find_first_token_location( int $type, ?string $search = null ): ?int {
		foreach ( $this->tokens as $i => $token ) {
			if ( is_array( $token ) && $type === $token[0] ) {
				if ( null === $search || false !== strpos( $token[1], $search ) ) {
					return $i;
				}
			}
		}
		return null;
	}

	/**
	 * Check if a token is whitespace or a comment.
	 *
	 * @param  array|string $token The token to check.
	 * @return bool                True if the token is whitespace or a comment.
	 */
	private function is_whitespace( $token ): bool {
		return is_array( $token )
			&& ( T_WHITESPACE === $token[0] || T_COMMENT === $token[0] || T_DOC_COMMENT === $token[0] );
	}
}
`;async function j(e,t){const i=o.joinPaths(t,"wp-config.php");!e.fileExists(i)&&e.fileExists(o.joinPaths(t,"wp-config-sample.php"))&&await e.writeFile(i,await e.readFileAsBuffer(o.joinPaths(t,"wp-config-sample.php"))),e.fileExists(i)&&await dt(e,i,{DB_NAME:"wordpress"})}async function ut(e,t,i){const n=o.phpVars({wpConfigPath:t,constants:i});if((await e.run({code:`${X}
		$wp_config_path = ${n.wpConfigPath};
		$transformer = WP_Config_Transformer::from_file($wp_config_path);
		$transformer->define_constants(${n.constants});
		$transformer->to_file($wp_config_path);
		`})).errors.length>0)throw new Error("Failed to rewrite constants in wp-config.php.")}async function dt(e,t,i){const n=Object.keys(i),s=o.phpVars({wpConfigPath:t,constantNames:n}),r=await e.run({code:`${X}
		$transformer = WP_Config_Transformer::from_file(${s.wpConfigPath});
		$missing = [];
		foreach (${s.constantNames} as $name) {
			if (!$transformer->constant_exists($name)) {
				$missing[] = $name;
			}
		}
		echo json_encode($missing);
		`});if(r.errors.length>0)throw new Error("Failed to check wp-config.php for constants.");let a;try{a=JSON.parse(r.text)}catch{throw new Error(`Failed to parse wp-config.php constant check output: ${r.text}`)}for(const l of a)await e.defineConstant(l,i[l])}async function E(e,{usesSqlite:t,hasCustomDatabasePath:i}){const n=await e.getPrimaryPhp();if(n.isFile("/internal/shared/preload/0-sqlite.php"))return;const s=o.joinPaths(e.documentRoot,"wp-content/mu-plugins/sqlite-database-integration");if(!n.isDir(s)&&!t&&!i&&!pt(n))throw new Error("Error connecting to the MySQL database.")}function pt(e){const t=o.joinPaths(e.documentRoot,"wp-config.php");if(!e.isFile(t))return!1;const i=e.readFileAsText(t),n=i.match(/define\s*\(\s*['"]DB_NAME['"]\s*,\s*['"]([^'"]*)['"]/),s=i.match(/define\s*\(\s*['"]DB_USER['"]\s*,\s*['"]([^'"]*)['"]/);return!n||!s?!1:n[1]!=="database_name_here"&&s[1]!=="username_here"}const V=["fsockopen","pfsockopen","curl_init","curl_exec","curl_multi_exec","mail"];function ft(e,t){var r,a;if(!f.isLegacyPHPVersion(t.phpVersion))return;const i=(((r=t.phpIniEntries)==null?void 0:r.disable_functions)??"").split(",").map(l=>l.trim()).filter(l=>l),s={disable_functions:Array.from(new Set([...i,...V])).join(","),allow_url_fopen:"0"};(a=t.phpIniEntries)!=null&&a["date.timezone"]||(s["date.timezone"]="UTC"),f.setPhpIniEntries(e,s)}async function ht(e,t){var a,l;const i=await e.getPrimaryPhp();if((a=t.hooks)!=null&&a.beforeWordPressFiles&&await t.hooks.beforeWordPressFiles(i),t.wordPressZip&&await O(i,await t.wordPressZip,{expectedFileCount:t.wordPressBundleFileCount}),t.constants)for(const _ in t.constants)i.defineConstant(_,t.constants[_]);i.defineConstant("WP_HOME",t.siteUrl),i.defineConstant("WP_SITEURL",t.siteUrl),await $t(i,e.documentRoot),await Le(i,e.documentRoot),(l=t.hooks)!=null&&l.beforeDatabaseSetup&&await t.hooks.beforeDatabaseSetup(i);let n=!1;t.sqliteIntegrationPluginZip&&(n=!0,await b(i,await t.sqliteIntegrationPluginZip,{phpVersion:t.phpVersion}),await gt(i,e.documentRoot));const s=t.wordpressInstallMode??"download-and-install",r=!!t.dataSqlPath;return(s==="download-and-install"||s==="install-from-existing-files"||s==="install-from-existing-files-if-needed")&&(await E(e,{usesSqlite:n,hasCustomDatabasePath:r}),await mt(i,e)),e}async function $t(e,t){const i=o.joinPaths(t,"wp-config.php"),n=o.joinPaths(t,"wp-config-sample.php");!e.fileExists(i)&&e.fileExists(n)&&await e.writeFile(i,await e.readFileAsBuffer(n))}async function gt(e,t){const i=o.joinPaths(t,"wp-content"),n=o.joinPaths(i,"db.php");e.isDir(i)&&!e.fileExists(n)&&await e.writeFile(n,Ze())}async function mt(e,t){try{await Tt(e)}catch(i){$.logger.warn("Legacy PHP WordPress installation error:",i)}await Je(e,t.absoluteUrl)}async function Tt(e){var r,a,l,_,c,d;const t=Lt(e,e.documentRoot);if(t!==null){const p=parseFloat(t);if(p<2.1)return;if(p<=3){await wt(e);return}}const i={disable_functions:V.join(","),allow_url_fopen:"0",error_reporting:String(we)},n=await f.withPHPIniValues(e,i,async()=>await e.request({url:"/wp-admin/install.php?step=2",method:"POST",body:{language:"en",prefix:"wp_",weblog_title:"My WordPress Website",user_name:"admin",admin_password:"password",admin_password2:"password",Submit:"Install WordPress",pw_weak:"1",admin_email:"admin@localhost.com"}}));if(!(((r=n.text)==null?void 0:r.includes("Success"))||((a=n.text)==null?void 0:a.includes("successful"))||((l=n.text)==null?void 0:l.includes("Finished"))||((_=n.text)==null?void 0:_.includes("Already Installed"))||((c=n.text)==null?void 0:c.includes("already have WordPress installed"))||!1))throw new Error(`Failed to install WordPress – installer responded with "${(d=n.text)==null?void 0:d.substring(0,100)}"`);await Et(e)}async function Et(e){try{(await e.run({code:`<?php
				$db_dir = getenv('DOCUMENT_ROOT') . '/wp-content/database/';
				$db_path = $db_dir . '.ht.sqlite';
				if (!file_exists($db_path)) { echo '0'; exit; }
				$pdo = new PDO('sqlite:' . $db_path);
				$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
				$nice_permalinks = '/%year%/%monthnum%/%day%/%postname%/';
				$stmt = $pdo->prepare(
					"UPDATE wp_options SET option_value = :val WHERE option_name = 'permalink_structure'"
				);
				$stmt->execute(array(':val' => $nice_permalinks));
				if ($stmt->rowCount() === 0) {
					$stmt = $pdo->prepare(
						"INSERT INTO wp_options (option_name, option_value, autoload) VALUES ('permalink_structure', :val, 'yes')"
					);
					$stmt->execute(array(':val' => $nice_permalinks));
				}
				$check = $pdo->query(
					"SELECT option_value FROM wp_options WHERE option_name = 'permalink_structure'"
				)->fetchColumn();
				echo $check === $nice_permalinks ? '1' : '0';
			`,env:{DOCUMENT_ROOT:e.documentRoot}})).text!=="1"&&$.logger.warn("Failed to default to pretty permalinks after WP install.")}catch{$.logger.warn("Failed to set pretty permalinks after WP install (non-fatal).")}}async function wt(e){try{await e.run({code:`<?php
				define('WP_INSTALLING', true);
				error_reporting(${L});
				ini_set('display_errors', '0');
				ob_start();
				require getenv('DOCUMENT_ROOT') . '/wp-load.php';
				ob_clean();
				if (file_exists(ABSPATH . 'wp-admin/includes/upgrade.php')) {
					require_once ABSPATH . 'wp-admin/includes/upgrade.php';
				} elseif (file_exists(ABSPATH . 'wp-admin/upgrade-functions.php')) {
					require_once ABSPATH . 'wp-admin/upgrade-functions.php';
				}
				if (function_exists('make_db_current_silent')) {
					make_db_current_silent();
				}
				// Seed essential options/roles when the loader exposes
				// them. The PDO fallback in runPostInstallLegacyFixups
				// backfills anything missing if either call dies.
				if (function_exists('populate_options')) populate_options();
				if (function_exists('populate_roles')) populate_roles();
				echo 'OK';
			`,env:{DOCUMENT_ROOT:e.documentRoot}})}catch(t){$.logger.warn("runDbDeltaOnly failed (non-fatal):",t)}}function Lt(e,t){const i=o.joinPaths(t,"wp-includes/version.php");if(!e.fileExists(i))return null;const s=e.readFileAsText(i).match(/\$wp_version\s*=\s*['"]([^'"]+)['"]/);return s?s[1]:null}async function yt(e){const t=await K(e);return await Y(t,e),t}async function Y(e,t){var a,l;if(f.isLegacyPHPVersion(t.phpVersion))return ht(e,t);const i=await e.getPrimaryPhp();if((a=t.hooks)!=null&&a.beforeWordPressFiles&&await t.hooks.beforeWordPressFiles(i),t.wordPressZip&&await O(i,await t.wordPressZip,{expectedFileCount:t.wordPressBundleFileCount}),t.constants)for(const _ in t.constants)i.defineConstant(_,t.constants[_]);t.dataSqlPath&&(i.defineConstant("DB_DIR",o.dirname(t.dataSqlPath)),i.defineConstant("DB_FILE",o.basename(t.dataSqlPath))),i.defineConstant("WP_HOME",t.siteUrl),i.defineConstant("WP_SITEURL",t.siteUrl),await j(i,e.documentRoot),(l=t.hooks)!=null&&l.beforeDatabaseSetup&&await t.hooks.beforeDatabaseSetup(i);let n=!1;t.sqliteIntegrationPluginZip&&(n=!0,await b(i,await t.sqliteIntegrationPluginZip,{phpVersion:t.phpVersion}),await Ee(i,e.documentRoot));const s=t.wordpressInstallMode??"download-and-install",r=!!t.dataSqlPath;if(["download-and-install","install-from-existing-files"].includes(s)){await E(e,{usesSqlite:n,hasCustomDatabasePath:r});try{await S(i)}catch(_){throw r||await T(e),_}r||await T(e)}else if(s==="install-from-existing-files-if-needed"){if(await E(e,{usesSqlite:n,hasCustomDatabasePath:r}),!await Q(i))try{await S(i)}catch(_){throw r||await T(e),_}r||await T(e)}return e}async function T(e){const t=await e.getPrimaryPhp();if(await Ot(t))return;if(t.isFile("/internal/shared/preload/0-sqlite.php"))throw new Error("Error connecting to the SQLite database.");const n=o.joinPaths(e.documentRoot,"wp-content/mu-plugins/sqlite-database-integration");throw t.isDir(n)?new Error("Error connecting to the SQLite database."):new Error("Error connecting to the MySQL database.")}async function K(e){bt(e);const t=e.spawnHandler??f.sandboxedSpawnHandlerFactory;async function i(s,r=!1){const a=await e.createPhpRuntime(r),l=new f.PHP(a);if(e.sapiName&&l.setSapiName(e.sapiName),s&&(l.requestHandler=s),e.phpIniEntries&&f.setPhpIniEntries(l,e.phpIniEntries),ft(l,{phpVersion:e.phpVersion,phpIniEntries:e.phpIniEntries}),l.defineConstant("WP_SQLITE_AST_DRIVER",!0),e.constants)for(const _ in e.constants)l.defineConstant(_,e.constants[_]);return r&&!l.isFile("/internal/.boot-files-written")&&(await ee(l,{phpVersion:e.phpVersion}),await f.writeFiles(l,"/",e.createFiles||{}),await te(l,o.joinPaths(new URL(e.siteUrl).pathname,"phpinfo.php")),await f.writeFiles(l,"/internal",{".boot-files-written":""})),t&&await l.setSpawnHandler(t(s?()=>s.instanceManager.acquirePHPInstance():void 0)),l.enableRuntimeRotation({recreateRuntime:e.createPhpRuntime,maxRequests:400}),e.onPHPInstanceCreated&&await e.onPHPInstanceCreated(l,{isPrimary:r}),l}const n=new f.PHPRequestHandler({documentRoot:e.documentRoot||"/wordpress",absoluteUrl:e.siteUrl,rewriteRules:J,pathAliases:e.pathAliases,getFileNotFoundAction:e.getFileNotFoundAction??z,cookieStore:e.cookieStore,php:e.maxPhpInstances===1?await i(void 0,!0):void 0,phpFactory:e.maxPhpInstances!==1?async({isPrimary:s})=>i(n,s):void 0,maxPhpInstances:e.maxPhpInstances});return n}function bt(e){"SQLITE_JOURNAL_MODE"in(e.constants??{})||(e.constants={...e.constants,SQLITE_JOURNAL_MODE:"DELETE"})}async function Q(e){return(await e.run({code:`<?php
			ob_start();
			$wp_load = getenv('DOCUMENT_ROOT') . '/wp-load.php';
			if (!file_exists($wp_load)) {
				echo '-1';
				exit;
			}
			require $wp_load;
			ob_clean();
			echo is_blog_installed() ? '1' : '0';
			ob_end_flush();
		`,env:{DOCUMENT_ROOT:e.documentRoot}})).text==="1"}async function S(e){var n;const t=await f.withPHPIniValues(e,{disable_functions:"fsockopen",allow_url_fopen:"0"},async()=>await e.request({url:"/wp-admin/install.php?step=2",method:"POST",body:{language:"en",prefix:"wp_",weblog_title:"My WordPress Website",user_name:"admin",admin_password:"password",admin_password2:"password",Submit:"Install WordPress",pw_weak:"1",admin_email:"admin@localhost.com"}}));if(!await Q(e))throw new Error(`Failed to install WordPress – installer responded with "${(n=t.text)==null?void 0:n.substring(0,100)}"`);(await e.run({code:`<?php
			ob_start();
			$wp_load = getenv('DOCUMENT_ROOT') . '/wp-load.php';
			if (!file_exists($wp_load)) {
				echo '0';
				exit;
			}
			require $wp_load;
			$nice_permalinks = '/%year%/%monthnum%/%day%/%postname%/';
			$option_result = update_option(
				'permalink_structure',
				$nice_permalinks
			);
			ob_clean();
			if ( get_option( 'permalink_structure' ) === $nice_permalinks ) {
				echo '1';
			} else {
				echo '0';
			}
			ob_end_flush();
		`,env:{DOCUMENT_ROOT:e.documentRoot}})).text!=="1"&&$.logger.warn("Failed to default to pretty permalinks after WP install.")}function z(e){return{type:"internal-redirect",uri:"/index.php"}}async function Ot(e){return(await e.run({code:`<?php
			ob_start();
			$wp_load = getenv('DOCUMENT_ROOT') . '/wp-load.php';
			if (!file_exists($wp_load)) {
				echo '-1';
				exit;
			}
			require $wp_load;
			ob_clean();
			echo $wpdb->check_connection( false ) ? '1' : '0';
			ob_end_flush();
		`,env:{DOCUMENT_ROOT:e.documentRoot}})).text==="1"}async function Pt(e){const{php:t,reap:i}=await e.instanceManager.acquirePHPInstance();try{const s=(await t.run({code:`<?php
				require '${e.documentRoot}/wp-includes/version.php';
				echo $wp_version;
			`})).text;if(!s)throw new Error("Unable to read loaded WordPress version.");return Z(s)}finally{i()}}function Z(e){if(/-(alpha|beta|RC)\d*-\d+$/.test(e))return"trunk";if(/-(beta|RC)\d*$/.test(e))return"beta";const n=e.match(/^(\d+\.\d+)(?:\.\d+)?$/);return n!==null?n[1]:e}const Nt="https://api.wordpress.org/core/stable-check/1.0/";let R;async function At(){R??(R=m.createMemoizedFetch(globalThis.fetch));const e=await R(Nt);if(!e.ok)throw new Error("Could not load the WordPress release catalog: "+`${e.status} ${e.statusText}`.trim());const t=await e.json();if(!t||typeof t!="object"||Array.isArray(t))throw new Error("The WordPress release catalog returned invalid data.");return Object.keys(t)}const J=[{match:new RegExp("^(/[_0-9a-zA-Z-]+)?(/wp-(content|admin|includes)/.*)"),replacement:"$2"}];async function ee(e,t={}){if(f.isLegacyPHPVersion(t.phpVersion))return ie(e);await e.mkdir("/internal/shared/mu-plugins"),await e.writeFile("/internal/shared/preload/env.php",`<?php

        // Allow adding filters/actions prior to loading WordPress.
        // $function_to_add MUST be a string.
        function playground_add_filter( $tag, $function_to_add, $priority = 10, $accepted_args = 1 ) {
            global $wp_filter;
            $wp_filter[$tag][$priority][$function_to_add] = array('function' => $function_to_add, 'accepted_args' => $accepted_args);
        }
        function playground_add_action( $tag, $function_to_add, $priority = 10, $accepted_args = 1 ) {
            playground_add_filter( $tag, $function_to_add, $priority, $accepted_args );
        }

        // Load our mu-plugins after customer mu-plugins
        // NOTE: this means our mu-plugins can't use the muplugins_loaded action!
        playground_add_action( 'muplugins_loaded', 'playground_load_mu_plugins', 0 );
        function playground_load_mu_plugins() {
            // Load all PHP files from /internal/shared/mu-plugins, sorted by filename
            $mu_plugins_dir = '/internal/shared/mu-plugins';
            if(!is_dir($mu_plugins_dir)){
                return;
            }
            $mu_plugins = glob( $mu_plugins_dir . '/*.php' );
            sort( $mu_plugins );
            foreach ( $mu_plugins as $mu_plugin ) {
                require_once $mu_plugin;
            }
        }
    `),await e.writeFile("/internal/shared/mu-plugins/1-auto-login.php",`<?php
		/**
		 * Returns the username to auto-login as, if any.
		 * @return string|false
		 */
		function playground_get_username_for_auto_login() {
			/**
			 * Allow users to auto-login as a specific user on their first visit.
			 *
			 * Prevent the auto-login if it already happened by checking for the
			 * playground_auto_login_already_happened cookie.
			 * This is used to allow the user to logout.
			 */
			if ( defined('PLAYGROUND_AUTO_LOGIN_AS_USER') && !isset($_COOKIE['playground_auto_login_already_happened']) ) {
				return PLAYGROUND_AUTO_LOGIN_AS_USER;
			}
			/**
			 * Allow users to auto-login as a specific user by passing the
			 * playground_force_auto_login_as_user GET parameter.
			 */
			if ( defined('PLAYGROUND_FORCE_AUTO_LOGIN_ENABLED') && isset($_GET['playground_force_auto_login_as_user']) ) {
				return $_GET['playground_force_auto_login_as_user'];
			}
			return false;
		}

		/**
		 * Logs the user in on their first visit if the Playground runtime told us to.
		 */
		function playground_auto_login() {
			/**
			 * The redirect should only run if the current PHP request is
			 * a HTTP request. If it's a PHP CLI run, we can't login the user
			 * because logins require cookies which aren't available in the CLI.
			 *
			 * Currently all Playground requests use the "cli" SAPI name
			 * to ensure support for WP-CLI, so the best way to distinguish
			 * between a CLI run and an HTTP request is by checking if the
			 * $_SERVER['REQUEST_URI'] global is set.
			 *
			 * If $_SERVER['REQUEST_URI'] is not set, we assume it's a CLI run.
			 */
			if (empty($_SERVER['REQUEST_URI'])) {
				return;
			}
			$user_name = playground_get_username_for_auto_login();
			if ( false === $user_name ) {
				return;
			}
			if (wp_doing_ajax() || defined('REST_REQUEST')) {
				return;
			}
			if ( is_user_logged_in() ) {
				return;
			}
			$user = get_user_by('login', $user_name);
			if (!$user) {
				return;
			}

			/**
			 * We're about to set cookies and redirect. It will log the user in
			 * if the headers haven't been sent yet.
			 *
			 * However, if they have been sent already – e.g. there a PHP
			 * notice was printed, we'll exit the script with a bunch of errors
			 * on the screen and without the user being logged in. This
			 * will happen on every page load and will effectively make Playground
			 * unusable.
			 *
			 * Therefore, we just won't auto-login if headers have been sent. Maybe
			 * we'll be able to finish the operation in one of the future requests
			 * or maybe not, but at least we won't end up with a permanent white screen.
			 */
			if (headers_sent()) {
				_doing_it_wrong('playground_auto_login', 'Headers already sent, the Playground runtime will not auto-login the user', '1.0.0');
				return;
			}

			/**
			 * This approach is described in a comment on
			 * https://developer.wordpress.org/reference/functions/wp_set_current_user/
			 */
			wp_set_current_user( $user->ID, $user->user_login );
			wp_set_auth_cookie( $user->ID );
			do_action( 'wp_login', $user->user_login, $user );

			setcookie('playground_auto_login_already_happened', '1');

			/**
			 * Confirm that nothing in WordPress, plugins, or filters have finalized
			 * the headers sending phase. See the comment above for more context.
			 */
			if (headers_sent()) {
				_doing_it_wrong('playground_auto_login', 'Headers already sent, the Playground runtime will not auto-login the user', '1.0.0');
				return;
			}

			/**
			 * Reload page to ensure the user is logged in correctly.
			 * WordPress uses cookies to determine if the user is logged in,
			 * so we need to reload the page to ensure the cookies are set.
			 */
			$redirect_url = $_SERVER['REQUEST_URI'];

			/**
			 * Intentionally do not use wp_redirect() here. It removes
			 * %0A and %0D sequences from the URL, which we don't want.
			 * There are valid use-cases for encoded newlines in the query string,
			 * for example html-api-debugger accepts markup with newlines
			 * encoded as %0A via the query string.
			 */
			header( "Location: $redirect_url", true, 302 );
			exit;
		}
		/**
		 * Autologin users from the wp-login.php page.
		 *
		 * The wp hook isn't triggered on
		 **/
		add_action('init', 'playground_auto_login', 1);

		/**
		 * Use an intermediate redirection step to ensure the login cookies
		 * are set before we redirecting to the landing page.
		 *
		 * /wp-admin/customize.php, and potentially other pages in WordPress,
		 * run authorization checks before running the init hook. If they're
		 * set as the landing page of the Blueprint, the user will be redirected
		 * to wp-login.php?reauth=1 before we have a chance to set the
		 * authorization cookie.
		 *
		 * To avoid this, we redirect to an intermediate page that will
		 * redirect the user to the landing page.
		 */
		function playground_auto_login_redirect_target() {
			if(strpos($_SERVER['REQUEST_URI'], '?playground-redirection-handler') !== false) {
				$next = $_GET['next'];
				header('Location: ' . $next, true, 302);
				exit;
			}
		}
		add_action('init', 'playground_auto_login_redirect_target', 1);

		/**
		 * Disable the Site Admin Email Verification Screen for any session started
		 * via autologin.
		 */
		add_filter('admin_email_check_interval', function($interval) {
			if(false === playground_get_username_for_auto_login()) {
				return 0;
			}
			return $interval;
		});
		`),await x(e),await e.writeFile("/internal/shared/preload/error-handler.php",`<?php
		(function() {
			$playground_consts = [];
			if(file_exists('/internal/shared/consts.json')) {
				$playground_consts = @json_decode(file_get_contents('/internal/shared/consts.json'), true) ?: [];
				$playground_consts = array_keys($playground_consts);
			}
			set_error_handler(function($severity, $message, $file, $line) use($playground_consts) {
				/**
				 * Networking support in Playground registers a http_api_transports filter.
				 *
				 * This filter is deprecated, and no longer actively used, but is needed for wp_http_supports().
				 * @see https://core.trac.wordpress.org/ticket/37708
				 */
				if (
					strpos($message, "http_api_transports") !== false &&
					strpos($message, "since version 6.4.0 with no alternative available") !== false
				) {
					return;
				}
				/**
				 * Playground defines some constants upfront, and some of them may be redefined
				 * in wp-config.php. For example, SITE_URL or WP_DEBUG. This is expected and
				 * we want Playground constants to take priority without showing warnings like:
				 *
				 * Warning: Constant SITE_URL already defined in
				 */
				if (strpos($message, "already defined") !== false) {
					foreach($playground_consts as $const) {
						if(strpos($message, "Constant $const already defined") !== false) {
							return;
						}
					}
				}
				/**
				 * Don't complain about network errors when not connected to the network.
				 */
				if (
					(
						! defined('USE_FETCH_FOR_REQUESTS') ||
						! USE_FETCH_FOR_REQUESTS
					) &&
					strpos($message, "WordPress could not establish a secure connection to WordPress.org") !== false)
				{
					return;
				}
				return false;
			});
		})();`)}async function te(e,t="/phpinfo.php"){await e.writeFile("/internal/shared/preload/phpinfo.php",`<?php
    // Render PHPInfo if the requested page is /phpinfo.php
    if ( isset($_SERVER['REQUEST_URI']) && ${o.phpVar(t)} === $_SERVER['REQUEST_URI'] ) {
        phpinfo();
        exit;
    }
    `)}async function b(e,t,i={}){if(f.isLegacyPHPVersion(i.phpVersion))return tt(e,t,i);await e.isDir("/tmp/sqlite-database-integration")&&await e.rmdir("/tmp/sqlite-database-integration",{recursive:!0}),await e.mkdir("/tmp/sqlite-database-integration"),await m.unzipFile(e,t,"/tmp/sqlite-database-integration");const n="/internal/shared/sqlite-database-integration",s=`/tmp/sqlite-database-integration/${(await e.listFiles("/tmp/sqlite-database-integration"))[0]}`;await e.mv(s,n);const r=o.joinPaths(n,"wp-includes/sqlite/class-wp-sqlite-db.php");if(await e.fileExists(r)){const p=await e.readFileAsText(r),u=p.replace("private $allow_unsafe_unquoted_parameters","protected $allow_unsafe_unquoted_parameters");u!==p&&await e.writeFile(r,u)}await e.defineConstant("SQLITE_MAIN_FILE","1");const l=(await e.readFileAsText(o.joinPaths(n,"db.copy"))).replace("'{SQLITE_IMPLEMENTATION_FOLDER_PATH}'",o.phpVar(n)).replace("'{SQLITE_PLUGIN}'",o.phpVar(o.joinPaths(n,"load.php"))),_=o.joinPaths(await e.documentRoot,"wp-content/db.php"),c=`
		// Do not preload this if WordPress comes with a custom db.php file.
		if(file_exists(${o.phpVar(_)})) {
			return;
		}
		`,d="/internal/shared/mu-plugins/sqlite-database-integration.php";await e.writeFile(d,`<?php${c}?>`+l),await e.writeFile("/internal/shared/preload/0-sqlite.php",Ut(c,d)),await e.writeFile("/internal/shared/mu-plugins/sqlite-test.php",`<?php
		global $wpdb;
		if(!($wpdb instanceof WP_SQLite_DB)) {
			var_dump(isset($wpdb));
			die("SQLite integration not loaded " . get_class($wpdb));
		}
		`)}function Ut(e,t){return`<?php
	if(!function_exists('mysqli_connect')) {
		function mysqli_connect() {}
	}

	${e}

	${v(`require_once ${o.phpVar(t)};`)}

		`}async function O(e,t,i={}){e.mkdir("/tmp/unzipped-wordpress");const n=new Uint8Array(await t.slice(0,4).arrayBuffer());if(C(n)){const a=await M(t.stream(),"zstd"),l=await B(a,e,"/tmp/unzipped-wordpress");if(i.expectedFileCount!=null&&l.fileCount!==i.expectedFileCount)throw new Error(`WordPress core bundle file-count parity check failed: extracted ${l.fileCount} files, expected ${i.expectedFileCount}. The download may be truncated or corrupt.`)}else await m.unzipFile(e,t,"/tmp/unzipped-wordpress");e.fileExists("/tmp/unzipped-wordpress/wordpress.zip")&&await m.unzipFile(e,"/tmp/unzipped-wordpress/wordpress.zip","/tmp/unzipped-wordpress");let s=e.fileExists("/tmp/unzipped-wordpress/wordpress")?"/tmp/unzipped-wordpress/wordpress":e.fileExists("/tmp/unzipped-wordpress/build")?"/tmp/unzipped-wordpress/build":"/tmp/unzipped-wordpress";if(!e.fileExists(o.joinPaths(s,"wp-config-sample.php"))){const a=e.listFiles(s);if(a.length){const l=a[0];e.fileExists(o.joinPaths(s,l,"wp-config-sample.php"))&&(s=o.joinPaths(s,l))}}const r=(a,l,_)=>{if(_.isDir(a)&&_.isDir(l))for(const c of _.listFiles(a)){const d=o.joinPaths(a,c),p=o.joinPaths(l,c);r(d,p,_)}else{if(_.fileExists(l)){const c=a.replace(/^\/tmp\/unzipped-wordpress\//,"/");$.logger.warn(`Cannot unzip WordPress files at ${l}: ${c} already exists.`);return}_.mv(a,l)}};r(s,e.documentRoot,e),e.fileExists(s)&&e.rmdir(s,{recursive:!0}),!e.fileExists(o.joinPaths(e.documentRoot,"wp-config.php"))&&e.fileExists(o.joinPaths(e.documentRoot,"wp-config-sample.php"))&&e.writeFile(o.joinPaths(e.documentRoot,"wp-config.php"),e.readFileAsText(o.joinPaths(e.documentRoot,"/wp-config-sample.php")))}const St=m.createMemoizedFetch(fetch),Rt="https://github.com/WordPress/WordPress/archive/refs/heads/master.zip";async function xt(e="latest"){if(e===null)e="latest";else if(e.startsWith("https://")||e.startsWith("http://")){const n=await crypto.subtle.digest("SHA-1",new TextEncoder().encode(e)),s=Array.from(new Uint8Array(n)).map(r=>r.toString(16).padStart(2,"0")).join("");return{releaseUrl:e,version:"custom-"+s.substring(0,8),source:"inferred"}}else if(e==="trunk"||e==="nightly"){const n=new Date().toISOString().split("T")[0];return{releaseUrl:`${Rt}?ts=${n}`,version:"trunk",source:"inferred"}}let i=await(await St("https://api.wordpress.org/core/version-check/1.7/?channel=beta")).json();i=i.offers.filter(n=>n.response==="autoupdate");for(const n of i){if(e==="beta"&&(n.version.includes("beta")||n.version.includes("RC")))return{releaseUrl:n.download,version:n.version,source:"api"};if(e==="latest"&&!n.version.includes("beta")&&!n.version.includes("RC"))return{releaseUrl:n.download,version:n.version,source:"api"};if(n.version.substring(0,e.length)===e)return{releaseUrl:n.download,version:n.version,source:"api"}}return e.match(/^\d+\.\d+\.0$/)&&(e=e.split(".").slice(0,2).join(".")),{releaseUrl:`https://wordpress.org/wordpress-${e}.zip`,version:e,source:"inferred"}}exports.StreamingTarParser=G;exports.bootRequestHandler=K;exports.bootWordPress=Y;exports.bootWordPressAndRequestHandler=yt;exports.createDecodedTarStream=M;exports.defineWpConfigConstants=ut;exports.ensureWpConfig=j;exports.extractTarStreamToPhp=B;exports.getFileNotFoundActionForWordPress=z;exports.getLoadedWordPressVersion=Pt;exports.getWordPressStableVersions=At;exports.isZipBundle=at;exports.isZstdBundle=C;exports.preloadPhpInfoRoute=te;exports.preloadSqliteIntegration=b;exports.resolveWordPressRelease=xt;exports.sanitizeTarPath=q;exports.setupPlatformLevelMuPlugins=ee;exports.unzipWordPress=O;exports.versionStringToLoadedWordPressVersion=Z;exports.wordPressRewriteRules=J;
