// @wp-playground/cli/run-cli-CwVKmNog.js
import { logger as h, LogSeverity as G, errorLogPath as de } from "@php-wasm/logger";
import { isLegacyPHPVersion as W, consumeAPI as xe, AllPHPVersions as ze, isPHPNextVersion as Ge, ProcessIdAllocator as Ye, describeError as Ee, printDebugDetails as Qe, HttpCookieStore as Ze, FileLockManagerInMemory as Xe, StreamedPHPResponse as ue, SupportedPHPVersions as Y, createObjectPoolProxy as Je, exposeAPI as Ke, PHPResponse as et, exposeSyncAPI as tt } from "@php-wasm/universal";
import { resolveRemoteBlueprint as rt, resolveRuntimeConfiguration as j, resolveBlueprintV2WordPressSource as nt, assertBlueprintV2WordPressVersionCompatibility as ot, compileBlueprintV2 as it, isBlueprintBundle as Ie, BlueprintReflection as K, compileBlueprintV1 as Me, runBlueprintV1Steps as pe } from "@wp-playground/blueprints";
import { RecommendedPHPVersion as B, zipDirectory as st } from "@wp-playground/common";
import w, { existsSync as ee, rmSync as at, mkdirSync as X, readdirSync as lt } from "fs";
import { MessageChannel as $e, Worker as ce } from "worker_threads";
import { createNodeFsMountHandler as dt } from "@php-wasm/node";
import c, { basename as N, join as te } from "path";
import { exec as ke } from "child_process";
import { promisify as ut } from "util";
import Ce from "express";
import { Readable as pt } from "stream";
import { pipeline as ct } from "stream/promises";
import ft from "yargs";
import { NodeJsFilesystem as ht, OverlayFilesystem as mt, InMemoryFilesystem as wt, ZipFilesystem as yt } from "@wp-playground/storage";
import { EmscriptenDownloadMonitor as Ae, ProgressTracker as De } from "@php-wasm/progress";
import { resolveWordPressRelease as Be } from "@wp-playground/wordpress";
import T from "fs-extra";
import { createRequire as bt } from "module";
import H from "os";
import { startBridge as gt } from "@php-wasm/xdebug-bridge";
import { dir as Pt, setGracefulCleanup as vt } from "tmp-promise";
import { CLIOutput as St, removeTempDirSymlink as xt, createTempDirSymlink as fe, makeXdebugConfig as Et, DEFAULT_PATH_SKIPPINGS as he, clearXdebugIDEConfig as It, addXdebugIDEConfig as Mt } from "@php-wasm/cli-util";
import { createHash as $t } from "crypto";
import { PHPMYADMIN_INSTALL_PATH as me, getPhpMyAdminInstallSteps as kt, PHPMYADMIN_ENTRY_PATH as Ct } from "@wp-playground/tools";
import { jspi as we } from "wasm-feature-detect";
function Q(e) {
  const t = [];
  for (const r of e) {
    const o = r.split(":");
    if (o.length !== 2)
      throw new Error(`Invalid mount format: ${r}.
				Expected format: /host/path:/vfs/path.
				If your path contains a colon, e.g. C:\\myplugin, use the --mount-dir option instead.
				Example: --mount-dir C:\\my-plugin /wordpress/wp-content/plugins/my-plugin`);
    const [i, n] = o;
    if (!ee(i))
      throw new Error(`Host path does not exist: ${i}`);
    t.push({ hostPath: i, vfsPath: n });
  }
  return t;
}
function ye(e) {
  if (e.length % 2 !== 0)
    throw new Error("Invalid mount format. Expected: /host/path /vfs/path");
  const t = [];
  for (let r = 0; r < e.length; r += 2) {
    const o = e[r], i = e[r + 1];
    if (!ee(o))
      throw new Error(`Host path does not exist: ${o}`);
    t.push({
      hostPath: c.resolve(process.cwd(), o),
      vfsPath: i
    });
  }
  return t;
}
async function Fr(e, t) {
  for (const r of t)
    try {
      await e.mount(
        r.vfsPath,
        dt(r.hostPath)
      );
    } catch (o) {
      const i = o instanceof Error ? o.message : String(o);
      throw new Error(
        `Error mounting path ${r.hostPath} at ${r.vfsPath}: ${i}`,
        { cause: o }
      );
    }
}
const be = {
  step: "runPHP",
  code: {
    filename: "activate-theme.php",
    // @TODO: Remove DOCROOT check after moving totally to Blueprints v2.
    content: `<?php
			$docroot = getenv('DOCROOT') ? getenv('DOCROOT') : '/wordpress';
			require_once "$docroot/wp-load.php";
			$theme = wp_get_theme();
			if (!$theme->exists()) {
				$themes = wp_get_themes();
				if (count($themes) > 0) {
					$themeName = array_keys($themes)[0];
					switch_theme($themeName);
				}
			}
		`
  }
};
function Te(e) {
  if (typeof e.autoMount != "string")
    return e;
  const t = e.autoMount, r = [...e.mount || []], o = [...e["mount-before-install"] || []], i = {
    ...e,
    mount: r,
    "mount-before-install": o,
    "additional-blueprint-steps": [
      ...e["additional-blueprint-steps"] || []
    ]
  };
  if (Tt(t)) {
    const s = `/wordpress/wp-content/plugins/${N(t)}`;
    r.push({
      hostPath: t,
      vfsPath: s,
      autoMounted: !0
    }), i["additional-blueprint-steps"].push({
      step: "activatePlugin",
      pluginPath: `/wordpress/wp-content/plugins/${N(t)}`
    });
  } else if (Bt(t)) {
    const n = N(t), s = `/wordpress/wp-content/themes/${n}`;
    r.push({
      hostPath: t,
      vfsPath: s,
      autoMounted: !0
    }), i["additional-blueprint-steps"].push({
      step: "activateTheme",
      themeFolderName: n
    });
  } else if (Dt(t)) {
    const n = w.readdirSync(t);
    for (const s of n)
      s !== "index.php" && r.push({
        hostPath: te(t, s),
        vfsPath: `/wordpress/wp-content/${s}`,
        autoMounted: !0
      });
    i["additional-blueprint-steps"].push(be);
  } else At(t) && (o.push({
    hostPath: t,
    vfsPath: "/wordpress",
    autoMounted: !0
  }), i.mode = "apply-to-existing-site", i["additional-blueprint-steps"].push(be), i.wordpressInstallMode || (i.wordpressInstallMode = "install-from-existing-files-if-needed"));
  return i;
}
function At(e) {
  const t = w.readdirSync(e);
  return t.includes("wp-admin") && t.includes("wp-includes") && t.includes("wp-content");
}
function Dt(e) {
  const t = w.readdirSync(e);
  return t.includes("themes") || t.includes("plugins") || t.includes("mu-plugins") || t.includes("uploads");
}
function Bt(e) {
  if (!w.readdirSync(e).includes("style.css"))
    return !1;
  const r = w.readFileSync(te(e, "style.css"), "utf8");
  return !!/^(?:[ \t]*<\?php)?[ \t/*#@]*Theme Name:(.*)$/im.exec(r);
}
function Tt(e) {
  const t = w.readdirSync(e), r = /^(?:[ \t]*<\?php)?[ \t/*#@]*Plugin Name:(.*)$/im;
  return !!t.filter((i) => i.endsWith(".php")).find((i) => {
    const n = w.readFileSync(te(e, i), "utf8");
    return !!r.exec(n);
  });
}
function Rt(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let r = 0; r < e.length; r += 2) {
    const o = e[r], i = e[r + 1];
    if (!o || !o.trim())
      throw new Error("Constant name cannot be empty");
    t[o.trim()] = i;
  }
  return t;
}
function Lt(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid boolean constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let r = 0; r < e.length; r += 2) {
    const o = e[r], i = e[r + 1].trim().toLowerCase();
    if (!o || !o.trim())
      throw new Error("Constant name cannot be empty");
    if (i === "true" || i === "1")
      t[o.trim()] = !0;
    else if (i === "false" || i === "0")
      t[o.trim()] = !1;
    else
      throw new Error(
        `Invalid boolean value for constant "${o}": "${i}". Must be "true", "false", "1", or "0".`
      );
  }
  return t;
}
function Wt(e) {
  if (e.length % 2 !== 0)
    throw new Error(
      "Invalid number constant definition format. Expected pairs of NAME value"
    );
  const t = {};
  for (let r = 0; r < e.length; r += 2) {
    const o = e[r], i = e[r + 1].trim();
    if (!o || !o.trim())
      throw new Error("Constant name cannot be empty");
    const n = Number(i);
    if (isNaN(n))
      throw new Error(
        `Invalid number value for constant "${o}": "${i}". Must be a valid number.`
      );
    t[o.trim()] = n;
  }
  return t;
}
function Vt(e = {}, t = {}, r = {}) {
  const o = {}, i = /* @__PURE__ */ new Set(), n = (s, a) => {
    for (const l in s) {
      if (i.has(l))
        throw new Error(
          `Constant "${l}" is defined multiple times across different --define-${a} flags`
        );
      i.add(l), o[l] = s[l];
    }
  };
  return n(e, "string"), n(t, "bool"), n(r, "number"), o;
}
function re(e) {
  return Vt(
    e.define,
    e["define-bool"],
    e["define-number"]
  );
}
const Ut = ut(ke);
function _t(e) {
  return new Promise((t) => {
    if (e === 0) return t(!1);
    const r = Ce().listen(e);
    r.once("listening", () => r.close(() => t(!1))), r.once(
      "error",
      (o) => t(o.code === "EADDRINUSE")
    );
  });
}
async function Ht(e) {
  const t = Ce(), r = await new Promise((s, a) => {
    const l = t.listen(e.port, () => {
      const u = l.address();
      u === null || typeof u == "string" ? a(new Error("Server address is not available")) : s(l);
    }).once("error", a);
  });
  t.use("/", async (s, a) => {
    try {
      const l = {
        url: s.url,
        headers: jt(s),
        method: s.method,
        body: await Ot(s)
      }, u = await e.handleRequest(l);
      await Ft(u, a);
    } catch (l) {
      h.error(l), a.headersSent || (a.statusCode = 500, a.end("Internal Server Error"));
    }
  });
  const i = r.address().port, n = process.env.CODESPACE_NAME;
  n && Nt(i, n);
  try {
    return await e.onBind(r, i);
  } catch (s) {
    throw r.listening && await new Promise((a) => {
      r.close(() => a()), r.closeAllConnections();
    }), s;
  }
}
async function Ft(e, t) {
  const [r, o] = await Promise.all([
    e.headers,
    e.httpStatusCode
  ]);
  t.statusCode = o;
  for (const n in r)
    t.setHeader(n, r[n]);
  const i = pt.fromWeb(e.stdout);
  try {
    await ct(i, t);
  } catch (n) {
    if (n instanceof Error && "code" in n && (n.code === "ERR_STREAM_PREMATURE_CLOSE" || n.code === "ERR_STREAM_UNABLE_TO_PIPE"))
      return;
    throw n;
  }
}
const Ot = async (e) => await new Promise((t) => {
  const r = [];
  e.on("data", (o) => {
    r.push(o);
  }), e.on("end", () => {
    t(new Uint8Array(Buffer.concat(r)));
  });
});
async function Nt(e, t) {
  h.log(`Publishing port ${e}...`);
  const r = `gh codespace ports visibility ${e}:public -c ${t}`;
  for (let o = 0; o < 10; o++)
    try {
      await Ut(r);
      return;
    } catch {
      await new Promise((i) => setTimeout(i, 2e3));
    }
}
const jt = (e) => {
  const t = {};
  if (e.rawHeaders && e.rawHeaders.length)
    for (let r = 0; r < e.rawHeaders.length; r += 2)
      t[e.rawHeaders[r].toLowerCase()] = e.rawHeaders[r + 1];
  return t;
};
function qt(e) {
  return /^latest$|^beta$|^trunk$|^nightly$|^(?:(\d+)\.(\d+)(?:\.(\d+))?)((?:-beta(?:\d+)?)|(?:-RC(?:\d+)?))?$/.test(e);
}
async function zt({
  sourceString: e,
  blueprintMayReadAdjacentFiles: t
}) {
  if (!e)
    return;
  if (e.startsWith("http://") || e.startsWith("https://"))
    return await rt(e);
  let r = c.resolve(process.cwd(), e);
  if (!w.existsSync(r))
    throw new Error(`Blueprint file does not exist: ${r}`);
  const o = w.statSync(r);
  if (o.isDirectory() && (r = c.join(r, "blueprint.json")), !o.isFile() && o.isSymbolicLink())
    throw new Error(
      `Blueprint path is neither a file nor a directory: ${r}`
    );
  const i = c.extname(r);
  switch (i) {
    case ".zip":
      return yt.fromArrayBuffer(
        w.readFileSync(r).buffer
      );
    case ".json": {
      const n = w.readFileSync(r, "utf-8");
      try {
        JSON.parse(n);
      } catch {
        throw new Error(
          `Blueprint file at ${r} is not a valid JSON file`
        );
      }
      const s = c.dirname(r), a = new ht(s);
      return new mt([
        new wt({
          "blueprint.json": n
        }),
        /**
         * Wrap the NodeJS filesystem to prevent access to local files
         * unless the user explicitly allowed it.
         */
        {
          read(l) {
            if (!t)
              throw new Error(
                `Error: Blueprint contained tried to read a local file at path "${l}" (via a resource of type "bundled"). Playground restricts access to local resources by default as a security measure. 

You can allow this Blueprint to read files from the same parent directory by explicitly adding the --blueprint-may-read-adjacent-files option to your command.`
              );
            return a.read(l);
          }
        }
      ]);
    }
    default:
      throw new Error(
        `Unsupported blueprint file extension: ${i}. Only .zip and .json files are supported.`
      );
  }
}
const J = c.join(H.homedir(), ".wordpress-playground");
async function Re(e = "trunk") {
  const t = typeof __dirname < "u" ? __dirname : import.meta.dirname;
  let r = c.join(t, "sqlite-database-integration.zip");
  if (!T.existsSync(r)) {
    const o = bt(import.meta.url), i = c.dirname(
      o.resolve("@wp-playground/wordpress-builds/package.json")
    );
    r = c.join(
      i,
      "src",
      "sqlite-database-integration",
      `sqlite-database-integration-${e}.zip`
    );
  }
  return new File([await T.readFile(r)], c.basename(r));
}
async function Le(e, t, r) {
  const o = c.join(J, t);
  return T.existsSync(o) || (T.ensureDirSync(J), await Gt(e, o, r)), We(o);
}
async function Gt(e, t, r) {
  const i = (await r.monitorFetch(fetch(e))).body.getReader(), n = `${t}.partial`, s = T.createWriteStream(n);
  for (; ; ) {
    const { done: a, value: l } = await i.read();
    if (l && s.write(l), a)
      break;
  }
  s.close(), s.closed || await new Promise((a, l) => {
    s.on("finish", () => {
      T.renameSync(n, t), a(null);
    }), s.on("error", (u) => {
      T.removeSync(n), l(u);
    });
  });
}
function We(e, t) {
  return new File([T.readFileSync(e)], N(e));
}
function Ve(e) {
  const t = [];
  e.intl && t.push("intl"), e.redis && t.push("redis"), e.memcached && t.push("memcached"), e.xdebug && t.push(
    typeof e.xdebug == "object" ? { name: "xdebug", options: e.xdebug } : "xdebug"
  );
  for (const r of e.phpExtension || [])
    t.push({
      source: {
        format: "manifest",
        manifestUrl: r
      }
    });
  return t;
}
class Yt {
  constructor(t, r) {
    this.args = t, this.siteUrl = r.siteUrl, this.cliOutput = r.cliOutput;
  }
  getWorkerType() {
    return "v2";
  }
  async bootWordPress(t, r) {
    let o;
    const i = await this.getWordPressInstallMode(), n = await this.getEffectiveBlueprint(), s = await j(
      n.declaration,
      {
        siteMode: O(i) ? "apply-to-existing-site" : "create-new-site"
      }
    );
    if (ge(s.phpVersion), i === "download-and-install" && (o = await nt(
      n.declaration,
      {
        streamBundledFile: n.streamBundledFile
      }
    ), !o)) {
      const l = new Ae();
      let u = !1;
      l.addEventListener("progress", (P) => {
        if (u)
          return;
        const { loaded: I, total: m } = P.detail, b = Math.floor(
          Math.min(100, 100 * I / m)
        );
        u = b === 100, this.cliOutput.updateProgress(
          "Downloading WordPress",
          b
        );
      });
      const g = await Be(
        s.wpVersion
      );
      o = await Le(
        g.releaseUrl,
        `${g.version}.zip`,
        l
      ), h.debug(
        `Resolved WordPress release URL: ${g?.releaseUrl}`
      );
    }
    if (O(i)) {
      const u = (await t.run({
        code: `<?php
					if (file_exists('/wordpress/wp-includes/version.php')) {
						require '/wordpress/wp-includes/version.php';
						echo $wp_version;
					}
				`
      })).text.trim();
      u && await ot(
        n.declaration,
        u
      );
    }
    let a;
    if (this.args.skipSqliteSetup || i === "do-not-attempt-installing")
      h.debug("Skipping SQLite integration plugin setup..."), a = void 0;
    else {
      this.cliOutput.updateProgress("Preparing SQLite database");
      const u = W(
        s.phpVersion
      ) ? "v3.0.0-rc.3-php52" : "trunk";
      a = await Re(u);
    }
    return this.cliOutput.updateProgress("Booting WordPress"), await t.bootWordPress(
      {
        phpVersion: s.phpVersion,
        siteUrl: this.siteUrl,
        wordpressInstallMode: i,
        networking: s.networking,
        wordPressZip: o && await o.arrayBuffer(),
        sqliteIntegrationPluginZip: await a?.arrayBuffer(),
        constants: Kt(
          this.args,
          s.phpVersion
        )
      },
      r
    ), t;
  }
  async bootRequestHandler({
    worker: t,
    fileLockManagerPort: r,
    nativeInternalDirPath: o
  }) {
    const i = xe(
      t.phpPort
    );
    await i.isConnected();
    const n = await this.getWordPressInstallMode(), s = await j(
      (await this.getEffectiveBlueprint()).declaration,
      {
        siteMode: O(n) ? "apply-to-existing-site" : "create-new-site"
      }
    );
    return ge(s.phpVersion), await i.useFileLockManager(r), await i.bootRequestHandler({
      phpVersion: s.phpVersion,
      siteUrl: this.siteUrl,
      networking: s.networking,
      mountsBeforeWpInstall: this.args["mount-before-install"] || [],
      mountsAfterWpInstall: this.args.mount || [],
      processId: t.processId,
      followSymlinks: this.args.followSymlinks === !0,
      trace: this.args.experimentalTrace === !0,
      extensions: Ve(
        Jt(
          {
            ...this.args,
            intl: s.intl
          },
          s.phpVersion
        )
      ),
      nativeInternalDirPath: o,
      pathAliases: this.args.pathAliases
    }), await i.isReady(), i;
  }
  async compileInputBlueprint(t) {
    const r = await this.getEffectiveBlueprint(
      t
    ), o = await this.getWordPressInstallMode(), i = new De();
    let n = "", s = !1;
    return i.addEventListener("progress", (l) => {
      if (s)
        return;
      s = l.detail.progress === 100;
      const u = Math.floor(l.detail.progress);
      n = l.detail.caption || n || "Running Blueprint", this.cliOutput.updateProgress(n.trim(), u);
    }), {
      ...await it(r.declaration, {
        progress: i,
        streamBundledFile: r.streamBundledFile,
        siteMode: O(o) ? "apply-to-existing-site" : "create-new-site"
      }),
      declaration: r.declaration
    };
  }
  /**
   * Resolves the worker install mode, including migrated v1 PHP-only sites.
   */
  getWordPressInstallMode() {
    return this.wordpressInstallModePromise ?? (this.wordpressInstallModePromise = Qt(
      this.args.blueprint
    ).then(
      (t) => Zt(this.args, t)
    )), this.wordpressInstallModePromise;
  }
  /**
   * Returns a v2 declaration after applying CLI-level defaults.
   *
   * Bundled Blueprints keep a `streamBundledFile` callback because v2 lowering
   * still needs the original archive to resolve bundled resources.
   */
  async getEffectiveBlueprint(t = []) {
    const r = this.args.blueprint || { version: 2 };
    if (Ie(r)) {
      const o = r, i = await K.create(o);
      return {
        declaration: Pe(
          ve(i.getDeclaration()),
          this.args,
          t
        ),
        streamBundledFile: (n) => o.read(n)
      };
    }
    return {
      declaration: Pe(
        ve(r),
        this.args,
        t
      ),
      streamBundledFile: void 0
    };
  }
}
async function Qt(e) {
  if (!e)
    return !1;
  const t = await K.create(e), r = t.getDeclaration();
  return t.getVersion() === 1 ? r.preferredVersions?.wp === !1 : r.wordpressVersion === "none";
}
function Zt(e, t = !1) {
  if (t) {
    if (e.mode && e.mode !== "mount-only")
      throw new Error(
        'Conflicting options: WordPress was requested, but the Blueprint sets `wordpressVersion: "none"`. Pick one.'
      );
    return "do-not-attempt-installing";
  }
  switch (e.mode) {
    case "apply-to-existing-site":
      return "install-from-existing-files-if-needed";
    case "mount-only":
      return "do-not-attempt-installing";
    case "create-new-site":
      return "download-and-install";
  }
  return e.wordpressInstallMode || "download-and-install";
}
function O(e) {
  return e === "install-from-existing-files" || e === "install-from-existing-files-if-needed";
}
function ge(e) {
  if (e === "next")
    throw new Error(
      'Blueprint v2 phpVersion "next" is supported by the web runtime only and cannot be used by Playground CLI. Choose a stable PHP version such as 8.3.'
    );
}
function Pe(e, t, r = []) {
  const o = { ...e };
  o.phpVersion === void 0 && (o.phpVersion = t.php || B), o.wordpressVersion === void 0 && (o.wordpressVersion = t.wp || "latest");
  const i = o.applicationOptions?.["wordpress-playground"];
  return i?.login === void 0 && t.login === !0 && (o.applicationOptions = {
    ...o.applicationOptions || {},
    "wordpress-playground": {
      ...i || {},
      login: t.login === !0
    }
  }), r.length > 0 && (o.additionalStepsAfterExecution = [
    ...o.additionalStepsAfterExecution || [],
    ...Xt(r)
  ]), o;
}
function Xt(e) {
  return e.flatMap((t) => t?.step === "activateTheme" && t.themeFolderName ? [
    {
      step: "activateTheme",
      themeDirectoryName: t.themeFolderName
    }
  ] : [t]);
}
function Jt(e, t) {
  return W(t) ? {
    ...e,
    intl: !1,
    redis: !1,
    memcached: !1,
    xdebug: !1
  } : e;
}
function Kt(e, t) {
  const r = { ...e["define-bool"] || {} };
  if (W(t))
    for (const o of e.defaultedDebugConstants || [])
      delete r[o];
  return re({
    ...e,
    "define-bool": r
  });
}
function ve(e) {
  if (e?.version === 2)
    return e;
  throw new Error(
    "The native Blueprint v2 handler requires a Blueprint v2 declaration."
  );
}
class er {
  constructor(t, r) {
    this.args = t, this.siteUrl = r.siteUrl, this.cliOutput = r.cliOutput;
  }
  getWorkerType() {
    return "v1";
  }
  async bootWordPress(t, r) {
    let o, i, n;
    const s = new Ae();
    if (this.args.wordpressInstallMode === "download-and-install") {
      let u = !1;
      s.addEventListener("progress", (g) => {
        if (u)
          return;
        const { loaded: P, total: I } = g.detail, m = Math.floor(
          Math.min(100, 100 * P / I)
        );
        u = m === 100, this.cliOutput.updateProgress(
          "Downloading WordPress",
          m
        );
      }), o = await Be(this.args.wp), n = c.join(
        J,
        `prebuilt-wp-content-for-wp-${o.version}.zip`
      ), i = w.existsSync(n) ? We(n) : await Le(
        o.releaseUrl,
        `${o.version}.zip`,
        s
      ), h.debug(
        `Resolved WordPress release URL: ${o?.releaseUrl}`
      );
    }
    let a;
    if (this.args.skipSqliteSetup)
      h.debug("Skipping SQLite integration plugin setup..."), a = void 0;
    else {
      this.cliOutput.updateProgress("Preparing SQLite database");
      const u = this.args.php || B, P = W(u) ? "v3.0.0-rc.3-php52" : "trunk";
      a = await Re(P);
    }
    this.cliOutput.updateProgress("Booting WordPress");
    const l = await j(
      this.getEffectiveBlueprint()
    );
    return await t.bootWordPress(
      {
        phpVersion: l.phpVersion,
        wpVersion: l.wpVersion,
        siteUrl: this.siteUrl,
        wordpressInstallMode: this.args.wordpressInstallMode || "download-and-install",
        wordPressZip: i && await i.arrayBuffer(),
        sqliteIntegrationPluginZip: await a?.arrayBuffer(),
        constants: re(this.args)
      },
      r
    ), n && !this.args["mount-before-install"] && !w.existsSync(n) && (this.cliOutput.updateProgress("Caching WordPress for next boot"), w.writeFileSync(
      n,
      // Comlink proxy is not assignable to UniversalPHP but
      // proxies all method calls transparently at runtime.
      await st(
        t,
        "/wordpress"
      )
    )), t;
  }
  async bootRequestHandler({
    worker: t,
    fileLockManagerPort: r,
    nativeInternalDirPath: o
  }) {
    const i = xe(
      t.phpPort
    );
    await i.isConnected();
    const n = await j(
      this.getEffectiveBlueprint()
    );
    return await i.useFileLockManager(r), await i.bootRequestHandler({
      phpVersion: n.phpVersion,
      siteUrl: this.siteUrl,
      mountsBeforeWpInstall: this.args["mount-before-install"] || [],
      mountsAfterWpInstall: this.args.mount || [],
      processId: t.processId,
      followSymlinks: this.args.followSymlinks === !0,
      trace: this.args.experimentalTrace === !0,
      extensions: Ve(this.args),
      nativeInternalDirPath: o,
      pathAliases: this.args.pathAliases
    }), await i.isReady(), i;
  }
  async compileInputBlueprint(t) {
    const r = this.getEffectiveBlueprint(), o = new De();
    let i = "", n = !1;
    return o.addEventListener("progress", (s) => {
      if (n)
        return;
      n = s.detail.progress === 100;
      const a = Math.floor(s.detail.progress);
      i = s.detail.caption || i || "Running Blueprint", this.cliOutput.updateProgress(i.trim(), a);
    }), await Me(r, {
      progress: o,
      additionalSteps: t
    });
  }
  getEffectiveBlueprint() {
    const t = this.args.blueprint;
    return Ie(t) ? t : {
      login: this.args.login,
      ...t || {},
      preferredVersions: {
        php: this.args.php ?? t?.preferredVersions?.php ?? B,
        wp: this.args.wp ?? t?.preferredVersions?.wp ?? "latest",
        ...t?.preferredVersions || {}
      }
    };
  }
}
async function tr(e, t = !0) {
  const o = `${c.basename(process.argv0)}${e}${process.pid}-`, i = await Pt({
    prefix: o,
    /*
     * Allow recursive cleanup on process exit.
     *
     * NOTE: I worried about whether this cleanup would follow symlinks
     * and delete target files instead of unlinking the symlink,
     * but this feature uses rimraf under the hood which respects symlinks:
     * https://github.com/raszi/node-tmp/blob/3d2fe387f3f91b13830b9182faa02c3231ea8258/lib/tmp.js#L318
     */
    unsafeCleanup: !0
  });
  return t && vt(), i;
}
async function rr(e, t, r) {
  const i = (await nr(
    e,
    t,
    r
  )).map(
    (n) => new Promise((s) => {
      w.rm(n, { recursive: !0 }, (a) => {
        a ? h.warn(
          `Failed to delete stale Playground temp dir: ${n}`,
          a
        ) : h.info(
          `Deleted stale Playground temp dir: ${n}`
        ), s();
      });
    })
  );
  await Promise.all(i);
}
async function nr(e, t, r) {
  try {
    const o = w.readdirSync(r).map((n) => c.join(r, n)), i = [];
    for (const n of o)
      await or(
        e,
        t,
        n
      ) && i.push(n);
    return i;
  } catch (o) {
    return h.warn(`Failed to find stale Playground temp dirs: ${o}`), [];
  }
}
async function or(e, t, r) {
  if (!w.lstatSync(r).isDirectory())
    return !1;
  const i = c.basename(r);
  if (!i.includes(e))
    return !1;
  const n = i.match(
    new RegExp(`^(.+)${e}(\\d+)-`)
  );
  if (!n)
    return !1;
  const s = {
    executableName: n[1],
    pid: n[2]
  };
  if (ir(s.pid))
    return !1;
  const a = Date.now() - t;
  return w.statSync(r).mtime.getTime() < a;
}
function ir(e, t) {
  try {
    return process.kill(Number(e), 0), !0;
  } catch (r) {
    const o = r, i = o && typeof o.code == "string" ? o.code : void 0;
    return i === "ESRCH" ? !1 : i === "EPERM" || i === "EACCES" ? (h.debug(
      `Permission denied while checking if process ${e} exists (code: ${i}).`,
      r
    ), !0) : (h.warn(
      `Could not determine if process ${e} exists due to unexpected error${i ? ` (code: ${i})` : ""}.`,
      r
    ), !0);
  }
}
function sr(e) {
  return process.env.CI === "true" || process.env.CI === "1" || process.env.GITHUB_ACTIONS === "true" || process.env.GITHUB_ACTIONS === "1" || (process.env.TERM || "").toLowerCase() === "dumb" ? !1 : e ? !!e.isTTY : process.stdout.isTTY;
}
class Ue extends St {
  constructor() {
    super(...arguments), this.lastProgressLine = "", this.progressActive = !1;
  }
  /**
   * Determines if progress updates should be rendered.
   *
   * Returns false when output is piped, redirected, or in CI environments.
   * This prevents progress spam in logs - users only see the final outcome.
   */
  get shouldRender() {
    return sr(this.writeStream);
  }
  printBanner() {
    if (this.isQuiet) return;
    const t = this.bold("WordPress Playground CLI");
    this.writeStream.write(`
${t}

`);
  }
  /**
   * Prints the configuration summary before starting the server.
   *
   * Displays PHP/WordPress versions, enabled extensions, all mounts
   * (with auto-mounts labeled), and any loaded blueprint. This gives
   * users a clear view of what's configured before the server boots.
   */
  printConfig(t) {
    if (this.isQuiet) return;
    const r = [];
    r.push(
      `${this.dim("PHP")} ${this.cyan(t.phpVersion)}  ${this.dim("WordPress")} ${this.cyan(t.wpVersion)}`
    );
    const o = [];
    if (t.intl && o.push("intl"), t.redis && o.push("redis"), t.memcached && o.push("memcached"), t.xdebug && o.push(this.yellow("xdebug")), o.length > 0 && r.push(`${this.dim("Extensions")} ${o.join(", ")}`), t.mounts.length > 0)
      for (const i of t.mounts) {
        const n = i.autoMounted ? ` ${this.dim("(auto-mount)")}` : "";
        r.push(
          `${this.dim("Mount")} ${i.hostPath} ${this.dim("→")} ${i.vfsPath}${n}`
        );
      }
    t.blueprint && r.push(`${this.dim("Blueprint")} ${t.blueprint}`), this.writeStream.write(r.join(`
`) + `

`);
  }
  /**
   * Starts a progress indicator that updates in-place.
   *
   * Subsequent updateProgress() calls rewrite the same line in TTY mode.
   * When output is piped or redirected, progress is completely skipped.
   */
  startProgress(t) {
    this.isQuiet || this.shouldRender && (this.progressActive = !0, this.updateProgress(t));
  }
  /**
   * Updates the current progress message and optional percentage.
   *
   * Rewrites the current line using ANSI cursor control in TTY mode.
   * Identical messages are skipped to prevent flickering. When piped,
   * this method does nothing (early return via shouldRender check).
   */
  updateProgress(t, r) {
    if (this.isQuiet || !this.shouldRender) return;
    this.progressActive || (this.progressActive = !0);
    let o = `${t}`;
    r !== void 0 && (o = `${t} ${this.dim(`${r}%`)}`), o !== this.lastProgressLine && (this.lastProgressLine = o, this.isTTY ? (this.writeStream.cursorTo(0), this.writeStream.write(o), this.writeStream.clearLine(1)) : this.writeStream.write(`${o}
`));
  }
  /**
   * Completes the progress indicator and moves to a new line.
   *
   * Optionally displays a final message before finishing. In TTY mode,
   * this ensures the cursor moves to the next line after the progress.
   */
  finishProgress(t) {
    this.isQuiet || this.shouldRender && (t && (this.isTTY ? (this.writeStream.cursorTo(0), this.writeStream.write(`${t}`), this.writeStream.clearLine(1)) : this.writeStream.write(`${t}
`)), this.isTTY && this.writeStream.write(`
`), this.progressActive = !1, this.lastProgressLine = "");
  }
  /**
   * Prints a status message, interrupting any active progress.
   *
   * Unlike progress updates, status messages are always printed on their
   * own line. Any active progress indicator is cleared before the message.
   */
  printStatus(t) {
    this.isQuiet || (this.progressActive && this.isTTY && (this.writeStream.cursorTo(0), this.writeStream.clearLine(0)), this.writeStream.write(`${t}
`), this.progressActive = !1, this.lastProgressLine = "");
  }
  /**
   * Prints an error message.
   *
   * Errors are always shown, even in quiet mode, and interrupt any
   * active progress display to ensure visibility.
   */
  printError(t) {
    this.progressActive && this.isTTY && (this.writeStream.cursorTo(0), this.writeStream.clearLine(0), this.progressActive = !1), this.writeStream.write(`${this.red("Error:")} ${t}
`);
  }
  /**
   * Prints the final "server ready" message with the URL.
   *
   * Note: The exact wording "WordPress is running on" is checked by
   * CI tests, so changes to this string will break test assertions.
   */
  printReady(t, r) {
    if (this.isQuiet) return;
    const o = r === 1 ? "worker" : "workers";
    this.writeStream.write(
      `
${this.green("Ready!")} WordPress is running on ${this.bold(t)} ${this.dim(`(${r} ${o})`)}

`
    );
  }
  /**
   * Prints the phpMyAdmin URL when the --phpmyadmin flag is enabled.
   */
  printPhpMyAdminUrl(t) {
    this.isQuiet || this.writeStream.write(
      `${this.cyan("phpMyAdmin")} available at ${this.bold(t)}

`
    );
  }
}
const ne = {
  Quiet: { name: "quiet", severity: G.Fatal },
  Normal: { name: "normal", severity: G.Info },
  Debug: { name: "debug", severity: G.Debug }
}, Se = ze.filter(
  (e) => !Ge(e)
);
class _ extends Error {
  constructor(t, r) {
    super(r), this.exitCode = t;
  }
}
async function ar(e) {
  try {
    const t = {
      "site-url": {
        describe: "Site URL to use for WordPress. Defaults to http://127.0.0.1:{port}",
        type: "string"
      },
      php: {
        describe: "PHP version to use.",
        type: "string",
        default: B,
        choices: Se
      },
      wp: {
        describe: "WordPress version to use.",
        type: "string",
        default: "latest"
      },
      define: {
        describe: 'Define PHP string constants (can be used multiple times). Format: NAME value. These constants are set via php.defineConstant() and only exist for the current request. Examples: --define API_KEY secret --define CON=ST "va=lu=e"',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: Rt
      },
      "define-bool": {
        describe: 'Define PHP boolean constants (can be used multiple times). Format: NAME value. Value must be "true", "false", "1", or "0". Examples: --define-bool WP_DEBUG true --define-bool MY_FEATURE false',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: Lt
      },
      "define-number": {
        describe: "Define PHP number constants (can be used multiple times). Format: NAME value. Examples: --define-number LIMIT 100 --define-number RATE 45.67",
        type: "string",
        nargs: 2,
        array: !0,
        coerce: Wt
      },
      // @TODO: Support read-only mounts, e.g. via WORKERFS, a custom
      // ReadOnlyNODEFS, or by copying the files into MEMFS
      mount: {
        describe: "Mount a directory to the PHP runtime (can be used multiple times). Format: /host/path:/vfs/path",
        type: "array",
        string: !0,
        nargs: 1,
        coerce: Q
      },
      "mount-before-install": {
        describe: "Mount a directory to the PHP runtime before WordPress installation (can be used multiple times). Format: /host/path:/vfs/path",
        type: "array",
        string: !0,
        nargs: 1,
        coerce: Q
      },
      "mount-dir": {
        describe: 'Mount a directory to the PHP runtime (can be used multiple times). Format: "/host/path" "/vfs/path"',
        type: "array",
        nargs: 2,
        array: !0,
        coerce: ye
      },
      "mount-dir-before-install": {
        describe: 'Mount a directory before WordPress installation (can be used multiple times). Format: "/host/path" "/vfs/path"',
        type: "string",
        nargs: 2,
        array: !0,
        coerce: ye
      },
      login: {
        describe: "Should log the user in",
        type: "boolean",
        default: !1
      },
      blueprint: {
        describe: "Blueprint to execute.",
        type: "string"
      },
      "blueprint-may-read-adjacent-files": {
        describe: 'Consent flag: Allow "bundled" resources in a local blueprint to read files in the same directory as the blueprint file.',
        type: "boolean",
        default: !1
      },
      "wordpress-install-mode": {
        describe: "Control how Playground prepares WordPress before booting.",
        type: "string",
        defaultDescription: "download-and-install",
        choices: [
          "download-and-install",
          "install-from-existing-files",
          "install-from-existing-files-if-needed",
          "do-not-attempt-installing"
        ]
      },
      "skip-wordpress-install": {
        describe: "[Deprecated] Use --wordpress-install-mode instead.",
        type: "boolean",
        hidden: !0
      },
      "skip-sqlite-setup": {
        describe: "Skip the SQLite integration plugin setup to allow the WordPress site to use MySQL.",
        type: "boolean",
        default: !1
      },
      // Hidden - Deprecated in favor of verbosity
      quiet: {
        describe: "Do not output logs and progress messages.",
        type: "boolean",
        default: !1,
        hidden: !0
      },
      verbosity: {
        describe: "Output logs and progress messages.",
        type: "string",
        choices: Object.values(ne).map(
          (d) => d.name
        ),
        default: "normal"
      },
      debug: {
        describe: "Print PHP error log content if an error occurs during Playground boot.",
        type: "boolean",
        default: !1,
        // Hide this deprecated option. Use verbosity=debug instead.
        hidden: !0
      },
      "auto-mount": {
        describe: "Automatically mount the specified directory. If no path is provided, mount the current working directory. You can mount a WordPress directory, a plugin directory, a theme directory, a wp-content directory, or any directory containing PHP and HTML files.",
        type: "string"
      },
      "follow-symlinks": {
        describe: `Allow Playground to follow symlinks by automatically mounting symlinked directories and files encountered in mounted directories. 
Warning: Following symlinks will expose files outside mounted directories to Playground and could be a security risk.`,
        type: "boolean",
        default: !1
      },
      "experimental-trace": {
        describe: "Print detailed messages about system behavior to the console. Useful for troubleshooting.",
        type: "boolean",
        default: !1,
        // Hide this option because we want to replace with a more general log-level flag.
        hidden: !0
      },
      "internal-cookie-store": {
        describe: "Enable internal cookie handling. When enabled, Playground will manage cookies internally using an HttpCookieStore that persists cookies across requests. When disabled, cookies are handled externally (e.g., by a browser in Node.js environments).",
        type: "boolean",
        default: !1
      },
      intl: {
        describe: "Enable Intl.",
        type: "boolean",
        default: !0
      },
      redis: {
        describe: "Enable Redis (requires JSPI support).",
        type: "boolean"
        // No default - will be determined at runtime based on JSPI availability
      },
      memcached: {
        describe: "Enable Memcached.",
        type: "boolean"
        // No default - will be determined at runtime based on JSPI availability
      },
      xdebug: {
        describe: "Enable Xdebug.",
        type: "boolean",
        default: !1
      },
      "php-extension": {
        describe: "Load a custom PHP.wasm extension manifest before PHP starts. Can be a local path, file: URL, or http(s) URL. Can be used multiple times.",
        type: "array",
        string: !0,
        nargs: 1
      },
      "experimental-unsafe-ide-integration": {
        describe: "Enable experimental IDE development tools. This option edits IDE config files to set Xdebug path mappings and web server details. CAUTION: If there are bugs, this feature may break your IDE config files. Please consider backing up your IDE configs before using this feature.",
        type: "string",
        // The empty value means the option is enabled for all
        // supported IDEs and, if needed, will create the relevant
        // config file for each.
        choices: ["", "vscode", "phpstorm"],
        coerce: (d) => d === "" ? ["vscode", "phpstorm"] : [d]
      },
      mode: {
        describe: "Choose whether Blueprint v2 creates a site, applies to an existing site, or only mounts files.",
        type: "string",
        choices: [
          "create-new-site",
          "apply-to-existing-site",
          "mount-only"
        ]
      },
      phpmyadmin: {
        describe: "Install phpMyAdmin for database management. The phpMyAdmin URL will be printed after boot. Optionally specify a custom URL path (default: /phpmyadmin).",
        type: "string",
        coerce: (d) => d === "" ? "/phpmyadmin" : d
      }
    }, r = {
      port: {
        describe: "Port to listen on when serving. Defaults to 9400 when available.",
        type: "number"
      },
      workers: {
        describe: 'Number of request-handling worker threads. Accepts a positive integer, or "auto" to use one per CPU core (minus one). Defaults to min(6, cpus-1).',
        type: "string",
        coerce: (d) => {
          if (d === void 0)
            return;
          if (d === "auto")
            return "auto";
          const y = Number(d);
          if (!Number.isInteger(y) || y < 1)
            throw new Error(
              `Invalid --workers value "${d}": expected a positive integer or "auto".`
            );
          return y;
        }
      },
      "experimental-multi-worker": {
        deprecated: "Use --workers=<n|auto> instead. The value of this flag is ignored.",
        describe: "Deprecated. Use --workers=<n|auto> to control the number of request-handling worker threads.",
        type: "number"
      },
      "experimental-devtools": {
        describe: "Enable experimental browser development tools.",
        type: "boolean"
      }
    }, o = {
      path: {
        describe: "Path to the project directory. Playground will auto-detect if this is a plugin, theme, wp-content, or WordPress directory.",
        type: "string",
        default: process.cwd()
      },
      php: {
        describe: "PHP version to use.",
        type: "string",
        default: B,
        choices: Se
      },
      wp: {
        describe: "WordPress version to use.",
        type: "string",
        default: "latest"
      },
      port: {
        describe: "Port to listen on. Defaults to 9400 when available.",
        type: "number"
      },
      blueprint: {
        describe: "Path to a Blueprint JSON file to execute on startup.",
        type: "string"
      },
      login: {
        describe: "Auto-login as the admin user.",
        type: "boolean",
        default: !0
      },
      xdebug: {
        describe: "Enable Xdebug for debugging.",
        type: "boolean",
        default: !1
      },
      "php-extension": t["php-extension"],
      "experimental-unsafe-ide-integration": t["experimental-unsafe-ide-integration"],
      "skip-browser": {
        describe: "Do not open the site in your default browser on startup.",
        type: "boolean",
        default: !1
      },
      quiet: {
        describe: "Suppress non-essential output.",
        type: "boolean",
        default: !1
      },
      // Advanced options for power users who need more control
      "site-url": {
        describe: "Override the site URL. By default, derived from the port (http://127.0.0.1:<port>).",
        type: "string"
      },
      mount: {
        describe: "Mount a directory to the PHP runtime (can be used multiple times). Format: /host/path:/vfs/path. Use this for additional mounts beyond auto-detection.",
        type: "array",
        string: !0,
        coerce: Q
      },
      reset: {
        describe: "Deletes the stored site directory and starts a new site from scratch.",
        type: "boolean",
        default: !1
      },
      "auto-mount": {
        describe: "Automatically detect project type (plugin, theme, wp-content, or WordPress) and mount accordingly. Use --no-auto-mount to disable and --mount to manually specify mounts instead.",
        type: "boolean",
        default: !0
      },
      // Define constants
      define: t.define,
      "define-bool": t["define-bool"],
      "define-number": t["define-number"],
      // Tools
      phpmyadmin: t.phpmyadmin
    }, i = {
      outfile: {
        describe: "When building, write to this output file.",
        type: "string",
        default: "wordpress.zip"
      }
    }, n = ft(e).usage("Usage: wp-playground <command> [options]").command(
      "start",
      "Start a local WordPress server with automatic project detection (recommended)",
      (d) => d.usage(
        `Usage: wp-playground start [options]

The easiest way to run WordPress locally. Automatically detects
if your directory contains a plugin, theme, wp-content, or
WordPress installation and configures everything for you.

Examples:
  wp-playground start                    # Start in current directory
  wp-playground start --path=./my-plugin # Start with a specific path
  wp-playground start --wp=6.7 --php=8.3 # Use specific versions
  wp-playground start --skip-browser     # Skip opening browser
  wp-playground start --no-auto-mount    # Disable auto-detection`
      ).options(o)
    ).command(
      "server",
      "Start a local WordPress server (advanced, low-level)",
      (d) => d.options({
        ...t,
        ...r
      })
    ).command(
      "run-blueprint",
      "Execute a Blueprint without starting a server",
      (d) => d.options({ ...t })
    ).command(
      "build-snapshot",
      "Build a ZIP snapshot of a WordPress site based on a Blueprint",
      (d) => d.options({
        ...t,
        ...i
      })
    ).command(
      "php",
      "Run a PHP script",
      (d) => d.options({ ...t })
    ).demandCommand(1, "Please specify a command").strictCommands().conflicts(
      "experimental-unsafe-ide-integration",
      "experimental-devtools"
    ).showHelpOnFail(!1).fail((d, y, A) => {
      throw y || (d && d.includes("Please specify a command") ? (A.showHelp(), console.error(`
` + d)) : console.error(d), new _(1));
    }).strictOptions().check(async (d) => {
      if (d["skip-wordpress-install"] === !0 && (d["wordpress-install-mode"] = "do-not-attempt-installing", d.wordpressInstallMode = "do-not-attempt-installing"), d.phpmyadmin && d["skip-sqlite-setup"])
        throw new Error(
          "The --phpmyadmin option requires SQLite setup. Remove --skip-sqlite-setup to use phpMyAdmin."
        );
      if (d.wp !== void 0 && typeof d.wp == "string" && !qt(d.wp))
        try {
          new URL(d.wp);
        } catch {
          throw new Error(
            'Unrecognized WordPress version. Please use "latest", "beta", "trunk", "nightly", a URL, or a numeric version such as "6.2", "6.0.1", "6.2-beta1", or "6.2-RC1" (see --help for details).'
          );
        }
      const y = d["site-url"];
      if (typeof y == "string" && y.trim() !== "")
        try {
          new URL(y);
        } catch {
          throw new Error(
            `Invalid site-url "${y}". Please provide a valid URL (e.g., http://localhost:8080 or https://example.com)`
          );
        }
      const A = d.autoMount;
      if (d._[0] !== "start" && typeof A == "string" && A) {
        let k = !1;
        try {
          k = w.statSync(A).isDirectory();
        } catch {
          k = !1;
        }
        if (!k)
          throw new Error(
            `The specified --auto-mount path is not a directory: '${A}'.`
          );
      }
      const D = e.some(
        (k) => k === "--mode" || k.startsWith("--mode=")
      );
      return d.hasExplicitBlueprintsV2Mode = D, !0;
    });
    n.wrap(n.terminalWidth());
    const s = await n.argv, a = s._[0];
    if (![
      "start",
      "run-blueprint",
      "server",
      "build-snapshot",
      "php"
    ].includes(a))
      throw n.showHelp(), new _(1);
    const l = s.define || {}, u = s["define-bool"] || {}, g = s["define-number"] || {}, P = (d) => d in l || d in u || d in g, I = s.php || B, m = W(I), b = [];
    m || (P("WP_DEBUG") || (u.WP_DEBUG = !0, b.push("WP_DEBUG")), P("WP_DEBUG_LOG") || (u.WP_DEBUG_LOG = !0, b.push("WP_DEBUG_LOG")), P("WP_DEBUG_DISPLAY") || (u.WP_DEBUG_DISPLAY = !1, b.push("WP_DEBUG_DISPLAY")));
    const M = {
      ...s,
      define: l,
      "define-bool": u,
      "define-number": g,
      defaultedDebugConstants: b,
      command: a,
      mount: [
        ...s.mount || [],
        ...s["mount-dir"] || []
      ],
      "mount-before-install": [
        ...s["mount-before-install"] || [],
        ...s["mount-dir-before-install"] || []
      ]
    }, $ = await He(M);
    return typeof $ == "number" ? { exitCode: $ } : $ === void 0 ? { exitCode: 0 } : {
      [Symbol.asyncDispose]: () => $[Symbol.asyncDispose](),
      [oe]: { cliServer: $ }
    };
  } catch (t) {
    if (t instanceof _)
      return { exitCode: t.exitCode };
    const r = process.argv.includes("--debug");
    throw new Ue({
      verbosity: r ? "debug" : "normal",
      writeStream: process.stderr
    }).printError(Ee(t)), r && await Qe(t), t;
  }
}
function q(e, t) {
  return e.find(
    (r) => r.vfsPath.replace(/\/$/, "") === t.replace(/\/$/, "")
  );
}
function _e(e) {
  const t = Math.max(1, H.cpus().length - 1);
  return e === void 0 ? Math.min(6, t) : e === "auto" ? t : e;
}
const oe = Symbol("playground-cli-testing");
async function He(e) {
  let t;
  const r = e.internalCookieStore ? new Ze() : void 0, o = [], i = /* @__PURE__ */ new Map(), n = new Ue({
    verbosity: e.verbosity || "normal"
  }), s = e.hasExplicitBlueprintsV2Mode ?? e.mode !== void 0;
  if (e.command === "start" && (e = ur(e, n)), e.autoMount !== void 0 && (e.autoMount === "" && (e = { ...e, autoMount: process.cwd() }), e = Te(e)), e.quiet && (e.verbosity = "quiet", delete e.quiet), e.debug && (e.verbosity = "debug", delete e.debug), e.verbosity) {
    const m = Object.values(ne).find(
      (b) => b.name === e.verbosity
    ).severity;
    h.setSeverityFilterLevel(m);
  }
  e.intl || (e.intl = !0), e.redis === void 0 && (e.redis = await we()), e.memcached === void 0 && (e.memcached = await we()), W(e.php || B) && (e.intl = !1, e.redis = !1, e.memcached = !1, e.xdebug = !1), e.phpmyadmin && (e.phpmyadmin === !0 && (e.phpmyadmin = "/phpmyadmin"), e.pathAliases = [
    {
      urlPrefix: e.phpmyadmin,
      fsPath: me
    }
  ]), e.command === "server" && (n.printBanner(), n.printConfig({
    phpVersion: e.php || B,
    wpVersion: e.wp || "latest",
    port: e.port ?? 9400,
    xdebug: !!e.xdebug,
    intl: !!e.intl,
    redis: !!e.redis,
    memcached: !!e.memcached,
    mounts: [
      ...e.mount || [],
      ...e["mount-before-install"] || []
    ],
    blueprint: typeof e.blueprint == "string" ? e.blueprint : void 0
  }));
  const l = e.command === "server" ? e.port ?? 9400 : 0, u = new Xe();
  let g = !1, P = !0;
  const I = await Ht({
    port: e.port ? e.port : await _t(l) ? 0 : l,
    onBind: async (m, b) => {
      const M = "127.0.0.1", $ = `http://${M}:${b}`, d = e["site-url"] || $;
      e["experimental-multi-worker"] !== void 0 && h.warn(
        "--experimental-multi-worker is deprecated and its value is ignored. Use --workers=<n|auto> instead."
      );
      const y = _e(e.workers);
      if (y < 6) {
        const S = "Running fewer than 6 workers may increase the likelihood of deadlock due to workers blocking on file locks.";
        e.workers === void 0 ? h.warn(
          `The default worker count has been reduced to ${y} because this machine has only ${H.cpus().length} CPU(s). ` + S
        ) : h.warn(
          `Worker count (${y}) is below the recommended threshold (6). ` + S
        );
      }
      const A = "-playground-cli-site-", D = await tr(A);
      h.debug(`Native temp dir for VFS root: ${D.path}`);
      let k = !1;
      const L = async function() {
        k || (k = !0, await Promise.all(
          o.map(async (C) => {
            await i.get(C)?.dispose(), await C.worker.terminate();
          })
        ), m && await new Promise((C) => {
          m.close(C), m.closeAllConnections();
        }), await D.cleanup());
      };
      try {
        const S = "WP Playground CLI - Listen for Xdebug", C = ".playground-xdebug-root", V = c.join(process.cwd(), C);
        if (await xt(V), e.xdebug) {
          const p = {
            hostPath: c.join(".", c.sep, C),
            vfsPath: "/"
          }, x = e.php || B;
          if (Y.includes(x) && Y.indexOf(x) <= Y.indexOf("8.5"))
            await fe(
              D.path,
              V,
              process.platform
            ), e.xdebug = Et({
              cwd: process.cwd(),
              mounts: [
                p,
                ...e["mount-before-install"] || [],
                ...e.mount || []
              ],
              pathSkippings: [...he]
            }), n.print(
              n.bold("Xdebug configured successfully")
            ), n.print(
              n.highlight("Playground source root: ") + ".playground-xdebug-root" + n.italic(
                n.dim(
                  " – you can set breakpoints and preview Playground's VFS structure in there."
                )
              )
            );
          else if (e.experimentalUnsafeIdeIntegration) {
            await fe(
              D.path,
              V,
              process.platform
            );
            try {
              await It(
                S,
                process.cwd()
              );
              const f = typeof e.xdebug == "object" ? e.xdebug : {}, v = await Mt(
                {
                  name: S,
                  host: M,
                  port: b,
                  ides: e.experimentalUnsafeIdeIntegration,
                  cwd: process.cwd(),
                  mounts: [
                    p,
                    ...e["mount-before-install"] || [],
                    ...e.mount || []
                  ],
                  pathSkippings: [
                    ...he
                  ],
                  ideKey: f.ideKey || "WPPLAYGROUNDCLI"
                }
              ), F = e.experimentalUnsafeIdeIntegration, U = F.includes("vscode"), ae = F.includes("phpstorm"), le = Object.values(v);
              n.print(""), le.length > 0 ? (n.print(
                n.bold(
                  "Xdebug configured successfully"
                )
              ), n.print(
                n.highlight(
                  "Updated IDE config: "
                ) + le.join(" ")
              ), n.print(
                n.highlight(
                  "Playground source root: "
                ) + ".playground-xdebug-root" + n.italic(
                  n.dim(
                    " – you can set breakpoints and preview Playground's VFS structure in there."
                  )
                )
              )) : (n.print(
                n.bold(
                  "Xdebug configuration failed."
                )
              ), n.print(
                "No IDE-specific project settings directory was found in the current working directory."
              )), n.print(""), U && v.vscode && (n.print(
                n.bold(
                  "VS Code / Cursor instructions:"
                )
              ), n.print(
                "  1. Ensure you have installed an IDE extension for PHP Debugging"
              ), n.print(
                `     (The ${n.bold("PHP Debug")} extension by ${n.bold(
                  "Xdebug"
                )} has been a solid option)`
              ), n.print(
                "  2. Open the Run and Debug panel on the left sidebar"
              ), n.print(
                `  3. Select "${n.italic(
                  S
                )}" from the dropdown`
              ), n.print(
                ' 4. Click "start debugging"'
              ), n.print(
                "  5. Set a breakpoint. For example, in .playground-xdebug-root/wordpress/index.php"
              ), n.print(
                "  6. Visit Playground in your browser to hit the breakpoint"
              ), ae && n.print("")), ae && v.phpstorm && (n.print(
                n.bold("PhpStorm instructions:")
              ), n.print(
                `  1. Choose "${n.italic(
                  S
                )}" debug configuration in the toolbar`
              ), n.print(
                "  2. Click the debug button (bug icon)"
              ), n.print(
                "  3. Set a breakpoint. For example, in .playground-xdebug-root/wordpress/index.php"
              ), n.print(
                "  4. Visit Playground in your browser to hit the breakpoint"
              )), n.print("");
            } catch (f) {
              throw new Error("Could not configure Xdebug", {
                cause: f
              });
            }
          }
        }
        const Oe = c.dirname(D.path), Ne = 2 * 24 * 60 * 60 * 1e3;
        rr(
          A,
          Ne,
          Oe
        );
        const ie = c.join(
          D.path,
          "internal"
        );
        X(ie);
        const je = [
          "wordpress",
          "tools",
          // Note: These dirs are from Emscripten's "default dirs" list:
          // https://github.com/emscripten-core/emscripten/blob/f431ec220e472e1f8d3db6b52fe23fb377facf30/src/lib/libfs.js#L1400-L1402
          //
          // Any Playground process with multiple workers may assume
          // these are part of a shared filesystem, so let's recognize
          // them explicitly here.
          "tmp",
          "home"
        ];
        for (const p of je) {
          const x = (f) => f.vfsPath === `/${p}`;
          if (!(e["mount-before-install"]?.some(
            x
          ) || e.mount?.some(x))) {
            const f = c.join(
              D.path,
              p
            );
            X(f), e["mount-before-install"] === void 0 && (e["mount-before-install"] = []), e["mount-before-install"].unshift({
              vfsPath: `/${p}`,
              hostPath: f
            });
          }
        }
        if (e["mount-before-install"])
          for (const p of e["mount-before-install"])
            h.debug(
              `Mount before WP install: ${p.vfsPath} -> ${p.hostPath}`
            );
        if (e.mount)
          for (const p of e.mount)
            h.debug(
              `Mount after WP install: ${p.vfsPath} -> ${p.hostPath}`
            );
        typeof e.blueprint == "string" && (e.blueprint = await zt({
          sourceString: e.blueprint,
          blueprintMayReadAdjacentFiles: e["blueprint-may-read-adjacent-files"] === !0
        }));
        let R;
        const z = await lr(
          e,
          s
        );
        z && dr(
          e,
          s
        ), e.wordpressInstallMode === void 0 && (e.wordpressInstallMode = "download-and-install"), z ? R = new Yt(e, {
          siteUrl: d,
          cliOutput: n
        }) : R = new er(e, {
          siteUrl: d,
          cliOutput: n
        });
        const se = [], qe = R.getWorkerType();
        for (let p = 0; p < y; p++) {
          const x = Fe(qe, {
            onExit: (E) => {
              k || E !== 0 && h.error(
                `Worker ${p} exited with code ${E}
`
              );
            }
          }).then(
            async (E) => {
              o.push(E);
              const f = await pr(u), v = await R.bootRequestHandler({
                worker: E,
                fileLockManagerPort: f,
                nativeInternalDirPath: ie
              });
              return i.set(
                E,
                v
              ), [E, v];
            }
          );
          se.push(x), p === 0 && await x;
        }
        await Promise.all(se), t = Je(
          o.map(
            (p) => i.get(p)
          )
        );
        {
          const p = new $e(), x = p.port1, E = p.port2;
          if (await Ke(
            {
              applyPostInstallMountsToAllWorkers: async () => {
                for (const f of i.values())
                  await f.mountAfterWordPressInstall(
                    e.mount || []
                  );
              }
            },
            void 0,
            x
          ), await R.bootWordPress(
            t,
            E
          ), x.close(), g = !0, z) {
            const f = await R.compileInputBlueprint(
              e["additional-blueprint-steps"] || []
            );
            f && await f.run(t);
          } else {
            const f = await R.compileInputBlueprint(
              e["additional-blueprint-steps"] || []
            );
            f && await pe(
              f,
              t
            );
          }
          if (e.phpmyadmin && !await t.fileExists(
            `${me}/index.php`
          )) {
            const f = await kt(), v = await Me({ steps: f });
            await pe(v, t);
          }
          if (e.command === "build-snapshot") {
            await fr(t, e.outfile), n.printStatus(`Exported to ${e.outfile}`), await L();
            return;
          } else if (e.command === "run-blueprint") {
            n.finishProgress("Done"), await L();
            return;
          } else if (e.command === "php") {
            const f = [
              // @TODO: Import this from somewhere?
              // Hardcoding it feels fragile.
              "/internal/shared/bin/php",
              ...(e._ || []).slice(1)
            ], v = await t.cli(f), [F] = await Promise.all([
              v.exitCode,
              v.stdout.pipeTo(
                new WritableStream({
                  write(U) {
                    process.stdout.write(U);
                  }
                })
              ),
              v.stderr.pipeTo(
                new WritableStream({
                  write(U) {
                    process.stderr.write(U);
                  }
                })
              )
            ]);
            return await L(), F;
          }
        }
        if (n.finishProgress(), n.printReady($, y), e.phpmyadmin) {
          const p = c.join(
            e.phpmyadmin,
            Ct
          );
          n.printPhpMyAdminUrl(
            new URL(p, $).toString()
          );
        }
        return e.xdebug && e.experimentalDevtools && (await gt({
          phpInstance: t,
          phpRoot: "/wordpress",
          excludedPaths: ["/internal"]
        })).start(), {
          playground: t,
          server: m,
          serverUrl: $,
          [Symbol.asyncDispose]: L,
          [oe]: {
            workerThreadCount: y
          }
        };
      } catch (S) {
        let C = "";
        e.verbosity === "debug" && await t?.fileExists(de) && (C = await t.readFileAsText(de));
        try {
          await L();
        } catch (V) {
          h.warn(
            `Failed to clean up after a boot failure: ${Ee(
              V
            )}`
          );
        }
        throw e.verbosity === "debug" ? new Error(C, { cause: S }) : S;
      }
    },
    async handleRequest(m) {
      if (!g)
        return ue.forHttpCode(
          502,
          "WordPress is not ready yet"
        );
      if (P) {
        P = !1;
        const M = {
          "Content-Type": ["text/plain"],
          "Content-Length": ["0"],
          Location: [m.url]
        };
        return m.headers?.cookie?.includes(
          "playground_auto_login_already_happened"
        ) && (M["Set-Cookie"] = [
          "playground_auto_login_already_happened=1; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Path=/"
        ]), ue.fromPHPResponse(
          new et(302, M, new Uint8Array())
        );
      }
      r && (m = {
        ...m,
        headers: {
          ...m.headers,
          // While we have an internal cookie store, we
          // completely replace the incoming request's Cookie
          // header with the cookies from our store. This avoids
          // getting into a strange state where both browser and
          // server are managing cookies.
          cookie: r.getCookieRequestHeader()
        }
      });
      const b = await t.requestStreamed(m);
      if (r) {
        const M = await b.headers;
        r.rememberCookiesFromResponseHeaders(M), delete M["set-cookie"];
      }
      return b;
    }
  });
  return I && typeof I == "object" && e.command === "start" && !e.skipBrowser && cr(I.serverUrl), I;
}
async function lr(e, t) {
  return t ? !0 : e.blueprint ? (await K.create(e.blueprint)).getVersion() === 2 : !1;
}
function dr(e, t) {
  if (t) {
    const r = q(
      e["mount-before-install"] || [],
      "/wordpress"
    ) || q(e.mount || [], "/wordpress");
    if (typeof e.autoMount == "string" || r?.autoMounted)
      throw new Error(
        "The --mode option cannot be used with --auto-mount because --auto-mount automatically sets the mode."
      );
    if (e.wordpressInstallMode !== void 0)
      throw new Error(
        "The --wordpress-install-mode option cannot be used with the --mode option. Use one or the other."
      );
    if (e.skipSqliteSetup === !0)
      throw new Error(
        "The --skipSqliteSetup option is not supported in Blueprint V2 mode."
      );
    return;
  }
  e.wordpressInstallMode === "do-not-attempt-installing" && (e.mode = "mount-only");
}
function ur(e, t) {
  let r = { ...e, command: "server" };
  const o = e.autoMount !== !1;
  delete r.autoMount, delete r["auto-mount"], o && (r.autoMount = c.resolve(process.cwd(), r.path ?? ""), r = Te(r), delete r.autoMount);
  const i = q(
    r["mount-before-install"] || [],
    "/wordpress"
  ) || q(r.mount || [], "/wordpress");
  if (i?.autoMounted && r.mode === "create-new-site")
    throw new Error(
      "Cannot use --mode=create-new-site with an auto-detected WordPress directory. Use --no-auto-mount, or choose --mode=apply-to-existing-site or --mode=mount-only."
    );
  if (i) {
    if (t.print(
      `Site files stored at: ${i?.hostPath}`
    ), e.reset)
      throw t.print(""), t.print(
        t.red(
          "This site is not managed by Playground CLI and cannot be reset."
        )
      ), t.print(
        "(It's not stored in the ~/.wordpress-playground/sites/<site-id> directory.)"
      ), t.print(""), t.print(
        "You may still remove the site's directory manually if you wish."
      ), new _(
        1,
        "This site is not managed by Playground CLI and cannot be reset."
      );
  } else {
    const n = r.autoMount || process.cwd(), s = $t("sha256").update(n).digest("hex"), a = H.homedir(), l = c.join(
      a,
      ".wordpress-playground/sites",
      s
    );
    t.print(`Site files stored at: ${l}`), ee(l) && e.reset && (t.print("Resetting site..."), at(l, { recursive: !0, force: !0 })), X(l, { recursive: !0 }), r["mount-before-install"] = [
      ...r["mount-before-install"] || [],
      { vfsPath: "/wordpress", hostPath: l }
    ], r.wordpressInstallMode = lt(l).length === 0 ? (
      // Only download WordPress on the first run when the site directory is still
      // empty.
      "download-and-install"
    ) : (
      // After that, reuse the WordPress installation from the initial run.
      "install-from-existing-files-if-needed"
    );
  }
  return r;
}
const Z = new Ye();
function Fe(e, { onExit: t } = {}) {
  let r;
  return e === "v1" ? r = new ce(new URL("./worker-thread-v1.js", import.meta.url)) : r = new ce(new URL("./worker-thread-v2.js", import.meta.url)), new Promise((o, i) => {
    const n = Z.claim();
    r.once("message", function(a) {
      a.command === "worker-script-initialized" && o({
        processId: n,
        worker: r,
        phpPort: a.phpPort
      });
    }), r.once("error", function(a) {
      Z.release(n), console.error(a);
      const l = a?.message || (a ? String(a) : "unknown error"), u = new Error(
        `Worker failed to load. Original error: ${l}`,
        { cause: a }
      );
      i(u);
    });
    let s = !1;
    r.once("spawn", () => {
      s = !0;
    }), r.once("exit", (a) => {
      Z.release(n), s || i(new Error(`Worker exited before spawning: ${a}`)), t?.(a);
    });
  });
}
async function pr(e) {
  const { port1: t, port2: r } = new $e();
  return await tt(e, t), r;
}
function cr(e) {
  const t = H.platform();
  let r;
  switch (t) {
    case "darwin":
      r = `open "${e}"`;
      break;
    case "win32":
      r = `start "" "${e}"`;
      break;
    default:
      r = `xdg-open "${e}"`;
      break;
  }
  ke(r, (o) => {
    o && h.debug(`Could not open browser: ${o.message}`);
  });
}
async function fr(e, t) {
  await e.run({
    code: `<?php
		$zip = new ZipArchive();
		if(false === $zip->open('/tmp/build.zip', ZipArchive::CREATE | ZipArchive::OVERWRITE)) {
			throw new Exception('Failed to create ZIP');
		}
		$files = new RecursiveIteratorIterator(
			new RecursiveDirectoryIterator('/wordpress')
		);
		foreach ($files as $file) {
			echo $file . PHP_EOL;
			if (!$file->isFile()) {
				continue;
			}
			$zip->addFile($file->getPathname(), $file->getPathname());
		}
		$zip->close();

	`
  });
  const r = await e.readFileAsBuffer("/tmp/build.zip");
  w.writeFileSync(t, r);
}
const Or = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  CLIArgsValidationError: _,
  LogVerbosity: ne,
  internalsKeyForTesting: oe,
  mergeDefinedConstants: re,
  parseOptionsAndRunCLI: ar,
  resolveWorkerCount: _e,
  runCLI: He,
  spawnWorkerThread: Fe
}, Symbol.toStringTag, { value: "Module" }));
export {
  _ as C,
  ne as L,
  He as a,
  Fr as b,
  Or as c,
  oe as i,
  re as m,
  ar as p,
  _e as r,
  Fe as s
};
