// @wp-playground/cli/worker-thread-v1.js
import { loadNodeRuntime as w } from "@php-wasm/node";
import { EmscriptenDownloadMonitor as f } from "@php-wasm/progress";
import { exposeAPI as y, PHPWorker as b, consumeAPISync as g, consumeAPI as h, releaseApiProxy as I, sandboxedSpawnHandlerFactory as W } from "@php-wasm/universal";
import { sprintf as H } from "@php-wasm/util";
import { RecommendedPHPVersion as k } from "@wp-playground/common";
import { bootWordPress as R, bootRequestHandler as q } from "@wp-playground/wordpress";
import { rootCertificates as A } from "tls";
import { MessageChannel as _, parentPort as v } from "worker_threads";
import { b as n, s as M } from "./run-cli-CwVKmNog.js";
import { logger as S } from "@php-wasm/logger";
function x(r, e, ...t) {
  console.log(
    performance.now().toFixed(6).padStart(15, "0"),
    r.toString().padStart(16, "0"),
    H(e, ...t)
  );
}
class F extends b {
  constructor(e) {
    super(void 0, e), this.bootedRequestHandler = !1, this.bootedWordPress = !1;
  }
  /**
   * Call this method before boot() to use file locking.
   *
   * This method is separate from boot() to simplify the related Comlink.transferHandlers
   * setup – if an argument is a MessagePort, we're transferring it, not copying it.
   *
   * @see comlink-sync.ts
   * @see phpwasm-emscripten-library-file-locking-for-node.js
   */
  async useFileLockManager(e) {
    this.fileLockManager = await g(e);
  }
  async bootWordPress(e, t) {
    if (this.bootedWordPress)
      throw new Error("WordPress already booted");
    this.bootedWordPress = !0;
    const {
      siteUrl: o,
      phpVersion: s,
      wordpressInstallMode: u,
      wordPressZip: l,
      sqliteIntegrationPluginZip: d,
      dataSqlPath: P,
      constants: m
    } = e;
    try {
      await R(this.__internal_getRequestHandler(), {
        siteUrl: o,
        phpVersion: s,
        wordpressInstallMode: u,
        wordPressZip: l !== void 0 ? new File([l], "wordpress.zip") : void 0,
        sqliteIntegrationPluginZip: d !== void 0 ? new File(
          [d],
          "sqlite-integration-plugin.zip"
        ) : void 0,
        // TODO: Are these redundant creations?
        createFiles: {
          "/internal/shared/ca-bundle.crt": A.join(`
`)
        },
        phpIniEntries: {
          "openssl.cafile": "/internal/shared/ca-bundle.crt",
          "curl.cainfo": "/internal/shared/ca-bundle.crt",
          allow_url_fopen: "1",
          disable_functions: ""
        },
        dataSqlPath: P,
        constants: m
      });
      const a = h(t);
      await a.applyPostInstallMountsToAllWorkers(), a[I](), c();
    } catch (a) {
      throw p(a), a;
    }
  }
  async bootRequestHandler(e) {
    if (this.bootedRequestHandler)
      throw new Error("Playground already booted");
    this.bootedRequestHandler = !0;
    try {
      const t = await q({
        siteUrl: e.siteUrl,
        phpVersion: e.phpVersion,
        maxPhpInstances: 1,
        createPhpRuntime: V(
          e,
          this.fileLockManager
        ),
        onPHPInstanceCreated: async (s) => {
          await n(s, e.mountsBeforeWpInstall), this.bootedWordPress && await n(s, e.mountsAfterWpInstall);
        },
        sapiName: "cli",
        cookieStore: !1,
        pathAliases: e.pathAliases,
        spawnHandler: () => W(() => {
          let s = e;
          return this.bootedWordPress || (s = {
            ...e,
            mountsAfterWpInstall: []
          }), C(
            s,
            this.fileLockManager
          );
        })
      });
      this.__internal_setRequestHandler(t);
      const o = await t.getPrimaryPhp();
      await this.setPrimaryPHP(o), c();
    } catch (t) {
      throw p(t), t;
    }
  }
  async mountAfterWordPressInstall(e) {
    this.bootedWordPress = !0, await n(this.__internal_getPHP(), e);
  }
  // Provide a named disposal method that can be invoked via comlink.
  async dispose() {
    await this[Symbol.asyncDispose]();
  }
}
function V(r, e) {
  return async () => await w(
    r.phpVersion || k,
    {
      fileLockManager: e,
      emscriptenOptions: {
        processId: r.processId,
        trace: r.trace ? x : void 0,
        nativeInternalDirPath: r.nativeInternalDirPath
      },
      followSymlinks: r.followSymlinks,
      extensions: r.extensions
    }
  );
}
async function C(r, e) {
  const t = await M("v1"), o = h(
    t.phpPort
  );
  return o.useFileLockManager(e), await o.bootRequestHandler({
    ...r,
    processId: t.processId
  }), {
    php: o,
    reap: () => {
      try {
        o.dispose();
      } catch {
      }
      try {
        t.worker.terminate();
      } catch {
      }
    }
  };
}
process.on("unhandledRejection", (r) => {
  S.error("Unhandled rejection:", r);
});
const i = new _(), [c, p] = y(
  new F(new f()),
  void 0,
  i.port1
);
v?.postMessage(
  {
    command: "worker-script-initialized",
    phpPort: i.port2
  },
  [i.port2]
);
export {
  F as PlaygroundCliBlueprintV1Worker
};
