// @wp-playground/cli/worker-thread-v2.js
import { loadNodeRuntime as I } from "@php-wasm/node";
import { EmscriptenDownloadMonitor as b } from "@php-wasm/progress";
import { exposeAPI as k, PHPWorker as W, consumeAPISync as H, consumeAPI as P, releaseApiProxy as _, sandboxedSpawnHandlerFactory as R, exposeSyncAPI as q } from "@php-wasm/universal";
import { sprintf as A } from "@php-wasm/util";
import { RecommendedPHPVersion as x } from "@wp-playground/common";
import { bootWordPress as v, bootRequestHandler as M } from "@wp-playground/wordpress";
import { rootCertificates as d } from "tls";
import { MessageChannel as m, parentPort as S } from "worker_threads";
import { b as a, s as F } from "./run-cli-CwVKmNog.js";
import { logger as E } from "@php-wasm/logger";
function C(r, e, ...t) {
  console.log(
    performance.now().toFixed(6).padStart(15, "0"),
    r.toString().padStart(16, "0"),
    A(e, ...t)
  );
}
function p(r) {
  return {
    "openssl.cafile": "/internal/shared/ca-bundle.crt",
    "curl.cainfo": "/internal/shared/ca-bundle.crt",
    allow_url_fopen: r ? "1" : "0",
    disable_functions: r ? "" : "fsockopen,pfsockopen,curl_init,curl_exec,curl_multi_exec,mail"
  };
}
class L extends W {
  constructor(e) {
    super(void 0, e), this.bootedRequestHandler = !1, this.bootedWordPress = !1;
  }
  /**
   * Call this method before boot() to use file locking.
   *
   * This method is separate from boot() to simplify the related Comlink.transferHandlers
   * setup – if an argument is a MessagePort, we're transferring it, not copying it.
   *
   * @see comlink-sync.ts
   * @see phpwasm-emscripten-library-file-locking-for-node.js
   */
  async useFileLockManager(e) {
    this.fileLockManager = await H(e);
  }
  async bootWordPress(e, t) {
    if (this.bootedWordPress)
      throw new Error("WordPress already booted");
    this.bootedWordPress = !0;
    const {
      siteUrl: n,
      phpVersion: o,
      wordpressInstallMode: w,
      wordPressZip: l,
      networking: f = !0,
      sqliteIntegrationPluginZip: c,
      dataSqlPath: y,
      constants: g
    } = e;
    try {
      await v(this.__internal_getRequestHandler(), {
        siteUrl: n,
        phpVersion: o,
        wordpressInstallMode: w,
        wordPressZip: l !== void 0 ? new File([l], "wordpress.zip") : void 0,
        sqliteIntegrationPluginZip: c !== void 0 ? new File(
          [c],
          "sqlite-integration-plugin.zip"
        ) : void 0,
        // TODO: Are these redundant creations?
        createFiles: {
          "/internal/shared/ca-bundle.crt": d.join(`
`)
        },
        phpIniEntries: p(f),
        dataSqlPath: y,
        constants: g
      });
      const s = P(t);
      await s.applyPostInstallMountsToAllWorkers(), s[_](), h();
    } catch (s) {
      throw u(s), s;
    }
  }
  async bootRequestHandler(e) {
    if (this.bootedRequestHandler)
      throw new Error("Playground already booted");
    this.bootedRequestHandler = !0;
    try {
      const t = await M({
        siteUrl: e.siteUrl,
        phpVersion: e.phpVersion,
        maxPhpInstances: 1,
        createFiles: {
          "/internal/shared/ca-bundle.crt": d.join(`
`)
        },
        phpIniEntries: p(
          e.networking ?? !0
        ),
        createPhpRuntime: V(
          e,
          this.fileLockManager
        ),
        onPHPInstanceCreated: async (o) => {
          await a(o, e.mountsBeforeWpInstall), this.bootedWordPress && await a(o, e.mountsAfterWpInstall);
        },
        sapiName: "cli",
        cookieStore: !1,
        pathAliases: e.pathAliases,
        spawnHandler: () => R(() => {
          let o = e;
          return this.bootedWordPress || (o = {
            ...e,
            mountsAfterWpInstall: []
          }), j(
            o,
            this.fileLockManager
          );
        })
      });
      this.__internal_setRequestHandler(t);
      const n = await t.getPrimaryPhp();
      await this.setPrimaryPHP(n), h();
    } catch (t) {
      throw u(t), t;
    }
  }
  async mountAfterWordPressInstall(e) {
    this.bootedWordPress = !0, await a(this.__internal_getPHP(), e);
  }
  // Provide a named disposal method that can be invoked via comlink.
  async dispose() {
    await this[Symbol.asyncDispose]();
  }
}
function V(r, e) {
  return async () => await I(
    r.phpVersion || x,
    {
      fileLockManager: e,
      emscriptenOptions: {
        processId: r.processId,
        trace: r.trace ? C : void 0,
        nativeInternalDirPath: r.nativeInternalDirPath
      },
      followSymlinks: r.followSymlinks,
      extensions: r.extensions
    }
  );
}
async function j(r, e) {
  const t = await F("v2"), n = P(
    t.phpPort
  ), o = new m();
  return await q(e, o.port1), await n.useFileLockManager(o.port2), await n.bootRequestHandler({
    ...r,
    processId: t.processId
  }), {
    php: n,
    reap: () => {
      try {
        n.dispose();
      } catch {
      }
      try {
        t.worker.terminate();
      } catch {
      }
    }
  };
}
process.on("unhandledRejection", (r) => {
  E.error("Unhandled rejection:", r);
});
const i = new m(), [h, u] = k(
  new L(new b()),
  void 0,
  i.port1
);
S?.postMessage(
  {
    command: "worker-script-initialized",
    phpPort: i.port2
  },
  [i.port2]
);
export {
  L as PlaygroundCliBlueprintV2Worker
};
