// @wp-playground/storage/index.js
import { Semaphore as F, normalizePath as u, joinPaths as v } from "@php-wasm/util";
import { Octokit as k } from "octokit";
import { StreamedFile as f } from "@php-wasm/stream-compression";
import { ZipReader as m, BlobReader as g, BlobWriter as R } from "@zip.js/zip.js";
function j(i) {
  return new k({
    auth: i
  });
}
function L(i, t = "") {
  t.length && !t.endsWith("/") && (t += "/");
  const e = {};
  for (const r of i)
    r.path.startsWith(t) && (e[r.path.substring(t.length)] = r.content);
  return e;
}
async function A(i, t, e, r, n, s = {}) {
  s.progress || (s.progress = {
    foundFiles: 0,
    downloadedFiles: 0
  });
  const { onProgress: a } = s, o = [], c = [], { data: l } = await i.rest.repos.getContent({
    owner: t,
    repo: e,
    path: n,
    ref: r
  });
  if (!Array.isArray(l))
    throw new Error(
      `Expected the list of files to be an array, but got ${typeof l}`
    );
  for (const h of l)
    h.type === "file" ? (++s.progress.foundFiles, a == null || a(s.progress), o.push(
      T(i, t, e, r, h).then(($) => (++s.progress.downloadedFiles, a == null || a(s.progress), $))
    )) : h.type === "dir" && c.push(
      A(
        i,
        t,
        e,
        r,
        h.path,
        s
      )
    );
  const d = await Promise.all(o), b = (await Promise.all(c)).flatMap(
    (h) => h
  );
  return [...d, ...b];
}
const S = new F({ concurrency: 15 });
async function T(i, t, e, r, n) {
  const s = await S.acquire();
  try {
    const { data: a } = await i.rest.repos.getContent({
      owner: t,
      repo: e,
      ref: r,
      path: n.path
    });
    if (!("content" in a))
      throw new Error(`No content found for ${n.path}`);
    return {
      name: n.name,
      path: n.path,
      content: B(a.content)
    };
  } finally {
    s();
  }
}
function B(i) {
  const t = window.atob(i), e = t.length, r = new Uint8Array(e);
  for (let n = 0; n < e; n++)
    r[n] = t.charCodeAt(n);
  return r;
}
async function W(i, t, e, r, n) {
  var d;
  const { data: s } = await i.rest.pulls.get({
    owner: t,
    repo: e,
    pull_number: r
  }), o = (d = (await i.rest.actions.listWorkflowRuns({
    owner: t,
    repo: e,
    branch: s.head.ref,
    workflow_id: n
  })).data.workflow_runs[0]) == null ? void 0 : d.id, c = await i.rest.actions.listWorkflowRunArtifacts({
    owner: t,
    repo: e,
    run_id: o
  });
  return (await i.rest.actions.downloadArtifact({
    owner: t,
    repo: e,
    artifact_id: c.data.artifacts[0].id,
    archive_format: "zip"
  })).data;
}
async function J(i, t, e) {
  var s;
  const { data: r, headers: n } = await i.request(
    "GET /repos/{owner}/{repo}",
    {
      owner: t,
      repo: e
    }
  );
  return !(!n["x-oauth-scopes"] || !((s = r.permissions) != null && s.push));
}
async function V(i, t, e, r, n) {
  await i.request("GET /repos/{owner}/{repo}/branches/{branch}", {
    owner: t,
    repo: e,
    branch: r
  }).then(
    () => !0,
    () => !1
  ) ? await i.request("PATCH /repos/{owner}/{repo}/git/refs/{ref}", {
    owner: t,
    repo: e,
    sha: n,
    ref: `heads/${r}`
  }) : await i.request("POST /repos/{owner}/{repo}/git/refs", {
    owner: t,
    repo: e,
    sha: n,
    ref: `refs/heads/${r}`
  });
}
async function _(i, t, e) {
  const r = await i.request("GET /user");
  return (await i.request("GET /repos/{owner}/{repo}/forks", {
    owner: t,
    repo: e
  })).data.find(
    (a) => a.owner && a.owner.login === r.data.login
  ) || await i.request("POST /repos/{owner}/{repo}/forks", {
    owner: t,
    repo: e
  }), r.data.login;
}
async function K(i, t, e, r, n, s) {
  const {
    data: { sha: a }
  } = await i.request("POST /repos/{owner}/{repo}/git/commits", {
    owner: t,
    repo: e,
    message: r,
    tree: s,
    parents: [n]
  });
  return a;
}
async function Q(i, t, e, r, n) {
  const s = await P(
    i,
    t,
    e,
    r,
    n
  );
  if (s.length === 0)
    return null;
  const {
    data: { sha: a }
  } = await i.request("POST /repos/{owner}/{repo}/git/trees", {
    owner: t,
    repo: e,
    base_tree: r,
    tree: s
  });
  return a;
}
async function P(i, t, e, r, n) {
  const s = [];
  for (const [a, o] of n.create)
    s.push(E(i, t, e, a, o));
  for (const [a, o] of n.update)
    s.push(E(i, t, e, a, o));
  for (const a of n.delete)
    s.push(U(i, t, e, r, a));
  return Promise.all(s).then(
    (a) => a.filter((o) => !!o)
  );
}
const D = new F({ concurrency: 10 });
async function E(i, t, e, r, n) {
  const s = await D.acquire();
  try {
    if (ArrayBuffer.isView(n))
      try {
        const a = new TextDecoder("utf-8", {
          fatal: !0
        }).decode(n);
        return {
          path: r,
          content: a,
          mode: "100644"
        };
      } catch {
        const {
          data: { sha: a }
        } = await i.rest.git.createBlob({
          owner: t,
          repo: e,
          encoding: "base64",
          content: x(n)
        });
        return {
          path: r,
          sha: a,
          mode: "100644"
        };
      }
    else
      return {
        path: r,
        content: n,
        mode: "100644"
      };
  } finally {
    s();
  }
}
async function U(i, t, e, r, n) {
  const s = await D.acquire();
  try {
    return await i.request("HEAD /repos/{owner}/{repo}/contents/:path", {
      owner: t,
      repo: e,
      ref: r,
      path: n
    }), {
      path: n,
      mode: "100644",
      sha: null
    };
  } catch {
    return;
  } finally {
    s();
  }
}
function x(i) {
  const t = [], e = i.byteLength;
  for (let r = 0; r < e; r++)
    t.push(String.fromCharCode(i[r]));
  return window.btoa(t.join(""));
}
async function* X(i, t, { exceptPaths: e = [] } = {}) {
  if (t = u(t), !await i.isDir(t)) {
    await i.fileExists(t) && (yield {
      path: t,
      read: async () => await i.readFileAsBuffer(t)
    });
    return;
  }
  const r = [t];
  for (; r.length; ) {
    const n = r.pop();
    if (!n)
      return;
    const s = await i.listFiles(n);
    for (const a of s) {
      const o = v(n, a);
      e.includes(o.substring(t.length + 1)) || (await i.isDir(o) ? r.push(o) : yield {
        path: o,
        read: async () => await i.readFileAsBuffer(o)
      });
    }
  }
}
async function Y(i, t) {
  const e = {
    create: /* @__PURE__ */ new Map(),
    update: /* @__PURE__ */ new Map(),
    delete: /* @__PURE__ */ new Set()
  }, r = /* @__PURE__ */ new Set();
  for await (const n of t) {
    r.add(n.path);
    const s = i.get(n.path), a = await n.read();
    s ? H(s, a) || e.update.set(n.path, a) : e.create.set(n.path, a);
  }
  for (const n of i.keys())
    r.has(n) || e.delete.add(n);
  return e;
}
function H(i, t) {
  return i.length === t.length && i.every((e, r) => e === t[r]);
}
async function q(i) {
  return i.type === "local-fs" ? i.handle : C(i.path);
}
async function C(i) {
  const t = i.split("/").filter((r) => r.length > 0);
  let e = await navigator.storage.getDirectory();
  for (const r of t)
    e = await e.getDirectoryHandle(r, { create: !0 });
  return e;
}
async function Z(i) {
  const e = await (await navigator.storage.getDirectory()).resolve(i);
  if (e === null)
    throw new DOMException(
      "Unable to resolve path of OPFS directory handle.",
      "NotFoundError"
    );
  return "/" + e.join("/");
}
async function tt(i) {
  const t = await q(i);
  for await (const e of t.keys())
    await t.removeEntry(e, {
      recursive: !0
    });
}
function et(i, t) {
  t = u(t);
  const e = ["", ".", "/"].includes(t);
  let r = i;
  if (e)
    t = "";
  else {
    const a = t.split("/");
    for (const o of a) {
      const c = r == null ? void 0 : r.find(
        (l) => l.name === o
      );
      if ((c == null ? void 0 : c.type) === "folder")
        r = c.children;
      else return c ? [c.name] : [];
    }
  }
  const n = [], s = [{ tree: r, path: t }];
  for (; s.length > 0; ) {
    const { tree: a, path: o } = s.pop();
    for (const c of a) {
      const l = `${o}${o ? "/" : ""}${c.name}`;
      c.type === "folder" ? s.push({
        tree: c.children,
        path: l
      }) : n.push(l);
    }
  }
  return n;
}
function rt(i, t) {
  return i.startsWith(t) ? i.substring(t.length) : i;
}
async function nt(i, t) {
  await t.clear();
  const e = async (r) => {
    const n = await i.listFiles(r);
    for (const s of n) {
      const a = r === "/" ? `/${s}` : `${r}/${s}`;
      if (await i.isDir(a))
        await t.mkdir(a), await e(a);
      else {
        const o = await i.read(a), c = new Uint8Array(await o.arrayBuffer());
        await t.writeFile(a, c);
      }
    }
  };
  await e("/");
}
class st extends EventTarget {
  constructor(t) {
    super(), this.encoder = new TextEncoder(), this.decoder = new TextDecoder(), this.backend = t;
  }
  async isDir(t) {
    return this.backend.isDir(t);
  }
  async fileExists(t) {
    return this.backend.fileExists(t);
  }
  async read(t) {
    return this.backend.read(t);
  }
  async readFileAsText(t) {
    const r = await (await this.read(t)).arrayBuffer();
    return this.decoder.decode(r);
  }
  async listFiles(t) {
    return this.backend.listFiles(t);
  }
  async writeFile(t, e) {
    const r = typeof e == "string" ? this.encoder.encode(e) : e;
    await this.backend.writeFile(t, r), this.dispatchEvent(new Event("change"));
  }
  async mkdir(t, e) {
    await this.backend.mkdir(t, (e == null ? void 0 : e.recursive) ?? !1), this.dispatchEvent(new Event("change"));
  }
  async rmdir(t, e) {
    await this.backend.rmdir(t, (e == null ? void 0 : e.recursive) ?? !1), this.dispatchEvent(new Event("change"));
  }
  async mv(t, e) {
    t !== e && (await this.backend.mv(t, e), this.dispatchEvent(new Event("change")));
  }
  async unlink(t) {
    await this.backend.unlink(t), this.dispatchEvent(new Event("change"));
  }
  async clear() {
    await this.backend.clear(), this.dispatchEvent(new Event("change"));
  }
}
class it {
  constructor(t) {
    this.fileTree = t;
  }
  async read(t) {
    let e = this.getEntryAtPath(t);
    if (typeof e == "string")
      e = new TextEncoder().encode(e);
    else if (!(e instanceof Uint8Array))
      throw new Error(`Unsupported content type: ${typeof e}`);
    const r = new ReadableStream({
      start(n) {
        n.enqueue(e), n.close();
      }
    });
    return new f(r, t, {
      filesize: e.byteLength
    });
  }
  getEntryAtPath(t) {
    let e = t.replace(/^\//, ""), r = this.fileTree;
    for (; e; ) {
      if (r[e])
        return r[e];
      const n = e.split("/"), s = n.shift();
      if (!s || !r[s])
        break;
      r = r[s], e = n.join("/");
    }
    throw new Error(`File not found at ${t}`);
  }
}
class y {
  constructor(t) {
    this.entries = /* @__PURE__ */ new Map(), this.zipReader = t;
  }
  static fromStream(t) {
    const e = new m(
      new g(new f(t, "archive.zip"))
    );
    return new y(e);
  }
  static fromArrayBuffer(t) {
    const e = new m(
      new g(new Blob([t]))
    );
    return new y(e);
  }
  async read(t) {
    const e = await this.getEntry(t), r = await e.getData(new R());
    return new f(r.stream(), t, {
      filesize: e.uncompressedSize
    });
  }
  async getEntry(t) {
    const e = await this.getEntries(), r = u(t).replace(/^\//, ""), n = e.get(r);
    if (!n)
      throw new Error(`File ${t} not found in the zip.`);
    return n;
  }
  async getEntries() {
    if (this.entries.size === 0) {
      const t = await this.zipReader.getEntries();
      for (const e of t)
        this.entries.set(e.filename, e);
    }
    return this.entries;
  }
  /**
   * Returns the paths of all entries in the zip (file and directory names).
   */
  async getAllFilePaths() {
    const t = await this.getEntries();
    return Array.from(t.keys());
  }
}
class at {
  constructor(t, e) {
    this.chroot = t, this.backend = e;
  }
  async read(t) {
    const e = v(this.chroot, t);
    return this.backend.read(e);
  }
}
class ot {
  /**
   * Creates a new OverlayFilesystem.
   *
   * @param filesystems An array of Filesystem instances to cascade through.
   *                    The order determines the priority - earlier filesystems
   *                    are checked first.
   */
  constructor(t) {
    if (!t.length)
      throw new Error(
        "OverlayFilesystem requires at least one filesystem"
      );
    this.filesystems = t;
  }
  /**
   * Reads a file by trying each filesystem in order until one succeeds.
   *
   * @param path The path to the file to read.
   * @returns A Promise that resolves to a StreamedFile from the first
   *          filesystem that successfully resolves the path.
   * @throws Error if all filesystems fail to resolve the path.
   */
  async read(t) {
    const e = [];
    for (const n of this.filesystems)
      try {
        return await n.read(t);
      } catch (s) {
        e.push(
          s instanceof Error ? s : new Error(String(s))
        );
      }
    const r = e.map((n) => n.message).join("; ");
    throw new Error(
      `Failed to read ${t} from any filesystem: ${r}`,
      { cause: e }
    );
  }
}
class ct {
  constructor(t) {
    if (this.baseUrl = "", this.options = t, this.isDataUrl = t.baseUrl.startsWith("data:"), this.isDataUrl)
      return;
    const e = new URL("./", t.baseUrl);
    if (e.protocol !== "http:" && e.protocol !== "https:")
      throw new Error(
        "Unsupported protocol: " + e.protocol + ". Only HTTP and HTTPS are supported."
      );
    this.baseUrl = e.origin + e.pathname;
  }
  async read(t) {
    if (this.isDataUrl)
      throw new Error(
        "FetchFilesystem cannot fetch files from data URLs"
      );
    t = u(t);
    const e = t.replace(/^\//, ""), r = new URL(e, this.baseUrl).toString();
    if (!r.startsWith(this.baseUrl))
      throw new Error(
        `Refused to read a file outside of the base URL: ${r}`
      );
    const n = this.options.corsProxy ? `${this.options.corsProxy}${encodeURIComponent(r)}` : r, s = await fetch(n);
    if (!s.ok)
      throw new Error(
        `Failed to fetch file at ${t}: ${s.statusText}`
      );
    const a = s.headers.get("content-length") ? parseInt(s.headers.get("content-length"), 10) : void 0;
    return new f(s.body, t, { filesize: a });
  }
}
class lt {
  constructor(t) {
    this.root = t;
  }
  async ensureNodeModules() {
    if (!this.fs || !this.path) {
      try {
        this.fs = await import("fs"), this.path = await import("path");
      } catch {
        this.fs = require("fs"), this.path = require("path");
      }
      this.root = this.path.resolve(this.root) + this.path.sep;
    }
  }
  async read(t) {
    if (await this.ensureNodeModules(), t = this.path.resolve(
      this.root,
      this.path.normalize(t.replace(/^\//, ""))
    ), !t.startsWith(this.root))
      throw new Error(
        `Refused to read a file outside of the root directory: ${t}`
      );
    return new Promise((e, r) => {
      const n = this.path.resolve(t), s = this.fs.createReadStream(n);
      s.on("error", (a) => {
        r(
          new Error(
            `Failed to read file at ${t}: ${a.message}`
          )
        );
      }), this.fs.stat(n, (a, o) => {
        a ? r(
          new Error(
            `Failed to read file at ${t}: ${a.message}`
          )
        ) : e(
          new f(
            z(s),
            t,
            {
              filesize: o.size
            }
          )
        );
      });
    });
  }
}
function z(i) {
  return new ReadableStream({
    start(e) {
      i.on("data", (r) => e.enqueue(r)), i.on("end", () => e.close());
    }
  });
}
class p {
  constructor(t) {
    this.opfsRoot = t;
  }
  /**
   * Create a backend for a specific OPFS directory handle.
   */
  static fromDirectoryHandle(t) {
    return new p(t);
  }
  /**
   * Create a backend for a specific path in OPFS.
   * The path will be created if `create` is true.
   * @throws Error if OPFS is not available or path doesn't exist (when create=false)
   */
  static async fromPath(t, e = !1) {
    if (typeof navigator > "u")
      throw new Error("OPFS not available: navigator is undefined");
    if (!navigator.storage || !navigator.storage.getDirectory)
      throw new Error("OPFS not available: storage API not supported");
    let r = await navigator.storage.getDirectory();
    const n = t.split("/").filter(Boolean);
    for (const s of n)
      r = await r.getDirectoryHandle(s, { create: e });
    return new p(r);
  }
  async clear() {
    for await (const [t] of this.opfsRoot.entries())
      try {
        await this.opfsRoot.removeEntry(t, { recursive: !0 });
      } catch {
      }
  }
  // ReadableFilesystemBackend interface
  async read(t) {
    const e = await this.readFileAsBuffer(t), r = new ReadableStream({
      start(n) {
        n.enqueue(e), n.close();
      }
    });
    return new f(r, t, {
      filesize: e.byteLength
    });
  }
  async isDir(t) {
    if (t === "/")
      return !0;
    try {
      const e = t.split("/").filter(Boolean);
      let r = this.opfsRoot;
      for (const n of e)
        r = await r.getDirectoryHandle(n);
      return !0;
    } catch {
      return !1;
    }
  }
  async fileExists(t) {
    const e = t.split("/").filter(Boolean);
    if (e.length === 0)
      return !0;
    const r = e.pop();
    try {
      let n = this.opfsRoot;
      for (const s of e)
        n = await n.getDirectoryHandle(s);
      try {
        return await n.getFileHandle(r), !0;
      } catch {
        return await n.getDirectoryHandle(r), !0;
      }
    } catch {
      return !1;
    }
  }
  async listFiles(t) {
    let e = this.opfsRoot;
    if (t !== "/") {
      const n = t.split("/").filter(Boolean);
      for (const s of n)
        e = await e.getDirectoryHandle(s);
    }
    const r = [];
    for await (const [n] of e.entries())
      r.push(n);
    return r;
  }
  async writeFile(t, e) {
    const r = t.split("/").filter(Boolean), n = r.pop();
    if (!n)
      throw new Error(`Invalid file path: ${t}`);
    let s = this.opfsRoot;
    for (const c of r)
      s = await s.getDirectoryHandle(c);
    const o = await (await s.getFileHandle(n, { create: !0 })).createWritable();
    await o.write(e), await o.close();
  }
  async mkdir(t, e = !1) {
    const r = t.split("/").filter(Boolean);
    let n = this.opfsRoot;
    for (let s = 0; s < r.length; s++) {
      const a = r[s], o = s === r.length - 1;
      n = await n.getDirectoryHandle(a, {
        create: e || o
      });
    }
  }
  async rmdir(t, e) {
    const r = t.split("/").filter(Boolean), n = r.pop();
    if (!n)
      return;
    let s = this.opfsRoot;
    for (const a of r)
      s = await s.getDirectoryHandle(a);
    await s.removeEntry(n, { recursive: e });
  }
  async mv(t, e) {
    if (await this.isDir(t))
      await this.copyDir(t, e), await this.rmdir(t, !0);
    else {
      const n = await this.readFileAsBuffer(t);
      await this.writeFile(e, n), await this.unlink(t);
    }
  }
  async unlink(t) {
    const e = t.split("/").filter(Boolean), r = e.pop();
    if (!r)
      return;
    let n = this.opfsRoot;
    for (const s of e)
      n = await n.getDirectoryHandle(s);
    try {
      await n.removeEntry(r);
    } catch {
    }
  }
  // --- Internal helpers ---
  async readFileAsBuffer(t) {
    const e = t.split("/").filter(Boolean), r = e.pop();
    if (!r)
      throw new Error(`Invalid file path: ${t}`);
    let n = this.opfsRoot;
    for (const o of e)
      n = await n.getDirectoryHandle(o);
    const a = await (await n.getFileHandle(r)).getFile();
    return new Uint8Array(await a.arrayBuffer());
  }
  async copyDir(t, e) {
    await this.mkdir(e);
    const r = await this.listFiles(t);
    for (const n of r) {
      const s = t === "/" ? `/${n}` : `${t}/${n}`, a = e === "/" ? `/${n}` : `${e}/${n}`;
      if (await this.isDir(s))
        await this.copyDir(s, a);
      else {
        const o = await this.readFileAsBuffer(s);
        await this.writeFile(a, o);
      }
    }
  }
}
class ft {
  constructor(t = {}) {
    this.root = { type: "dir", children: {} };
    for (const [e, r] of Object.entries(t))
      this.writeFileSync(e, r);
  }
  async read(t) {
    const r = this.getFile(t).content, n = new ReadableStream({
      start(s) {
        s.enqueue(r), s.close();
      }
    });
    return new f(n, t, {
      filesize: r.byteLength
    });
  }
  async isDir(t) {
    const e = this.getNode(t);
    return !!e && e.type === "dir";
  }
  async fileExists(t) {
    return !!this.getNode(t);
  }
  async listFiles(t) {
    const e = this.getNode(t);
    return !e || e.type !== "dir" ? [] : Object.keys(e.children);
  }
  async writeFile(t, e) {
    this.writeFileSync(t, e);
  }
  async mkdir(t, e = !1) {
    if (t === "/")
      return;
    const { parent: r, name: n } = e ? this.getOrCreateParent(t) : this.getParent(t);
    r.children[n] || (r.children[n] = { type: "dir", children: {} });
  }
  async rmdir(t, e) {
    const { parent: r, name: n } = this.getParent(t), s = r.children[n];
    if (!(!s || s.type !== "dir")) {
      if (!e && Object.keys(s.children).length > 0)
        throw new Error("Directory not empty");
      delete r.children[n];
    }
  }
  async mv(t, e) {
    const { parent: r, name: n } = this.getParent(t), s = r.children[n];
    if (!s)
      throw new Error(`Source not found: ${t}`);
    const { parent: a, name: o } = this.getParent(e);
    a.children[o] = s, delete r.children[n];
  }
  async unlink(t) {
    const { parent: e, name: r } = this.getParent(t), n = e.children[r];
    n && n.type === "file" && delete e.children[r];
  }
  async clear() {
    this.root = { type: "dir", children: {} };
  }
  // --- Internal helpers ---
  writeFileSync(t, e) {
    const { parent: r, name: n } = this.getParent(t);
    r.children[n] = {
      type: "file",
      content: new Uint8Array(e)
    };
  }
  getNode(t) {
    if (t === "/")
      return this.root;
    const e = t.split("/").filter(Boolean);
    let r = this.root;
    for (const n of e) {
      if (r.type !== "dir")
        return;
      const s = r.children[n];
      if (!s)
        return;
      r = s;
    }
    return r;
  }
  getDir(t) {
    const e = this.getNode(t);
    if (!e || e.type !== "dir")
      throw new Error(`Directory not found: ${t}`);
    return e;
  }
  getFile(t) {
    const e = this.getNode(t);
    if (!e || e.type !== "file")
      throw new Error(`File not found: ${t}`);
    return e;
  }
  /**
   * Get parent directory, throwing if it doesn't exist.
   */
  getParent(t) {
    const e = t.split("/").filter(Boolean), r = e.pop();
    if (!r)
      throw new Error(`Invalid path: ${t}`);
    const n = e.length ? `/${e.join("/")}` : "/", s = this.getNode(n);
    if (!s || s.type !== "dir")
      throw new Error(`Parent directory not found: ${n}`);
    return { parent: s, name: r };
  }
  /**
   * Get parent directory, creating it if it doesn't exist.
   */
  getOrCreateParent(t) {
    const e = t.split("/").filter(Boolean), r = e.pop();
    if (!r)
      throw new Error(`Invalid path: ${t}`);
    const n = e.length ? `/${e.join("/")}` : "/";
    return { parent: this.ensureDir(n), name: r };
  }
  ensureDir(t) {
    if (t === "/")
      return this.root;
    const e = t.split("/").filter(Boolean);
    let r = this.root;
    for (const n of e) {
      const s = r.children[n];
      if (!s) {
        const a = { type: "dir", children: {} };
        r.children[n] = a, r = a;
        continue;
      }
      if (s.type !== "dir")
        throw new Error(
          `Path segment "${n}" is not a directory in ${t}`
        );
      r = s;
    }
    return r;
  }
}
class ht extends Error {
  constructor(t, e) {
    super(
      `Authentication required to access private repository: ${t}`
    ), this.name = "GitAuthenticationError", this.repoUrl = t, this.status = e;
  }
}
function w() {
  return import("./git-sparse-checkout-Bigbax-u.js");
}
function O() {
  return import("./git-create-dotgit-directory-CvcNS4KR.js");
}
async function dt(...i) {
  return (await w()).sparseCheckout(...i);
}
async function ut(...i) {
  return (await w()).listGitFiles(...i);
}
async function wt(...i) {
  return (await w()).resolveCommitHash(...i);
}
async function yt(...i) {
  return (await w()).listGitRefs(...i);
}
async function pt(...i) {
  return (await O()).createDotGitDirectory(
    ...i
  );
}
export {
  at as ChrootFilesystem,
  st as EventedFilesystem,
  ct as FetchFilesystem,
  ht as GitAuthenticationError,
  it as InMemoryFilesystem,
  ft as InMemoryFilesystemBackend,
  lt as NodeJsFilesystem,
  p as OpfsFilesystemBackend,
  ot as OverlayFilesystem,
  y as ZipFilesystem,
  Y as changeset,
  tt as clearContentsFromMountDevice,
  nt as copyFilesystem,
  j as createClient,
  K as createCommit,
  pt as createDotGitDirectory,
  V as createOrUpdateBranch,
  Q as createTree,
  E as createTreeNode,
  P as createTreeNodes,
  U as deleteFile,
  q as directoryHandleFromMountDevice,
  Z as directoryHandleToOpfsPath,
  L as filesListToObject,
  _ as fork,
  W as getArtifact,
  A as getFilesFromDirectory,
  X as iterateFiles,
  et as listDescendantFiles,
  ut as listGitFiles,
  yt as listGitRefs,
  J as mayPush,
  C as opfsPathToDirectoryHandle,
  rt as removePathPrefix,
  wt as resolveCommitHash,
  dt as sparseCheckout
};
