// @wp-playground/storage/isomorphic-git-internals-DwkB8rfP.js
import Qn from "sha.js";
import "async-lock";
import ti from "crc-32";
import Je from "pako";
import "diff3";
var Ke = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {};
function Ze(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var br = {}, It = {};
It.byteLength = ni;
It.toByteArray = oi;
It.fromByteArray = ci;
var z = [], L = [], ri = typeof Uint8Array < "u" ? Uint8Array : Array, Rt = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
for (var K = 0, ei = Rt.length; K < ei; ++K)
  z[K] = Rt[K], L[Rt.charCodeAt(K)] = K;
L[45] = 62;
L[95] = 63;
function Qe(e) {
  var t = e.length;
  if (t % 4 > 0)
    throw new Error("Invalid string. Length must be a multiple of 4");
  var r = e.indexOf("=");
  r === -1 && (r = t);
  var n = r === t ? 0 : 4 - r % 4;
  return [r, n];
}
function ni(e) {
  var t = Qe(e), r = t[0], n = t[1];
  return (r + n) * 3 / 4 - n;
}
function ii(e, t, r) {
  return (t + r) * 3 / 4 - r;
}
function oi(e) {
  var t, r = Qe(e), n = r[0], a = r[1], c = new ri(ii(e, n, a)), u = 0, s = a > 0 ? n - 4 : n, h;
  for (h = 0; h < s; h += 4)
    t = L[e.charCodeAt(h)] << 18 | L[e.charCodeAt(h + 1)] << 12 | L[e.charCodeAt(h + 2)] << 6 | L[e.charCodeAt(h + 3)], c[u++] = t >> 16 & 255, c[u++] = t >> 8 & 255, c[u++] = t & 255;
  return a === 2 && (t = L[e.charCodeAt(h)] << 2 | L[e.charCodeAt(h + 1)] >> 4, c[u++] = t & 255), a === 1 && (t = L[e.charCodeAt(h)] << 10 | L[e.charCodeAt(h + 1)] << 4 | L[e.charCodeAt(h + 2)] >> 2, c[u++] = t >> 8 & 255, c[u++] = t & 255), c;
}
function ai(e) {
  return z[e >> 18 & 63] + z[e >> 12 & 63] + z[e >> 6 & 63] + z[e & 63];
}
function si(e, t, r) {
  for (var n, a = [], c = t; c < r; c += 3)
    n = (e[c] << 16 & 16711680) + (e[c + 1] << 8 & 65280) + (e[c + 2] & 255), a.push(ai(n));
  return a.join("");
}
function ci(e) {
  for (var t, r = e.length, n = r % 3, a = [], c = 16383, u = 0, s = r - n; u < s; u += c)
    a.push(si(e, u, u + c > s ? s : u + c));
  return n === 1 ? (t = e[r - 1], a.push(
    z[t >> 2] + z[t << 4 & 63] + "=="
  )) : n === 2 && (t = (e[r - 2] << 8) + e[r - 1], a.push(
    z[t >> 10] + z[t >> 4 & 63] + z[t << 2 & 63] + "="
  )), a.join("");
}
var vr = {};
/*! ieee754. BSD-3-Clause License. Feross Aboukhadijeh <https://feross.org/opensource> */
vr.read = function(e, t, r, n, a) {
  var c, u, s = a * 8 - n - 1, h = (1 << s) - 1, d = h >> 1, m = -7, y = r ? a - 1 : 0, b = r ? -1 : 1, E = e[t + y];
  for (y += b, c = E & (1 << -m) - 1, E >>= -m, m += s; m > 0; c = c * 256 + e[t + y], y += b, m -= 8)
    ;
  for (u = c & (1 << -m) - 1, c >>= -m, m += n; m > 0; u = u * 256 + e[t + y], y += b, m -= 8)
    ;
  if (c === 0)
    c = 1 - d;
  else {
    if (c === h)
      return u ? NaN : (E ? -1 : 1) * (1 / 0);
    u = u + Math.pow(2, n), c = c - d;
  }
  return (E ? -1 : 1) * u * Math.pow(2, c - n);
};
vr.write = function(e, t, r, n, a, c) {
  var u, s, h, d = c * 8 - a - 1, m = (1 << d) - 1, y = m >> 1, b = a === 23 ? Math.pow(2, -24) - Math.pow(2, -77) : 0, E = n ? 0 : c - 1, O = n ? 1 : -1, F = t < 0 || t === 0 && 1 / t < 0 ? 1 : 0;
  for (t = Math.abs(t), isNaN(t) || t === 1 / 0 ? (s = isNaN(t) ? 1 : 0, u = m) : (u = Math.floor(Math.log(t) / Math.LN2), t * (h = Math.pow(2, -u)) < 1 && (u--, h *= 2), u + y >= 1 ? t += b / h : t += b * Math.pow(2, 1 - y), t * h >= 2 && (u++, h /= 2), u + y >= m ? (s = 0, u = m) : u + y >= 1 ? (s = (t * h - 1) * Math.pow(2, a), u = u + y) : (s = t * Math.pow(2, y - 1) * Math.pow(2, a), u = 0)); a >= 8; e[r + E] = s & 255, E += O, s /= 256, a -= 8)
    ;
  for (u = u << a | s, d += a; d > 0; e[r + E] = u & 255, E += O, u /= 256, d -= 8)
    ;
  e[r + E - O] |= F * 128;
};
/*!
 * The buffer module from node.js, for the browser.
 *
 * @author   Feross Aboukhadijeh <https://feross.org>
 * @license  MIT
 */
(function(e) {
  var t = It, r = vr, n = typeof Symbol == "function" && typeof Symbol.for == "function" ? Symbol.for("nodejs.util.inspect.custom") : null;
  e.Buffer = s, e.SlowBuffer = _, e.INSPECT_MAX_BYTES = 50;
  var a = 2147483647;
  e.kMaxLength = a, s.TYPED_ARRAY_SUPPORT = c(), !s.TYPED_ARRAY_SUPPORT && typeof console < "u" && typeof console.error == "function" && console.error(
    "This browser lacks typed array (Uint8Array) support which is required by `buffer` v5.x. Use `buffer` v4.x if you require old browser support."
  );
  function c() {
    try {
      var f = new Uint8Array(1), i = { foo: function() {
        return 42;
      } };
      return Object.setPrototypeOf(i, Uint8Array.prototype), Object.setPrototypeOf(f, i), f.foo() === 42;
    } catch {
      return !1;
    }
  }
  Object.defineProperty(s.prototype, "parent", {
    enumerable: !0,
    get: function() {
      if (s.isBuffer(this))
        return this.buffer;
    }
  }), Object.defineProperty(s.prototype, "offset", {
    enumerable: !0,
    get: function() {
      if (s.isBuffer(this))
        return this.byteOffset;
    }
  });
  function u(f) {
    if (f > a)
      throw new RangeError('The value "' + f + '" is invalid for option "size"');
    var i = new Uint8Array(f);
    return Object.setPrototypeOf(i, s.prototype), i;
  }
  function s(f, i, o) {
    if (typeof f == "number") {
      if (typeof i == "string")
        throw new TypeError(
          'The "string" argument must be of type string. Received type number'
        );
      return y(f);
    }
    return h(f, i, o);
  }
  s.poolSize = 8192;
  function h(f, i, o) {
    if (typeof f == "string")
      return b(f, i);
    if (ArrayBuffer.isView(f))
      return O(f);
    if (f == null)
      throw new TypeError(
        "The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type " + typeof f
      );
    if (q(f, ArrayBuffer) || f && q(f.buffer, ArrayBuffer) || typeof SharedArrayBuffer < "u" && (q(f, SharedArrayBuffer) || f && q(f.buffer, SharedArrayBuffer)))
      return F(f, i, o);
    if (typeof f == "number")
      throw new TypeError(
        'The "value" argument must not be of type number. Received type number'
      );
    var l = f.valueOf && f.valueOf();
    if (l != null && l !== f)
      return s.from(l, i, o);
    var p = A(f);
    if (p) return p;
    if (typeof Symbol < "u" && Symbol.toPrimitive != null && typeof f[Symbol.toPrimitive] == "function")
      return s.from(
        f[Symbol.toPrimitive]("string"),
        i,
        o
      );
    throw new TypeError(
      "The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type " + typeof f
    );
  }
  s.from = function(f, i, o) {
    return h(f, i, o);
  }, Object.setPrototypeOf(s.prototype, Uint8Array.prototype), Object.setPrototypeOf(s, Uint8Array);
  function d(f) {
    if (typeof f != "number")
      throw new TypeError('"size" argument must be of type number');
    if (f < 0)
      throw new RangeError('The value "' + f + '" is invalid for option "size"');
  }
  function m(f, i, o) {
    return d(f), f <= 0 ? u(f) : i !== void 0 ? typeof o == "string" ? u(f).fill(i, o) : u(f).fill(i) : u(f);
  }
  s.alloc = function(f, i, o) {
    return m(f, i, o);
  };
  function y(f) {
    return d(f), u(f < 0 ? 0 : x(f) | 0);
  }
  s.allocUnsafe = function(f) {
    return y(f);
  }, s.allocUnsafeSlow = function(f) {
    return y(f);
  };
  function b(f, i) {
    if ((typeof i != "string" || i === "") && (i = "utf8"), !s.isEncoding(i))
      throw new TypeError("Unknown encoding: " + i);
    var o = v(f, i) | 0, l = u(o), p = l.write(f, i);
    return p !== o && (l = l.slice(0, p)), l;
  }
  function E(f) {
    for (var i = f.length < 0 ? 0 : x(f.length) | 0, o = u(i), l = 0; l < i; l += 1)
      o[l] = f[l] & 255;
    return o;
  }
  function O(f) {
    if (q(f, Uint8Array)) {
      var i = new Uint8Array(f);
      return F(i.buffer, i.byteOffset, i.byteLength);
    }
    return E(f);
  }
  function F(f, i, o) {
    if (i < 0 || f.byteLength < i)
      throw new RangeError('"offset" is outside of buffer bounds');
    if (f.byteLength < i + (o || 0))
      throw new RangeError('"length" is outside of buffer bounds');
    var l;
    return i === void 0 && o === void 0 ? l = new Uint8Array(f) : o === void 0 ? l = new Uint8Array(f, i) : l = new Uint8Array(f, i, o), Object.setPrototypeOf(l, s.prototype), l;
  }
  function A(f) {
    if (s.isBuffer(f)) {
      var i = x(f.length) | 0, o = u(i);
      return o.length === 0 || f.copy(o, 0, 0, i), o;
    }
    if (f.length !== void 0)
      return typeof f.length != "number" || Ut(f.length) ? u(0) : E(f);
    if (f.type === "Buffer" && Array.isArray(f.data))
      return E(f.data);
  }
  function x(f) {
    if (f >= a)
      throw new RangeError("Attempt to allocate Buffer larger than maximum size: 0x" + a.toString(16) + " bytes");
    return f | 0;
  }
  function _(f) {
    return +f != f && (f = 0), s.alloc(+f);
  }
  s.isBuffer = function(i) {
    return i != null && i._isBuffer === !0 && i !== s.prototype;
  }, s.compare = function(i, o) {
    if (q(i, Uint8Array) && (i = s.from(i, i.offset, i.byteLength)), q(o, Uint8Array) && (o = s.from(o, o.offset, o.byteLength)), !s.isBuffer(i) || !s.isBuffer(o))
      throw new TypeError(
        'The "buf1", "buf2" arguments must be one of type Buffer or Uint8Array'
      );
    if (i === o) return 0;
    for (var l = i.length, p = o.length, g = 0, w = Math.min(l, p); g < w; ++g)
      if (i[g] !== o[g]) {
        l = i[g], p = o[g];
        break;
      }
    return l < p ? -1 : p < l ? 1 : 0;
  }, s.isEncoding = function(i) {
    switch (String(i).toLowerCase()) {
      case "hex":
      case "utf8":
      case "utf-8":
      case "ascii":
      case "latin1":
      case "binary":
      case "base64":
      case "ucs2":
      case "ucs-2":
      case "utf16le":
      case "utf-16le":
        return !0;
      default:
        return !1;
    }
  }, s.concat = function(i, o) {
    if (!Array.isArray(i))
      throw new TypeError('"list" argument must be an Array of Buffers');
    if (i.length === 0)
      return s.alloc(0);
    var l;
    if (o === void 0)
      for (o = 0, l = 0; l < i.length; ++l)
        o += i[l].length;
    var p = s.allocUnsafe(o), g = 0;
    for (l = 0; l < i.length; ++l) {
      var w = i[l];
      if (q(w, Uint8Array))
        g + w.length > p.length ? s.from(w).copy(p, g) : Uint8Array.prototype.set.call(
          p,
          w,
          g
        );
      else if (s.isBuffer(w))
        w.copy(p, g);
      else
        throw new TypeError('"list" argument must be an Array of Buffers');
      g += w.length;
    }
    return p;
  };
  function v(f, i) {
    if (s.isBuffer(f))
      return f.length;
    if (ArrayBuffer.isView(f) || q(f, ArrayBuffer))
      return f.byteLength;
    if (typeof f != "string")
      throw new TypeError(
        'The "string" argument must be one of type string, Buffer, or ArrayBuffer. Received type ' + typeof f
      );
    var o = f.length, l = arguments.length > 2 && arguments[2] === !0;
    if (!l && o === 0) return 0;
    for (var p = !1; ; )
      switch (i) {
        case "ascii":
        case "latin1":
        case "binary":
          return o;
        case "utf8":
        case "utf-8":
          return Ot(f).length;
        case "ucs2":
        case "ucs-2":
        case "utf16le":
        case "utf-16le":
          return o * 2;
        case "hex":
          return o >>> 1;
        case "base64":
          return ce(f).length;
        default:
          if (p)
            return l ? -1 : Ot(f).length;
          i = ("" + i).toLowerCase(), p = !0;
      }
  }
  s.byteLength = v;
  function P(f, i, o) {
    var l = !1;
    if ((i === void 0 || i < 0) && (i = 0), i > this.length || ((o === void 0 || o > this.length) && (o = this.length), o <= 0) || (o >>>= 0, i >>>= 0, o <= i))
      return "";
    for (f || (f = "utf8"); ; )
      switch (f) {
        case "hex":
          return Wn(this, i, o);
        case "utf8":
        case "utf-8":
          return ne(this, i, o);
        case "ascii":
          return qn(this, i, o);
        case "latin1":
        case "binary":
          return zn(this, i, o);
        case "base64":
          return Ln(this, i, o);
        case "ucs2":
        case "ucs-2":
        case "utf16le":
        case "utf-16le":
          return Vn(this, i, o);
        default:
          if (l) throw new TypeError("Unknown encoding: " + f);
          f = (f + "").toLowerCase(), l = !0;
      }
  }
  s.prototype._isBuffer = !0;
  function j(f, i, o) {
    var l = f[i];
    f[i] = f[o], f[o] = l;
  }
  s.prototype.swap16 = function() {
    var i = this.length;
    if (i % 2 !== 0)
      throw new RangeError("Buffer size must be a multiple of 16-bits");
    for (var o = 0; o < i; o += 2)
      j(this, o, o + 1);
    return this;
  }, s.prototype.swap32 = function() {
    var i = this.length;
    if (i % 4 !== 0)
      throw new RangeError("Buffer size must be a multiple of 32-bits");
    for (var o = 0; o < i; o += 4)
      j(this, o, o + 3), j(this, o + 1, o + 2);
    return this;
  }, s.prototype.swap64 = function() {
    var i = this.length;
    if (i % 8 !== 0)
      throw new RangeError("Buffer size must be a multiple of 64-bits");
    for (var o = 0; o < i; o += 8)
      j(this, o, o + 7), j(this, o + 1, o + 6), j(this, o + 2, o + 5), j(this, o + 3, o + 4);
    return this;
  }, s.prototype.toString = function() {
    var i = this.length;
    return i === 0 ? "" : arguments.length === 0 ? ne(this, 0, i) : P.apply(this, arguments);
  }, s.prototype.toLocaleString = s.prototype.toString, s.prototype.equals = function(i) {
    if (!s.isBuffer(i)) throw new TypeError("Argument must be a Buffer");
    return this === i ? !0 : s.compare(this, i) === 0;
  }, s.prototype.inspect = function() {
    var i = "", o = e.INSPECT_MAX_BYTES;
    return i = this.toString("hex", 0, o).replace(/(.{2})/g, "$1 ").trim(), this.length > o && (i += " ... "), "<Buffer " + i + ">";
  }, n && (s.prototype[n] = s.prototype.inspect), s.prototype.compare = function(i, o, l, p, g) {
    if (q(i, Uint8Array) && (i = s.from(i, i.offset, i.byteLength)), !s.isBuffer(i))
      throw new TypeError(
        'The "target" argument must be one of type Buffer or Uint8Array. Received type ' + typeof i
      );
    if (o === void 0 && (o = 0), l === void 0 && (l = i ? i.length : 0), p === void 0 && (p = 0), g === void 0 && (g = this.length), o < 0 || l > i.length || p < 0 || g > this.length)
      throw new RangeError("out of range index");
    if (p >= g && o >= l)
      return 0;
    if (p >= g)
      return -1;
    if (o >= l)
      return 1;
    if (o >>>= 0, l >>>= 0, p >>>= 0, g >>>= 0, this === i) return 0;
    for (var w = g - p, I = l - o, k = Math.min(w, I), T = this.slice(p, g), N = i.slice(o, l), U = 0; U < k; ++U)
      if (T[U] !== N[U]) {
        w = T[U], I = N[U];
        break;
      }
    return w < I ? -1 : I < w ? 1 : 0;
  };
  function re(f, i, o, l, p) {
    if (f.length === 0) return -1;
    if (typeof o == "string" ? (l = o, o = 0) : o > 2147483647 ? o = 2147483647 : o < -2147483648 && (o = -2147483648), o = +o, Ut(o) && (o = p ? 0 : f.length - 1), o < 0 && (o = f.length + o), o >= f.length) {
      if (p) return -1;
      o = f.length - 1;
    } else if (o < 0)
      if (p) o = 0;
      else return -1;
    if (typeof i == "string" && (i = s.from(i, l)), s.isBuffer(i))
      return i.length === 0 ? -1 : ee(f, i, o, l, p);
    if (typeof i == "number")
      return i = i & 255, typeof Uint8Array.prototype.indexOf == "function" ? p ? Uint8Array.prototype.indexOf.call(f, i, o) : Uint8Array.prototype.lastIndexOf.call(f, i, o) : ee(f, [i], o, l, p);
    throw new TypeError("val must be string, number or Buffer");
  }
  function ee(f, i, o, l, p) {
    var g = 1, w = f.length, I = i.length;
    if (l !== void 0 && (l = String(l).toLowerCase(), l === "ucs2" || l === "ucs-2" || l === "utf16le" || l === "utf-16le")) {
      if (f.length < 2 || i.length < 2)
        return -1;
      g = 2, w /= 2, I /= 2, o /= 2;
    }
    function k(ue, fe) {
      return g === 1 ? ue[fe] : ue.readUInt16BE(fe * g);
    }
    var T;
    if (p) {
      var N = -1;
      for (T = o; T < w; T++)
        if (k(f, T) === k(i, N === -1 ? 0 : T - N)) {
          if (N === -1 && (N = T), T - N + 1 === I) return N * g;
        } else
          N !== -1 && (T -= T - N), N = -1;
    } else
      for (o + I > w && (o = w - I), T = o; T >= 0; T--) {
        for (var U = !0, dt = 0; dt < I; dt++)
          if (k(f, T + dt) !== k(i, dt)) {
            U = !1;
            break;
          }
        if (U) return T;
      }
    return -1;
  }
  s.prototype.includes = function(i, o, l) {
    return this.indexOf(i, o, l) !== -1;
  }, s.prototype.indexOf = function(i, o, l) {
    return re(this, i, o, l, !0);
  }, s.prototype.lastIndexOf = function(i, o, l) {
    return re(this, i, o, l, !1);
  };
  function jn(f, i, o, l) {
    o = Number(o) || 0;
    var p = f.length - o;
    l ? (l = Number(l), l > p && (l = p)) : l = p;
    var g = i.length;
    l > g / 2 && (l = g / 2);
    for (var w = 0; w < l; ++w) {
      var I = parseInt(i.substr(w * 2, 2), 16);
      if (Ut(I)) return w;
      f[o + w] = I;
    }
    return w;
  }
  function Nn(f, i, o, l) {
    return yt(Ot(i, f.length - o), f, o, l);
  }
  function Dn(f, i, o, l) {
    return yt(Jn(i), f, o, l);
  }
  function Mn(f, i, o, l) {
    return yt(ce(i), f, o, l);
  }
  function Gn(f, i, o, l) {
    return yt(Kn(i, f.length - o), f, o, l);
  }
  s.prototype.write = function(i, o, l, p) {
    if (o === void 0)
      p = "utf8", l = this.length, o = 0;
    else if (l === void 0 && typeof o == "string")
      p = o, l = this.length, o = 0;
    else if (isFinite(o))
      o = o >>> 0, isFinite(l) ? (l = l >>> 0, p === void 0 && (p = "utf8")) : (p = l, l = void 0);
    else
      throw new Error(
        "Buffer.write(string, encoding, offset[, length]) is no longer supported"
      );
    var g = this.length - o;
    if ((l === void 0 || l > g) && (l = g), i.length > 0 && (l < 0 || o < 0) || o > this.length)
      throw new RangeError("Attempt to write outside buffer bounds");
    p || (p = "utf8");
    for (var w = !1; ; )
      switch (p) {
        case "hex":
          return jn(this, i, o, l);
        case "utf8":
        case "utf-8":
          return Nn(this, i, o, l);
        case "ascii":
        case "latin1":
        case "binary":
          return Dn(this, i, o, l);
        case "base64":
          return Mn(this, i, o, l);
        case "ucs2":
        case "ucs-2":
        case "utf16le":
        case "utf-16le":
          return Gn(this, i, o, l);
        default:
          if (w) throw new TypeError("Unknown encoding: " + p);
          p = ("" + p).toLowerCase(), w = !0;
      }
  }, s.prototype.toJSON = function() {
    return {
      type: "Buffer",
      data: Array.prototype.slice.call(this._arr || this, 0)
    };
  };
  function Ln(f, i, o) {
    return i === 0 && o === f.length ? t.fromByteArray(f) : t.fromByteArray(f.slice(i, o));
  }
  function ne(f, i, o) {
    o = Math.min(f.length, o);
    for (var l = [], p = i; p < o; ) {
      var g = f[p], w = null, I = g > 239 ? 4 : g > 223 ? 3 : g > 191 ? 2 : 1;
      if (p + I <= o) {
        var k, T, N, U;
        switch (I) {
          case 1:
            g < 128 && (w = g);
            break;
          case 2:
            k = f[p + 1], (k & 192) === 128 && (U = (g & 31) << 6 | k & 63, U > 127 && (w = U));
            break;
          case 3:
            k = f[p + 1], T = f[p + 2], (k & 192) === 128 && (T & 192) === 128 && (U = (g & 15) << 12 | (k & 63) << 6 | T & 63, U > 2047 && (U < 55296 || U > 57343) && (w = U));
            break;
          case 4:
            k = f[p + 1], T = f[p + 2], N = f[p + 3], (k & 192) === 128 && (T & 192) === 128 && (N & 192) === 128 && (U = (g & 15) << 18 | (k & 63) << 12 | (T & 63) << 6 | N & 63, U > 65535 && U < 1114112 && (w = U));
        }
      }
      w === null ? (w = 65533, I = 1) : w > 65535 && (w -= 65536, l.push(w >>> 10 & 1023 | 55296), w = 56320 | w & 1023), l.push(w), p += I;
    }
    return Hn(l);
  }
  var ie = 4096;
  function Hn(f) {
    var i = f.length;
    if (i <= ie)
      return String.fromCharCode.apply(String, f);
    for (var o = "", l = 0; l < i; )
      o += String.fromCharCode.apply(
        String,
        f.slice(l, l += ie)
      );
    return o;
  }
  function qn(f, i, o) {
    var l = "";
    o = Math.min(f.length, o);
    for (var p = i; p < o; ++p)
      l += String.fromCharCode(f[p] & 127);
    return l;
  }
  function zn(f, i, o) {
    var l = "";
    o = Math.min(f.length, o);
    for (var p = i; p < o; ++p)
      l += String.fromCharCode(f[p]);
    return l;
  }
  function Wn(f, i, o) {
    var l = f.length;
    (!i || i < 0) && (i = 0), (!o || o < 0 || o > l) && (o = l);
    for (var p = "", g = i; g < o; ++g)
      p += Zn[f[g]];
    return p;
  }
  function Vn(f, i, o) {
    for (var l = f.slice(i, o), p = "", g = 0; g < l.length - 1; g += 2)
      p += String.fromCharCode(l[g] + l[g + 1] * 256);
    return p;
  }
  s.prototype.slice = function(i, o) {
    var l = this.length;
    i = ~~i, o = o === void 0 ? l : ~~o, i < 0 ? (i += l, i < 0 && (i = 0)) : i > l && (i = l), o < 0 ? (o += l, o < 0 && (o = 0)) : o > l && (o = l), o < i && (o = i);
    var p = this.subarray(i, o);
    return Object.setPrototypeOf(p, s.prototype), p;
  };
  function R(f, i, o) {
    if (f % 1 !== 0 || f < 0) throw new RangeError("offset is not uint");
    if (f + i > o) throw new RangeError("Trying to access beyond buffer length");
  }
  s.prototype.readUintLE = s.prototype.readUIntLE = function(i, o, l) {
    i = i >>> 0, o = o >>> 0, l || R(i, o, this.length);
    for (var p = this[i], g = 1, w = 0; ++w < o && (g *= 256); )
      p += this[i + w] * g;
    return p;
  }, s.prototype.readUintBE = s.prototype.readUIntBE = function(i, o, l) {
    i = i >>> 0, o = o >>> 0, l || R(i, o, this.length);
    for (var p = this[i + --o], g = 1; o > 0 && (g *= 256); )
      p += this[i + --o] * g;
    return p;
  }, s.prototype.readUint8 = s.prototype.readUInt8 = function(i, o) {
    return i = i >>> 0, o || R(i, 1, this.length), this[i];
  }, s.prototype.readUint16LE = s.prototype.readUInt16LE = function(i, o) {
    return i = i >>> 0, o || R(i, 2, this.length), this[i] | this[i + 1] << 8;
  }, s.prototype.readUint16BE = s.prototype.readUInt16BE = function(i, o) {
    return i = i >>> 0, o || R(i, 2, this.length), this[i] << 8 | this[i + 1];
  }, s.prototype.readUint32LE = s.prototype.readUInt32LE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), (this[i] | this[i + 1] << 8 | this[i + 2] << 16) + this[i + 3] * 16777216;
  }, s.prototype.readUint32BE = s.prototype.readUInt32BE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), this[i] * 16777216 + (this[i + 1] << 16 | this[i + 2] << 8 | this[i + 3]);
  }, s.prototype.readIntLE = function(i, o, l) {
    i = i >>> 0, o = o >>> 0, l || R(i, o, this.length);
    for (var p = this[i], g = 1, w = 0; ++w < o && (g *= 256); )
      p += this[i + w] * g;
    return g *= 128, p >= g && (p -= Math.pow(2, 8 * o)), p;
  }, s.prototype.readIntBE = function(i, o, l) {
    i = i >>> 0, o = o >>> 0, l || R(i, o, this.length);
    for (var p = o, g = 1, w = this[i + --p]; p > 0 && (g *= 256); )
      w += this[i + --p] * g;
    return g *= 128, w >= g && (w -= Math.pow(2, 8 * o)), w;
  }, s.prototype.readInt8 = function(i, o) {
    return i = i >>> 0, o || R(i, 1, this.length), this[i] & 128 ? (255 - this[i] + 1) * -1 : this[i];
  }, s.prototype.readInt16LE = function(i, o) {
    i = i >>> 0, o || R(i, 2, this.length);
    var l = this[i] | this[i + 1] << 8;
    return l & 32768 ? l | 4294901760 : l;
  }, s.prototype.readInt16BE = function(i, o) {
    i = i >>> 0, o || R(i, 2, this.length);
    var l = this[i + 1] | this[i] << 8;
    return l & 32768 ? l | 4294901760 : l;
  }, s.prototype.readInt32LE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), this[i] | this[i + 1] << 8 | this[i + 2] << 16 | this[i + 3] << 24;
  }, s.prototype.readInt32BE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), this[i] << 24 | this[i + 1] << 16 | this[i + 2] << 8 | this[i + 3];
  }, s.prototype.readFloatLE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), r.read(this, i, !0, 23, 4);
  }, s.prototype.readFloatBE = function(i, o) {
    return i = i >>> 0, o || R(i, 4, this.length), r.read(this, i, !1, 23, 4);
  }, s.prototype.readDoubleLE = function(i, o) {
    return i = i >>> 0, o || R(i, 8, this.length), r.read(this, i, !0, 52, 8);
  }, s.prototype.readDoubleBE = function(i, o) {
    return i = i >>> 0, o || R(i, 8, this.length), r.read(this, i, !1, 52, 8);
  };
  function D(f, i, o, l, p, g) {
    if (!s.isBuffer(f)) throw new TypeError('"buffer" argument must be a Buffer instance');
    if (i > p || i < g) throw new RangeError('"value" argument is out of bounds');
    if (o + l > f.length) throw new RangeError("Index out of range");
  }
  s.prototype.writeUintLE = s.prototype.writeUIntLE = function(i, o, l, p) {
    if (i = +i, o = o >>> 0, l = l >>> 0, !p) {
      var g = Math.pow(2, 8 * l) - 1;
      D(this, i, o, l, g, 0);
    }
    var w = 1, I = 0;
    for (this[o] = i & 255; ++I < l && (w *= 256); )
      this[o + I] = i / w & 255;
    return o + l;
  }, s.prototype.writeUintBE = s.prototype.writeUIntBE = function(i, o, l, p) {
    if (i = +i, o = o >>> 0, l = l >>> 0, !p) {
      var g = Math.pow(2, 8 * l) - 1;
      D(this, i, o, l, g, 0);
    }
    var w = l - 1, I = 1;
    for (this[o + w] = i & 255; --w >= 0 && (I *= 256); )
      this[o + w] = i / I & 255;
    return o + l;
  }, s.prototype.writeUint8 = s.prototype.writeUInt8 = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 1, 255, 0), this[o] = i & 255, o + 1;
  }, s.prototype.writeUint16LE = s.prototype.writeUInt16LE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 2, 65535, 0), this[o] = i & 255, this[o + 1] = i >>> 8, o + 2;
  }, s.prototype.writeUint16BE = s.prototype.writeUInt16BE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 2, 65535, 0), this[o] = i >>> 8, this[o + 1] = i & 255, o + 2;
  }, s.prototype.writeUint32LE = s.prototype.writeUInt32LE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 4, 4294967295, 0), this[o + 3] = i >>> 24, this[o + 2] = i >>> 16, this[o + 1] = i >>> 8, this[o] = i & 255, o + 4;
  }, s.prototype.writeUint32BE = s.prototype.writeUInt32BE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 4, 4294967295, 0), this[o] = i >>> 24, this[o + 1] = i >>> 16, this[o + 2] = i >>> 8, this[o + 3] = i & 255, o + 4;
  }, s.prototype.writeIntLE = function(i, o, l, p) {
    if (i = +i, o = o >>> 0, !p) {
      var g = Math.pow(2, 8 * l - 1);
      D(this, i, o, l, g - 1, -g);
    }
    var w = 0, I = 1, k = 0;
    for (this[o] = i & 255; ++w < l && (I *= 256); )
      i < 0 && k === 0 && this[o + w - 1] !== 0 && (k = 1), this[o + w] = (i / I >> 0) - k & 255;
    return o + l;
  }, s.prototype.writeIntBE = function(i, o, l, p) {
    if (i = +i, o = o >>> 0, !p) {
      var g = Math.pow(2, 8 * l - 1);
      D(this, i, o, l, g - 1, -g);
    }
    var w = l - 1, I = 1, k = 0;
    for (this[o + w] = i & 255; --w >= 0 && (I *= 256); )
      i < 0 && k === 0 && this[o + w + 1] !== 0 && (k = 1), this[o + w] = (i / I >> 0) - k & 255;
    return o + l;
  }, s.prototype.writeInt8 = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 1, 127, -128), i < 0 && (i = 255 + i + 1), this[o] = i & 255, o + 1;
  }, s.prototype.writeInt16LE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 2, 32767, -32768), this[o] = i & 255, this[o + 1] = i >>> 8, o + 2;
  }, s.prototype.writeInt16BE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 2, 32767, -32768), this[o] = i >>> 8, this[o + 1] = i & 255, o + 2;
  }, s.prototype.writeInt32LE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 4, 2147483647, -2147483648), this[o] = i & 255, this[o + 1] = i >>> 8, this[o + 2] = i >>> 16, this[o + 3] = i >>> 24, o + 4;
  }, s.prototype.writeInt32BE = function(i, o, l) {
    return i = +i, o = o >>> 0, l || D(this, i, o, 4, 2147483647, -2147483648), i < 0 && (i = 4294967295 + i + 1), this[o] = i >>> 24, this[o + 1] = i >>> 16, this[o + 2] = i >>> 8, this[o + 3] = i & 255, o + 4;
  };
  function oe(f, i, o, l, p, g) {
    if (o + l > f.length) throw new RangeError("Index out of range");
    if (o < 0) throw new RangeError("Index out of range");
  }
  function ae(f, i, o, l, p) {
    return i = +i, o = o >>> 0, p || oe(f, i, o, 4), r.write(f, i, o, l, 23, 4), o + 4;
  }
  s.prototype.writeFloatLE = function(i, o, l) {
    return ae(this, i, o, !0, l);
  }, s.prototype.writeFloatBE = function(i, o, l) {
    return ae(this, i, o, !1, l);
  };
  function se(f, i, o, l, p) {
    return i = +i, o = o >>> 0, p || oe(f, i, o, 8), r.write(f, i, o, l, 52, 8), o + 8;
  }
  s.prototype.writeDoubleLE = function(i, o, l) {
    return se(this, i, o, !0, l);
  }, s.prototype.writeDoubleBE = function(i, o, l) {
    return se(this, i, o, !1, l);
  }, s.prototype.copy = function(i, o, l, p) {
    if (!s.isBuffer(i)) throw new TypeError("argument should be a Buffer");
    if (l || (l = 0), !p && p !== 0 && (p = this.length), o >= i.length && (o = i.length), o || (o = 0), p > 0 && p < l && (p = l), p === l || i.length === 0 || this.length === 0) return 0;
    if (o < 0)
      throw new RangeError("targetStart out of bounds");
    if (l < 0 || l >= this.length) throw new RangeError("Index out of range");
    if (p < 0) throw new RangeError("sourceEnd out of bounds");
    p > this.length && (p = this.length), i.length - o < p - l && (p = i.length - o + l);
    var g = p - l;
    return this === i && typeof Uint8Array.prototype.copyWithin == "function" ? this.copyWithin(o, l, p) : Uint8Array.prototype.set.call(
      i,
      this.subarray(l, p),
      o
    ), g;
  }, s.prototype.fill = function(i, o, l, p) {
    if (typeof i == "string") {
      if (typeof o == "string" ? (p = o, o = 0, l = this.length) : typeof l == "string" && (p = l, l = this.length), p !== void 0 && typeof p != "string")
        throw new TypeError("encoding must be a string");
      if (typeof p == "string" && !s.isEncoding(p))
        throw new TypeError("Unknown encoding: " + p);
      if (i.length === 1) {
        var g = i.charCodeAt(0);
        (p === "utf8" && g < 128 || p === "latin1") && (i = g);
      }
    } else typeof i == "number" ? i = i & 255 : typeof i == "boolean" && (i = Number(i));
    if (o < 0 || this.length < o || this.length < l)
      throw new RangeError("Out of range index");
    if (l <= o)
      return this;
    o = o >>> 0, l = l === void 0 ? this.length : l >>> 0, i || (i = 0);
    var w;
    if (typeof i == "number")
      for (w = o; w < l; ++w)
        this[w] = i;
    else {
      var I = s.isBuffer(i) ? i : s.from(i, p), k = I.length;
      if (k === 0)
        throw new TypeError('The value "' + i + '" is invalid for argument "value"');
      for (w = 0; w < l - o; ++w)
        this[w + o] = I[w % k];
    }
    return this;
  };
  var Xn = /[^+/0-9A-Za-z-_]/g;
  function Yn(f) {
    if (f = f.split("=")[0], f = f.trim().replace(Xn, ""), f.length < 2) return "";
    for (; f.length % 4 !== 0; )
      f = f + "=";
    return f;
  }
  function Ot(f, i) {
    i = i || 1 / 0;
    for (var o, l = f.length, p = null, g = [], w = 0; w < l; ++w) {
      if (o = f.charCodeAt(w), o > 55295 && o < 57344) {
        if (!p) {
          if (o > 56319) {
            (i -= 3) > -1 && g.push(239, 191, 189);
            continue;
          } else if (w + 1 === l) {
            (i -= 3) > -1 && g.push(239, 191, 189);
            continue;
          }
          p = o;
          continue;
        }
        if (o < 56320) {
          (i -= 3) > -1 && g.push(239, 191, 189), p = o;
          continue;
        }
        o = (p - 55296 << 10 | o - 56320) + 65536;
      } else p && (i -= 3) > -1 && g.push(239, 191, 189);
      if (p = null, o < 128) {
        if ((i -= 1) < 0) break;
        g.push(o);
      } else if (o < 2048) {
        if ((i -= 2) < 0) break;
        g.push(
          o >> 6 | 192,
          o & 63 | 128
        );
      } else if (o < 65536) {
        if ((i -= 3) < 0) break;
        g.push(
          o >> 12 | 224,
          o >> 6 & 63 | 128,
          o & 63 | 128
        );
      } else if (o < 1114112) {
        if ((i -= 4) < 0) break;
        g.push(
          o >> 18 | 240,
          o >> 12 & 63 | 128,
          o >> 6 & 63 | 128,
          o & 63 | 128
        );
      } else
        throw new Error("Invalid code point");
    }
    return g;
  }
  function Jn(f) {
    for (var i = [], o = 0; o < f.length; ++o)
      i.push(f.charCodeAt(o) & 255);
    return i;
  }
  function Kn(f, i) {
    for (var o, l, p, g = [], w = 0; w < f.length && !((i -= 2) < 0); ++w)
      o = f.charCodeAt(w), l = o >> 8, p = o % 256, g.push(p), g.push(l);
    return g;
  }
  function ce(f) {
    return t.toByteArray(Yn(f));
  }
  function yt(f, i, o, l) {
    for (var p = 0; p < l && !(p + o >= i.length || p >= f.length); ++p)
      i[p + o] = f[p];
    return p;
  }
  function q(f, i) {
    return f instanceof i || f != null && f.constructor != null && f.constructor.name != null && f.constructor.name === i.name;
  }
  function Ut(f) {
    return f !== f;
  }
  var Zn = function() {
    for (var f = "0123456789abcdef", i = new Array(256), o = 0; o < 16; ++o)
      for (var l = o * 16, p = 0; p < 16; ++p)
        i[l + p] = f[o] + f[p];
    return i;
  }();
})(br);
var cr = { exports: {} };
typeof Object.create == "function" ? cr.exports = function(t, r) {
  r && (t.super_ = r, t.prototype = Object.create(r.prototype, {
    constructor: {
      value: t,
      enumerable: !1,
      writable: !0,
      configurable: !0
    }
  }));
} : cr.exports = function(t, r) {
  if (r) {
    t.super_ = r;
    var n = function() {
    };
    n.prototype = r.prototype, t.prototype = new n(), t.prototype.constructor = t;
  }
};
var ui = cr.exports, ur = { exports: {} };
/*! safe-buffer. MIT License. Feross Aboukhadijeh <https://feross.org/opensource> */
(function(e, t) {
  var r = br, n = r.Buffer;
  function a(u, s) {
    for (var h in u)
      s[h] = u[h];
  }
  n.from && n.alloc && n.allocUnsafe && n.allocUnsafeSlow ? e.exports = r : (a(r, t), t.Buffer = c);
  function c(u, s, h) {
    return n(u, s, h);
  }
  c.prototype = Object.create(n.prototype), a(n, c), c.from = function(u, s, h) {
    if (typeof u == "number")
      throw new TypeError("Argument must not be a number");
    return n(u, s, h);
  }, c.alloc = function(u, s, h) {
    if (typeof u != "number")
      throw new TypeError("Argument must be a number");
    var d = n(u);
    return s !== void 0 ? typeof h == "string" ? d.fill(s, h) : d.fill(s) : d.fill(0), d;
  }, c.allocUnsafe = function(u) {
    if (typeof u != "number")
      throw new TypeError("Argument must be a number");
    return n(u);
  }, c.allocUnsafeSlow = function(u) {
    if (typeof u != "number")
      throw new TypeError("Argument must be a number");
    return r.SlowBuffer(u);
  };
})(ur, ur.exports);
var Er = ur.exports, fi = {}.toString, li = Array.isArray || function(e) {
  return fi.call(e) == "[object Array]";
}, at = TypeError, tn = Object, hi = Error, pi = EvalError, yi = RangeError, di = ReferenceError, rn = SyntaxError, mi = URIError, gi = Math.abs, wi = Math.floor, bi = Math.max, vi = Math.min, Ei = Math.pow, xi = Math.round, Ai = Number.isNaN || function(t) {
  return t !== t;
}, Si = Ai, Bi = function(t) {
  return Si(t) || t === 0 ? t : t < 0 ? -1 : 1;
}, _i = Object.getOwnPropertyDescriptor, bt = _i;
if (bt)
  try {
    bt([], "length");
  } catch {
    bt = null;
  }
var st = bt, vt = Object.defineProperty || !1;
if (vt)
  try {
    vt({}, "a", { value: 1 });
  } catch {
    vt = !1;
  }
var Ft = vt, Ct, le;
function en() {
  return le || (le = 1, Ct = function() {
    if (typeof Symbol != "function" || typeof Object.getOwnPropertySymbols != "function")
      return !1;
    if (typeof Symbol.iterator == "symbol")
      return !0;
    var t = {}, r = Symbol("test"), n = Object(r);
    if (typeof r == "string" || Object.prototype.toString.call(r) !== "[object Symbol]" || Object.prototype.toString.call(n) !== "[object Symbol]")
      return !1;
    var a = 42;
    t[r] = a;
    for (var c in t)
      return !1;
    if (typeof Object.keys == "function" && Object.keys(t).length !== 0 || typeof Object.getOwnPropertyNames == "function" && Object.getOwnPropertyNames(t).length !== 0)
      return !1;
    var u = Object.getOwnPropertySymbols(t);
    if (u.length !== 1 || u[0] !== r || !Object.prototype.propertyIsEnumerable.call(t, r))
      return !1;
    if (typeof Object.getOwnPropertyDescriptor == "function") {
      var s = (
        /** @type {PropertyDescriptor} */
        Object.getOwnPropertyDescriptor(t, r)
      );
      if (s.value !== a || s.enumerable !== !0)
        return !1;
    }
    return !0;
  }), Ct;
}
var jt, he;
function Ii() {
  if (he) return jt;
  he = 1;
  var e = typeof Symbol < "u" && Symbol, t = en();
  return jt = function() {
    return typeof e != "function" || typeof Symbol != "function" || typeof e("foo") != "symbol" || typeof Symbol("bar") != "symbol" ? !1 : t();
  }, jt;
}
var Nt, pe;
function nn() {
  return pe || (pe = 1, Nt = typeof Reflect < "u" && Reflect.getPrototypeOf || null), Nt;
}
var Dt, ye;
function on() {
  if (ye) return Dt;
  ye = 1;
  var e = tn;
  return Dt = e.getPrototypeOf || null, Dt;
}
var Fi = "Function.prototype.bind called on incompatible ", $i = Object.prototype.toString, Pi = Math.max, ki = "[object Function]", de = function(t, r) {
  for (var n = [], a = 0; a < t.length; a += 1)
    n[a] = t[a];
  for (var c = 0; c < r.length; c += 1)
    n[c + t.length] = r[c];
  return n;
}, Ti = function(t, r) {
  for (var n = [], a = r, c = 0; a < t.length; a += 1, c += 1)
    n[c] = t[a];
  return n;
}, Oi = function(e, t) {
  for (var r = "", n = 0; n < e.length; n += 1)
    r += e[n], n + 1 < e.length && (r += t);
  return r;
}, Ui = function(t) {
  var r = this;
  if (typeof r != "function" || $i.apply(r) !== ki)
    throw new TypeError(Fi + r);
  for (var n = Ti(arguments, 1), a, c = function() {
    if (this instanceof a) {
      var m = r.apply(
        this,
        de(n, arguments)
      );
      return Object(m) === m ? m : this;
    }
    return r.apply(
      t,
      de(n, arguments)
    );
  }, u = Pi(0, r.length - n.length), s = [], h = 0; h < u; h++)
    s[h] = "$" + h;
  if (a = Function("binder", "return function (" + Oi(s, ",") + "){ return binder.apply(this,arguments); }")(c), r.prototype) {
    var d = function() {
    };
    d.prototype = r.prototype, a.prototype = new d(), d.prototype = null;
  }
  return a;
}, Ri = Ui, ct = Function.prototype.bind || Ri, xr = Function.prototype.call, Mt, me;
function Ar() {
  return me || (me = 1, Mt = Function.prototype.apply), Mt;
}
var Ci = typeof Reflect < "u" && Reflect && Reflect.apply, ji = ct, Ni = Ar(), Di = xr, Mi = Ci, an = Mi || ji.call(Di, Ni), Gi = ct, Li = at, Hi = xr, qi = an, Sr = function(t) {
  if (t.length < 1 || typeof t[0] != "function")
    throw new Li("a function is required");
  return qi(Gi, Hi, t);
}, Gt, ge;
function zi() {
  if (ge) return Gt;
  ge = 1;
  var e = Sr, t = st, r;
  try {
    r = /** @type {{ __proto__?: typeof Array.prototype }} */
    [].__proto__ === Array.prototype;
  } catch (u) {
    if (!u || typeof u != "object" || !("code" in u) || u.code !== "ERR_PROTO_ACCESS")
      throw u;
  }
  var n = !!r && t && t(
    Object.prototype,
    /** @type {keyof typeof Object.prototype} */
    "__proto__"
  ), a = Object, c = a.getPrototypeOf;
  return Gt = n && typeof n.get == "function" ? e([n.get]) : typeof c == "function" ? (
    /** @type {import('./get')} */
    function(s) {
      return c(s == null ? s : a(s));
    }
  ) : !1, Gt;
}
var Lt, we;
function sn() {
  if (we) return Lt;
  we = 1;
  var e = nn(), t = on(), r = zi();
  return Lt = e ? function(a) {
    return e(a);
  } : t ? function(a) {
    if (!a || typeof a != "object" && typeof a != "function")
      throw new TypeError("getProto: not an object");
    return t(a);
  } : r ? function(a) {
    return r(a);
  } : null, Lt;
}
var Ht, be;
function Wi() {
  if (be) return Ht;
  be = 1;
  var e = Function.prototype.call, t = Object.prototype.hasOwnProperty, r = ct;
  return Ht = r.call(e, t), Ht;
}
var S, Vi = tn, Xi = hi, Yi = pi, Ji = yi, Ki = di, et = rn, rt = at, Zi = mi, Qi = gi, to = wi, ro = bi, eo = vi, no = Ei, io = xi, oo = Bi, cn = Function, qt = function(e) {
  try {
    return cn('"use strict"; return (' + e + ").constructor;")();
  } catch {
  }
}, it = st, ao = Ft, zt = function() {
  throw new rt();
}, so = it ? function() {
  try {
    return arguments.callee, zt;
  } catch {
    try {
      return it(arguments, "callee").get;
    } catch {
      return zt;
    }
  }
}() : zt, Z = Ii()(), C = sn(), co = on(), uo = nn(), un = Ar(), ut = xr, tt = {}, fo = typeof Uint8Array > "u" || !C ? S : C(Uint8Array), J = {
  __proto__: null,
  "%AggregateError%": typeof AggregateError > "u" ? S : AggregateError,
  "%Array%": Array,
  "%ArrayBuffer%": typeof ArrayBuffer > "u" ? S : ArrayBuffer,
  "%ArrayIteratorPrototype%": Z && C ? C([][Symbol.iterator]()) : S,
  "%AsyncFromSyncIteratorPrototype%": S,
  "%AsyncFunction%": tt,
  "%AsyncGenerator%": tt,
  "%AsyncGeneratorFunction%": tt,
  "%AsyncIteratorPrototype%": tt,
  "%Atomics%": typeof Atomics > "u" ? S : Atomics,
  "%BigInt%": typeof BigInt > "u" ? S : BigInt,
  "%BigInt64Array%": typeof BigInt64Array > "u" ? S : BigInt64Array,
  "%BigUint64Array%": typeof BigUint64Array > "u" ? S : BigUint64Array,
  "%Boolean%": Boolean,
  "%DataView%": typeof DataView > "u" ? S : DataView,
  "%Date%": Date,
  "%decodeURI%": decodeURI,
  "%decodeURIComponent%": decodeURIComponent,
  "%encodeURI%": encodeURI,
  "%encodeURIComponent%": encodeURIComponent,
  "%Error%": Xi,
  "%eval%": eval,
  // eslint-disable-line no-eval
  "%EvalError%": Yi,
  "%Float16Array%": typeof Float16Array > "u" ? S : Float16Array,
  "%Float32Array%": typeof Float32Array > "u" ? S : Float32Array,
  "%Float64Array%": typeof Float64Array > "u" ? S : Float64Array,
  "%FinalizationRegistry%": typeof FinalizationRegistry > "u" ? S : FinalizationRegistry,
  "%Function%": cn,
  "%GeneratorFunction%": tt,
  "%Int8Array%": typeof Int8Array > "u" ? S : Int8Array,
  "%Int16Array%": typeof Int16Array > "u" ? S : Int16Array,
  "%Int32Array%": typeof Int32Array > "u" ? S : Int32Array,
  "%isFinite%": isFinite,
  "%isNaN%": isNaN,
  "%IteratorPrototype%": Z && C ? C(C([][Symbol.iterator]())) : S,
  "%JSON%": typeof JSON == "object" ? JSON : S,
  "%Map%": typeof Map > "u" ? S : Map,
  "%MapIteratorPrototype%": typeof Map > "u" || !Z || !C ? S : C((/* @__PURE__ */ new Map())[Symbol.iterator]()),
  "%Math%": Math,
  "%Number%": Number,
  "%Object%": Vi,
  "%Object.getOwnPropertyDescriptor%": it,
  "%parseFloat%": parseFloat,
  "%parseInt%": parseInt,
  "%Promise%": typeof Promise > "u" ? S : Promise,
  "%Proxy%": typeof Proxy > "u" ? S : Proxy,
  "%RangeError%": Ji,
  "%ReferenceError%": Ki,
  "%Reflect%": typeof Reflect > "u" ? S : Reflect,
  "%RegExp%": RegExp,
  "%Set%": typeof Set > "u" ? S : Set,
  "%SetIteratorPrototype%": typeof Set > "u" || !Z || !C ? S : C((/* @__PURE__ */ new Set())[Symbol.iterator]()),
  "%SharedArrayBuffer%": typeof SharedArrayBuffer > "u" ? S : SharedArrayBuffer,
  "%String%": String,
  "%StringIteratorPrototype%": Z && C ? C(""[Symbol.iterator]()) : S,
  "%Symbol%": Z ? Symbol : S,
  "%SyntaxError%": et,
  "%ThrowTypeError%": so,
  "%TypedArray%": fo,
  "%TypeError%": rt,
  "%Uint8Array%": typeof Uint8Array > "u" ? S : Uint8Array,
  "%Uint8ClampedArray%": typeof Uint8ClampedArray > "u" ? S : Uint8ClampedArray,
  "%Uint16Array%": typeof Uint16Array > "u" ? S : Uint16Array,
  "%Uint32Array%": typeof Uint32Array > "u" ? S : Uint32Array,
  "%URIError%": Zi,
  "%WeakMap%": typeof WeakMap > "u" ? S : WeakMap,
  "%WeakRef%": typeof WeakRef > "u" ? S : WeakRef,
  "%WeakSet%": typeof WeakSet > "u" ? S : WeakSet,
  "%Function.prototype.call%": ut,
  "%Function.prototype.apply%": un,
  "%Object.defineProperty%": ao,
  "%Object.getPrototypeOf%": co,
  "%Math.abs%": Qi,
  "%Math.floor%": to,
  "%Math.max%": ro,
  "%Math.min%": eo,
  "%Math.pow%": no,
  "%Math.round%": io,
  "%Math.sign%": oo,
  "%Reflect.getPrototypeOf%": uo
};
if (C)
  try {
    null.error;
  } catch (e) {
    var lo = C(C(e));
    J["%Error.prototype%"] = lo;
  }
var ho = function e(t) {
  var r;
  if (t === "%AsyncFunction%")
    r = qt("async function () {}");
  else if (t === "%GeneratorFunction%")
    r = qt("function* () {}");
  else if (t === "%AsyncGeneratorFunction%")
    r = qt("async function* () {}");
  else if (t === "%AsyncGenerator%") {
    var n = e("%AsyncGeneratorFunction%");
    n && (r = n.prototype);
  } else if (t === "%AsyncIteratorPrototype%") {
    var a = e("%AsyncGenerator%");
    a && C && (r = C(a.prototype));
  }
  return J[t] = r, r;
}, ve = {
  __proto__: null,
  "%ArrayBufferPrototype%": ["ArrayBuffer", "prototype"],
  "%ArrayPrototype%": ["Array", "prototype"],
  "%ArrayProto_entries%": ["Array", "prototype", "entries"],
  "%ArrayProto_forEach%": ["Array", "prototype", "forEach"],
  "%ArrayProto_keys%": ["Array", "prototype", "keys"],
  "%ArrayProto_values%": ["Array", "prototype", "values"],
  "%AsyncFunctionPrototype%": ["AsyncFunction", "prototype"],
  "%AsyncGenerator%": ["AsyncGeneratorFunction", "prototype"],
  "%AsyncGeneratorPrototype%": ["AsyncGeneratorFunction", "prototype", "prototype"],
  "%BooleanPrototype%": ["Boolean", "prototype"],
  "%DataViewPrototype%": ["DataView", "prototype"],
  "%DatePrototype%": ["Date", "prototype"],
  "%ErrorPrototype%": ["Error", "prototype"],
  "%EvalErrorPrototype%": ["EvalError", "prototype"],
  "%Float32ArrayPrototype%": ["Float32Array", "prototype"],
  "%Float64ArrayPrototype%": ["Float64Array", "prototype"],
  "%FunctionPrototype%": ["Function", "prototype"],
  "%Generator%": ["GeneratorFunction", "prototype"],
  "%GeneratorPrototype%": ["GeneratorFunction", "prototype", "prototype"],
  "%Int8ArrayPrototype%": ["Int8Array", "prototype"],
  "%Int16ArrayPrototype%": ["Int16Array", "prototype"],
  "%Int32ArrayPrototype%": ["Int32Array", "prototype"],
  "%JSONParse%": ["JSON", "parse"],
  "%JSONStringify%": ["JSON", "stringify"],
  "%MapPrototype%": ["Map", "prototype"],
  "%NumberPrototype%": ["Number", "prototype"],
  "%ObjectPrototype%": ["Object", "prototype"],
  "%ObjProto_toString%": ["Object", "prototype", "toString"],
  "%ObjProto_valueOf%": ["Object", "prototype", "valueOf"],
  "%PromisePrototype%": ["Promise", "prototype"],
  "%PromiseProto_then%": ["Promise", "prototype", "then"],
  "%Promise_all%": ["Promise", "all"],
  "%Promise_reject%": ["Promise", "reject"],
  "%Promise_resolve%": ["Promise", "resolve"],
  "%RangeErrorPrototype%": ["RangeError", "prototype"],
  "%ReferenceErrorPrototype%": ["ReferenceError", "prototype"],
  "%RegExpPrototype%": ["RegExp", "prototype"],
  "%SetPrototype%": ["Set", "prototype"],
  "%SharedArrayBufferPrototype%": ["SharedArrayBuffer", "prototype"],
  "%StringPrototype%": ["String", "prototype"],
  "%SymbolPrototype%": ["Symbol", "prototype"],
  "%SyntaxErrorPrototype%": ["SyntaxError", "prototype"],
  "%TypedArrayPrototype%": ["TypedArray", "prototype"],
  "%TypeErrorPrototype%": ["TypeError", "prototype"],
  "%Uint8ArrayPrototype%": ["Uint8Array", "prototype"],
  "%Uint8ClampedArrayPrototype%": ["Uint8ClampedArray", "prototype"],
  "%Uint16ArrayPrototype%": ["Uint16Array", "prototype"],
  "%Uint32ArrayPrototype%": ["Uint32Array", "prototype"],
  "%URIErrorPrototype%": ["URIError", "prototype"],
  "%WeakMapPrototype%": ["WeakMap", "prototype"],
  "%WeakSetPrototype%": ["WeakSet", "prototype"]
}, ft = ct, xt = Wi(), po = ft.call(ut, Array.prototype.concat), yo = ft.call(un, Array.prototype.splice), Ee = ft.call(ut, String.prototype.replace), At = ft.call(ut, String.prototype.slice), mo = ft.call(ut, RegExp.prototype.exec), go = /[^%.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|%$))/g, wo = /\\(\\)?/g, bo = function(t) {
  var r = At(t, 0, 1), n = At(t, -1);
  if (r === "%" && n !== "%")
    throw new et("invalid intrinsic syntax, expected closing `%`");
  if (n === "%" && r !== "%")
    throw new et("invalid intrinsic syntax, expected opening `%`");
  var a = [];
  return Ee(t, go, function(c, u, s, h) {
    a[a.length] = s ? Ee(h, wo, "$1") : u || c;
  }), a;
}, vo = function(t, r) {
  var n = t, a;
  if (xt(ve, n) && (a = ve[n], n = "%" + a[0] + "%"), xt(J, n)) {
    var c = J[n];
    if (c === tt && (c = ho(n)), typeof c > "u" && !r)
      throw new rt("intrinsic " + t + " exists, but is not available. Please file an issue!");
    return {
      alias: a,
      name: n,
      value: c
    };
  }
  throw new et("intrinsic " + t + " does not exist!");
}, fn = function(t, r) {
  if (typeof t != "string" || t.length === 0)
    throw new rt("intrinsic name must be a non-empty string");
  if (arguments.length > 1 && typeof r != "boolean")
    throw new rt('"allowMissing" argument must be a boolean');
  if (mo(/^%?[^%]*%?$/, t) === null)
    throw new et("`%` may not be present anywhere but at the beginning and end of the intrinsic name");
  var n = bo(t), a = n.length > 0 ? n[0] : "", c = vo("%" + a + "%", r), u = c.name, s = c.value, h = !1, d = c.alias;
  d && (a = d[0], yo(n, po([0, 1], d)));
  for (var m = 1, y = !0; m < n.length; m += 1) {
    var b = n[m], E = At(b, 0, 1), O = At(b, -1);
    if ((E === '"' || E === "'" || E === "`" || O === '"' || O === "'" || O === "`") && E !== O)
      throw new et("property names with quotes must have matching quotes");
    if ((b === "constructor" || !y) && (h = !0), a += "." + b, u = "%" + a + "%", xt(J, u))
      s = J[u];
    else if (s != null) {
      if (!(b in s)) {
        if (!r)
          throw new rt("base intrinsic for " + t + " exists, but the property is not available.");
        return;
      }
      if (it && m + 1 >= n.length) {
        var F = it(s, b);
        y = !!F, y && "get" in F && !("originalValue" in F.get) ? s = F.get : s = s[b];
      } else
        y = xt(s, b), s = s[b];
      y && !h && (J[u] = s);
    }
  }
  return s;
}, ln = fn, hn = Sr, Eo = hn([ln("%String.prototype.indexOf%")]), pn = function(t, r) {
  var n = (
    /** @type {(this: unknown, ...args: unknown[]) => unknown} */
    ln(t, !!r)
  );
  return typeof n == "function" && Eo(t, ".prototype.") > -1 ? hn(
    /** @type {const} */
    [n]
  ) : n;
}, Wt, xe;
function xo() {
  if (xe) return Wt;
  xe = 1;
  var e = Function.prototype.toString, t = typeof Reflect == "object" && Reflect !== null && Reflect.apply, r, n;
  if (typeof t == "function" && typeof Object.defineProperty == "function")
    try {
      r = Object.defineProperty({}, "length", {
        get: function() {
          throw n;
        }
      }), n = {}, t(function() {
        throw 42;
      }, null, r);
    } catch (_) {
      _ !== n && (t = null);
    }
  else
    t = null;
  var a = /^\s*class\b/, c = function(v) {
    try {
      var P = e.call(v);
      return a.test(P);
    } catch {
      return !1;
    }
  }, u = function(v) {
    try {
      return c(v) ? !1 : (e.call(v), !0);
    } catch {
      return !1;
    }
  }, s = Object.prototype.toString, h = "[object Object]", d = "[object Function]", m = "[object GeneratorFunction]", y = "[object HTMLAllCollection]", b = "[object HTML document.all class]", E = "[object HTMLCollection]", O = typeof Symbol == "function" && !!Symbol.toStringTag, F = !(0 in [,]), A = function() {
    return !1;
  };
  if (typeof document == "object") {
    var x = document.all;
    s.call(x) === s.call(document.all) && (A = function(v) {
      if ((F || !v) && (typeof v > "u" || typeof v == "object"))
        try {
          var P = s.call(v);
          return (P === y || P === b || P === E || P === h) && v("") == null;
        } catch {
        }
      return !1;
    });
  }
  return Wt = t ? function(v) {
    if (A(v))
      return !0;
    if (!v || typeof v != "function" && typeof v != "object")
      return !1;
    try {
      t(v, null, r);
    } catch (P) {
      if (P !== n)
        return !1;
    }
    return !c(v) && u(v);
  } : function(v) {
    if (A(v))
      return !0;
    if (!v || typeof v != "function" && typeof v != "object")
      return !1;
    if (O)
      return u(v);
    if (c(v))
      return !1;
    var P = s.call(v);
    return P !== d && P !== m && !/^\[object HTML/.test(P) ? !1 : u(v);
  }, Wt;
}
var Vt, Ae;
function Ao() {
  if (Ae) return Vt;
  Ae = 1;
  var e = xo(), t = Object.prototype.toString, r = Object.prototype.hasOwnProperty, n = function(h, d, m) {
    for (var y = 0, b = h.length; y < b; y++)
      r.call(h, y) && (m == null ? d(h[y], y, h) : d.call(m, h[y], y, h));
  }, a = function(h, d, m) {
    for (var y = 0, b = h.length; y < b; y++)
      m == null ? d(h.charAt(y), y, h) : d.call(m, h.charAt(y), y, h);
  }, c = function(h, d, m) {
    for (var y in h)
      r.call(h, y) && (m == null ? d(h[y], y, h) : d.call(m, h[y], y, h));
  };
  function u(s) {
    return t.call(s) === "[object Array]";
  }
  return Vt = function(h, d, m) {
    if (!e(d))
      throw new TypeError("iterator must be a function");
    var y;
    arguments.length >= 3 && (y = m), u(h) ? n(h, d, y) : typeof h == "string" ? a(h, d, y) : c(h, d, y);
  }, Vt;
}
var Xt, Se;
function So() {
  return Se || (Se = 1, Xt = [
    "Float16Array",
    "Float32Array",
    "Float64Array",
    "Int8Array",
    "Int16Array",
    "Int32Array",
    "Uint8Array",
    "Uint8ClampedArray",
    "Uint16Array",
    "Uint32Array",
    "BigInt64Array",
    "BigUint64Array"
  ]), Xt;
}
var Yt, Be;
function Bo() {
  if (Be) return Yt;
  Be = 1;
  var e = So(), t = typeof globalThis > "u" ? Ke : globalThis;
  return Yt = function() {
    for (var n = [], a = 0; a < e.length; a++)
      typeof t[e[a]] == "function" && (n[n.length] = e[a]);
    return n;
  }, Yt;
}
var Jt = { exports: {} }, Kt, _e;
function _o() {
  if (_e) return Kt;
  _e = 1;
  var e = Ft, t = rn, r = at, n = st;
  return Kt = function(c, u, s) {
    if (!c || typeof c != "object" && typeof c != "function")
      throw new r("`obj` must be an object or a function`");
    if (typeof u != "string" && typeof u != "symbol")
      throw new r("`property` must be a string or a symbol`");
    if (arguments.length > 3 && typeof arguments[3] != "boolean" && arguments[3] !== null)
      throw new r("`nonEnumerable`, if provided, must be a boolean or null");
    if (arguments.length > 4 && typeof arguments[4] != "boolean" && arguments[4] !== null)
      throw new r("`nonWritable`, if provided, must be a boolean or null");
    if (arguments.length > 5 && typeof arguments[5] != "boolean" && arguments[5] !== null)
      throw new r("`nonConfigurable`, if provided, must be a boolean or null");
    if (arguments.length > 6 && typeof arguments[6] != "boolean")
      throw new r("`loose`, if provided, must be a boolean");
    var h = arguments.length > 3 ? arguments[3] : null, d = arguments.length > 4 ? arguments[4] : null, m = arguments.length > 5 ? arguments[5] : null, y = arguments.length > 6 ? arguments[6] : !1, b = !!n && n(c, u);
    if (e)
      e(c, u, {
        configurable: m === null && b ? b.configurable : !m,
        enumerable: h === null && b ? b.enumerable : !h,
        value: s,
        writable: d === null && b ? b.writable : !d
      });
    else if (y || !h && !d && !m)
      c[u] = s;
    else
      throw new t("This environment does not support defining a property as non-configurable, non-writable, or non-enumerable.");
  }, Kt;
}
var Zt, Ie;
function Io() {
  if (Ie) return Zt;
  Ie = 1;
  var e = Ft, t = function() {
    return !!e;
  };
  return t.hasArrayLengthDefineBug = function() {
    if (!e)
      return null;
    try {
      return e([], "length", { value: 1 }).length !== 1;
    } catch {
      return !0;
    }
  }, Zt = t, Zt;
}
var Qt, Fe;
function Fo() {
  if (Fe) return Qt;
  Fe = 1;
  var e = fn, t = _o(), r = Io()(), n = st, a = at, c = e("%Math.floor%");
  return Qt = function(s, h) {
    if (typeof s != "function")
      throw new a("`fn` is not a function");
    if (typeof h != "number" || h < 0 || h > 4294967295 || c(h) !== h)
      throw new a("`length` must be a positive 32-bit integer");
    var d = arguments.length > 2 && !!arguments[2], m = !0, y = !0;
    if ("length" in s && n) {
      var b = n(s, "length");
      b && !b.configurable && (m = !1), b && !b.writable && (y = !1);
    }
    return (m || y || !d) && (r ? t(
      /** @type {Parameters<define>[0]} */
      s,
      "length",
      h,
      !0,
      !0
    ) : t(
      /** @type {Parameters<define>[0]} */
      s,
      "length",
      h
    )), s;
  }, Qt;
}
var tr, $e;
function $o() {
  if ($e) return tr;
  $e = 1;
  var e = ct, t = Ar(), r = an;
  return tr = function() {
    return r(e, t, arguments);
  }, tr;
}
var Pe;
function Po() {
  return Pe || (Pe = 1, function(e) {
    var t = Fo(), r = Ft, n = Sr, a = $o();
    e.exports = function(u) {
      var s = n(arguments), h = 1 + u.length - (arguments.length - 1);
      return t(
        s,
        h > 0 ? h : 0,
        !0
      );
    }, r ? r(e.exports, "apply", { value: a }) : e.exports.apply = a;
  }(Jt)), Jt.exports;
}
var rr, ke;
function ko() {
  if (ke) return rr;
  ke = 1;
  var e = en();
  return rr = function() {
    return e() && !!Symbol.toStringTag;
  }, rr;
}
var er, Te;
function To() {
  if (Te) return er;
  Te = 1;
  var e = Ao(), t = Bo(), r = Po(), n = pn, a = st, c = sn(), u = n("Object.prototype.toString"), s = ko()(), h = typeof globalThis > "u" ? Ke : globalThis, d = t(), m = n("String.prototype.slice"), y = n("Array.prototype.indexOf", !0) || function(A, x) {
    for (var _ = 0; _ < A.length; _ += 1)
      if (A[_] === x)
        return _;
    return -1;
  }, b = { __proto__: null };
  s && a && c ? e(d, function(F) {
    var A = new h[F]();
    if (Symbol.toStringTag in A && c) {
      var x = c(A), _ = a(x, Symbol.toStringTag);
      if (!_ && x) {
        var v = c(x);
        _ = a(v, Symbol.toStringTag);
      }
      if (_ && _.get) {
        var P = r(_.get);
        b[
          /** @type {`$${import('.').TypedArrayName}`} */
          "$" + F
        ] = P;
      }
    }
  }) : e(d, function(F) {
    var A = new h[F](), x = A.slice || A.set;
    if (x) {
      var _ = (
        /** @type {import('./types').BoundSlice | import('./types').BoundSet} */
        // @ts-expect-error TODO FIXME
        r(x)
      );
      b[
        /** @type {`$${import('.').TypedArrayName}`} */
        "$" + F
      ] = _;
    }
  });
  var E = function(A) {
    var x = !1;
    return e(
      /** @type {Record<`\$${import('.').TypedArrayName}`, Getter>} */
      b,
      /** @type {(getter: Getter, name: `\$${import('.').TypedArrayName}`) => void} */
      function(_, v) {
        if (!x)
          try {
            "$" + _(A) === v && (x = /** @type {import('.').TypedArrayName} */
            m(v, 1));
          } catch {
          }
      }
    ), x;
  }, O = function(A) {
    var x = !1;
    return e(
      /** @type {Record<`\$${import('.').TypedArrayName}`, Getter>} */
      b,
      /** @type {(getter: Getter, name: `\$${import('.').TypedArrayName}`) => void} */
      function(_, v) {
        if (!x)
          try {
            _(A), x = /** @type {import('.').TypedArrayName} */
            m(v, 1);
          } catch {
          }
      }
    ), x;
  };
  return er = function(A) {
    if (!A || typeof A != "object")
      return !1;
    if (!s) {
      var x = m(u(A), 8, -1);
      return y(d, x) > -1 ? x : x !== "Object" ? !1 : O(A);
    }
    return a ? E(A) : null;
  }, er;
}
var nr, Oe;
function Oo() {
  if (Oe) return nr;
  Oe = 1;
  var e = To();
  return nr = function(r) {
    return !!e(r);
  }, nr;
}
var Uo = at, Ro = pn, Co = Ro("TypedArray.prototype.buffer", !0), jo = Oo(), No = Co || function(t) {
  if (!jo(t))
    throw new Uo("Not a Typed Array");
  return t.buffer;
}, H = Er.Buffer, Do = li, Mo = No, Go = ArrayBuffer.isView || function(t) {
  try {
    return Mo(t), !0;
  } catch {
    return !1;
  }
}, Lo = typeof Uint8Array < "u", yn = typeof ArrayBuffer < "u" && typeof Uint8Array < "u", Ho = yn && (H.prototype instanceof Uint8Array || H.TYPED_ARRAY_SUPPORT), qo = function(t, r) {
  if (H.isBuffer(t))
    return t.constructor && !("isBuffer" in t) ? H.from(t) : t;
  if (typeof t == "string")
    return H.from(t, r);
  if (yn && Go(t)) {
    if (t.byteLength === 0)
      return H.alloc(0);
    if (Ho) {
      var n = H.from(t.buffer, t.byteOffset, t.byteLength);
      if (n.byteLength === t.byteLength)
        return n;
    }
    var a = t instanceof Uint8Array ? t : new Uint8Array(t.buffer, t.byteOffset, t.byteLength), c = H.from(a);
    if (c.length === t.byteLength)
      return c;
  }
  if (Lo && t instanceof Uint8Array)
    return H.from(t);
  var u = Do(t);
  if (u)
    for (var s = 0; s < t.length; s += 1) {
      var h = t[s];
      if (typeof h != "number" || h < 0 || h > 255 || ~~h !== h)
        throw new RangeError("Array items must be numbers in the range 0-255.");
    }
  if (u || H.isBuffer(t) && t.constructor && typeof t.constructor.isBuffer == "function" && t.constructor.isBuffer(t))
    return H.from(t);
  throw new TypeError('The "data" argument must be a string, an Array, a Buffer, a Uint8Array, or a DataView.');
}, zo = Er.Buffer, Wo = qo;
function $t(e, t) {
  this._block = zo.alloc(e), this._finalSize = t, this._blockSize = e, this._len = 0;
}
$t.prototype.update = function(e, t) {
  e = Wo(e, t || "utf8");
  for (var r = this._block, n = this._blockSize, a = e.length, c = this._len, u = 0; u < a; ) {
    for (var s = c % n, h = Math.min(a - u, n - s), d = 0; d < h; d++)
      r[s + d] = e[u + d];
    c += h, u += h, c % n === 0 && this._update(r);
  }
  return this._len += a, this;
};
$t.prototype.digest = function(e) {
  var t = this._len % this._blockSize;
  this._block[t] = 128, this._block.fill(0, t + 1), t >= this._finalSize && (this._update(this._block), this._block.fill(0));
  var r = this._len * 8;
  if (r <= 4294967295)
    this._block.writeUInt32BE(r, this._blockSize - 4);
  else {
    var n = (r & 4294967295) >>> 0, a = (r - n) / 4294967296;
    this._block.writeUInt32BE(a, this._blockSize - 8), this._block.writeUInt32BE(n, this._blockSize - 4);
  }
  this._update(this._block);
  var c = this._hash();
  return e ? c.toString(e) : c;
};
$t.prototype._update = function() {
  throw new Error("_update must be implemented by subclass");
};
var Vo = $t, Xo = ui, dn = Vo, Yo = Er.Buffer, Jo = [
  1518500249,
  1859775393,
  -1894007588,
  -899497514
], Ko = new Array(80);
function lt() {
  this.init(), this._w = Ko, dn.call(this, 64, 56);
}
Xo(lt, dn);
lt.prototype.init = function() {
  return this._a = 1732584193, this._b = 4023233417, this._c = 2562383102, this._d = 271733878, this._e = 3285377520, this;
};
function Zo(e) {
  return e << 1 | e >>> 31;
}
function Qo(e) {
  return e << 5 | e >>> 27;
}
function ta(e) {
  return e << 30 | e >>> 2;
}
function ra(e, t, r, n) {
  return e === 0 ? t & r | ~t & n : e === 2 ? t & r | t & n | r & n : t ^ r ^ n;
}
lt.prototype._update = function(e) {
  for (var t = this._w, r = this._a | 0, n = this._b | 0, a = this._c | 0, c = this._d | 0, u = this._e | 0, s = 0; s < 16; ++s)
    t[s] = e.readInt32BE(s * 4);
  for (; s < 80; ++s)
    t[s] = Zo(t[s - 3] ^ t[s - 8] ^ t[s - 14] ^ t[s - 16]);
  for (var h = 0; h < 80; ++h) {
    var d = ~~(h / 20), m = Qo(r) + ra(d, n, a, c) + u + t[h] + Jo[d] | 0;
    u = c, c = a, a = ta(n), n = r, r = m;
  }
  this._a = r + this._a | 0, this._b = n + this._b | 0, this._c = a + this._c | 0, this._d = c + this._d | 0, this._e = u + this._e | 0;
};
lt.prototype._hash = function() {
  var e = Yo.allocUnsafe(20);
  return e.writeInt32BE(this._a | 0, 0), e.writeInt32BE(this._b | 0, 4), e.writeInt32BE(this._c | 0, 8), e.writeInt32BE(this._d | 0, 12), e.writeInt32BE(this._e | 0, 16), e;
};
var ea = lt;
const na = /* @__PURE__ */ Ze(ea), Ue = (e, t) => function(...r) {
  const n = t.promiseModule;
  return new n((a, c) => {
    t.multiArgs ? r.push((...u) => {
      t.errorFirst ? u[0] ? c(u) : (u.shift(), a(u)) : a(u);
    }) : t.errorFirst ? r.push((u, s) => {
      u ? c(u) : a(s);
    }) : r.push(a), e.apply(this, r);
  });
};
var ia = (e, t) => {
  t = Object.assign({
    exclude: [/.+(Sync|Stream)$/],
    errorFirst: !0,
    promiseModule: Promise
  }, t);
  const r = typeof e;
  if (!(e !== null && (r === "object" || r === "function")))
    throw new TypeError(`Expected \`input\` to be a \`Function\` or \`Object\`, got \`${e === null ? "null" : r}\``);
  const n = (c) => {
    const u = (s) => typeof s == "string" ? c === s : s.test(c);
    return t.include ? t.include.some(u) : !t.exclude.some(u);
  };
  let a;
  r === "function" ? a = function(...c) {
    return t.excludeMain ? e(...c) : Ue(e, t).apply(this, c);
  } : a = Object.create(Object.getPrototypeOf(e));
  for (const c in e) {
    const u = e[c];
    a[c] = typeof u == "function" && n(c) ? Ue(u, t) : u;
  }
  return a;
};
const mt = /* @__PURE__ */ Ze(ia);
function Re(e) {
  return Array.isArray(e) ? e : [e];
}
const fr = "", Ce = " ", ir = "\\", oa = /^\s+$/, aa = /(?:[^\\]|^)\\$/, sa = /^\\!/, ca = /^\\#/, ua = /\r?\n/g, fa = /^\.*\/|^\.+$/, or = "/";
let mn = "node-ignore";
typeof Symbol < "u" && (mn = Symbol.for("node-ignore"));
const je = mn, la = (e, t, r) => Object.defineProperty(e, t, { value: r }), ha = /([0-z])-([0-z])/g, gn = () => !1, pa = (e) => e.replace(
  ha,
  (t, r, n) => r.charCodeAt(0) <= n.charCodeAt(0) ? t : fr
), ya = (e) => {
  const { length: t } = e;
  return e.slice(0, t - t % 2);
}, da = [
  [
    // remove BOM
    // TODO:
    // Other similar zero-width characters?
    /^\uFEFF/,
    () => fr
  ],
  // > Trailing spaces are ignored unless they are quoted with backslash ("\")
  [
    // (a\ ) -> (a )
    // (a  ) -> (a)
    // (a ) -> (a)
    // (a \ ) -> (a  )
    /((?:\\\\)*?)(\\?\s+)$/,
    (e, t, r) => t + (r.indexOf("\\") === 0 ? Ce : fr)
  ],
  // replace (\ ) with ' '
  // (\ ) -> ' '
  // (\\ ) -> '\\ '
  // (\\\ ) -> '\\ '
  [
    /(\\+?)\s/g,
    (e, t) => {
      const { length: r } = t;
      return t.slice(0, r - r % 2) + Ce;
    }
  ],
  // Escape metacharacters
  // which is written down by users but means special for regular expressions.
  // > There are 12 characters with special meanings:
  // > - the backslash \,
  // > - the caret ^,
  // > - the dollar sign $,
  // > - the period or dot .,
  // > - the vertical bar or pipe symbol |,
  // > - the question mark ?,
  // > - the asterisk or star *,
  // > - the plus sign +,
  // > - the opening parenthesis (,
  // > - the closing parenthesis ),
  // > - and the opening square bracket [,
  // > - the opening curly brace {,
  // > These special characters are often called "metacharacters".
  [
    /[\\$.|*+(){^]/g,
    (e) => `\\${e}`
  ],
  [
    // > a question mark (?) matches a single character
    /(?!\\)\?/g,
    () => "[^/]"
  ],
  // leading slash
  [
    // > A leading slash matches the beginning of the pathname.
    // > For example, "/*.c" matches "cat-file.c" but not "mozilla-sha1/sha1.c".
    // A leading slash matches the beginning of the pathname
    /^\//,
    () => "^"
  ],
  // replace special metacharacter slash after the leading slash
  [
    /\//g,
    () => "\\/"
  ],
  [
    // > A leading "**" followed by a slash means match in all directories.
    // > For example, "**/foo" matches file or directory "foo" anywhere,
    // > the same as pattern "foo".
    // > "**/foo/bar" matches file or directory "bar" anywhere that is directly
    // >   under directory "foo".
    // Notice that the '*'s have been replaced as '\\*'
    /^\^*\\\*\\\*\\\//,
    // '**/foo' <-> 'foo'
    () => "^(?:.*\\/)?"
  ],
  // starting
  [
    // there will be no leading '/'
    //   (which has been replaced by section "leading slash")
    // If starts with '**', adding a '^' to the regular expression also works
    /^(?=[^^])/,
    function() {
      return /\/(?!$)/.test(this) ? "^" : "(?:^|\\/)";
    }
  ],
  // two globstars
  [
    // Use lookahead assertions so that we could match more than one `'/**'`
    /\\\/\\\*\\\*(?=\\\/|$)/g,
    // Zero, one or several directories
    // should not use '*', or it will be replaced by the next replacer
    // Check if it is not the last `'/**'`
    (e, t, r) => t + 6 < r.length ? "(?:\\/[^\\/]+)*" : "\\/.+"
  ],
  // normal intermediate wildcards
  [
    // Never replace escaped '*'
    // ignore rule '\*' will match the path '*'
    // 'abc.*/' -> go
    // 'abc.*'  -> skip this rule,
    //    coz trailing single wildcard will be handed by [trailing wildcard]
    /(^|[^\\]+)(\\\*)+(?=.+)/g,
    // '*.js' matches '.js'
    // '*.js' doesn't match 'abc'
    (e, t, r) => {
      const n = r.replace(/\\\*/g, "[^\\/]*");
      return t + n;
    }
  ],
  [
    // unescape, revert step 3 except for back slash
    // For example, if a user escape a '\\*',
    // after step 3, the result will be '\\\\\\*'
    /\\\\\\(?=[$.|*+(){^])/g,
    () => ir
  ],
  [
    // '\\\\' -> '\\'
    /\\\\/g,
    () => ir
  ],
  [
    // > The range notation, e.g. [a-zA-Z],
    // > can be used to match one of the characters in a range.
    // `\` is escaped by step 3
    /(\\)?\[([^\]/]*?)(\\*)($|\])/g,
    (e, t, r, n, a) => t === ir ? `\\[${r}${ya(n)}${a}` : a === "]" && n.length % 2 === 0 ? `[${pa(r)}${n}]` : "[]"
  ],
  // ending
  [
    // 'js' will not match 'js.'
    // 'ab' will not match 'abc'
    /(?:[^*])$/,
    // WTF!
    // https://git-scm.com/docs/gitignore
    // changes in [2.22.1](https://git-scm.com/docs/gitignore/2.22.1)
    // which re-fixes #24, #38
    // > If there is a separator at the end of the pattern then the pattern
    // > will only match directories, otherwise the pattern can match both
    // > files and directories.
    // 'js*' will not match 'a.js'
    // 'js/' will not match 'a.js'
    // 'js' will match 'a.js' and 'a.js/'
    (e) => /\/$/.test(e) ? `${e}$` : `${e}(?=$|\\/$)`
  ],
  // trailing wildcard
  [
    /(\^|\\\/)?\\\*$/,
    (e, t) => `${t ? `${t}[^/]+` : "[^/]*"}(?=$|\\/$)`
  ]
], Ne = /* @__PURE__ */ Object.create(null), ma = (e, t) => {
  let r = Ne[e];
  return r || (r = da.reduce(
    (n, [a, c]) => n.replace(a, c.bind(e)),
    e
  ), Ne[e] = r), t ? new RegExp(r, "i") : new RegExp(r);
}, Br = (e) => typeof e == "string", ga = (e) => e && Br(e) && !oa.test(e) && !aa.test(e) && e.indexOf("#") !== 0, wa = (e) => e.split(ua);
class ba {
  constructor(t, r, n, a) {
    this.origin = t, this.pattern = r, this.negative = n, this.regex = a;
  }
}
const va = (e, t) => {
  const r = e;
  let n = !1;
  e.indexOf("!") === 0 && (n = !0, e = e.substr(1)), e = e.replace(sa, "!").replace(ca, "#");
  const a = ma(e, t);
  return new ba(
    r,
    e,
    n,
    a
  );
}, Ea = (e, t) => {
  throw new t(e);
}, X = (e, t, r) => Br(e) ? e ? X.isNotRelative(e) ? r(
  `path should be a \`path.relative()\`d string, but got "${t}"`,
  RangeError
) : !0 : r("path must not be empty", TypeError) : r(
  `path must be a string, but got \`${t}\``,
  TypeError
), wn = (e) => fa.test(e);
X.isNotRelative = wn;
X.convert = (e) => e;
class xa {
  constructor({
    ignorecase: t = !0,
    ignoreCase: r = t,
    allowRelativePaths: n = !1
  } = {}) {
    la(this, je, !0), this._rules = [], this._ignoreCase = r, this._allowRelativePaths = n, this._initCache();
  }
  _initCache() {
    this._ignoreCache = /* @__PURE__ */ Object.create(null), this._testCache = /* @__PURE__ */ Object.create(null);
  }
  _addPattern(t) {
    if (t && t[je]) {
      this._rules = this._rules.concat(t._rules), this._added = !0;
      return;
    }
    if (ga(t)) {
      const r = va(t, this._ignoreCase);
      this._added = !0, this._rules.push(r);
    }
  }
  // @param {Array<string> | string | Ignore} pattern
  add(t) {
    return this._added = !1, Re(
      Br(t) ? wa(t) : t
    ).forEach(this._addPattern, this), this._added && this._initCache(), this;
  }
  // legacy
  addPattern(t) {
    return this.add(t);
  }
  //          |           ignored : unignored
  // negative |   0:0   |   0:1   |   1:0   |   1:1
  // -------- | ------- | ------- | ------- | --------
  //     0    |  TEST   |  TEST   |  SKIP   |    X
  //     1    |  TESTIF |  SKIP   |  TEST   |    X
  // - SKIP: always skip
  // - TEST: always test
  // - TESTIF: only test if checkUnignored
  // - X: that never happen
  // @param {boolean} whether should check if the path is unignored,
  //   setting `checkUnignored` to `false` could reduce additional
  //   path matching.
  // @returns {TestResult} true if a file is ignored
  _testOne(t, r) {
    let n = !1, a = !1;
    return this._rules.forEach((c) => {
      const { negative: u } = c;
      if (a === u && n !== a || u && !n && !a && !r)
        return;
      c.regex.test(t) && (n = !u, a = u);
    }), {
      ignored: n,
      unignored: a
    };
  }
  // @returns {TestResult}
  _test(t, r, n, a) {
    const c = t && X.convert(t);
    return X(
      c,
      t,
      this._allowRelativePaths ? gn : Ea
    ), this._t(c, r, n, a);
  }
  _t(t, r, n, a) {
    if (t in r)
      return r[t];
    if (a || (a = t.split(or)), a.pop(), !a.length)
      return r[t] = this._testOne(t, n);
    const c = this._t(
      a.join(or) + or,
      r,
      n,
      a
    );
    return r[t] = c.ignored ? c : this._testOne(t, n);
  }
  ignores(t) {
    return this._test(t, this._ignoreCache, !1).ignored;
  }
  createFilter() {
    return (t) => !this.ignores(t);
  }
  filter(t) {
    return Re(t).filter(this.createFilter());
  }
  // @returns {TestResult}
  test(t) {
    return this._test(t, this._testCache, !0);
  }
}
const lr = (e) => new xa(e), Aa = (e) => X(e && X.convert(e), e, gn);
lr.isPathValid = Aa;
lr.default = lr;
if (
  // Detect `process` so that it can run in browsers.
  typeof process < "u" && (process.env && process.env.IGNORE_TEST_WIN32 || process.platform === "win32")
) {
  const e = (r) => /^\\\\\?\\/.test(r) || /["<>|\u0000-\u001F]+/u.test(r) ? r : r.replace(/\\/g, "/");
  X.convert = e;
  const t = /^[a-z]:\//i;
  X.isNotRelative = (r) => t.test(r) || wn(r);
}
class B extends Error {
  constructor(t) {
    super(t), this.caller = "";
  }
  toJSON() {
    return {
      code: this.code,
      data: this.data,
      caller: this.caller,
      message: this.message,
      stack: this.stack
    };
  }
  fromJSON(t) {
    const r = new B(t.message);
    return r.code = t.code, r.data = t.data, r.caller = t.caller, r.stack = t.stack, r;
  }
  get isIsomorphicGitError() {
    return !0;
  }
}
class _r extends B {
  /**
   * @param {Array<string>} filepaths
   */
  constructor(t) {
    super(
      `Modifying the index is not possible because you have unmerged files: ${t.toString}. Fix them up in the work tree, and then use 'git add/rm as appropriate to mark resolution and make a commit.`
    ), this.code = this.name = _r.code, this.data = { filepaths: t };
  }
}
_r.code = "UnmergedPathsError";
class $ extends B {
  /**
   * @param {string} message
   */
  constructor(t) {
    super(
      `An internal error caused this command to fail.

If you're not a developer, report the bug to the developers of the application you're using. If this is a bug in isomorphic-git then you should create a proper bug yourselves. The bug should include a minimal reproduction and details about the version and environment.

Please file a bug report at https://github.com/isomorphic-git/isomorphic-git/issues with this error message: ${t}`
    ), this.code = this.name = $.code, this.data = { message: t };
  }
}
$.code = "InternalError";
class Pt extends B {
  /**
   * @param {string} filepath
   */
  constructor(t) {
    super(`The filepath "${t}" contains unsafe character sequences`), this.code = this.name = Pt.code, this.data = { filepath: t };
  }
}
Pt.code = "UnsafeFilepathError";
class Y {
  constructor(t) {
    this.buffer = t, this._start = 0;
  }
  eof() {
    return this._start >= this.buffer.length;
  }
  tell() {
    return this._start;
  }
  seek(t) {
    this._start = t;
  }
  slice(t) {
    const r = this.buffer.slice(this._start, this._start + t);
    return this._start += t, r;
  }
  toString(t, r) {
    const n = this.buffer.toString(t, this._start, this._start + r);
    return this._start += r, n;
  }
  write(t, r, n) {
    const a = this.buffer.write(t, this._start, r, n);
    return this._start += r, a;
  }
  copy(t, r, n) {
    const a = t.copy(this.buffer, this._start, r, n);
    return this._start += a, a;
  }
  readUInt8() {
    const t = this.buffer.readUInt8(this._start);
    return this._start += 1, t;
  }
  writeUInt8(t) {
    const r = this.buffer.writeUInt8(t, this._start);
    return this._start += 1, r;
  }
  readUInt16BE() {
    const t = this.buffer.readUInt16BE(this._start);
    return this._start += 2, t;
  }
  writeUInt16BE(t) {
    const r = this.buffer.writeUInt16BE(t, this._start);
    return this._start += 2, r;
  }
  readUInt32BE() {
    const t = this.buffer.readUInt32BE(this._start);
    return this._start += 4, t;
  }
  writeUInt32BE(t) {
    const r = this.buffer.writeUInt32BE(t, this._start);
    return this._start += 4, r;
  }
}
function Ir(e, t) {
  return -(e < t) || +(e > t);
}
function Sa(e, t) {
  return Ir(e.path, t.path);
}
function Ba(e) {
  let t = "";
  for (const r of new Uint8Array(e))
    r < 16 && (t += "0"), t += r.toString(16);
  return t;
}
let ar = null;
async function St(e) {
  return ar === null && (ar = await Ia()), ar ? bn(e) : _a(e);
}
function _a(e) {
  return new na().update(e).digest("hex");
}
async function bn(e) {
  const t = await crypto.subtle.digest("SHA-1", e);
  return Ba(t);
}
async function Ia() {
  try {
    return await bn(new Uint8Array([])) === "da39a3ee5e6b4b0d3255bfef95601890afd80709";
  } catch {
  }
  return !1;
}
function hr(e) {
  const t = Math.max(e.lastIndexOf("/"), e.lastIndexOf("\\"));
  return t === -1 ? "." : t === 0 ? "/" : e.slice(0, t);
}
class ht extends B {
  /**
   * @param {string} what
   */
  constructor(t) {
    super(`Could not find ${t}.`), this.code = this.name = ht.code, this.data = { what: t };
  }
}
ht.code = "NotFoundError";
let kt = class vn extends B {
  /**
   * @param {string} oid
   * @param {'blob'|'commit'|'tag'|'tree'} actual
   * @param {'blob'|'commit'|'tag'|'tree'} expected
   * @param {string} [filepath]
   */
  constructor(t, r, n, a) {
    super(
      `Object ${t} ${a ? `at ${a}` : ""}was anticipated to be a ${n} but it is a ${r}.`
    ), this.code = this.name = vn.code, this.data = { oid: t, actual: r, expected: n, filepath: a };
  }
};
kt.code = "ObjectTypeError";
class Fr extends B {
  /**
   * @param {string} value
   */
  constructor(t) {
    super(`Expected a 40-char hex object id but saw "${t}".`), this.code = this.name = Fr.code, this.data = { value: t };
  }
}
Fr.code = "InvalidOidError";
class $r extends B {
  /**
   * @param {string} remote
   */
  constructor(t) {
    super(`Could not find a fetch refspec for remote "${t}". Make sure the config file has an entry like the following:
[remote "${t}"]
	fetch = +refs/heads/*:refs/remotes/origin/*
`), this.code = this.name = $r.code, this.data = { remote: t };
  }
}
$r.code = "NoRefspecError";
/*!
 * This code for `path.join` is directly copied from @zenfs/core/path for bundle size improvements.
 * SPDX-License-Identifier: LGPL-3.0-or-later
 * Copyright (c) James Prevett and other ZenFS contributors.
 */
function Fa(e, t) {
  let r = "", n = 0, a = -1, c = 0, u = "\0";
  for (let s = 0; s <= e.length; ++s) {
    if (s < e.length) u = e[s];
    else {
      if (u === "/") break;
      u = "/";
    }
    if (u === "/") {
      if (!(a === s - 1 || c === 1)) if (c === 2) {
        if (r.length < 2 || n !== 2 || r.at(-1) !== "." || r.at(-2) !== ".") {
          if (r.length > 2) {
            const h = r.lastIndexOf("/");
            h === -1 ? (r = "", n = 0) : (r = r.slice(0, h), n = r.length - 1 - r.lastIndexOf("/")), a = s, c = 0;
            continue;
          } else if (r.length !== 0) {
            r = "", n = 0, a = s, c = 0;
            continue;
          }
        }
        t && (r += r.length > 0 ? "/.." : "..", n = 2);
      } else
        r.length > 0 ? r += "/" + e.slice(a + 1, s) : r = e.slice(a + 1, s), n = s - a - 1;
      a = s, c = 0;
    } else u === "." && c !== -1 ? ++c : c = -1;
  }
  return r;
}
function $a(e) {
  if (!e.length) return ".";
  const t = e[0] === "/", r = e.at(-1) === "/";
  return e = Fa(e, !t), e.length ? (r && (e += "/"), t ? `/${e}` : e) : t ? "/" : r ? "./" : ".";
}
function nt(...e) {
  if (e.length === 0) return ".";
  let t;
  for (let r = 0; r < e.length; ++r) {
    const n = e[r];
    n.length > 0 && (t === void 0 ? t = n : t += "/" + n);
  }
  return t === void 0 ? "." : $a(t);
}
function Pa(e, t) {
  return Ir(De(e), De(t));
}
function De(e) {
  return e.mode === "040000" ? e.path + "/" : e.path;
}
function En(e) {
  switch (e) {
    case "040000":
      return "tree";
    case "100644":
      return "blob";
    case "100755":
      return "blob";
    case "120000":
      return "blob";
    case "160000":
      return "commit";
  }
  throw new $(`Unexpected GitTree entry mode: ${e}`);
}
function ka(e) {
  const t = [];
  let r = 0;
  for (; r < e.length; ) {
    const n = e.indexOf(32, r);
    if (n === -1)
      throw new $(
        `GitTree: Error parsing buffer at byte location ${r}: Could not find the next space character.`
      );
    const a = e.indexOf(0, r);
    if (a === -1)
      throw new $(
        `GitTree: Error parsing buffer at byte location ${r}: Could not find the next null character.`
      );
    let c = e.slice(r, n).toString("utf8");
    c === "40000" && (c = "040000");
    const u = En(c), s = e.slice(n + 1, a).toString("utf8");
    if (s.includes("\\") || s.includes("/"))
      throw new Pt(s);
    const h = e.slice(a + 1, a + 21).toString("hex");
    r = a + 21, t.push({ mode: c, path: s, oid: h, type: u });
  }
  return t;
}
function Ta(e) {
  if (typeof e == "number" && (e = e.toString(8)), e.match(/^0?4.*/)) return "040000";
  if (e.match(/^1006.*/)) return "100644";
  if (e.match(/^1007.*/)) return "100755";
  if (e.match(/^120.*/)) return "120000";
  if (e.match(/^160.*/)) return "160000";
  throw new $(`Could not understand file mode: ${e}`);
}
function Oa(e) {
  return !e.oid && e.sha && (e.oid = e.sha), e.mode = Ta(e.mode), e.type || (e.type = En(e.mode)), e;
}
let Bt = class xn {
  constructor(t) {
    if (Buffer.isBuffer(t))
      this._entries = ka(t);
    else if (Array.isArray(t))
      this._entries = t.map(Oa);
    else
      throw new $("invalid type passed to GitTree constructor");
    this._entries.sort(Sa);
  }
  static from(t) {
    return new xn(t);
  }
  render() {
    return this._entries.map((t) => `${t.mode} ${t.type} ${t.oid}    ${t.path}`).join(`
`);
  }
  toObject() {
    const t = [...this._entries];
    return t.sort(Pa), Buffer.concat(
      t.map((r) => {
        const n = Buffer.from(r.mode.replace(/^0/, "")), a = Buffer.from(" "), c = Buffer.from(r.path, "utf8"), u = Buffer.from([0]), s = Buffer.from(r.oid, "hex");
        return Buffer.concat([n, a, c, u, s]);
      })
    );
  }
  /**
   * @returns {TreeEntry[]}
   */
  entries() {
    return this._entries;
  }
  *[Symbol.iterator]() {
    for (const t of this._entries)
      yield t;
  }
};
class An {
  /**
   * Wraps a raw object with a Git header.
   *
   * @param {Object} params - The parameters for wrapping.
   * @param {string} params.type - The type of the Git object (e.g., 'blob', 'tree', 'commit').
   * @param {Uint8Array} params.object - The raw object data to wrap.
   * @returns {Uint8Array} The wrapped Git object as a single buffer.
   */
  static wrap({ type: t, object: r }) {
    const n = `${t} ${r.length}\0`, a = n.length, c = a + r.length, u = new Uint8Array(c);
    for (let s = 0; s < a; s++)
      u[s] = n.charCodeAt(s);
    return u.set(r, a), u;
  }
  /**
   * Unwraps a Git object buffer into its type and raw object data.
   *
   * @param {Buffer|Uint8Array} buffer - The buffer containing the wrapped Git object.
   * @returns {{ type: string, object: Buffer }} An object containing the type and the raw object data.
   * @throws {InternalError} If the length specified in the header does not match the actual object length.
   */
  static unwrap(t) {
    const r = t.indexOf(32), n = t.indexOf(0), a = t.slice(0, r).toString("utf8"), c = t.slice(r + 1, n).toString("utf8"), u = t.length - (n + 1);
    if (parseInt(c) !== u)
      throw new $(
        `Length mismatch: expected ${c} bytes but got ${u} instead.`
      );
    return {
      type: a,
      object: Buffer.from(t.slice(n + 1))
    };
  }
}
async function Ua({ fs: e, gitdir: t, oid: r }) {
  const n = `objects/${r.slice(0, 2)}/${r.slice(2)}`, a = await e.read(`${t}/${n}`);
  return a ? { object: a, format: "deflated", source: n } : null;
}
function Ra(e, t) {
  const r = new Y(e), n = Me(r);
  if (n !== t.byteLength)
    throw new $(
      `applyDelta expected source buffer to be ${n} bytes but the provided buffer was ${t.length} bytes`
    );
  const a = Me(r);
  let c;
  const u = Le(r, t);
  if (u.byteLength === a)
    c = u;
  else {
    c = Buffer.alloc(a);
    const s = new Y(c);
    for (s.copy(u); !r.eof(); )
      s.copy(Le(r, t));
    const h = s.tell();
    if (a !== h)
      throw new $(
        `applyDelta expected target buffer to be ${a} bytes but the resulting buffer was ${h} bytes`
      );
  }
  return c;
}
function Me(e) {
  let t = 0, r = 0, n = null;
  do
    n = e.readUInt8(), t |= (n & 127) << r, r += 7;
  while (n & 128);
  return t;
}
function Ge(e, t, r) {
  let n = 0, a = 0;
  for (; r--; )
    t & 1 && (n |= e.readUInt8() << a), t >>= 1, a += 8;
  return n;
}
function Le(e, t) {
  const r = e.readUInt8(), n = 128, a = 15, c = 112;
  if (r & n) {
    const u = Ge(e, r & a, 4);
    let s = Ge(e, (r & c) >> 4, 3);
    return s === 0 && (s = 65536), t.slice(u, u + s);
  } else
    return e.slice(r);
}
function Ca(e) {
  let t = [e];
  return {
    next() {
      return Promise.resolve({ done: t.length === 0, value: t.pop() });
    },
    return() {
      return t = [], {};
    },
    [Symbol.asyncIterator]() {
      return this;
    }
  };
}
function ja(e) {
  return e[Symbol.asyncIterator] ? e[Symbol.asyncIterator]() : e[Symbol.iterator] ? e[Symbol.iterator]() : e.next ? e : Ca(e);
}
class Na {
  constructor(t) {
    if (typeof Buffer > "u")
      throw new Error("Missing Buffer dependency");
    this.stream = ja(t), this.buffer = null, this.cursor = 0, this.undoCursor = 0, this.started = !1, this._ended = !1, this._discardedBytes = 0;
  }
  eof() {
    return this._ended && this.cursor === this.buffer.length;
  }
  tell() {
    return this._discardedBytes + this.cursor;
  }
  async byte() {
    if (!this.eof() && (this.started || await this._init(), !(this.cursor === this.buffer.length && (await this._loadnext(), this._ended))))
      return this._moveCursor(1), this.buffer[this.undoCursor];
  }
  async chunk() {
    if (!this.eof() && (this.started || await this._init(), !(this.cursor === this.buffer.length && (await this._loadnext(), this._ended))))
      return this._moveCursor(this.buffer.length), this.buffer.slice(this.undoCursor, this.cursor);
  }
  async read(t) {
    if (!this.eof())
      return this.started || await this._init(), this.cursor + t > this.buffer.length && (this._trim(), await this._accumulate(t)), this._moveCursor(t), this.buffer.slice(this.undoCursor, this.cursor);
  }
  async skip(t) {
    this.eof() || (this.started || await this._init(), this.cursor + t > this.buffer.length && (this._trim(), await this._accumulate(t)), this._moveCursor(t));
  }
  async undo() {
    this.cursor = this.undoCursor;
  }
  async _next() {
    this.started = !0;
    let { done: t, value: r } = await this.stream.next();
    return t && (this._ended = !0, !r) ? Buffer.alloc(0) : (r && (r = Buffer.from(r)), r);
  }
  _trim() {
    this.buffer = this.buffer.slice(this.undoCursor), this.cursor -= this.undoCursor, this._discardedBytes += this.undoCursor, this.undoCursor = 0;
  }
  _moveCursor(t) {
    this.undoCursor = this.cursor, this.cursor += t, this.cursor > this.buffer.length && (this.cursor = this.buffer.length);
  }
  async _accumulate(t) {
    if (this._ended) return;
    const r = [this.buffer];
    for (; this.cursor + t > Da(r); ) {
      const n = await this._next();
      if (this._ended) break;
      r.push(n);
    }
    this.buffer = Buffer.concat(r);
  }
  async _loadnext() {
    this._discardedBytes += this.buffer.length, this.undoCursor = 0, this.cursor = 0, this.buffer = await this._next();
  }
  async _init() {
    this.buffer = await this._next();
  }
}
function Da(e) {
  return e.reduce((t, r) => t + r.length, 0);
}
async function Ma(e, t) {
  const r = new Na(e);
  let n = await r.read(4);
  if (n = n.toString("utf8"), n !== "PACK")
    throw new $(`Invalid PACK header '${n}'`);
  let a = await r.read(4);
  if (a = a.readUInt32BE(0), a !== 2)
    throw new $(`Invalid packfile version: ${a}`);
  let c = await r.read(4);
  if (c = c.readUInt32BE(0), !(c < 1))
    for (; !r.eof() && c--; ) {
      const u = r.tell(), { type: s, length: h, ofs: d, reference: m } = await Ga(r), y = new Je.Inflate();
      for (; !y.result; ) {
        const b = await r.chunk();
        if (!b) break;
        if (y.push(b, !1), y.err)
          throw new $(`Pako error: ${y.msg}`);
        if (y.result) {
          if (y.result.length !== h)
            throw new $(
              "Inflated object size is different from that stated in packfile."
            );
          await r.undo(), await r.read(b.length - y.strm.avail_in);
          const E = r.tell();
          await t({
            data: y.result,
            type: s,
            num: c,
            offset: u,
            end: E,
            reference: m,
            ofs: d
          });
        }
      }
    }
}
async function Ga(e) {
  let t = await e.byte();
  const r = t >> 4 & 7;
  let n = t & 15;
  if (t & 128) {
    let u = 4;
    do
      t = await e.byte(), n |= (t & 127) << u, u += 7;
    while (t & 128);
  }
  let a, c;
  if (r === 6) {
    let u = 0;
    a = 0;
    const s = [];
    do
      t = await e.byte(), a |= (t & 127) << u, u += 7, s.push(t);
    while (t & 128);
    c = Buffer.from(s);
  }
  return r === 7 && (c = await e.read(20)), { type: r, length: n, ofs: a, reference: c };
}
async function Sn(e) {
  return Je.inflate(e);
}
function La(e) {
  const t = [];
  let r = 0, n = 0;
  do {
    r = e.readUInt8();
    const a = r & 127;
    t.push(a), n = r & 128;
  } while (n);
  return t.reduce((a, c) => a + 1 << 7 | c, -1);
}
function Ha(e, t) {
  let r = t, n = 4, a = null;
  do
    a = e.readUInt8(), r |= (a & 127) << n, n += 7;
  while (a & 128);
  return r;
}
let Bn = class pr {
  constructor(t) {
    Object.assign(this, t), this.offsetCache = {};
  }
  static async fromIdx({ idx: t, getExternalRefDelta: r }) {
    const n = new Y(t);
    if (n.slice(4).toString("hex") !== "ff744f63")
      return;
    const c = n.readUInt32BE();
    if (c !== 2)
      throw new $(
        `Unable to read version ${c} packfile IDX. (Only version 2 supported)`
      );
    if (t.byteLength > 2048 * 1024 * 1024)
      throw new $(
        "To keep implementation simple, I haven't implemented the layer 5 feature needed to support packfiles > 2GB in size."
      );
    n.seek(n.tell() + 4 * 255);
    const u = n.readUInt32BE(), s = [];
    for (let m = 0; m < u; m++) {
      const y = n.slice(20).toString("hex");
      s[m] = y;
    }
    n.seek(n.tell() + 4 * u);
    const h = /* @__PURE__ */ new Map();
    for (let m = 0; m < u; m++)
      h.set(s[m], n.readUInt32BE());
    const d = n.slice(20).toString("hex");
    return new pr({
      hashes: s,
      crcs: {},
      offsets: h,
      packfileSha: d,
      getExternalRefDelta: r
    });
  }
  static async fromPack({ pack: t, getExternalRefDelta: r, onProgress: n }) {
    const a = {
      1: "commit",
      2: "tree",
      3: "blob",
      4: "tag",
      6: "ofs-delta",
      7: "ref-delta"
    }, c = {}, u = t.slice(-20).toString("hex"), s = [], h = {}, d = /* @__PURE__ */ new Map();
    let m = null, y = null;
    await Ma([t], async ({ data: A, type: x, reference: _, offset: v, num: P }) => {
      m === null && (m = P);
      const j = Math.floor(
        (m - P) * 100 / m
      );
      j !== y && n && await n({
        phase: "Receiving objects",
        loaded: m - P,
        total: m
      }), y = j, x = a[x], ["commit", "tree", "blob", "tag"].includes(x) ? c[v] = {
        type: x,
        offset: v
      } : x === "ofs-delta" ? c[v] = {
        type: x,
        offset: v
      } : x === "ref-delta" && (c[v] = {
        type: x,
        offset: v
      });
    });
    const b = Object.keys(c).map(Number);
    for (const [A, x] of b.entries()) {
      const _ = A + 1 === b.length ? t.byteLength - 20 : b[A + 1], v = c[x], P = ti.buf(t.slice(x, _)) >>> 0;
      v.end = _, v.crc = P;
    }
    const E = new pr({
      pack: Promise.resolve(t),
      packfileSha: u,
      crcs: h,
      hashes: s,
      offsets: d,
      getExternalRefDelta: r
    });
    y = null;
    let O = 0;
    const F = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    for (let A in c) {
      A = Number(A);
      const x = Math.floor(O * 100 / m);
      x !== y && n && await n({
        phase: "Resolving deltas",
        loaded: O,
        total: m
      }), O++, y = x;
      const _ = c[A];
      if (!_.oid)
        try {
          E.readDepth = 0, E.externalReadDepth = 0;
          const { type: v, object: P } = await E.readSlice({ start: A });
          F[E.readDepth] += 1;
          const j = await St(An.wrap({ type: v, object: P }));
          _.oid = j, s.push(j), d.set(j, A), h[j] = _.crc;
        } catch {
          continue;
        }
    }
    return s.sort(), E;
  }
  async toBuffer() {
    const t = [], r = (d, m) => {
      t.push(Buffer.from(d, m));
    };
    r("ff744f63", "hex"), r("00000002", "hex");
    const n = new Y(Buffer.alloc(256 * 4));
    for (let d = 0; d < 256; d++) {
      let m = 0;
      for (const y of this.hashes)
        parseInt(y.slice(0, 2), 16) <= d && m++;
      n.writeUInt32BE(m);
    }
    t.push(n.buffer);
    for (const d of this.hashes)
      r(d, "hex");
    const a = new Y(Buffer.alloc(this.hashes.length * 4));
    for (const d of this.hashes)
      a.writeUInt32BE(this.crcs[d]);
    t.push(a.buffer);
    const c = new Y(Buffer.alloc(this.hashes.length * 4));
    for (const d of this.hashes)
      c.writeUInt32BE(this.offsets.get(d));
    t.push(c.buffer), r(this.packfileSha, "hex");
    const u = Buffer.concat(t), s = await St(u), h = Buffer.alloc(20);
    return h.write(s, "hex"), Buffer.concat([u, h]);
  }
  async load({ pack: t }) {
    this.pack = t;
  }
  async unload() {
    this.pack = null;
  }
  async read({ oid: t }) {
    if (!this.offsets.get(t)) {
      if (this.getExternalRefDelta)
        return this.externalReadDepth++, this.getExternalRefDelta(t);
      throw new $(`Could not read object ${t} from packfile`);
    }
    const r = this.offsets.get(t);
    return this.readSlice({ start: r });
  }
  async readSlice({ start: t }) {
    if (this.offsetCache[t])
      return Object.assign({}, this.offsetCache[t]);
    this.readDepth++;
    const r = {
      16: "commit",
      32: "tree",
      48: "blob",
      64: "tag",
      96: "ofs_delta",
      112: "ref_delta"
    }, n = await this.pack;
    if (!n)
      throw new $(
        "Could not read packfile data. The packfile may be missing, corrupted, or too large to read into memory."
      );
    const a = n.slice(t), c = new Y(a), u = c.readUInt8(), s = u & 112;
    let h = r[s];
    if (h === void 0)
      throw new $("Unrecognized type: 0b" + s.toString(2));
    const d = u & 15;
    let m = d;
    u & 128 && (m = Ha(c, d));
    let b = null, E = null;
    if (h === "ofs_delta") {
      const F = La(c), A = t - F;
      ({ object: b, type: h } = await this.readSlice({ start: A }));
    }
    if (h === "ref_delta") {
      const F = c.slice(20).toString("hex");
      ({ object: b, type: h } = await this.read({ oid: F }));
    }
    const O = a.slice(c.tell());
    if (E = Buffer.from(await Sn(O)), E.byteLength !== m)
      throw new $(
        `Packfile told us object would have length ${m} but it had length ${E.byteLength}`
      );
    return b && (E = Buffer.from(Ra(E, b))), this.readDepth > 3 && (this.offsetCache[t] = { type: h, object: E }), { type: h, format: "content", object: E };
  }
};
const gt = Symbol("PackfileCache");
async function qa({
  fs: e,
  filename: t,
  getExternalRefDelta: r,
  emitter: n,
  emitterPrefix: a
}) {
  const c = await e.read(t);
  return Bn.fromIdx({ idx: c, getExternalRefDelta: r });
}
function za({
  fs: e,
  cache: t,
  filename: r,
  getExternalRefDelta: n,
  emitter: a,
  emitterPrefix: c
}) {
  t[gt] || (t[gt] = /* @__PURE__ */ new Map());
  let u = t[gt].get(r);
  return u || (u = qa({
    fs: e,
    filename: r,
    getExternalRefDelta: n,
    emitter: a,
    emitterPrefix: c
  }), t[gt].set(r, u)), u;
}
async function Wa(e, { start: t = 0, end: r = e.length } = {}) {
  return St(e.subarray(t, r));
}
async function Va({
  fs: e,
  cache: t,
  gitdir: r,
  oid: n,
  format: a = "content",
  getExternalRefDelta: c
}) {
  let u = await e.readdir(nt(r, "objects/pack"));
  u = u.filter((s) => s.endsWith(".idx"));
  for (const s of u) {
    const h = `${r}/objects/pack/${s}`, d = await za({
      fs: e,
      cache: t,
      filename: h,
      getExternalRefDelta: c
    });
    if (d.error) throw new $(d.error);
    if (d.offsets.has(n)) {
      const m = h.replace(/idx$/, "pack");
      d.pack || (d.pack = e.read(m));
      const y = await d.pack;
      if (!y)
        throw d.pack = null, new $(
          `Could not read packfile at ${m}. The file may be missing, corrupted, or too large to read into memory.`
        );
      if (!d._checksumVerified) {
        const E = d.packfileSha, O = y.subarray(-20), F = Array.from(O).map((x) => x.toString(16).padStart(2, "0")).join("");
        if (F !== E)
          throw new $(
            `Packfile trailer mismatch: expected ${E}, got ${F}. The packfile may be corrupted.`
          );
        const A = await Wa(y, {
          start: 0,
          end: y.length - 20
        });
        if (A !== E)
          throw new $(
            `Packfile payload corrupted: calculated ${A} but expected ${E}. The packfile may have been tampered with.`
          );
        d._checksumVerified = !0;
      }
      const b = await d.read({ oid: n, getExternalRefDelta: c });
      return b.format = "content", b.source = `objects/pack/${s.replace(/idx$/, "pack")}`, b;
    }
  }
  return null;
}
async function pt({
  fs: e,
  cache: t,
  gitdir: r,
  oid: n,
  format: a = "content"
}) {
  const c = (m) => pt({ fs: e, cache: t, gitdir: r, oid: m });
  let u;
  if (n === "4b825dc642cb6eb9a060e54bf8d69288fbee4904" && (u = { format: "wrapped", object: Buffer.from("tree 0\0") }), u || (u = await Ua({ fs: e, gitdir: r, oid: n })), !u) {
    if (u = await Va({
      fs: e,
      cache: t,
      gitdir: r,
      oid: n,
      getExternalRefDelta: c
    }), !u)
      throw new ht(n);
    return u;
  }
  if (a === "deflated" || (u.format === "deflated" && (u.object = Buffer.from(await Sn(u.object)), u.format = "wrapped"), a === "wrapped"))
    return u;
  const s = await St(u.object);
  if (s !== n)
    throw new $(
      `SHA check failed! Expected ${n}, computed ${s}`
    );
  const { object: h, type: d } = An.unwrap(u.object);
  if (u.type = d, u.object = h, u.format = "content", a === "content")
    return u;
  throw new $(`invalid requested format "${a}"`);
}
class Pr extends B {
  /**
   * @param {'note'|'remote'|'tag'|'branch'} noun
   * @param {string} where
   * @param {boolean} canForce
   */
  constructor(t, r, n = !0) {
    super(
      `Failed to create ${t} at ${r} because it already exists.${n ? ` (Hint: use 'force: true' parameter to overwrite existing ${t}.)` : ""}`
    ), this.code = this.name = Pr.code, this.data = { noun: t, where: r, canForce: n };
  }
}
Pr.code = "AlreadyExistsError";
class kr extends B {
  /**
   * @param {'oids'|'refs'} nouns
   * @param {string} short
   * @param {string[]} matches
   */
  constructor(t, r, n) {
    super(
      `Found multiple ${t} matching "${r}" (${n.join(
        ", "
      )}). Use a longer abbreviation length to disambiguate them.`
    ), this.code = this.name = kr.code, this.data = { nouns: t, short: r, matches: n };
  }
}
kr.code = "AmbiguousError";
class Tr extends B {
  /**
   * @param {string[]} filepaths
   */
  constructor(t) {
    super(
      `Your local changes to the following files would be overwritten by checkout: ${t.join(
        ", "
      )}`
    ), this.code = this.name = Tr.code, this.data = { filepaths: t };
  }
}
Tr.code = "CheckoutConflictError";
class Or extends B {
  /**
   * @param {string} oid
   * @param {number} parentCount
   */
  constructor(t, r) {
    super(
      `Cannot cherry-pick merge commit ${t}. Merge commits have ${r} parents and require specifying which parent to use as the base.`
    ), this.code = this.name = Or.code, this.data = { oid: t, parentCount: r };
  }
}
Or.code = "CherryPickMergeCommitError";
class Ur extends B {
  /**
   * @param {string} oid
   */
  constructor(t) {
    super(
      `Cannot cherry-pick root commit ${t}. Root commits have no parents.`
    ), this.code = this.name = Ur.code, this.data = { oid: t };
  }
}
Ur.code = "CherryPickRootCommitError";
class Rr extends B {
  /**
   * @param {string} ref
   * @param {string} oid
   */
  constructor(t, r) {
    super(
      `Failed to checkout "${t}" because commit ${r} is not available locally. Do a git fetch to make the branch available locally.`
    ), this.code = this.name = Rr.code, this.data = { ref: t, oid: r };
  }
}
Rr.code = "CommitNotFetchedError";
class Cr extends B {
  constructor() {
    super("Empty response from git server."), this.code = this.name = Cr.code, this.data = {};
  }
}
Cr.code = "EmptyServerResponseError";
class jr extends B {
  constructor() {
    super("A simple fast-forward merge was not possible."), this.code = this.name = jr.code, this.data = {};
  }
}
jr.code = "FastForwardError";
class Nr extends B {
  /**
   * @param {string} prettyDetails
   * @param {PushResult} result
   */
  constructor(t, r) {
    super(`One or more branches were not updated: ${t}`), this.code = this.name = Nr.code, this.data = { prettyDetails: t, result: r };
  }
}
Nr.code = "GitPushError";
class Dr extends B {
  /**
   * @param {number} statusCode
   * @param {string} statusMessage
   * @param {string} response
   */
  constructor(t, r, n) {
    super(`HTTP Error: ${t} ${r}`), this.code = this.name = Dr.code, this.data = { statusCode: t, statusMessage: r, response: n };
  }
}
Dr.code = "HttpError";
class ot extends B {
  /**
   * @param {'leading-slash'|'trailing-slash'|'directory'} [reason]
   */
  constructor(t) {
    let r = "invalid filepath";
    t === "leading-slash" || t === "trailing-slash" ? r = '"filepath" parameter should not include leading or trailing directory separators because these can cause problems on some platforms.' : t === "directory" && (r = '"filepath" should not be a directory.'), super(r), this.code = this.name = ot.code, this.data = { reason: t };
  }
}
ot.code = "InvalidFilepathError";
class Mr extends B {
  /**
   * @param {string} ref
   * @param {string} suggestion
   * @param {boolean} canForce
   */
  constructor(t, r) {
    super(
      `"${t}" would be an invalid git reference. (Hint: a valid alternative would be "${r}".)`
    ), this.code = this.name = Mr.code, this.data = { ref: t, suggestion: r };
  }
}
Mr.code = "InvalidRefNameError";
class Gr extends B {
  /**
   * @param {number} depth
   */
  constructor(t) {
    super(`Maximum search depth of ${t} exceeded.`), this.code = this.name = Gr.code, this.data = { depth: t };
  }
}
Gr.code = "MaxDepthError";
class Lr extends B {
  constructor() {
    super("Merges with conflicts are not supported yet."), this.code = this.name = Lr.code, this.data = {};
  }
}
Lr.code = "MergeNotSupportedError";
class Hr extends B {
  /**
   * @param {Array<string>} filepaths
   * @param {Array<string>} bothModified
   * @param {Array<string>} deleteByUs
   * @param {Array<string>} deleteByTheirs
   */
  constructor(t, r, n, a) {
    super(
      `Automatic merge failed with one or more merge conflicts in the following files: ${t.toString()}. Fix conflicts then commit the result.`
    ), this.code = this.name = Hr.code, this.data = { filepaths: t, bothModified: r, deleteByUs: n, deleteByTheirs: a };
  }
}
Hr.code = "MergeConflictError";
class qr extends B {
  /**
   * @param {'author'|'committer'|'tagger'} role
   */
  constructor(t) {
    super(
      `No name was provided for ${t} in the argument or in the .git/config file.`
    ), this.code = this.name = qr.code, this.data = { role: t };
  }
}
qr.code = "MissingNameError";
class Tt extends B {
  /**
   * @param {string} parameter
   */
  constructor(t) {
    super(
      `The function requires a "${t}" parameter but none was provided.`
    ), this.code = this.name = Tt.code, this.data = { parameter: t };
  }
}
Tt.code = "MissingParameterError";
class zr extends B {
  /**
   * @param {Error[]} errors
   * @param {string} message
   */
  constructor(t) {
    super(
      'There are multiple errors that were thrown by the method. Please refer to the "errors" property to see more'
    ), this.code = this.name = zr.code, this.data = { errors: t }, this.errors = t;
  }
}
zr.code = "MultipleGitError";
class Wr extends B {
  /**
   * @param {string} expected
   * @param {string} actual
   */
  constructor(t, r) {
    super(`Expected "${t}" but received "${r}".`), this.code = this.name = Wr.code, this.data = { expected: t, actual: r };
  }
}
Wr.code = "ParseError";
class Vr extends B {
  /**
   * @param {'not-fast-forward'|'tag-exists'} reason
   */
  constructor(t) {
    let r = "";
    t === "not-fast-forward" ? r = " because it was not a simple fast-forward" : t === "tag-exists" && (r = " because tag already exists"), super(`Push rejected${r}. Use "force: true" to override.`), this.code = this.name = Vr.code, this.data = { reason: t };
  }
}
Vr.code = "PushRejectedError";
class Xr extends B {
  /**
   * @param {'shallow'|'deepen-since'|'deepen-not'|'deepen-relative'} capability
   * @param {'depth'|'since'|'exclude'|'relative'} parameter
   */
  constructor(t, r) {
    super(
      `Remote does not support the "${t}" so the "${r}" parameter cannot be used.`
    ), this.code = this.name = Xr.code, this.data = { capability: t, parameter: r };
  }
}
Xr.code = "RemoteCapabilityError";
class Yr extends B {
  /**
   * @param {string} preview
   * @param {string} response
   */
  constructor(t, r) {
    super(
      `Remote did not reply using the "smart" HTTP protocol. Expected "001e# service=git-upload-pack" but received: ${t}`
    ), this.code = this.name = Yr.code, this.data = { preview: t, response: r };
  }
}
Yr.code = "SmartHttpError";
class Jr extends B {
  /**
   * @param {string} url
   * @param {string} transport
   * @param {string} [suggestion]
   */
  constructor(t, r, n) {
    super(
      `Git remote "${t}" uses an unrecognized transport protocol: "${r}"`
    ), this.code = this.name = Jr.code, this.data = { url: t, transport: r, suggestion: n };
  }
}
Jr.code = "UnknownTransportError";
class Kr extends B {
  /**
   * @param {string} url
   */
  constructor(t) {
    super(`Cannot parse remote URL: "${t}"`), this.code = this.name = Kr.code, this.data = { url: t };
  }
}
Kr.code = "UrlParseError";
class Zr extends B {
  constructor() {
    super("The operation was canceled."), this.code = this.name = Zr.code, this.data = {};
  }
}
Zr.code = "UserCanceledError";
class Qr extends B {
  /**
   * @param {Array<string>} filepaths
   */
  constructor(t) {
    super(
      `Could not merge index: Entry for '${t}' is not up to date. Either reset the index entry to HEAD, or stage your unstaged changes.`
    ), this.code = this.name = Qr.code, this.data = { filepath: t };
  }
}
Qr.code = "IndexResetError";
class te extends B {
  /**
   * @param {string} ref
   */
  constructor(t) {
    super(
      `"${t}" does not point to any commit. You're maybe working on a repository with no commits yet. `
    ), this.code = this.name = te.code, this.data = { ref: t };
  }
}
te.code = "NoCommitError";
function yr({ name: e, email: t, timestamp: r, timezoneOffset: n }) {
  return n = Xa(n), `${e} <${t}> ${r} ${n}`;
}
function Xa(e) {
  const t = Ya(Ja(e));
  e = Math.abs(e);
  const r = Math.floor(e / 60);
  e -= r * 60;
  let n = String(r), a = String(e);
  return n.length < 2 && (n = "0" + n), a.length < 2 && (a = "0" + a), (t === -1 ? "-" : "+") + n + a;
}
function Ya(e) {
  return Math.sign(e) || (Object.is(e, -0) ? -1 : 1);
}
function Ja(e) {
  return e === 0 ? e : -e;
}
function W(e) {
  return e = e.replace(/\r/g, ""), e = e.replace(/^\n+/, ""), e = e.replace(/\n+$/, "") + `
`, e;
}
function _t(e) {
  const [, t, r, n, a] = e.match(
    /^(.*) <(.*)> (.*) (.*)$/
  );
  return {
    name: t,
    email: r,
    timestamp: Number(n),
    timezoneOffset: Ka(a)
  };
}
function Ka(e) {
  let [, t, r, n] = e.match(/(\+|-)(\d\d)(\d\d)/);
  return n = (t === "+" ? 1 : -1) * (Number(r) * 60 + Number(n)), Za(n);
}
function Za(e) {
  return e === 0 ? e : -e;
}
let _n = class Et {
  constructor(t) {
    if (typeof t == "string")
      this._tag = t;
    else if (Buffer.isBuffer(t))
      this._tag = t.toString("utf8");
    else if (typeof t == "object")
      this._tag = Et.render(t);
    else
      throw new $(
        "invalid type passed to GitAnnotatedTag constructor"
      );
  }
  static from(t) {
    return new Et(t);
  }
  static render(t) {
    return `object ${t.object}
type ${t.type}
tag ${t.tag}
tagger ${yr(t.tagger)}

${t.message}
${t.gpgsig ? t.gpgsig : ""}`;
  }
  justHeaders() {
    return this._tag.slice(0, this._tag.indexOf(`

`));
  }
  message() {
    const t = this.withoutSignature();
    return t.slice(t.indexOf(`

`) + 2);
  }
  parse() {
    return Object.assign(this.headers(), {
      message: this.message(),
      gpgsig: this.gpgsig()
    });
  }
  render() {
    return this._tag;
  }
  headers() {
    const t = this.justHeaders().split(`
`), r = [];
    for (const a of t)
      a[0] === " " ? r[r.length - 1] += `
` + a.slice(1) : r.push(a);
    const n = {};
    for (const a of r) {
      const c = a.slice(0, a.indexOf(" ")), u = a.slice(a.indexOf(" ") + 1);
      Array.isArray(n[c]) ? n[c].push(u) : n[c] = u;
    }
    return n.tagger && (n.tagger = _t(n.tagger)), n.committer && (n.committer = _t(n.committer)), n;
  }
  withoutSignature() {
    const t = W(this._tag);
    return t.indexOf(`
-----BEGIN PGP SIGNATURE-----`) === -1 ? t : t.slice(0, t.lastIndexOf(`
-----BEGIN PGP SIGNATURE-----`));
  }
  gpgsig() {
    if (this._tag.indexOf(`
-----BEGIN PGP SIGNATURE-----`) === -1) return;
    const t = this._tag.slice(
      this._tag.indexOf("-----BEGIN PGP SIGNATURE-----"),
      this._tag.indexOf("-----END PGP SIGNATURE-----") + 27
    );
    return W(t);
  }
  payload() {
    return this.withoutSignature() + `
`;
  }
  toObject() {
    return Buffer.from(this._tag, "utf8");
  }
  static async sign(t, r, n) {
    const a = t.payload();
    let { signature: c } = await r({ payload: a, secretKey: n });
    c = W(c);
    const u = a + c;
    return Et.from(u);
  }
};
function sr(e) {
  return e.trim().split(`
`).map((t) => " " + t).join(`
`) + `
`;
}
function Qa(e) {
  return e.split(`
`).map((t) => t.replace(/^ /, "")).join(`
`);
}
let In = class G {
  constructor(t) {
    if (typeof t == "string")
      this._commit = t;
    else if (Buffer.isBuffer(t))
      this._commit = t.toString("utf8");
    else if (typeof t == "object")
      this._commit = G.render(t);
    else
      throw new $("invalid type passed to GitCommit constructor");
  }
  static fromPayloadSignature({ payload: t, signature: r }) {
    const n = G.justHeaders(t), a = G.justMessage(t), c = W(
      n + `
gpgsig` + sr(r) + `
` + a
    );
    return new G(c);
  }
  static from(t) {
    return new G(t);
  }
  toObject() {
    return Buffer.from(this._commit, "utf8");
  }
  // Todo: allow setting the headers and message
  headers() {
    return this.parseHeaders();
  }
  // Todo: allow setting the headers and message
  message() {
    return G.justMessage(this._commit);
  }
  parse() {
    return Object.assign({ message: this.message() }, this.headers());
  }
  static justMessage(t) {
    return W(t.slice(t.indexOf(`

`) + 2));
  }
  static justHeaders(t) {
    return t.slice(0, t.indexOf(`

`));
  }
  parseHeaders() {
    const t = G.justHeaders(this._commit).split(`
`), r = [];
    for (const a of t)
      a[0] === " " ? r[r.length - 1] += `
` + a.slice(1) : r.push(a);
    const n = {
      parent: []
    };
    for (const a of r) {
      const c = a.slice(0, a.indexOf(" ")), u = a.slice(a.indexOf(" ") + 1);
      Array.isArray(n[c]) ? n[c].push(u) : n[c] = u;
    }
    return n.author && (n.author = _t(n.author)), n.committer && (n.committer = _t(n.committer)), n;
  }
  static renderHeaders(t) {
    let r = "";
    if (t.tree ? r += `tree ${t.tree}
` : r += `tree 4b825dc642cb6eb9a060e54bf8d69288fbee4904
`, t.parent) {
      if (t.parent.length === void 0)
        throw new $("commit 'parent' property should be an array");
      for (const c of t.parent)
        r += `parent ${c}
`;
    }
    const n = t.author;
    r += `author ${yr(n)}
`;
    const a = t.committer || t.author;
    return r += `committer ${yr(a)}
`, t.gpgsig && (r += "gpgsig" + sr(t.gpgsig)), r;
  }
  static render(t) {
    return G.renderHeaders(t) + `
` + W(t.message);
  }
  render() {
    return this._commit;
  }
  withoutSignature() {
    const t = W(this._commit);
    if (t.indexOf(`
gpgsig`) === -1) return t;
    const r = t.slice(0, t.indexOf(`
gpgsig`)), n = t.slice(
      t.indexOf(`-----END PGP SIGNATURE-----
`) + 28
    );
    return W(r + `
` + n);
  }
  isolateSignature() {
    const t = this._commit.slice(
      this._commit.indexOf("-----BEGIN PGP SIGNATURE-----"),
      this._commit.indexOf("-----END PGP SIGNATURE-----") + 27
    );
    return Qa(t);
  }
  static async sign(t, r, n) {
    const a = t.withoutSignature(), c = G.justMessage(t._commit);
    let { signature: u } = await r({ payload: a, secretKey: n });
    u = W(u);
    const h = G.justHeaders(t._commit) + `
gpgsig` + sr(u) + `
` + c;
    return G.from(h);
  }
};
async function dr({ fs: e, cache: t, gitdir: r, oid: n }) {
  if (n === "4b825dc642cb6eb9a060e54bf8d69288fbee4904")
    return { tree: Bt.from([]), oid: n };
  const { type: a, object: c } = await pt({ fs: e, cache: t, gitdir: r, oid: n });
  if (a === "tag")
    return n = _n.from(c).parse().object, dr({ fs: e, cache: t, gitdir: r, oid: n });
  if (a === "commit")
    return n = In.from(c).parse().tree, dr({ fs: e, cache: t, gitdir: r, oid: n });
  if (a !== "tree")
    throw new kt(n, a, "tree");
  return { tree: Bt.from(c), oid: n };
}
async function mr(e, t) {
  const r = await e.readdir(t);
  r == null ? await e.rm(t) : r.length ? await Promise.all(
    r.map((n) => {
      const a = nt(t, n);
      return e.lstat(a).then((c) => {
        if (c)
          return c.isDirectory() ? mr(e, a) : e.rm(a);
      });
    })
  ).then(() => e.rmdir(t)) : await e.rmdir(t);
}
function ts(e) {
  return rs(e) && He(e.then) && He(e.catch);
}
function rs(e) {
  return e && typeof e == "object";
}
function He(e) {
  return typeof e == "function";
}
function qe(e) {
  return ts(((r) => {
    try {
      return r.readFile().catch((n) => n);
    } catch (n) {
      return n;
    }
  })(e));
}
const ze = [
  "readFile",
  "writeFile",
  "mkdir",
  "rmdir",
  "unlink",
  "stat",
  "lstat",
  "readdir",
  "readlink",
  "symlink"
];
function We(e, t) {
  if (qe(t))
    for (const r of ze)
      e[`_${r}`] = t[r].bind(t);
  else
    for (const r of ze)
      e[`_${r}`] = mt(t[r].bind(t));
  qe(t) ? (t.cp && (e._cp = t.cp.bind(t)), t.rm ? e._rm = t.rm.bind(t) : t.rmdir.length > 1 ? e._rm = t.rmdir.bind(t) : e._rm = mr.bind(null, e)) : (t.cp && (e._cp = mt(t.cp.bind(t))), t.rm ? e._rm = mt(t.rm.bind(t)) : t.rmdir.length > 2 ? e._rm = mt(t.rmdir.bind(t)) : e._rm = mr.bind(null, e));
}
class Fn {
  /**
   * Creates an instance of FileSystem.
   *
   * @param {Object} fs - A file system implementation to wrap.
   */
  constructor(t) {
    if (typeof t._original_unwrapped_fs < "u") return t;
    const r = Object.getOwnPropertyDescriptor(t, "promises");
    r && r.enumerable ? We(this, t.promises) : We(this, t), this._original_unwrapped_fs = t;
  }
  /**
   * Return true if a file exists, false if it doesn't exist.
   * Rethrows errors that aren't related to file existence.
   *
   * @param {string} filepath - The path to the file.
   * @param {Object} [options] - Additional options.
   * @returns {Promise<boolean>} - `true` if the file exists, `false` otherwise.
   */
  async exists(t, r = {}) {
    try {
      return await this._stat(t), !0;
    } catch (n) {
      if (n.code === "ENOENT" || n.code === "ENOTDIR" || (n.code || "").includes("ENS"))
        return !1;
      throw console.log('Unhandled error in "FileSystem.exists()" function', n), n;
    }
  }
  /**
   * Return the contents of a file if it exists, otherwise returns null.
   *
   * @param {string} filepath - The path to the file.
   * @param {Object} [options] - Options for reading the file.
   * @returns {Promise<Buffer|string|null>} - The file contents, or `null` if the file doesn't exist.
   */
  async read(t, r = {}) {
    try {
      let n = await this._readFile(t, r);
      if (r.autocrlf === "true")
        try {
          n = new TextDecoder("utf8", { fatal: !0 }).decode(n), n = n.replace(/\r\n/g, `
`), n = new TextEncoder().encode(n);
        } catch {
        }
      return typeof n != "string" && (n = Buffer.from(n)), n;
    } catch {
      return null;
    }
  }
  /**
   * Write a file (creating missing directories if need be) without throwing errors.
   *
   * @param {string} filepath - The path to the file.
   * @param {Buffer|Uint8Array|string} contents - The data to write.
   * @param {Object|string} [options] - Options for writing the file.
   * @returns {Promise<void>}
   */
  async write(t, r, n = {}) {
    try {
      await this._writeFile(t, r, n);
    } catch {
      await this.mkdir(hr(t)), await this._writeFile(t, r, n);
    }
  }
  /**
   * Make a directory (or series of nested directories) without throwing an error if it already exists.
   *
   * @param {string} filepath - The path to the directory.
   * @param {boolean} [_selfCall=false] - Internal flag to prevent infinite recursion.
   * @returns {Promise<void>}
   */
  async mkdir(t, r = !1) {
    try {
      await this._mkdir(t);
    } catch (n) {
      if (n === null || n.code === "EEXIST") return;
      if (r) throw n;
      if (n.code === "ENOENT") {
        const a = hr(t);
        if (a === "." || a === "/" || a === t) throw n;
        await this.mkdir(a), await this.mkdir(t, !0);
      }
    }
  }
  /**
   * Delete a file without throwing an error if it is already deleted.
   *
   * @param {string} filepath - The path to the file.
   * @returns {Promise<void>}
   */
  async rm(t) {
    try {
      await this._unlink(t);
    } catch (r) {
      if (r.code !== "ENOENT") throw r;
    }
  }
  /**
   * Delete a directory without throwing an error if it is already deleted.
   *
   * @param {string} filepath - The path to the directory.
   * @param {Object} [opts] - Options for deleting the directory.
   * @returns {Promise<void>}
   */
  async rmdir(t, r) {
    try {
      r && r.recursive ? await this._rm(t, r) : await this._rmdir(t);
    } catch (n) {
      if (n.code !== "ENOENT") throw n;
    }
  }
  /**
   * Read a directory without throwing an error is the directory doesn't exist
   *
   * @param {string} filepath - The path to the directory.
   * @returns {Promise<string[]|null>} - An array of file names, or `null` if the path is not a directory.
   */
  async readdir(t) {
    try {
      const r = await this._readdir(t);
      return r.sort(Ir), r;
    } catch (r) {
      return r.code === "ENOTDIR" ? null : [];
    }
  }
  /**
   * Return a flat list of all the files nested inside a directory
   *
   * Based on an elegant concurrent recursive solution from SO
   * https://stackoverflow.com/a/45130990/2168416
   *
   * @param {string} dir - The directory to read.
   * @returns {Promise<string[]>} - A flat list of all files in the directory.
   */
  async readdirDeep(t) {
    const r = await this._readdir(t);
    return (await Promise.all(
      r.map(async (a) => {
        const c = t + "/" + a;
        return (await this._stat(c)).isDirectory() ? this.readdirDeep(c) : c;
      })
    )).reduce((a, c) => a.concat(c), []);
  }
  /**
   * Return the Stats of a file/symlink if it exists, otherwise returns null.
   * Rethrows errors that aren't related to file existence.
   *
   * @param {string} filename - The path to the file or symlink.
   * @returns {Promise<Object|null>} - The stats object, or `null` if the file doesn't exist.
   */
  async lstat(t) {
    try {
      return await this._lstat(t);
    } catch (r) {
      if (r.code === "ENOENT" || (r.code || "").includes("ENS"))
        return null;
      throw r;
    }
  }
  /**
   * Reads the contents of a symlink if it exists, otherwise returns null.
   * Rethrows errors that aren't related to file existence.
   *
   * @param {string} filename - The path to the symlink.
   * @param {Object} [opts={ encoding: 'buffer' }] - Options for reading the symlink.
   * @returns {Promise<Buffer|null>} - The symlink target, or `null` if it doesn't exist.
   */
  async readlink(t, r = { encoding: "buffer" }) {
    try {
      const n = await this._readlink(t, r);
      return Buffer.isBuffer(n) ? n : Buffer.from(n);
    } catch (n) {
      if (n.code === "ENOENT" || (n.code || "").includes("ENS"))
        return null;
      throw n;
    }
  }
  /**
   * Write the contents of buffer to a symlink.
   *
   * @param {string} filename - The path to the symlink.
   * @param {Buffer} buffer - The symlink target.
   * @returns {Promise<void>}
   */
  async writelink(t, r) {
    return this._symlink(r.toString("utf8"), t);
  }
}
function V(e, t) {
  if (t === void 0)
    throw new Tt(e);
}
function es(e) {
  return e.startsWith("/") || /^[a-zA-Z]:[\\/]/.test(e);
}
async function $n({ fsp: e, dotgit: t }) {
  V("fsp", e), V("dotgit", t);
  const r = await e._stat(t).catch(() => ({ isFile: () => !1, isDirectory: () => !1 }));
  return r.isDirectory() ? t : r.isFile() ? e._readFile(t, "utf8").then((n) => n.trimRight().substr(8)).then((n) => es(n) ? n : nt(hr(t), n)) : t;
}
async function ns({ fs: e, cache: t, gitdir: r, oid: n, filepath: a }) {
  if (a.startsWith("/"))
    throw new ot("leading-slash");
  if (a.endsWith("/"))
    throw new ot("trailing-slash");
  const c = n, u = await dr({ fs: e, cache: t, gitdir: r, oid: n }), s = u.tree;
  if (a === "")
    n = u.oid;
  else {
    const h = a.split("/");
    n = await Pn({
      fs: e,
      cache: t,
      gitdir: r,
      tree: s,
      pathArray: h,
      oid: c,
      filepath: a
    });
  }
  return n;
}
async function Pn({
  fs: e,
  cache: t,
  gitdir: r,
  tree: n,
  pathArray: a,
  oid: c,
  filepath: u
}) {
  const s = a.shift();
  for (const h of n)
    if (h.path === s) {
      if (a.length === 0)
        return h.oid;
      {
        const { type: d, object: m } = await pt({
          fs: e,
          cache: t,
          gitdir: r,
          oid: h.oid
        });
        if (d !== "tree")
          throw new kt(c, d, "tree", u);
        return n = Bt.from(m), Pn({
          fs: e,
          cache: t,
          gitdir: r,
          tree: n,
          pathArray: a,
          oid: c,
          filepath: u
        });
      }
    }
  throw new ht(`file or directory found at "${c}:${u}"`);
}
async function is({
  fs: e,
  cache: t,
  onProgress: r,
  dir: n,
  gitdir: a,
  filepath: c
}) {
  try {
    c = nt(n, c);
    const u = await e.read(c), s = (d) => pt({ fs: e, cache: t, gitdir: a, oid: d }), h = await Bn.fromPack({
      pack: u,
      getExternalRefDelta: s,
      onProgress: r
    });
    return await e.write(c.replace(/\.pack$/, ".idx"), await h.toBuffer()), {
      oids: [...h.hashes]
    };
  } catch (u) {
    throw u.caller = "git.indexPack", u;
  }
}
async function os({
  fs: e,
  onProgress: t,
  dir: r,
  gitdir: n = nt(r, ".git"),
  filepath: a,
  cache: c = {}
}) {
  try {
    V("fs", e), V("dir", r), V("gitdir", r), V("filepath", a);
    const u = new Fn(e), s = await $n({ fsp: u, dotgit: n });
    return await is({
      fs: u,
      cache: c,
      onProgress: t,
      dir: r,
      gitdir: s,
      filepath: a
    });
  } catch (u) {
    throw u.caller = "git.indexPack", u;
  }
}
async function as({
  fs: e,
  dir: t,
  gitdir: r = nt(t, ".git"),
  oid: n,
  format: a = "parsed",
  filepath: c = void 0,
  encoding: u = void 0,
  cache: s = {}
}) {
  try {
    V("fs", e), V("gitdir", r), V("oid", n);
    const h = new Fn(e), d = await $n({ fsp: h, dotgit: r });
    c !== void 0 && (n = await ns({
      fs: h,
      cache: s,
      gitdir: d,
      oid: n,
      filepath: c
    }));
    const y = await pt({
      fs: h,
      cache: s,
      gitdir: d,
      oid: n,
      format: a === "parsed" ? "content" : a
    });
    if (y.oid = n, a === "parsed")
      switch (y.format = "parsed", y.type) {
        case "commit":
          y.object = In.from(y.object).parse();
          break;
        case "tree":
          y.object = Bt.from(y.object).entries();
          break;
        case "blob":
          u ? y.object = y.object.toString(u) : (y.object = new Uint8Array(y.object), y.format = "content");
          break;
        case "tag":
          y.object = _n.from(y.object).parse();
          break;
        default:
          throw new kt(
            y.oid,
            y.type,
            "blob|commit|tag|tree"
          );
      }
    else (y.format === "deflated" || y.format === "wrapped") && (y.type = y.format);
    return y;
  } catch (h) {
    throw h.caller = "git.readObject", h;
  }
}
typeof globalThis.Buffer > "u" && (globalThis.Buffer = br.Buffer);
class As extends Error {
  constructor(t, r, n) {
    super(`Object ${t} has type ${r}, but expected ${n}.`), this.name = "ObjectTypeError";
  }
}
function ss(e, t) {
  const r = t.toString(16);
  return "0".repeat(e - r.length) + r;
}
class Ss {
  static flush() {
    return Buffer.from("0000", "utf8");
  }
  static delim() {
    return Buffer.from("0001", "utf8");
  }
  static encode(t) {
    const r = typeof t == "string" ? Buffer.from(t) : t;
    return Buffer.concat([
      Buffer.from(ss(4, r.length + 4), "utf8"),
      Buffer.from(r)
    ]);
  }
  static decode(t) {
    const r = parseInt(t.subarray(0, 4).toString("utf8"), 16);
    return t.subarray(4, r).toString("utf8");
  }
}
function cs(e) {
  return Symbol.asyncIterator in e ? e[Symbol.asyncIterator]() : e[Symbol.iterator]();
}
async function us(e) {
  let t = 0;
  const r = [], n = cs(e);
  for (; ; ) {
    const { value: u, done: s } = await n.next();
    if (s)
      break;
    u && (r.push(u), t += u.byteLength);
  }
  n.return && await n.return();
  const a = new Uint8Array(t);
  let c = 0;
  for (const u of r)
    a.set(u, c), c += u.byteLength;
  return a;
}
async function Bs(e) {
  const t = Buffer.from(await us(e)), r = [], n = [], a = [], c = [], u = [], s = [];
  let h = !1, d = 0;
  for (; d + 4 <= t.length; ) {
    const m = parseInt(
      t.subarray(d, d + 4).toString("utf8"),
      16
    );
    if (d += 4, m === 0 || m === 1)
      continue;
    if (Number.isNaN(m) || m < 4)
      throw new Error("Invalid git packet line length.");
    const y = t.subarray(d, d + m - 4);
    d += m - 4;
    const b = y[0];
    if (b === 1) {
      n.push(y.subarray(1));
      continue;
    }
    if (b === 2) {
      a.push(y.subarray(1));
      continue;
    }
    if (b === 3)
      throw new Error(y.subarray(1).toString("utf8"));
    r.push(y);
  }
  for (const m of r) {
    const y = m.toString("utf8").trim();
    if (y.startsWith("shallow"))
      c.push(y.slice(-40).trim());
    else if (y.startsWith("unshallow"))
      u.push(y.slice(-40).trim());
    else if (y.startsWith("ACK")) {
      const [, b, E] = y.split(" ");
      s.push({ oid: b, status: E });
    } else y.startsWith("NAK") && (h = !0);
  }
  return {
    shallows: c,
    unshallows: u,
    acks: s,
    nak: h,
    packfile: n,
    progress: a
  };
}
function M(e) {
  const t = e.startsWith("/"), r = [];
  for (const a of e.split(/[\\/]+/))
    if (!(!a || a === ".")) {
      if (a === "..") {
        r.pop();
        continue;
      }
      r.push(a);
    }
  const n = r.join("/");
  return t ? `/${n}` : n || ".";
}
function wt(e) {
  const t = M(e), r = t.lastIndexOf("/");
  return r <= 0 ? t.startsWith("/") ? "/" : "." : t.slice(0, r);
}
function Ve(e) {
  return {
    isFile: () => e === "file",
    isDirectory: () => e === "directory",
    isSymbolicLink: () => e === "symlink"
  };
}
function kn(e, t) {
  const r = new Error(t);
  return r.code = e, r;
}
function Q(e) {
  return kn(
    "ENOENT",
    `ENOENT: no such file or directory, ${e}`
  );
}
class fs {
  constructor() {
    this.entries = /* @__PURE__ */ new Map([
      ["/", { type: "directory" }]
    ]), this.mkdirSync("/repo/.git/objects/pack"), this.writeFileSync("/repo/.git/HEAD", `ref: refs/heads/main
`);
  }
  mkdirSync(t) {
    const r = M(t);
    if (r === "/") {
      this.entries.set("/", { type: "directory" });
      return;
    }
    let n = "";
    for (const a of r.split("/").filter(Boolean))
      n += `/${a}`, this.entries.set(n, { type: "directory" });
  }
  writeFileSync(t, r) {
    const n = M(t);
    this.mkdirSync(wt(n)), this.entries.set(n, {
      type: "file",
      data: Buffer.from(r)
    });
  }
  async readFile(t, r) {
    const n = M(t), a = this.getFileEntry(n);
    if (!a || a.type !== "file")
      throw Q(n);
    const c = typeof r == "string" ? r : r == null ? void 0 : r.encoding;
    return c ? a.data.toString(c) : Buffer.from(a.data);
  }
  async writeFile(t, r, n) {
    const a = M(t);
    this.mkdirSync(wt(a));
    const c = typeof n == "string" ? n : n == null ? void 0 : n.encoding;
    this.entries.set(a, {
      type: "file",
      data: typeof r == "string" ? Buffer.from(r, c) : Buffer.from(r)
    });
  }
  async mkdir(t) {
    this.mkdirSync(t);
  }
  async readdir(t) {
    const r = M(t), n = this.entries.get(r);
    if (!n || n.type !== "directory")
      throw Q(r);
    const a = /* @__PURE__ */ new Set(), c = r === "/" ? "/" : `${r}/`;
    for (const u of this.entries.keys()) {
      if (!u.startsWith(c) || u === r)
        continue;
      const s = u.slice(c.length).split("/")[0];
      s && a.add(s);
    }
    return [...a].sort();
  }
  async stat(t) {
    const r = M(t), n = this.getResolvedEntry(r);
    if (!n)
      throw Q(r);
    return Ve(n.type);
  }
  async lstat(t) {
    const r = M(t), n = this.entries.get(r);
    if (!n)
      throw Q(r);
    return Ve(n.type);
  }
  async unlink(t) {
    const r = M(t), n = this.entries.get(r);
    if (!n || n.type !== "file" && n.type !== "symlink")
      throw Q(r);
    this.entries.delete(r);
  }
  async rmdir(t) {
    const r = M(t), n = r === "/" ? "/" : `${r}/`;
    for (const a of this.entries.keys())
      if (a.startsWith(n) && a !== r)
        throw new Error(
          `ENOTEMPTY: directory not empty, ${r}`
        );
    this.entries.delete(r);
  }
  async rm(t) {
    this.entries.delete(M(t));
  }
  async readlink(t) {
    const r = M(t), n = this.entries.get(r);
    if (!n || n.type !== "symlink")
      throw Q(r);
    return n.target;
  }
  async symlink(t, r) {
    const n = M(r);
    this.mkdirSync(wt(n)), this.entries.set(n, {
      type: "symlink",
      target: t
    });
  }
  getFileEntry(t) {
    const r = this.entries.get(t);
    return (r == null ? void 0 : r.type) !== "symlink" ? r : this.getResolvedEntry(t);
  }
  getResolvedEntry(t, r = /* @__PURE__ */ new Set()) {
    const n = this.entries.get(t);
    if ((n == null ? void 0 : n.type) !== "symlink")
      return n;
    if (r.has(t))
      throw kn(
        "ELOOP",
        `ELOOP: too many symbolic links, ${t}`
      );
    r.add(t);
    const a = n.target.startsWith("/") ? n.target : M(`${wt(t)}/${n.target}`);
    return this.getResolvedEntry(a, r);
  }
}
function ls(e) {
  const t = e.readUInt32BE(1028), r = [], n = /* @__PURE__ */ new Map();
  let a = 8 + 256 * 4;
  for (let s = 0; s < t; s++)
    r.push(e.subarray(a, a + 20).toString("hex")), a += 20;
  a += t * 4;
  const c = a;
  a += t * 4;
  const u = a;
  for (let s = 0; s < t; s++) {
    const h = e.readUInt32BE(c + s * 4);
    if (h & 2147483648) {
      const d = h & 2147483647, m = Number(
        e.readBigUInt64BE(u + d * 8)
      );
      n.set(r[s], m);
    } else
      n.set(r[s], h);
  }
  return {
    hashes: r,
    offsets: n,
    packfileSha: e.subarray(e.length - 40, e.length - 20).toString("hex")
  };
}
const Xe = "/repo", Ye = "/repo/.git", hs = "/repo/.git/objects/pack/pack.playground.pack", ps = ".git/objects/pack/pack.playground.pack", ys = "/repo/.git/objects/pack/pack.playground.idx";
class gr {
  constructor({
    fs: t,
    indexBuffer: r,
    hashes: n,
    offsets: a,
    packfileSha: c,
    offsetToOid: u
  }) {
    this.cache = {}, this.fs = t, this.indexBuffer = r, this.hashes = n, this.offsets = a, this.packfileSha = c, this.offsetToOid = u ?? new Map([...a].map(([s, h]) => [h, s]));
  }
  static async fromPack({ pack: t }) {
    const r = new fs(), n = Buffer.from(t);
    if (await r.writeFile(hs, n), n.byteLength === 0)
      return new gr({
        fs: r,
        indexBuffer: Buffer.alloc(0),
        hashes: [],
        offsets: /* @__PURE__ */ new Map(),
        packfileSha: ""
      });
    await os({
      fs: r,
      dir: Xe,
      gitdir: Ye,
      filepath: ps,
      cache: {}
    });
    const a = await r.readFile(ys), c = ls(a);
    return new gr({
      fs: r,
      indexBuffer: a,
      ...c
    });
  }
  async read({ oid: t }) {
    const r = await as({
      fs: this.fs,
      dir: Xe,
      gitdir: Ye,
      oid: t,
      format: "content",
      cache: this.cache
    });
    return {
      oid: t,
      type: r.type,
      format: r.format,
      object: Buffer.from(r.object)
    };
  }
  async readSlice({ start: t }) {
    const r = this.offsetToOid.get(t);
    if (r)
      return this.read({ oid: r });
    throw new Error(`Could not read object at packfile offset ${t}.`);
  }
  async toBuffer() {
    return Buffer.from(this.indexBuffer);
  }
}
function ds(e, t) {
  return e < t ? -1 : e > t ? 1 : 0;
}
function Tn(e, t) {
  return ds(e.path, t.path);
}
class _s {
  constructor() {
    this.entries = /* @__PURE__ */ new Map();
  }
  insert({ filepath: t, oid: r, stats: n }) {
    const a = Buffer.from(t);
    this.entries.set(t, {
      filepath: t,
      path: t,
      oid: r,
      stats: n,
      flags: {
        assumeValid: !1,
        extended: !1,
        stage: 0,
        nameLength: a.length < 4095 ? a.length : 4095
      }
    });
  }
  renderFlags(t) {
    return (t.flags.assumeValid ? 32768 : 0) + (t.flags.extended ? 16384 : 0) + ((t.flags.stage & 3) << 12) + (t.flags.nameLength & 4095);
  }
  entryToBuffer(t) {
    const r = Buffer.from(t.path), n = Math.ceil((62 + r.length + 1) / 8) * 8, a = Buffer.alloc(n);
    let c = 0;
    const u = (s) => {
      a.writeUInt32BE(s, c), c += 4;
    };
    return u(t.stats.ctimeSeconds), u(t.stats.ctimeNanoseconds), u(t.stats.mtimeSeconds), u(t.stats.mtimeNanoseconds), u(t.stats.dev), u(t.stats.ino), u(t.stats.mode || 33188), u(t.stats.uid), u(t.stats.gid), u(t.stats.size), Buffer.from(t.oid, "hex").copy(a, c), c += 20, a.writeUInt16BE(this.renderFlags(t), c), c += 2, r.copy(a, c), a;
  }
  async toObject() {
    const t = Buffer.alloc(12);
    t.write("DIRC", 0, 4, "utf8"), t.writeUInt32BE(2, 4), t.writeUInt32BE(this.entries.size, 8);
    const r = Buffer.concat(
      [...this.entries.values()].sort(Tn).map((c) => this.entryToBuffer(c))
    ), n = Buffer.concat([t, r]), a = ms(n);
    return Buffer.concat([n, a]);
  }
}
function ms(e) {
  return Buffer.from(Qn("sha1").update(e).digest("hex"), "hex");
}
function wr(e) {
  const t = e == null ? void 0 : e.match(/^(.*) <(.*)> (.*) (.*)$/);
  return t ? {
    name: t[1],
    email: t[2],
    timestamp: Number(t[3]),
    timezoneOffset: 0
  } : {
    name: "",
    email: "",
    timestamp: 0,
    timezoneOffset: 0
  };
}
function On(e) {
  const [t, ...r] = e.split(`

`), n = [];
  for (const c of t.split(`
`))
    c.startsWith(" ") && n.length > 0 ? n[n.length - 1] += `
${c.slice(1)}` : n.push(c);
  const a = {};
  for (const c of n) {
    const u = c.indexOf(" ");
    if (u === -1)
      continue;
    const s = c.slice(0, u), h = c.slice(u + 1);
    if (s === "parent") {
      const d = a.parent;
      a.parent = Array.isArray(d) ? [...d, h] : d ? [d, h] : [h];
    } else
      a[s] = h;
  }
  return {
    headers: a,
    message: r.join(`

`)
  };
}
class Un {
  constructor(t) {
    this.payload = Buffer.from(t).toString("utf8");
  }
  static from(t) {
    return new Un(t);
  }
  parse() {
    const { headers: t, message: r } = On(this.payload), n = t.parent;
    return {
      tree: String(t.tree ?? ""),
      parent: Array.isArray(n) ? n : n ? [String(n)] : [],
      author: wr(String(t.author ?? "")),
      committer: wr(String(t.committer ?? "")),
      message: r,
      gpgsig: typeof t.gpgsig == "string" ? t.gpgsig : void 0
    };
  }
}
function gs(e) {
  return e === "160000" ? "commit" : e.match(/^0?4/) ? "tree" : "blob";
}
class Rn {
  constructor(t) {
    const r = [], n = Buffer.from(t);
    let a = 0;
    for (; a < n.length; ) {
      const c = n.indexOf(32, a), u = n.indexOf(0, a);
      if (c === -1 || u === -1)
        throw new Error("Invalid Git tree object.");
      let s = n.subarray(a, c).toString("utf8");
      s === "40000" && (s = "040000");
      const h = n.subarray(c + 1, u).toString("utf8"), d = n.subarray(u + 1, u + 21).toString("hex");
      r.push({
        mode: s,
        path: h,
        oid: d,
        type: gs(s)
      }), a = u + 21;
    }
    this.object = r.sort(Tn);
  }
  static from(t) {
    return new Rn(t);
  }
  entries() {
    return this.object;
  }
}
class Cn {
  constructor(t) {
    this.payload = Buffer.from(t).toString("utf8");
  }
  static from(t) {
    return new Cn(t);
  }
  parse() {
    const { headers: t, message: r } = On(this.payload);
    return {
      object: String(t.object ?? ""),
      type: String(t.type ?? ""),
      tag: String(t.tag ?? ""),
      tagger: wr(String(t.tagger ?? "")),
      message: r,
      signature: typeof t.signature == "string" ? t.signature : void 0
    };
  }
}
export {
  Ss as G,
  As as O,
  gr as a,
  Cn as b,
  us as c,
  Rn as d,
  Un as e,
  br as f,
  _s as g,
  Bs as p
};
