// fs-ext-extra-prebuilt/dist/fs-ext.js
"use strict";
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = void 0;
exports.flock = flock;
exports.flockSync = flockSync;
exports.fcntl = fcntl;
exports.fcntlSync = fcntlSync;
exports.seek = seek;
exports.seekSync = seekSync;
exports.statVFS = statVFS;
exports.lockFileEx = lockFileEx;
exports.lockFileExSync = lockFileExSync;
exports.unlockFileEx = unlockFileEx;
exports.unlockFileExSync = unlockFileExSync;
exports.useNativeModule = useNativeModule;
exports.getNativeModuleSource = getNativeModuleSource;
// eslint-disable-next-line @typescript-eslint/no-var-requires
const loader = require('../load-prebuilt');
// Load native module (prebuilt first, then local build)
let binding = loader.loadNativeModule();
if (!binding) {
    throw new Error('Failed to load fs-ext native module. ' +
        'No prebuilt binary found for ' +
        process.platform +
        '-' +
        process.arch +
        ' (Node ' +
        process.version +
        '), and no local build available.');
}
// Convert string flock flags to numeric constants
function stringToFlockFlags(flag) {
    if (typeof flag !== 'string') {
        return flag;
    }
    switch (flag) {
        case 'sh':
            return binding.constants.LOCK_SH;
        case 'ex':
            return binding.constants.LOCK_EX;
        case 'shnb':
            return binding.constants.LOCK_SH | binding.constants.LOCK_NB;
        case 'exnb':
            return binding.constants.LOCK_EX | binding.constants.LOCK_NB;
        case 'un':
            return binding.constants.LOCK_UN;
        default:
            throw new Error('Unknown flock flag: ' + flag);
    }
}
// Convert string fcntl commands to numeric constants
function stringToFcntlFlags(flag) {
    if (typeof flag !== 'string') {
        return flag;
    }
    switch (flag) {
        case 'getfd':
            return binding.constants.F_GETFD;
        case 'setfd':
            return binding.constants.F_SETFD;
        case 'setlk':
            return binding.constants.F_SETLK;
        case 'setlkw':
            return binding.constants.F_SETLKW;
        case 'getlk':
            return binding.constants.F_GETLK;
        default:
            throw new Error('Unknown fcntl flag: ' + flag);
    }
}
function noop() { }
/**
 * Apply or remove an advisory lock on an open file (async).
 *
 * @param fd - File descriptor
 * @param flags - Lock operation: 'sh', 'ex', 'shnb', 'exnb', 'un', or numeric flags
 * @param callback - Called when operation completes
 */
function flock(fd, flags, callback) {
    const cb = typeof callback === 'function' ? callback : noop;
    const oper = stringToFlockFlags(flags);
    binding.flock(fd, oper, cb);
}
/**
 * Apply or remove an advisory lock on an open file (sync).
 *
 * @param fd - File descriptor
 * @param flags - Lock operation: 'sh', 'ex', 'shnb', 'exnb', 'un', or numeric flags
 * @returns 0 on success
 */
function flockSync(fd, flags) {
    const oper = stringToFlockFlags(flags);
    return binding.flock(fd, oper);
}
/**
 * Perform file control operations (async).
 *
 * For F_SETLK and F_SETLKW, start and len specify the byte range to lock.
 * If start and len are omitted, they default to 0, which locks the entire file.
 *
 * @param fd - File descriptor
 * @param cmd - Command: 'getfd', 'setfd', 'setlk', 'setlkw', 'getlk', or numeric
 * @param arg - Command argument (e.g., F_RDLCK, F_WRLCK, F_UNLCK for locks)
 * @param start - Start offset for byte-range locks (optional)
 * @param len - Length for byte-range locks (optional)
 * @param callback - Called when operation completes
 */
function fcntl(fd, cmd, arg, start, len, callback) {
    const numCmd = stringToFcntlFlags(cmd);
    // Handle different argument patterns:
    // fcntl(fd, cmd, callback) - arg defaults to 0
    // fcntl(fd, cmd, arg, callback) - no range
    // fcntl(fd, cmd, arg, start, len) - sync with range
    // fcntl(fd, cmd, arg, start, len, callback) - async with range
    let numArg;
    let numStart;
    let numLen;
    let cb;
    if (typeof arg === 'function') {
        cb = arg;
        numArg = 0;
    }
    else if (typeof start === 'function') {
        cb = start;
        numArg = arg ?? 0;
    }
    else if (typeof len === 'function') {
        cb = len;
        numArg = arg ?? 0;
        numStart = start;
    }
    else {
        cb = callback;
        numArg = arg ?? 0;
        numStart = start;
        numLen = len;
    }
    // If start and len are provided, pass them to the binding
    if (numStart !== undefined && numLen !== undefined) {
        if (typeof cb === 'function') {
            binding.fcntl(fd, numCmd, numArg, numStart, numLen, cb);
            return;
        }
        binding.fcntl(fd, numCmd, numArg, numStart, numLen);
        return;
    }
    if (typeof cb === 'function') {
        binding.fcntl(fd, numCmd, numArg, cb);
        return;
    }
    binding.fcntl(fd, numCmd, numArg);
}
/**
 * Perform file control operations (sync).
 *
 * For F_SETLK and F_SETLKW, start and len specify the byte range to lock.
 * If start and len are omitted, they default to 0, which locks the entire file.
 *
 * @param fd - File descriptor
 * @param cmd - Command: 'getfd', 'setfd', 'setlk', 'setlkw', 'getlk', or numeric
 * @param arg - Command argument (e.g., F_RDLCK, F_WRLCK, F_UNLCK for locks)
 * @param start - Start offset for byte-range locks (optional)
 * @param len - Length for byte-range locks (optional)
 * @returns Result of the fcntl call
 */
function fcntlSync(fd, cmd, arg, start, len) {
    const numCmd = stringToFcntlFlags(cmd);
    const numArg = arg ?? 0;
    // If start and len are provided, pass them to the binding
    if (start !== undefined && len !== undefined) {
        return binding.fcntl(fd, numCmd, numArg, start, len);
    }
    return binding.fcntl(fd, numCmd, numArg);
}
/**
 * Reposition read/write file offset (async).
 *
 * @param fd - File descriptor
 * @param position - Offset in bytes
 * @param whence - SEEK_SET, SEEK_CUR, or SEEK_END
 * @param callback - Called with (err, newPosition)
 */
function seek(fd, position, whence, callback) {
    const cb = typeof callback === 'function' ? callback : noop;
    binding.seek(fd, position, whence, cb);
}
/**
 * Reposition read/write file offset (sync).
 *
 * @param fd - File descriptor
 * @param position - Offset in bytes
 * @param whence - SEEK_SET, SEEK_CUR, or SEEK_END
 * @returns New file position
 */
function seekSync(fd, position, whence) {
    return binding.seek(fd, position, whence);
}
/**
 * Get filesystem statistics.
 *
 * @param path - Path to any file on the filesystem
 * @param callback - Called with (err, stats)
 */
function statVFS(path, callback) {
    const fsPath = path || '/';
    binding.statVFS(fsPath, callback);
}
/**
 * Lock a file region using Windows LockFileEx (async, Windows only).
 *
 * @param fd - File descriptor
 * @param flags - 0 for shared, LOCKFILE_EXCLUSIVE_LOCK for exclusive
 * @param offsetLow - Low 32 bits of offset
 * @param offsetHigh - High 32 bits of offset
 * @param lengthLow - Low 32 bits of length
 * @param lengthHigh - High 32 bits of length
 * @param callback - Called when operation completes
 */
function lockFileEx(fd, flags, offsetLow, offsetHigh, lengthLow, lengthHigh, callback) {
    if (!binding.lockFileEx) {
        throw new Error('lockFileEx is only available on Windows');
    }
    const cb = typeof callback === 'function' ? callback : noop;
    binding.lockFileEx(fd, flags, offsetLow, offsetHigh, lengthLow, lengthHigh, cb);
}
/**
 * Lock a file region using Windows LockFileEx (sync, Windows only).
 *
 * @param fd - File descriptor
 * @param flags - 0 for shared, LOCKFILE_EXCLUSIVE_LOCK for exclusive
 * @param offsetLow - Low 32 bits of offset
 * @param offsetHigh - High 32 bits of offset
 * @param lengthLow - Low 32 bits of length
 * @param lengthHigh - High 32 bits of length
 * @returns 0 on success
 */
function lockFileExSync(fd, flags, offsetLow, offsetHigh, lengthLow, lengthHigh) {
    if (!binding.lockFileEx) {
        throw new Error('lockFileEx is only available on Windows');
    }
    return binding.lockFileEx(fd, flags, offsetLow, offsetHigh, lengthLow, lengthHigh);
}
/**
 * Unlock a file region using Windows UnlockFileEx (async, Windows only).
 *
 * @param fd - File descriptor
 * @param offsetLow - Low 32 bits of offset
 * @param offsetHigh - High 32 bits of offset
 * @param lengthLow - Low 32 bits of length
 * @param lengthHigh - High 32 bits of length
 * @param callback - Called when operation completes
 */
function unlockFileEx(fd, offsetLow, offsetHigh, lengthLow, lengthHigh, callback) {
    if (!binding.unlockFileEx) {
        throw new Error('unlockFileEx is only available on Windows');
    }
    const cb = typeof callback === 'function' ? callback : noop;
    binding.unlockFileEx(fd, offsetLow, offsetHigh, lengthLow, lengthHigh, cb);
}
/**
 * Unlock a file region using Windows UnlockFileEx (sync, Windows only).
 *
 * @param fd - File descriptor
 * @param offsetLow - Low 32 bits of offset
 * @param offsetHigh - High 32 bits of offset
 * @param lengthLow - Low 32 bits of length
 * @param lengthHigh - High 32 bits of length
 * @returns 0 on success
 */
function unlockFileExSync(fd, offsetLow, offsetHigh, lengthLow, lengthHigh) {
    if (!binding.unlockFileEx) {
        throw new Error('unlockFileEx is only available on Windows');
    }
    return binding.unlockFileEx(fd, offsetLow, offsetHigh, lengthLow, lengthHigh);
}
// Ensure Windows LockFileEx constants are available on all platforms
if (binding.constants.LOCKFILE_EXCLUSIVE_LOCK === undefined) {
    binding.constants.LOCKFILE_EXCLUSIVE_LOCK = 0x00000002;
}
if (binding.constants.LOCKFILE_FAIL_IMMEDIATELY === undefined) {
    binding.constants.LOCKFILE_FAIL_IMMEDIATELY = 0x00000001;
}
/**
 * Constants for use with flock, fcntl, seek, and Windows lock functions.
 */
exports.constants = binding.constants;
/**
 * Switch to a different native module source.
 * Useful for testing to explicitly use the local build instead of prebuilts.
 *
 * @param source - 'prebuilt' or 'local'
 * @throws Error if the requested source cannot be loaded
 */
function useNativeModule(source) {
    const newBinding = loader.useNativeModule(source);
    if (!newBinding) {
        throw new Error('Failed to load fs-ext native module from source: ' + source);
    }
    binding = newBinding;
    // Re-add Windows constants if needed
    if (binding.constants.LOCKFILE_EXCLUSIVE_LOCK === undefined) {
        binding.constants.LOCKFILE_EXCLUSIVE_LOCK = 0x00000002;
    }
    if (binding.constants.LOCKFILE_FAIL_IMMEDIATELY === undefined) {
        binding.constants.LOCKFILE_FAIL_IMMEDIATELY = 0x00000001;
    }
    // Update exported constants reference
    exports.constants = binding.constants;
}
/**
 * Get the current native module source.
 *
 * @returns 'prebuilt', 'local', or null if not loaded
 */
function getNativeModuleSource() {
    return loader.getNativeModuleSource();
}
