// @php-wasm/progress/index.js
var y = (i) => {
  throw TypeError(i);
};
var b = (i, e, t) => e.has(i) || y("Cannot " + t);
var n = (i, e, t) => (b(i, e, "read from private field"), t ? t.call(i) : e.get(i)), f = (i, e, t) => e.has(i) ? y("Cannot add the same private member more than once") : e instanceof WeakSet ? e.add(i) : e.set(i, t), _ = (i, e, t, s) => (b(i, e, "write to private field"), s ? s.call(i, t) : e.set(i, t), t), m = (i, e, t) => (b(i, e, "access private method"), t);
var w = (i, e, t, s) => ({
  set _(o) {
    _(i, e, o, t);
  },
  get _() {
    return n(i, e, s);
  }
});
import { logger as I } from "@php-wasm/logger";
const R = 5 * 1024 * 1024;
var l, g, v, C;
class F extends EventTarget {
  constructor() {
    super(...arguments);
    f(this, v);
    f(this, l, {});
    f(this, g, {});
  }
  expectAssets(t) {
    for (const [s, o] of Object.entries(t)) {
      const r = "http://example.com/", a = new URL(s, r).pathname.split("/").pop();
      a in n(this, l) || (n(this, l)[a] = o), a in n(this, g) || (n(this, g)[a] = 0);
    }
  }
  async monitorFetch(t) {
    const s = await t;
    return x(s, (r) => {
      m(this, v, C).call(this, s.url, r.detail.loaded, r.detail.total);
    });
  }
}
l = new WeakMap(), g = new WeakMap(), v = new WeakSet(), /**
 * Notifies about the download #progress of a file.
 *
 * @param  file   The file name.
 * @param  loaded The number of bytes of that file loaded so far.
 * @param  fileSize  The length number of bytes in the loaded file.
 */
C = function(t, s, o) {
  const r = new URL(t, "http://example.com").pathname.split("/").pop();
  o ? r in n(this, l) || (n(this, l)[r] = o, n(this, g)[r] = s) : o = n(this, l)[r], r in n(this, g) || I.warn(
    `Registered a download #progress of an unregistered file "${r}". This may cause a sudden **decrease** in the #progress percentage as the length number of bytes increases during the download.`
  ), n(this, g)[r] = s, this.dispatchEvent(
    new CustomEvent("progress", {
      detail: {
        loaded: T(n(this, g)),
        total: T(n(this, l))
      }
    })
  );
};
function T(i) {
  return Object.values(i).reduce((e, t) => e + t, 0);
}
function x(i, e) {
  const t = i.headers.get("content-length") || "", s = parseInt(t, 10) || R;
  return new Response(
    k(i.body, s, e),
    {
      status: i.status,
      statusText: i.statusText,
      headers: i.headers
    }
  );
}
function k(i, e, t) {
  let s = 0;
  function o(r, c, a) {
    const h = performance.now();
    !a && h - s < 500 || (s = h, t(
      new CustomEvent("progress", {
        detail: {
          loaded: r,
          total: c
        }
      })
    ));
  }
  return new ReadableStream({
    async start(r) {
      if (!i) {
        r.close();
        return;
      }
      const c = i.getReader();
      let a = 0;
      for (; ; )
        try {
          const { done: h, value: E } = await c.read();
          if (E && (a += E.byteLength), h) {
            o(a, a, h), r.close();
            break;
          } else
            o(a, e, h), r.enqueue(E);
        } catch (h) {
          I.error({ e: h }), r.error(h);
          break;
        }
    }
  });
}
var d, u, p, P;
class S extends EventTarget {
  constructor() {
    super(...arguments);
    f(this, p);
    f(this, d);
    f(this, u);
    _(this, d, {}), _(this, u, 0), this.progress = 0, this.mode = "REAL_TIME", this.caption = "";
  }
  partialObserver(t, s = "") {
    const o = ++w(this, u)._;
    return n(this, d)[o] = 0, (r) => {
      const { loaded: c, total: a } = (r == null ? void 0 : r.detail) || {};
      n(this, d)[o] = c / a * t, m(this, p, P).call(this, this.totalProgress, "REAL_TIME", s);
    };
  }
  slowlyIncrementBy(t) {
    const s = ++w(this, u)._;
    n(this, d)[s] = t, m(this, p, P).call(this, this.totalProgress, "SLOWLY_INCREMENT");
  }
  get totalProgress() {
    return Object.values(n(this, d)).reduce(
      (t, s) => t + s,
      0
    );
  }
}
d = new WeakMap(), u = new WeakMap(), p = new WeakSet(), P = function(t, s, o) {
  this.dispatchEvent(
    new CustomEvent("progress", {
      detail: {
        progress: t,
        mode: s,
        caption: o
      }
    })
  );
};
const L = 1e-5;
class O extends EventTarget {
  constructor({
    weight: e = 1,
    caption: t = "",
    fillTime: s = 4
  } = {}) {
    super(), this._selfWeight = 1, this._selfDone = !1, this._selfProgress = 0, this._selfCaption = "", this._isFilling = !1, this._subTrackers = [], this._weight = e, this._selfCaption = t, this._fillTime = s;
  }
  /**
   * Creates a new sub-tracker with a specific weight.
   *
   * The weight determines what percentage of the overall progress
   * the sub-tracker represents. For example, if the main tracker is
   * monitoring a process that has two stages, and the first stage
   * is expected to take twice as long as the second stage, you could
   * create the first sub-tracker with a weight of 0.67 and the second
   * sub-tracker with a weight of 0.33.
   *
   * The caption is an optional string that describes the current stage
   * of the operation. If provided, it will be used as the progress caption
   * for the sub-tracker. If not provided, the main tracker will look for
   * the next sub-tracker with a non-empty caption and use that as the progress
   * caption instead.
   *
   * Returns the newly-created sub-tracker.
   *
   * @throws {Error} If the weight of the new stage would cause the total weight of all stages to exceed 1.
   *
   * @param weight The weight of the new stage, as a decimal value between 0 and 1.
   * @param caption The caption for the new stage, which will be used as the progress caption for the sub-tracker.
   *
   * @example
   * ```ts
   * const tracker = new ProgressTracker();
   * const subTracker1 = tracker.stage(0.67, 'Slow stage');
   * const subTracker2 = tracker.stage(0.33, 'Fast stage');
   *
   * subTracker2.set(50);
   * subTracker1.set(75);
   * subTracker2.set(100);
   * subTracker1.set(100);
   * ```
   */
  stage(e, t = "") {
    if (e || (e = this._selfWeight), this._selfWeight - e < -L)
      throw new Error(
        `Cannot add a stage with weight ${e} as the total weight of registered stages would exceed 1.`
      );
    this._selfWeight -= e;
    const s = new O({
      caption: t,
      weight: e,
      fillTime: this._fillTime
    });
    return this._subTrackers.push(s), s.addEventListener("progress", () => this.notifyProgress()), s.addEventListener("done", () => {
      this.done && this.notifyDone();
    }), s;
  }
  /**
   * Fills the progress bar slowly over time, simulating progress.
   *
   * The progress bar is filled in a 100 steps, and each step, the progress
   * is increased by 1. If `stopBeforeFinishing` is true, the progress bar
   * will stop filling when it reaches 99% so that you can call `finish()`
   * explicitly.
   *
   * If the progress bar is filling or already filled, this method does nothing.
   *
   * @example
   * ```ts
   * const progress = new ProgressTracker({ caption: 'Processing...' });
   * progress.fillSlowly();
   * ```
   *
   * @param options Optional options.
   */
  fillSlowly({ stopBeforeFinishing: e = !0 } = {}) {
    if (this._isFilling)
      return;
    this._isFilling = !0;
    const s = this._fillTime / 100;
    this._fillInterval = setInterval(() => {
      this.set(this._selfProgress + 1), e && this._selfProgress >= 99 && clearInterval(this._fillInterval);
    }, s);
  }
  set(e) {
    this._selfProgress = Math.min(e, 100), this.notifyProgress(), this._selfProgress + L >= 100 && this.finish();
  }
  finish() {
    this._fillInterval && clearInterval(this._fillInterval), this._selfDone = !0, this._selfProgress = 100, this._isFilling = !1, this._fillInterval = void 0, this.notifyProgress(), this.notifyDone();
  }
  get caption() {
    for (let e = this._subTrackers.length - 1; e >= 0; e--)
      if (!this._subTrackers[e].done) {
        const t = this._subTrackers[e].caption;
        if (t)
          return t;
      }
    return this._selfCaption;
  }
  setCaption(e) {
    this._selfCaption = e, this.notifyProgress();
  }
  get done() {
    return this.progress + L >= 100;
  }
  get progress() {
    if (this._selfDone)
      return 100;
    const e = this._subTrackers.reduce(
      (t, s) => t + s.progress * s.weight,
      this._selfProgress * this._selfWeight
    );
    return Math.round(e * 1e4) / 1e4;
  }
  get weight() {
    return this._weight;
  }
  get observer() {
    return this._progressObserver || (this._progressObserver = (e) => {
      this.set(e);
    }), this._progressObserver;
  }
  get loadingListener() {
    return this._loadingListener || (this._loadingListener = (e) => {
      this.set(e.detail.loaded / e.detail.total * 100);
    }), this._loadingListener;
  }
  pipe(e) {
    e.setProgress({
      progress: this.progress,
      caption: this.caption
    }), this.addEventListener("progress", (t) => {
      e.setProgress({
        progress: t.detail.progress,
        caption: t.detail.caption
      });
    }), this.addEventListener("done", () => {
      e.setLoaded();
    });
  }
  addEventListener(e, t) {
    super.addEventListener(e, t);
  }
  removeEventListener(e, t) {
    super.removeEventListener(e, t);
  }
  notifyProgress() {
    const e = this;
    this.dispatchEvent(
      new CustomEvent("progress", {
        detail: {
          get progress() {
            return e.progress;
          },
          get caption() {
            return e.caption;
          }
        }
      })
    );
  }
  notifyDone() {
    this.dispatchEvent(new CustomEvent("done"));
  }
}
export {
  F as EmscriptenDownloadMonitor,
  S as ProgressObserver,
  O as ProgressTracker,
  x as cloneResponseMonitorProgress,
  k as cloneStreamMonitorProgress
};
