// @wp-playground/common/index.js
import { phpVars as c } from "@php-wasm/util";
function l(i = fetch) {
  const e = {};
  return async function(t, a) {
    if (!e[t]) {
      e[t] = {
        responsePromise: i(t, a),
        async nextResponse() {
          const o = await e[t].responsePromise, [n, $] = e[t].unlockedBodyStream.tee();
          return e[t].unlockedBodyStream = n, new Response($, {
            status: o.status,
            statusText: o.statusText,
            headers: o.headers
          });
        }
      };
      const s = await e[t].responsePromise;
      e[t].unlockedBodyStream = s.body;
    }
    return e[t].nextResponse();
  };
}
const p = "8.3", h = async (i, e, r, t = !0) => {
  const a = `/tmp/file-${Math.random()}.zip`;
  let s = !1;
  try {
    if (e instanceof File) {
      const n = e;
      e = a, s = !0, await i.writeFile(
        e,
        new Uint8Array(await n.arrayBuffer())
      );
    }
    const o = c({
      zipPath: e,
      extractToPath: r,
      overwriteFiles: t
    });
    await i.run({
      code: `<?php
        function unzip($zipPath, $extractTo, $overwriteFiles = true)
        {
            if (!is_dir($extractTo)) {
                mkdir($extractTo, 0777, true);
            }
            $zip = new ZipArchive;
            $res = $zip->open($zipPath);
            if ($res === TRUE) {
				try {
					if ($overwriteFiles) {
						if (!$zip->extractTo($extractTo)) {
							throw new Exception('Could not extract ZIP file.');
						}
					} else {
						for ($i = 0; $i < $zip->numFiles; $i++) {
							$filename = $zip->getNameIndex($i);
							$extractFilePath = rtrim($extractTo, '/') . '/' . $filename;
							// Check if file exists and $overwriteFiles is false
							if (!file_exists($extractFilePath)) {
								// Extract file
								$zip->extractTo($extractTo, $filename);
							}
						}
					}
				} catch (Exception $e) {
					$zip->close();
					throw $e;
				}
				$zip->close();
				chmod($extractTo, 0777);
            } else {
                $fileSize = file_exists($zipPath) ? filesize($zipPath) : 'unknown';
	                throw new Exception("Could not unzip file. Error code: " . $res . ". File size: " . $fileSize . " bytes.");
	            }
		}
        unzip(${o.zipPath}, ${o.extractToPath}, ${o.overwriteFiles});
        `
    });
  } finally {
    if (s)
      try {
        await i.fileExists(a) && await i.unlink(a);
      } catch {
      }
  }
}, u = async (i, e) => {
  const r = `/tmp/file${Math.random()}.zip`;
  try {
    const t = c({
      directoryPath: e,
      outputPath: r
    });
    return await i.run({
      code: `<?php
			/**
			 * Creates a ZIP archive from a Playground directory.
			 *
			 * The JavaScript wrapper removes the temporary archive in a finally
			 * block, so this function only owns archive creation and permissions.
			 */
			function zipDirectory($directoryPath, $outputPath) {
				$zip = new ZipArchive;
				$res = $zip->open($outputPath, ZipArchive::CREATE);
			if ($res !== TRUE) {
				throw new Exception('Failed to create ZIP');
			}
			$files = new RecursiveIteratorIterator(
				new RecursiveDirectoryIterator($directoryPath)
			);
			foreach ($files as $file) {
				$file = strval($file);
				if (is_dir($file)) {
					continue;
				}
				$zip->addFile($file, substr($file, strlen($directoryPath)));
			}
			$zip->close();
			chmod($outputPath, 0777);
		}
			zipDirectory(${t.directoryPath}, ${t.outputPath});
			`
    }), await i.readFileAsBuffer(r);
  } finally {
    try {
      await i.fileExists(r) && await i.unlink(r);
    } catch {
    }
  }
};
export {
  p as RecommendedPHPVersion,
  l as createMemoizedFetch,
  h as unzipFile,
  u as zipDirectory
};
